{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 7 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 10/31 by 11:59 pm***\n",
    "\n",
    "## Samtools\n",
    "\n",
    "Use Samtools to convert to BAM and sort the output SAMs produced by HISAT:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converting ../6_alignment/alignment/hisat/ERR1442621.sam to ERR1442621.bam and sorting.\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n",
      "Converting ../6_alignment/alignment/hisat/ERR1442622.sam to ERR1442622.bam and sorting.\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n",
      "Converting ../6_alignment/alignment/hisat/ERR1442623.sam to ERR1442623.bam and sorting.\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n",
      "Converting ../6_alignment/alignment/hisat/ERR1442631.sam to ERR1442631.bam and sorting.\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n",
      "Converting ../6_alignment/alignment/hisat/ERR1442632.sam to ERR1442632.bam and sorting.\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n",
      "Converting ../6_alignment/alignment/hisat/ERR1442633.sam to ERR1442633.bam and sorting.\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n",
      "Converting ../6_alignment/alignment/hisat/ERR1442641.sam to ERR1442641.bam and sorting.\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n",
      "Converting ../6_alignment/alignment/hisat/ERR1442642.sam to ERR1442642.bam and sorting.\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n",
      "Converting ../6_alignment/alignment/hisat/ERR1442643.sam to ERR1442643.bam and sorting.\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "!for sam in ../6_alignment/alignment/hisat/E*.sam; do \\\n",
    "    bn=$(basename \"$sam\" .sam); \\\n",
    "    echo \"Converting $sam to $bn.bam and sorting.\"; \\\n",
    "    samtools view -@ 32 -b $sam | samtools sort -@ 32 > ../6_alignment/alignment/hisat/$bn.bam; \\\n",
    "    done"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge the BAMs to a single BAM and generate an index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools merge -@ 32 ../6_alignment/alignment/hisat/merged.bam  ../6_alignment/alignment/hisat/*.bam \n",
    "!samtools index -@ 32 ../6_alignment/alignment/hisat/merged.bam "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use Samtools `flagstat` to generate QC metrics for the STAR and HISAT merged BAMs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36657911 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "3790977 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "32865834 + 0 mapped (89.66% : N/A)\n",
      "32866934 + 0 paired in sequencing\n",
      "16433467 + 0 read1\n",
      "16433467 + 0 read2\n",
      "27601850 + 0 properly paired (83.98% : N/A)\n",
      "27978204 + 0 with itself and mate mapped\n",
      "1096653 + 0 singletons (3.34% : N/A)\n",
      "60984 + 0 with mate mapped to a different chr\n",
      "41816 + 0 with mate mapped to a different chr (mapQ>=5)\n",
      "35951572 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "3084638 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "32819336 + 0 mapped (91.29% : N/A)\n",
      "32866934 + 0 paired in sequencing\n",
      "16433467 + 0 read1\n",
      "16433467 + 0 read2\n",
      "29734698 + 0 properly paired (90.47% : N/A)\n",
      "29734698 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (0.00% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
     ]
    }
   ],
   "source": [
    "!samtools flagstat ../6_alignment/alignment/hisat/merged.bam\n",
    "!samtools flagstat ../6_alignment/alignment/star/Aligned.sorted.bam"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Qualimap\n",
    "\n",
    "Use Qualimap to run `bamqc` and `rnaseq` to generate QC metrics for the STAR and HISAT merged BAMs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 2] No such file or directory: '../6_alignment/alignment/star'\n",
      "/data/users/heky1803/BIOL343/6_alignment/alignment/star\n"
     ]
    }
   ],
   "source": [
    "%cd ../6_alignment/alignment/star\n",
    "!samtools view -H Aligned.sorted.bam > header.sam\n",
    "!sed -i 's/^@RG\\tID:\\(.*\\)/@RG\\tID:\\1\\tSM:\\1/' header.sam\n",
    "!samtools reheader header.sam Aligned.sorted.bam > Aligned.fixed.bam\n",
    "!samtools index Aligned.fixed.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/users/heky1803/BIOL343/7_alignment_qc\n"
     ]
    }
   ],
   "source": [
    "%cd ~/BIOL343/7_alignment_qc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Java memory size is set to 1200M\n",
      "Launching application...\n",
      "\n",
      "[0.001s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /dev/cpuset.\n",
      "QualiMap v.2.3\n",
      "Built on 2023-05-19 16:57\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 35\n",
      "Max memory (Mb): 1258\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 1460\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 32\n",
      "Initializing regions from /data/users/heky1803/BIOL343/2_genome_exploration/genome/zebrafish_annotations.gtf.....\n",
      "Found 1390999 regions\n",
      "Filling region references... \n",
      "Processed 146 out of 1460 windows...\n",
      "Processed 292 out of 1460 windows...\n",
      "Processed 438 out of 1460 windows...\n",
      "Processed 584 out of 1460 windows...\n",
      "Processed 730 out of 1460 windows...\n",
      "Processed 876 out of 1460 windows...\n",
      "Processed 1022 out of 1460 windows...\n",
      "Processed 1168 out of 1460 windows...\n",
      "Processed 1314 out of 1460 windows...\n",
      "Processed 1460 out of 1460 windows...\n",
      "Total processed windows:1460\n",
      "Number of reads: 32866934\n",
      "Number of valid reads: 29734698\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 27091772\n",
      "Num mapped first of pair: 13527064\n",
      "Num mapped second of pair: 13564708\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 665\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 21972867446\n",
      "referenceSize: 1371719383\n",
      "numberOfSequencedBases: 2680236209\n",
      "numberOfAs: 714703687\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 673\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "Finished\n",
      "Java memory size is set to 1200M\n",
      "Launching application...\n",
      "\n",
      "[0.001s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /dev/cpuset.\n",
      "QualiMap v.2.3\n",
      "Built on 2023-05-19 16:57\n",
      "\n",
      "Selected tool: rnaseq\n",
      "Thu Oct 30 16:31:16 CDT 2025\t\tWARNING\tOutput folder already exists, the results will be saved there\n",
      "\n",
      "Initializing regions from /data/users/heky1803/BIOL343/2_genome_exploration/genome/zebrafish_annotations.gtf...\n",
      "\n",
      "Initialized 100000 regions...\n",
      "Initialized 200000 regions...\n",
      "Initialized 300000 regions...\n",
      "Initialized 400000 regions...\n",
      "Initialized 500000 regions...\n",
      "Initialized 600000 regions...\n",
      "Initialized 700000 regions...\n",
      "Initialized 800000 regions...\n",
      "Initialized 900000 regions...\n",
      "Initialized 1000000 regions...\n",
      "Initialized 1100000 regions...\n",
      "Initialized 1200000 regions...\n",
      "Initialized 1300000 regions...\n",
      "\n",
      "Initialized 1390999 regions it total\n",
      "\n",
      "Starting constructing transcripts for RNA-seq stats...\n",
      "Warning! Transcript NM_001077752.1 has incorrect exon coordinates. Transcript skipped...\n",
      "Finished constructing transcripts\n",
      "\n",
      "Starting BAM file analysis\n",
      "\n",
      "Analyzed 500000 reads...\n",
      "Analyzed 1000000 reads...\n",
      "Analyzed 1500000 reads...\n",
      "Analyzed 2000000 reads...\n",
      "Analyzed 2500000 reads...\n",
      "Analyzed 3000000 reads...\n",
      "Analyzed 3500000 reads...\n",
      "Analyzed 4000000 reads...\n",
      "Analyzed 4500000 reads...\n",
      "Analyzed 6500000 reads...\n",
      "Analyzed 7000000 reads...\n",
      "Analyzed 8000000 reads...\n",
      "Analyzed 8500000 reads...\n",
      "Analyzed 9000000 reads...\n",
      "Analyzed 9500000 reads...\n",
      "Analyzed 10000000 reads...\n",
      "Analyzed 10500000 reads...\n",
      "Analyzed 11000000 reads...\n",
      "Analyzed 12500000 reads...\n",
      "Analyzed 13000000 reads...\n",
      "Analyzed 13500000 reads...\n",
      "Analyzed 14500000 reads...\n",
      "Analyzed 15000000 reads...\n",
      "Analyzed 15500000 reads...\n",
      "Analyzed 16000000 reads...\n",
      "Analyzed 16500000 reads...\n",
      "Analyzed 17500000 reads...\n",
      "Analyzed 18000000 reads...\n",
      "Analyzed 18500000 reads...\n",
      "Analyzed 19000000 reads...\n",
      "Analyzed 19500000 reads...\n",
      "Analyzed 20000000 reads...\n",
      "Analyzed 20500000 reads...\n",
      "Analyzed 21000000 reads...\n",
      "Analyzed 21500000 reads...\n",
      "Analyzed 22000000 reads...\n",
      "Analyzed 22500000 reads...\n",
      "Analyzed 23000000 reads...\n",
      "Analyzed 23500000 reads...\n",
      "Analyzed 24000000 reads...\n",
      "Analyzed 24500000 reads...\n",
      "Analyzed 25000000 reads...\n",
      "Analyzed 25500000 reads...\n",
      "Analyzed 26000000 reads...\n",
      "Analyzed 26500000 reads...\n",
      "Analyzed 27000000 reads...\n",
      "Analyzed 27500000 reads...\n",
      "Analyzed 28000000 reads...\n",
      "Analyzed 28500000 reads...\n",
      "Analyzed 29000000 reads...\n",
      "Analyzed 29500000 reads...\n",
      "Analyzed 30000000 reads...\n",
      "Analyzed 30500000 reads...\n",
      "Analyzed 31000000 reads...\n",
      "Analyzed 31500000 reads...\n",
      "Analyzed 32000000 reads...\n",
      "Analyzed 32500000 reads...\n",
      "\n",
      "WARNING! The following chromosomes from reads are not found in annotations:\n",
      "NW_003336881.1\n",
      "NW_003337184.1\n",
      "NW_003336429.1\n",
      "NW_003336346.1\n",
      "NW_003336323.1\n",
      "NW_003336783.1\n",
      "NW_003336361.1\n",
      "NW_008805409.1\n",
      "NW_003336760.1\n",
      "NW_008805567.1\n",
      "NW_008805398.1\n",
      "NW_003336990.1\n",
      "NW_003336432.1\n",
      "NW_008805544.1\n",
      "NW_003336478.1\n",
      "NW_003336625.1\n",
      "NW_008805529.1\n",
      "NW_008805485.1\n",
      "NW_003336479.1\n",
      "NW_003337196.1\n",
      "NW_001884441.4\n",
      "NW_008805401.1\n",
      "NW_003336372.1\n",
      "NW_008805364.1\n",
      "NW_008805578.1\n",
      "NW_003337052.1\n",
      "NW_008805532.1\n",
      "NW_003337150.1\n",
      "NW_003336820.1\n",
      "NW_003336805.1\n",
      "NW_003337037.1\n",
      "NW_003336843.1\n",
      "NW_003336421.1\n",
      "NW_003335184.2\n",
      "NW_008805570.1\n",
      "NW_003336866.1\n",
      "NW_008805517.1\n",
      "NW_003336598.1\n",
      "NW_003337168.1\n",
      "NW_003334517.2\n",
      "NW_003337062.1\n",
      "NW_003337122.1\n",
      "NW_008805566.1\n",
      "NW_003336095.2\n",
      "NW_003336454.2\n",
      "NW_003337217.1\n",
      "NW_003337160.1\n",
      "NW_003337047.1\n",
      "NW_008805581.1\n",
      "NW_008805486.1\n",
      "NW_003336769.1\n",
      "NW_003336541.1\n",
      "NW_003335052.1\n",
      "NW_003335922.2\n",
      "NW_003337111.1\n",
      "NW_003336443.1\n",
      "NW_003334032.1\n",
      "NW_003336651.1\n",
      "NW_003337229.1\n",
      "NW_003337200.1\n",
      "NW_008805433.1\n",
      "NW_003336962.2\n",
      "NW_008805471.1\n",
      "NW_008805494.1\n",
      "NW_003336755.1\n",
      "NW_003336574.1\n",
      "NW_003336924.1\n",
      "NW_003336548.1\n",
      "NW_003336525.1\n",
      "NW_008805407.1\n",
      "NW_008805396.1\n",
      "NW_003337163.1\n",
      "NW_008805546.1\n",
      "NW_003336815.1\n",
      "NW_003334254.1\n",
      "NW_003335242.1\n",
      "NW_003337198.1\n",
      "NW_003337250.1\n",
      "NW_003336959.1\n",
      "NW_003336766.1\n",
      "NW_003337137.1\n",
      "NW_008805419.1\n",
      "NW_003336514.1\n",
      "NW_008805385.1\n",
      "NW_003337031.1\n",
      "NW_003336826.1\n",
      "NW_003335375.1\n",
      "NW_003336610.1\n",
      "NW_008805519.1\n",
      "NW_003334182.1\n",
      "NW_008805472.1\n",
      "NW_003334519.2\n",
      "NW_003335206.1\n",
      "NW_003335935.1\n",
      "NW_003337101.1\n",
      "NW_003337041.1\n",
      "NW_003336097.2\n",
      "NW_008805522.1\n",
      "NW_003336415.1\n",
      "NW_003336830.1\n",
      "NW_003337174.1\n",
      "NW_008805446.1\n",
      "NW_003336706.1\n",
      "NW_003336721.1\n",
      "NW_008805461.1\n",
      "NW_003336998.2\n",
      "NW_003336744.1\n",
      "NW_003337136.1\n",
      "NW_008805348.1\n",
      "NW_008805386.1\n",
      "NW_008805510.1\n",
      "NW_008805556.1\n",
      "NW_003337099.1\n",
      "NW_003334034.1\n",
      "NW_003336888.1\n",
      "NW_003334628.1\n",
      "NW_003336425.1\n",
      "NW_003337225.1\n",
      "NW_003337202.1\n",
      "NW_003334452.1\n",
      "NW_003336738.1\n",
      "NW_003336365.1\n",
      "NW_003337139.1\n",
      "NW_003336764.1\n",
      "NW_003337127.1\n",
      "NW_003336629.1\n",
      "NW_003337044.1\n",
      "NW_003336282.1\n",
      "NW_003336474.1\n",
      "NW_003336451.1\n",
      "NW_003336813.1\n",
      "NW_003337192.1\n",
      "NW_003336414.1\n",
      "NW_008805466.1\n",
      "NW_003336399.1\n",
      "NW_003336437.1\n",
      "NW_008805525.1\n",
      "NW_003336561.2\n",
      "NW_003336376.1\n",
      "NW_003336330.2\n",
      "NW_003336957.1\n",
      "NW_003336809.1\n",
      "NW_003334852.2\n",
      "NW_003337071.1\n",
      "NW_003336278.1\n",
      "NW_008805368.1\n",
      "NW_003337154.1\n",
      "NW_003336327.1\n",
      "NW_003336801.1\n",
      "NW_003334814.2\n",
      "NW_008805383.1\n",
      "NW_003334499.2\n",
      "NW_003336654.1\n",
      "NW_003336824.1\n",
      "NW_003336840.1\n",
      "NW_003336424.1\n",
      "NW_003337247.1\n",
      "NW_008805432.1\n",
      "NW_003334683.1\n",
      "NW_003336387.2\n",
      "NW_003334956.2\n",
      "NW_008805559.1\n",
      "NW_003335276.1\n",
      "NW_003336780.1\n",
      "NW_003337187.1\n",
      "NW_003336716.1\n",
      "NW_003336510.1\n",
      "NW_003336754.1\n",
      "NW_003337126.1\n",
      "NW_003336742.1\n",
      "NW_003336995.1\n",
      "NW_003336568.1\n",
      "NW_003337043.1\n",
      "NW_008805372.1\n",
      "NW_003336281.1\n",
      "NW_003337029.1\n",
      "NW_003336765.1\n",
      "NW_003337070.1\n",
      "NW_003336349.1\n",
      "NW_008805418.1\n",
      "NW_003335007.1\n",
      "NW_003336670.1\n",
      "NW_008805535.1\n",
      "NW_003336848.1\n",
      "NW_008805384.1\n",
      "NW_008805361.1\n",
      "NW_003336632.1\n",
      "NW_003336386.1\n",
      "NW_003337182.1\n",
      "NW_003337265.1\n",
      "NW_003334684.1\n",
      "NW_008805516.1\n",
      "NW_008805490.1\n",
      "NW_003336966.1\n",
      "NW_003335250.1\n",
      "NW_003336302.1\n",
      "NW_003336593.1\n",
      "NW_003337144.1\n",
      "NW_003336288.1\n",
      "NW_003335691.1\n",
      "NW_003336529.1\n",
      "NW_003337069.1\n",
      "NW_003336664.1\n",
      "NW_003336834.1\n",
      "NW_008805487.1\n",
      "NW_003336871.1\n",
      "NW_008805464.1\n",
      "NW_008805504.1\n",
      "NW_003336649.1\n",
      "NW_003336458.1\n",
      "NW_003336435.2\n",
      "NW_008805403.1\n",
      "NW_003336747.1\n",
      "NW_003337179.1\n",
      "NW_003336374.1\n",
      "NW_003336773.1\n",
      "NW_003334854.1\n",
      "NW_003336325.1\n",
      "NW_003337205.1\n",
      "NW_003334455.1\n",
      "NW_003336675.1\n",
      "NW_008805530.1\n",
      "NW_003336385.1\n",
      "NW_008805453.1\n",
      "NW_003334453.1\n",
      "NW_003336324.1\n",
      "NW_008805404.1\n",
      "NW_008805427.1\n",
      "NW_003336433.1\n",
      "NW_003334404.1\n",
      "NW_003337008.2\n",
      "NW_003335322.1\n",
      "NW_003336665.1\n",
      "NW_003336494.1\n",
      "NW_003336895.1\n",
      "NW_003336688.1\n",
      "NW_003334988.2\n",
      "NW_008805488.1\n",
      "NW_001884442.4\n",
      "NW_003337106.1\n",
      "NW_003336350.1\n",
      "NW_003336702.1\n",
      "NW_003336956.1\n",
      "NW_003337095.1\n",
      "NW_003336347.1\n",
      "NW_008805367.1\n",
      "NW_003335920.1\n",
      "NW_003337132.1\n",
      "NW_003336846.1\n",
      "NW_003336290.1\n",
      "NW_008805382.1\n",
      "NW_003337034.1\n",
      "NW_003336422.1\n",
      "\n",
      "Processed 32801652 reads in total\n",
      "\n",
      "BAM file analysis finished\n",
      "Creating plots\n",
      "Writing HTML report...\n",
      "HTML report created successfully\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!qualimap bamqc -nt 32 -outdir qualimap/star/bam -bam ~/BIOL343/6_alignment/alignment/star/Aligned.fixed.bam --feature-file ~/BIOL343/2_genome_exploration/genome/zebrafish_annotations.gtf\n",
    "!qualimap rnaseq -outdir qualimap/star/rnaseq -bam ~/BIOL343/6_alignment/alignment/star/Aligned.fixed.bam -gtf ~/BIOL343/2_genome_exploration/genome/zebrafish_annotations.gtf\n",
    "#!qualimap bamqc -nt 32 -outdir qualimap/hisat/bam -bam ../6_alignment/alignment/hisat/merged.bam --feature-file ../2_genome_exploration/genome/zebrafish_annotations.gtf\n",
    "#!qualimap rnaseq -outdir qualimap/hisat/rnaseq -bam ../6_alignment/alignment/hisat/merged.bam -gtf ../2_genome_exploration/genome/zebrafish_annotations.gtf "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Report\n",
    "\n",
    "Use MuliQC to generate a single report that merges all Samtools and Qualimap metrics:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=313553;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2m| v1.17\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Prepending directory to sample names\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/heky1803/BIOL343/5_fastq/hw_fastq/qc\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/heky1803/BIOL343/5_fastq/hwtrimmed/qc\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/heky1803/BIOL343/6_alignment\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/heky1803/BIOL343/7_alignment_qc\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m325/325\u001b[0m  rnaseq/css/up-pressed.png\u001b[0mp.png\u001b[0m\n",
      "\u001b[?25h\u001b[34m|          qualimap\u001b[0m | Found 2 BamQC reports\n",
      "\u001b[34m|          qualimap\u001b[0m | Found 2 RNASeq reports\n",
      "\u001b[34m|          samtools\u001b[0m | Found 1 stats reports\n",
      "\u001b[34m|          samtools\u001b[0m | Found 2 flagstat reports\n",
      "\u001b[34m|              star\u001b[0m | Found 1 reports\n",
      "\u001b[34m|            hisat2\u001b[0m | Found 20 reports\n",
      "\u001b[34m|            fastqc\u001b[0m | Found 36 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "!multiqc --force -d ../5_fastq/hw_fastq/qc/ ../5_fastq/hwtrimmed/qc/ ../6_alignment ."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## JBrowse2\n",
    "\n",
    "Upload a screenshot of JBrowse2 that shows the reference genome, the annotations, and the two alignments as tracks."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Snakemake Pipeline (Optional)\n",
    "\n",
    "1. Copy your `Snakefile` and `config.yaml` from `6_alignment` to `7_alignment_qc`.  \n",
    "2. In the `Snakefile`, add the following rules:  \n",
    "    a. `convert_hisat` - use `samtools` to convert HISAT SAMs to BAMs; sort the BAMs.  \n",
    "    b. `merge_hisat` - use `samtools` to merge all HISAT BAMs.  \n",
    "    c. `flagstat` -  use `samtools` to generate `flagstat` QC.  \n",
    "    d. `qualimap` - use `qualimap` to generate `bamqc` and `rnaseq` QC.  \n",
    "    e. `report` - use `multiqc` to aggregate all QC data into a single report. This should be an update of the report from the previous week's homework.  \n",
    "\n",
    "This `Snakefile` should not use any `wrappers` because we already have the relevant software installed in the `biol343` conda environment. The pipeline should run to completion when any instructor or classmate runs `snakemake --use-conda`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biol343",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
