{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 6 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 10/24 by 11:59 pm***\n",
    "\n",
    "## STAR two-pass alignment\n",
    "\n",
    "Build the STAR SA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘alignment/star_for_homework/’: File exists\n",
      "mkdir: cannot create directory ‘../2_genome_exploration/genome_for_homework/star_for_homework’: File exists\n",
      "\tSTAR --runThreadN 32 --runMode genomeGenerate --genomeDir ../2_genome_exploration/genome_for_homework/star_for_homework --genomeFastaFiles ../2_genome_exploration/genome_for_homework/genome_for_homework.fa --sjdbGTFfile ../2_genome_exploration/genome_for_homework/annotations_for_homework.gtf --sjdbOverhang 50 --genomeSAindexNbases 13\n",
      "\tSTAR version: 2.7.10b   compiled: 2022-11-01T09:53:26-04:00 :/home/dobin/data/STAR/STARcode/STAR.master/source\n",
      "Oct 21 21:42:37 ..... started STAR run\n",
      "Oct 21 21:42:37 ... starting to generate Genome files\n",
      "Oct 21 21:43:14 ..... processing annotations GTF\n",
      "Oct 21 21:43:48 ... starting to sort Suffix Array. This may take a long time...\n",
      "Oct 21 21:44:01 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Oct 21 21:52:12 ... loading chunks from disk, packing SA...\n",
      "Oct 21 21:57:10 ... finished generating suffix array\n",
      "Oct 21 21:57:10 ... generating Suffix Array index\n",
      "Oct 21 21:59:12 ... completed Suffix Array index\n",
      "Oct 21 21:59:15 ..... inserting junctions into the genome indices\n",
      "Oct 21 22:00:56 ... writing Genome to disk ...\n",
      "Oct 21 22:01:39 ... writing Suffix Array to disk ...\n",
      "Oct 21 22:05:03 ... writing SAindex to disk\n",
      "Oct 21 22:05:09 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "%mkdir alignment/star_for_homework/\n",
    "%mkdir ../2_genome_exploration/genome_for_homework/star_for_homework\n",
    "\n",
    "!STAR \\\n",
    "--runThreadN 32 \\\n",
    "--runMode genomeGenerate \\\n",
    "--genomeDir ../2_genome_exploration/genome_for_homework/star_for_homework \\\n",
    "--genomeFastaFiles ../2_genome_exploration/genome_for_homework/genome_for_homework.fa \\\n",
    "--sjdbGTFfile ../2_genome_exploration/genome_for_homework/annotations_for_homework.gtf \\\n",
    "--sjdbOverhang 50 \\\n",
    "--genomeSAindexNbases 13"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use STAR and ***all*** of your FASTQs for a first-pass alignment. Be sure to update the `sjdbOverhang` parameter to a value that is appropriate for your dataset. Rename the log file so that you can compare the logs from the first and second alignments:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tSTAR --runThreadN 32 --runMode alignReads --genomeDir ../2_genome_exploration/genome_for_homework/star_for_homework --readFilesManifest manifest_HW.tsv --readFilesCommand zcat --outSAMtype BAM Unsorted --outSAMunmapped Within --outFileNamePrefix alignment/star_for_homework/first_pass_ --sjdbOverhang 50\n",
      "\tSTAR version: 2.7.10b   compiled: 2022-11-01T09:53:26-04:00 :/home/dobin/data/STAR/STARcode/STAR.master/source\n",
      "Oct 21 22:07:06 ..... started STAR run\n",
      "Oct 21 22:07:07 ..... loading genome\n",
      "Oct 21 22:09:43 ..... started mapping\n",
      "Oct 21 23:06:35 ..... finished mapping\n",
      "Oct 21 23:06:49 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "!STAR \\\n",
    "--runThreadN 32 \\\n",
    "--runMode alignReads \\\n",
    "--genomeDir ../2_genome_exploration/genome_for_homework/star_for_homework \\\n",
    "--readFilesManifest manifest_HW.tsv \\\n",
    "--readFilesCommand zcat \\\n",
    "--outSAMtype BAM Unsorted \\\n",
    "--outSAMunmapped Within \\\n",
    "--outFileNamePrefix alignment/star_for_homework/first_pass_ \\\n",
    "--sjdbOverhang 50"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use STAR, all of your FASTQs, and `SJ.out.tab` for the second-pass alignment. You will likely need to make your own `manifest.tsv`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tSTAR --runThreadN 32 --runMode alignReads --genomeDir ../2_genome_exploration/genome_for_homework/star_for_homework --readFilesManifest manifest_HW.tsv --readFilesCommand zcat --outSAMtype BAM Unsorted --outSAMunmapped Within --outSAMattributes NH HI AS nM RG --outFileNamePrefix alignment/star_for_homework/second_pass_ --sjdbFileChrStartEnd alignment/star_for_homework/first_pass_SJ.out.tab\n",
      "\tSTAR version: 2.7.10b   compiled: 2022-11-01T09:53:26-04:00 :/home/dobin/data/STAR/STARcode/STAR.master/source\n",
      "Oct 21 23:06:56 ..... started STAR run\n",
      "Oct 21 23:06:57 ..... loading genome\n",
      "Oct 21 23:09:40 ..... inserting junctions into the genome indices\n",
      "Oct 21 23:10:59 ..... started mapping\n",
      "Oct 22 00:09:09 ..... finished mapping\n",
      "Oct 22 00:09:20 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "!STAR \\\n",
    "--runThreadN 32 \\\n",
    "--runMode alignReads \\\n",
    "--genomeDir ../2_genome_exploration/genome_for_homework/star_for_homework \\\n",
    "--readFilesManifest manifest_HW.tsv \\\n",
    "--readFilesCommand zcat \\\n",
    "--outSAMtype BAM Unsorted \\\n",
    "--outSAMunmapped Within \\\n",
    "--outSAMattributes NH HI AS nM RG \\\n",
    "--outFileNamePrefix alignment/star_for_homework/second_pass_ \\\n",
    "--sjdbFileChrStartEnd alignment/star_for_homework/first_pass_SJ.out.tab\n",
    "\n",
    "!mv alignment/star/Log.final.out alignment/star/second-pass.final.out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Update this block to briefly describe the differences between the logs for the first and second alignments."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HISAT alignment\n",
    "\n",
    "(***Note***: you may need Dr. Wheeler to build the HISAT index of your genome. Ask him for assistance when you get to this part of the assignment.)\n",
    "\n",
    "Use a Bash `while` loop to align all of your FASTQs and save the output files using the same names as the input FASTQ:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916027.fastq.gz with sample Ventral_DG_granule_cells_A\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 34963364\n",
      "\t\tAligned 0 time: 4726720 (13.52%)\n",
      "\t\tAligned 1 time: 28779113 (82.31%)\n",
      "\t\tAligned >1 times: 1457531 (4.17%)\n",
      "\tOverall alignment rate: 86.48%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916028.fastq.gz with sample Ventral_DG_granule_cells_B\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 39970850\n",
      "\t\tAligned 0 time: 5262219 (13.17%)\n",
      "\t\tAligned 1 time: 33023830 (82.62%)\n",
      "\t\tAligned >1 times: 1684801 (4.22%)\n",
      "\tOverall alignment rate: 86.83%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916029.fastq.gz with sample Ventral_DG_granule_cells_C\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 34611761\n",
      "\t\tAligned 0 time: 4735015 (13.68%)\n",
      "\t\tAligned 1 time: 28238306 (81.59%)\n",
      "\t\tAligned >1 times: 1638440 (4.73%)\n",
      "\tOverall alignment rate: 86.32%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916030.fastq.gz with sample Dorsal_DG_granule_cells_A\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 20164564\n",
      "\t\tAligned 0 time: 4064155 (20.15%)\n",
      "\t\tAligned 1 time: 15201491 (75.39%)\n",
      "\t\tAligned >1 times: 898918 (4.46%)\n",
      "\tOverall alignment rate: 79.85%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916031.fastq.gz with sample Dorsal_DG_granule_cells_B\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 19267237\n",
      "\t\tAligned 0 time: 3994549 (20.73%)\n",
      "\t\tAligned 1 time: 14422779 (74.86%)\n",
      "\t\tAligned >1 times: 849909 (4.41%)\n",
      "\tOverall alignment rate: 79.27%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916032.fastq.gz with sample Dorsal_DG_granule_cells_C\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 19665120\n",
      "\t\tAligned 0 time: 4174634 (21.23%)\n",
      "\t\tAligned 1 time: 14633342 (74.41%)\n",
      "\t\tAligned >1 times: 857144 (4.36%)\n",
      "\tOverall alignment rate: 78.77%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916033.fastq.gz with sample Dorsal_DG_mossy_cells_A\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 48263112\n",
      "\t\tAligned 0 time: 9759994 (20.22%)\n",
      "\t\tAligned 1 time: 36094713 (74.79%)\n",
      "\t\tAligned >1 times: 2408405 (4.99%)\n",
      "\tOverall alignment rate: 79.78%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916034.fastq.gz with sample Dorsal_DG_mossy_cells_B\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 45959103\n",
      "\t\tAligned 0 time: 9314190 (20.27%)\n",
      "\t\tAligned 1 time: 34397051 (74.84%)\n",
      "\t\tAligned >1 times: 2247862 (4.89%)\n",
      "\tOverall alignment rate: 79.73%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916035.fastq.gz with sample Dorsal_DG_mossy_cells_C\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 40407931\n",
      "\t\tAligned 0 time: 9135616 (22.61%)\n",
      "\t\tAligned 1 time: 29591145 (73.23%)\n",
      "\t\tAligned >1 times: 1681170 (4.16%)\n",
      "\tOverall alignment rate: 77.39%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916036.fastq.gz with sample Dorsal_CA3_pyramidal_cells_A\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 39504550\n",
      "\t\tAligned 0 time: 5541549 (14.03%)\n",
      "\t\tAligned 1 time: 31645452 (80.11%)\n",
      "\t\tAligned >1 times: 2317549 (5.87%)\n",
      "\tOverall alignment rate: 85.97%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916037.fastq.gz with sample Dorsal_CA3_pyramidal_cells_B\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 43490322\n",
      "\t\tAligned 0 time: 5060149 (11.64%)\n",
      "\t\tAligned 1 time: 36451497 (83.82%)\n",
      "\t\tAligned >1 times: 1978676 (4.55%)\n",
      "\tOverall alignment rate: 88.36%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916038.fastq.gz with sample Dorsal_CA3_pyramidal_cells_C\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 42915276\n",
      "\t\tAligned 0 time: 5577581 (13.00%)\n",
      "\t\tAligned 1 time: 35273012 (82.19%)\n",
      "\t\tAligned >1 times: 2064683 (4.81%)\n",
      "\tOverall alignment rate: 87.00%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916039.fastq.gz with sample Ventral_CA3_pyramidal_cells_A\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 46276381\n",
      "\t\tAligned 0 time: 6005751 (12.98%)\n",
      "\t\tAligned 1 time: 37502855 (81.04%)\n",
      "\t\tAligned >1 times: 2767775 (5.98%)\n",
      "\tOverall alignment rate: 87.02%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916040.fastq.gz with sample Ventral_CA3_pyramidal_cells_B\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 47005885\n",
      "\t\tAligned 0 time: 6319590 (13.44%)\n",
      "\t\tAligned 1 time: 37823425 (80.47%)\n",
      "\t\tAligned >1 times: 2862870 (6.09%)\n",
      "\tOverall alignment rate: 86.56%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916041.fastq.gz with sample Ventral_CA3_pyramidal_cells_C\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 54489435\n",
      "\t\tAligned 0 time: 6680741 (12.26%)\n",
      "\t\tAligned 1 time: 44708022 (82.05%)\n",
      "\t\tAligned >1 times: 3100672 (5.69%)\n",
      "\tOverall alignment rate: 87.74%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916042.fastq.gz with sample Dorsal_CA2_pyramidal_cells_A\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 49070546\n",
      "\t\tAligned 0 time: 6050301 (12.33%)\n",
      "\t\tAligned 1 time: 40566305 (82.67%)\n",
      "\t\tAligned >1 times: 2453940 (5.00%)\n",
      "\tOverall alignment rate: 87.67%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916043.fastq.gz with sample Dorsal_CA2_pyramidal_cells_B\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 47934303\n",
      "\t\tAligned 0 time: 5882551 (12.27%)\n",
      "\t\tAligned 1 time: 39678900 (82.78%)\n",
      "\t\tAligned >1 times: 2372852 (4.95%)\n",
      "\tOverall alignment rate: 87.73%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916044.fastq.gz with sample Dorsal_CA2_pyramidal_cells_C\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 47142321\n",
      "\t\tAligned 0 time: 5675067 (12.04%)\n",
      "\t\tAligned 1 time: 39209373 (83.17%)\n",
      "\t\tAligned >1 times: 2257881 (4.79%)\n",
      "\tOverall alignment rate: 87.96%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916045.fastq.gz with sample Dorsal_CA1_pyramidal_cells_A\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 49892005\n",
      "\t\tAligned 0 time: 6409244 (12.85%)\n",
      "\t\tAligned 1 time: 40644983 (81.47%)\n",
      "\t\tAligned >1 times: 2837778 (5.69%)\n",
      "\tOverall alignment rate: 87.15%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916046.fastq.gz with sample Dorsal_CA1_pyramidal_cells_B\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 56428424\n",
      "\t\tAligned 0 time: 7036195 (12.47%)\n",
      "\t\tAligned 1 time: 46124608 (81.74%)\n",
      "\t\tAligned >1 times: 3267621 (5.79%)\n",
      "\tOverall alignment rate: 87.53%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916047.fastq.gz with sample Dorsal_CA1_pyramidal_cells_C\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 45461044\n",
      "\t\tAligned 0 time: 5794565 (12.75%)\n",
      "\t\tAligned 1 time: 37277630 (82.00%)\n",
      "\t\tAligned >1 times: 2388849 (5.25%)\n",
      "\tOverall alignment rate: 87.25%\n",
      "Aligning /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/SRR2916048.fastq.gz with sample Ventral_CA1_pyramidal_cells_A\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 16591196\n",
      "\t\tAligned 0 time: 3117253 (18.79%)\n",
      "\t\tAligned 1 time: 12749309 (76.84%)\n",
      "\t\tAligned >1 times: 724634 (4.37%)\n",
      "\tOverall alignment rate: 81.21%\n"
     ]
    }
   ],
   "source": [
    "!mkdir -p hisat\n",
    "\n",
    "!while IFS=$'\\t' read -r fq dash rg; do \\\n",
    "    rg=$(echo \"$rg\" | sed 's/^[[:space:]]*//; s/[[:space:]]*$//'); \\\n",
    "    bn=$(basename \"$fq\" .fastq.gz); \\\n",
    "    echo \"Aligning $fq with sample $rg\"; \\\n",
    "    hisat2 -x /data/users/zhouz6436/BIOL343/6_alignment/hisat/mouse_genome \\\n",
    "           -p 16 \\\n",
    "           -U \"$fq\" \\\n",
    "           --rg-id \"$rg\" --rg \"SM:$rg\" \\\n",
    "           --summary-file hisat/${bn}.log \\\n",
    "           --new-summary \\\n",
    "           -S hisat/${bn}.sam; \\\n",
    "done < manifest_HW.tsv\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Snakemake Pipeline\n",
    "\n",
    "In contrast to last week (FASTQ files), there aren't as many decisions that need to be made during alignment, so using a notebook to control and report code isn't as necessary. As you saw in in [6_alignment](6_alignment.ipynb), file and directory management can get complicated during alignment. This is a case where Snakemake scripting shines.\n",
    "\n",
    "1. Copy your `Snakefile` and `config.yaml` from `5_fastq` to `6_alignment`. \n",
    "2. In the `Snakefile`, add the following rules:  \n",
    "    a. `fetch_genome` - using a variation from the code from `2_genome_exploration`, download your reference genome and annotations.  \n",
    "    b. `star_sa` - use `STAR` to produce the suffix array of the reference genome.  \n",
    "    c. `star_align_first` - use `STAR` to perform first-pass alignment with all FASTQ files.  \n",
    "    d. `star_align_second` - use `STAR` to perform second-pass alignment with all FASTQ files.  \n",
    "    e. `hisat_align` - use `HISAT` to align all FASTQ files.  \n",
    "    f. `report` - use `multiqc` to aggregate all QC data into a single report.\n",
    "\n",
    "This `Snakefile` should not use any `wrappers` because we already have the relevant software installed in the `biol343` conda environment. The pipeline should run to completion when any instructor or classmate runs `snakemake --use-conda`."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biol343",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
