{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 1 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 9/19 by 11:59 pm***\n",
    "\n",
    "## Paper selection\n",
    "\n",
    "Link to paper:https://elifesciences.org/articles/14997#s4\n",
    "\n",
    "*Major research question:* The differential gene expressions in different types of excitatory neurons in the hippocampus.\n",
    "\n",
    "*Short synopsis of paper (50-75 words):* This paper is about using RNA-seq to study the differential gene expression in major excitatory neorons in the hippocampus. The author compared the dorsal and ventral regions, found a difference in cell type and region-specific. They found new mark genes and created Hipposeq which is a public database.This paper helps us understanding how the gene expression affects the function, memory and disease in hippocampus.  \n",
    "\n",
    "## RNA-seq\n",
    "\n",
    "*Description of the use of RNA-seq:* They extracted RNA from the labeled granule, mossy, and pyramidal cells from mice. They used RNA-seq to get the full information of gene expression accross the whole genome.\n",
    "\n",
    "### Experimental details\n",
    "\n",
    "*Species:* mice \n",
    "\n",
    "*Tissue source:* hippocampus of mice\n",
    "\n",
    "*Variables (i.e., treatments, multiples tissue sources, time course, etc.):* Cell type(granule, mossy, and pyramidal cell), region(dorsal and ventral)\n",
    "\n",
    "*Method of RNA extraction:* Using PicoPure RNA Isolation kit and with an on-column RNase-free DNase I treatment to remove DNA contamination.\n",
    "\n",
    "*Method of library prep:* ??? 4 barcoded libraries. Sequenceing library.\n",
    "\n",
    "*Type of sequencer used:* Illumina HiSeq 2500\n",
    "\n",
    "*Number and description of replicates:* 3\n",
    "\n",
    "### Computational details\n",
    "\n",
    "*Link to reference genome:* https://www.ensembl.org/Mus_musculus/Info/Index?redirect=no\n",
    "\n",
    "*Link to genome annotations (GTF):* Ensembl NCBIM37 transcript annotation file. I can't find the file that used in this paper because it's an old version. The link of the current version of GTF:   https://www.ensembl.org/Mus_musculus/Info/Index?redirect=no \n",
    "\n",
    "*Link to RNA-seq reads (BioProject or SRA):* http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?token=adsveykeprejbej&acc=GSE74985\n",
    "\n",
    "## Results\n",
    "\n",
    "*Include figures that report RNA-seq results.*\n",
    "\n",
    "To include figures in Markdown, copy the image (JPG, PNG, etc.) to `1_paper_selection/assets/` and include the Markdown code: `![Image alt text](assets/{filename})`. The image will be rendered when the Markdown document is processed. If you want to adjust the size of the image, you have to use HTML tags. Examples can be found by lookinga the Markdown syntax in `1_paper.selection.ipynb`, i.e., `<img src=\"assets/journal.ppat.1012268.g001.PNG\" alt=\"experiment data\" width=\"600\">`\n",
    "\n",
    "- *Figure 1* Figure 1\n",
    "\n",
    "- *Figure 2* Figure 2\n",
    "\n",
    "- *Etc.* \n",
    "\n",
    "*Short description of each figure, including the results and conclusions derived from the figure:*\n",
    "\n",
    "- *Figure 1* a) Hippocampus divided into dorsal and ventral regions. The same cell classes(CA1, CA2, CA3 pyramidal cells, dentate gyrus, granule cells, and mossy cells) were sampled across regions.\n",
    "b)workflow of this experiment. Specific RNA-seq data were generated for multiple hippocampal populations.\n",
    "c)All processed data were saved into Hipposeq, the RNA-seq database. It created a public database to explore hippocampal gene expression.\n",
    "\n",
    "- *Figure 2* RNA-seq can separate hippocampal neurons by class and region. The marker genes found in the study were checked using histology, which showed that different types of hippocampal neurons have differential gene expression in the dorsal and ventral regions.\n",
    "- *Etc.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
