{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 1 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 9/20 by 11:59 pm***\n",
    "\n",
    "## Paper selection\n",
    "\n",
    "Link to paper: \n",
    " https://journals.biologists.com/bio/article/10/6/bio058222/268993/RNA-Seq-reveals-changes-in-human-placental\n",
    "\n",
    "*Major research question:* \n",
    "Examine the effects of changes in womb from the first trimester to the second trimester with the levels of transcriptome and methylome.\n",
    "\n",
    "*Short synopsis of paper (50-75 words):* \n",
    "This paper aims to discuss the results and methods of evaluating the impact of changes in uterine enviornment on trophoblast development and function using RNA-sequencing and DNA-methylation.\n",
    "\n",
    "\n",
    "## RNA-seq\n",
    "\n",
    "*Description of the use of RNA-seq:* \n",
    "RNA was extracted from placental villi using RNeasy PLus Universal Mini Kit. equimolar pools were sequenced using pired end 100 base reads (PE100) using Illumina NextSeq. Quality control of sequencing was assessed using FastQC, fasq_screen and Picard Tools and summarized with MultiQC, extra QC was done with a custom rRNA and mtRNA counts script from GitHub. The reads were trimmed with TrimGalore! and aligned using GRCh38 (Human Genome) and STAR aligner. The genes were quantified using HTSeq-Counts. The genes counts were used for a differential gene analysis using the DESeq2 package in R.\n",
    "\n",
    "\n",
    ">\"Paired-end sequencing was performed on Illumina NextSeq Direct High\n",
    "Output with read lengths of 100 bp. QC of sequencing was assessed using\n",
    "FastQC (v 0.11.5), fastq_screen (v0.9.3) and Picard Tools (v 2.9.0) and\n",
    "summarised with MultiQC (v1.8dev). Reads were trimmed to remove\n",
    "adapters and low quality bases with TrimGalore! (v0.6.4) and aligned to the\n",
    "human genome (GRCh38) with STAR aligner (v2.5.1b), with a mean of\n",
    "90.4% reads uniquely mapping and mean of 56 M paired reads/sample.\n",
    "Gene quantification was determined with HTSeq-Counts (v0.6.1p1).\n",
    "Additional quality control was performed with a custom rRNA and\n",
    "mtRNA counts script ( provided on GitHub). Counts extracted with htseqcounts were used to perform differential gene analysis in R (version 3.5.2)\n",
    "using package DESeq2 (v.1.22.2).\" \n",
    "\n",
    "### Experimental details\n",
    "\n",
    "*Species:* \n",
    "Human (GRCh38)\n",
    "\n",
    "*Tissue source:* Placental villi first trimester and second trimester, obtained during pregnancy termination. \n",
    "\n",
    "*Variables (i.e., treatments, multiples tissue sources, time course, etc.):*  \n",
    "All samples were snap-frozen immediately in liquid nitrogen and stored at -80Â°C. Sex was assigned using sex-specific gene expression Xist, Rps4yl, Ddx3y, Usp9y and Sry. Time points first and second trimester. Also potential pathological conditions are a variable.\n",
    "\n",
    "*Method of RNA extraction:*  RNEasy Plus Universal Mini Kit used to extract the RNA from the placental villi\n",
    "\n",
    "*Method of library prep:* Used Illumina TruSeq Stranded mRNA Library Kit\n",
    "\n",
    "*Type of sequencer used:* Illumina NextSeq for high-throughput sequencing.\n",
    "\n",
    "*Number and description of replicates:* Of the 3242 genes none were ovelapping with the list (101 genes) \n",
    "8 samples from individuals in the first trimester\n",
    "6 samples from individuals in the second trimester\n",
    "\n",
    "### Computational details\n",
    "\n",
    "PCA was performed using rlog-transformed count data for all ages. DO and Kegg pathway analysis used clusterProfiler package on DEG with absolute log2 fold change \\>1 and adjusted P-value\\<0.05. The data matrix for scRNA-seq data were obtained. Regularized log transformation function was applied to counts for heatmaps, whiych rere generated with ComplexHeadmap R package. Selection of DEG on heatmaps based on highest significance lowest adjusted P-value and highest absolute log2 fold changes. DEGs with enriched motifs were analyzed with enrichR R package. GO terms was performed with rrvigo R package.\n",
    "\n",
    "\n",
    "*Link to reference genome:* https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.26_GRCh38/GCF_000001405.26_GRCh38_genomic.fna.gz\n",
    "\n",
    "*Link to genome annotations (GTF):* https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.26_GRCh38/GCF_000001405.26_GRCh38_genomic.gff.gz\n",
    "\n",
    "*Link to RNA-seq reads (BioProject or SRA):* https://www.ncbi.nlm.nih.gov/sra?linkname=bioproject_sra_all&from_uid=734550 \n",
    "\n",
    "## Results\n",
    "\n",
    "*Include figures that report RNA-seq results.*\n",
    "\n",
    "To include figures in Markdown, copy the image (JPG, PNG, etc.) to `1_paper_selection/assets/` and include the Markdown code: `![Image alt text](assets/{filename})`. The image will be rendered when the Markdown document is processed. If you want to adjust the size of the image, you have to use HTML tags. Examples can be found by lookinga the Markdown syntax in `1_paper.selection.ipynb`, i.e., `<img src=\"assets/journal.ppat.1012268.g001.PNG\" alt=\"experiment data\" width=\"600\">`\n",
    "\n",
    "- *Figure 1*\n",
    "\n",
    "<img src=\"https://cob.silverchair-cdn.com/cob/content_public/journal/bio/10/6/10.1242_bio.058222/2/m_bio05822201.png?Expires=1728624646&Signature=LhLRuhioFGMIr2Vy2WKcjjY2GGihxt1vcqPmssYBUPQ6KTdarchQKWxCCdTmwIyg0ux-OLpSrjoydMYZ7-C1odSNm0ajGzUjfxDHTxUEP7NBwXWLeQQH1T9A6nLItn9~sHcnVJ0N6q2ZPlTCv~rmyCI~LkaGMbT9FYqnHkiUsw514UXWh1h7g4Jfokn96S5G1sEvjqDUdz~ESaIVyntmDvOiKpkJ-Pa3H1DQ3OdeMafYEYQmFzUA-e0oQ3TWxp~W~5s9iYebm9dIvBfBuLMRP6HKf~k~c0OOEcsnAmeRJG0qjtLrkTWkkneO-IJxR2SV5ONkB4VMoy4o~57Wa53GwQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA\">\n",
    "\n",
    "- *Figure 2*\n",
    "\n",
    "<img src=\"https://cob.silverchair-cdn.com/cob/content_public/journal/bio/10/6/10.1242_bio.058222/2/m_bio05822202.png?Expires=1728624647&Signature=Y6i~vihjQIC3b3st-UHJTHv3BTnp3UqHnI5g~1KybYr0T1BIRouPXe-j8G5l8BX0glXM5sCPbQD9JbA6M-ARklaXdab99l4eu~Fl~XqdRQTxXZIUrqj47FX~EavEaNNOOgzLHjW7p-AVtZpMhpc3MNi8XmGUI6NY4vf--ULbUo~z3k9tnXRcW~0LkAsf9lZuSGpgFmoHD9hK2yR0FVjmVYRWqpt9xy-XMFdykuk6fU2CNxbkUQ39ayhZoPX7DzV0itDBrO-Ukw-jJOmqIAHQK3OPkhCMo87PVCFfNewKyvzBwbv4JvWyrx~80pG~wxmCMYrfp3MRPKp6ebiTcvSAfQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA\">\n",
    "\n",
    "- *Figure 3* \n",
    "<img src=\"https://cob.silverchair-cdn.com/cob/content_public/journal/bio/10/6/10.1242_bio.058222/2/m_bio05822203.png?Expires=1728624647&Signature=CB3YVkZ2UVTRUq10yBfoHmmdUohkbFSiFC7JzREBqQxNOutjxKGYi5UbcLlTpdcXu0ZKnWjMXVGuhCR8keyfFSO6jyRZ64KU9vo~xoKmwa~9PMYKo4o4bNN4YEBy6V581HD-CMPsBDUdC3s7oWxNIyuSURD94IulaoIZyrusUoy-lF0zKPcH4zEddUHlnbCY1E4gXS~CnTEgsjkNQvYPJCcvRYKxL~NP51VVH3C909fIsRxRQrFYZ5KB2v4XXnYJ6Dqj3BcHgTgSOg3a4wpme1RtD2T~CcAxhHYRybKjH4sxrDDLgI7HMmXbnV2RAn8Tz9GDBjJIv0oD1Hrfgqdw~w__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA\"> \n",
    "\n",
    "- *Figure 4* \n",
    "<img src=\"https://cob.silverchair-cdn.com/cob/content_public/journal/bio/10/6/10.1242_bio.058222/2/m_bio05822204.png?Expires=1728624647&Signature=DbvWfGNrF5yBSM1Ks992J~zoMoJKQPu9AucUzpTUojfrAaK7yX7XjJpWwb2P6drg~DdRl9TWVWKgUogTY6aBQqdxM8ssmiAeKJHMu209sYIY9jgKcL68~St13wSZuay4T8mrXQlRxr565QXKi1pwf-MxVv-uplRcsA-QfWjXe~13DY~-~5WqcMy8oUg5R05eR59TmoHqrqyfaDOSMS9DdZB2s9iUpmwewnXiIaKabNJ7pOMFohwcjmF1btNBotIpT80zMk00pe5aNNgBmpyyKxhf0c6zmcYdaRFDI13N-PIGHjd7gAYrbxqIotHq5GCFrp~jA9g7Jz6Z9UMAqu3Jrg__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA\">\n",
    "\n",
    "- *Figure 5* \n",
    "<img src=\"https://cob.silverchair-cdn.com/cob/content_public/journal/bio/10/6/10.1242_bio.058222/2/m_bio05822205.png?Expires=1728624647&Signature=alzaGQ2F-oTN1b2CENyBOAv~YM03N9W3oCTCPXmngIXDZm1MVoK8-p00pT5BqBk8WTWWPt1qZMEaz5AztFr-twkLi-LR5-PjleULxCTdR1J9Nw7XvS5ZK5vfOYBn3IeYudB6dEQ7hc7t7Oq7NBP5-TdRj1w6eHhbfPWEmgyODvXJsH9U4lE4DjrT8wHR2MhE7cvD5jeYoBZ-booIW-46o1K4bUYYOd1HkEKaLqmV6DORkfR0HHCTxv8tV1~MqyVmnU-m50yrugihlVC10f~K3G5Svgvf4vj9ELZarF047t32Nhr9u8rIm-Ehq~xsX1EfC-ffg6V4tnB6Uj1Wm9Wqig__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA\">\n",
    "\n",
    "- *Figure 6* \n",
    "<img src=\"https://cob.silverchair-cdn.com/cob/content_public/journal/bio/10/6/10.1242_bio.058222/2/m_bio05822206.png?Expires=1728624647&Signature=dlY8MzjDazZZ47aJ8GdeTqBrBW0u343KRa6Sl9U4K8HZejlpsdmk3vUEqlfYlViaeNVeX8Rt-5ujbn1EvI4v2~MrTEd9oU4OdqEEwvgjQpucce-CBtEA5ZJv3NGIIrtunRnHgKer1httE~8jHH-NoJvYepLpTsJ3h9PmKRHVhz0Pea-LEgShaDnHoNYDETeEGEv~73Z-oRkDYAiVhpKCWiB~eY-2kvO2WXvQqW4wAVeYpBUDbZ6UH7hW8Ufrdeoy2QrXyIaxGGKHFRau6T-d03zxc52HNRj6002Qgm7f8WewAzAL3M0K6E4CcuYUPhIY1XVc~FwHqjDethJakmMa-A__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA\">\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "*Short description of each figure, including the results and conclusions derived from the figure:*\n",
    "\n",
    "- *Figure 1*\n",
    "This figure contains the PCA separation between the expression of the gene during the first and second trimester samples. The volcano plot shows which genes were expressed more in the first trimester in red and more in the second trimester in blue. The barplots show the GO terms in relation to their bio processes, molecular functions, and cellular components. \n",
    "\n",
    "- *Figure 2*\n",
    "Heatmap A shows which genes are involved in glycolytic pathway comparing top differentiatally expressed genes from first trimester to the top differentially expressed genes in the second trimester and they were also compared to a previously published scRNA-seq data from first and second trimesters. The B figure represents immunostaining of first and second trimester sections with anti-pyruvate kinase antibody. D and E are also stains. C is a heatmap of the DEGs in the first trimester vs those in the second, specivfically ones involved in ER stress and processes and then those were compared to the previous studies dataset as well. G is a heatmap of DEGs related to oxidative stress.\n",
    "\n",
    "- *Figure 3*\n",
    "A is a heatmap comparing DEGs related to hormones from this study and scRNA-seq data. B western blots depicting transthyretin and leptim lecels in first and second trimester placental lysates. Ponceau S staining was used to normalize protein loading. C is immunostining with anti spexin antibody. D shows a volcano plot of transport-related genes with the same color scheme as in figure 1. E is a heatmap of top trasport related genes split by transport type. \n",
    "\n",
    "- *Figure 4*\n",
    "A is a heatmap of DEGs associated with WNT signalling. B is a heatmap of extracellular matrix and angiogenesis-related DEGs. E is an immuolocalization of meprin. D is an immunolocalization of CD31 and E is an immunolocalization of musculin. F is a heatmap of diffenetially experssed TFs. G is CBX3 motif is enriched in proportion of DEGs.\n",
    "\n",
    "- *Figure 5*\n",
    "A is a comparison of gestational ages between this study and other studies. B is a PCA that shows the separation between a gene that is expressed differently in the first trimester compared to the second trimester. C is Global methylation levels shown to be higher in trimester two since the shift in points is always toward the top left for the second trimester and for the first trimester they localize or shift toward the bottom right. D compares the RNA expression and DNA methylation at promoters and gene bodies, showing a correlation of reduced expression with increased methylation from first to second trimester. \n",
    "\n",
    "- *Figure 6*\n",
    "Summarizes the significant GO terms of the differentially expressed genes between the first and second trimesters. Clusters of terms indicate stronger relation. Larger clusters show relations for gene expression of transmembrane transporters, cell differentiation and development and leukocyte and adaptive immune response. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
