{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 1 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 9/20 by 11:59 pm***\n",
    "\n",
    "## Paper selection\n",
    "\n",
    "Link to paper: https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1007559#abstract0\n",
    "\n",
    "*Major research question:* Perform cell-specific transcriptional analysis of four major somatic tissues isolated from adult worms to understand the factors affecting adult animal gene expression and \"enriched\" genes that identify parallels between human and worm tissue function-specificity.\n",
    "\n",
    "*Short synopsis of paper (50-75 words):* The authors wanted to understand tissue-specific gene expression and how it relates to molecular mechanisms such as signaling-related genes involved in neural hormone networks, muscular tissue contraction (among other mechanisms), and further how the adult worms are distinguishable from larval tissue in their genes. They drew a comparison between human tissue and the worms also - noting transcriptional profile similarities, such as in the worm intestine compared to the human's large and small intestine.\n",
    "\n",
    "## RNA-seq\n",
    "\n",
    "*Description of the use of RNA-seq:* They used RNA-seq to determine highly expressed (\"enriched\") genes within the four types of sample tissues\n",
    " \n",
    "### Experimental details\n",
    "\n",
    "*Species:* Caenorhabditis elegans\n",
    "\n",
    "*Tissue source:* Neuron, intestine, hypodermis, and muscle cells from adult worms\n",
    "\n",
    "*Variables (i.e., treatments, multiples tissue sources, time course, etc.):* The neuron, hypodermis, and intestine tissues were fluorescently-marked with GFP. Above in tissue source (they used the four different tissues for analysis)\n",
    "\n",
    "*Method of RNA extraction:* RNA was extracted using standard Trizol/ chloroform/ isopropanol method, DNase digested, and cleaned using Qiagen RNEasy Minelute columnns\n",
    "\n",
    "*Method of library prep:* Either Nugen Encore NGS Library System or the Illumina TruSeq DNA Sample Prep (both were used)\n",
    "\n",
    "*Type of sequencer used:* Illumina HiSeq 2000 platform, \"35–200 million reads (average of 107,674,388 reads) were obtained for each sample and mapped to the C. elegans genome.\"\n",
    "\n",
    "*Number and description of replicates:* 27 Day 1 adult tissue samples (7 neuron, 5 intestine, 7 hypodermis, 8 muscle)\n",
    "\n",
    "### Computational details\n",
    "\n",
    "*Link to reference genome:* https://downloads.wormbase.org/species/c_elegans/sequence/genomic/c_elegans.PRJNA275000.current.genomic.fa.gz\n",
    "\n",
    "*Link to genome annotations (GTF):* https://downloads.wormbase.org/species/c_elegans/gff/c_elegans.canonical_bioproject.current.annotations.gff3.gz\n",
    "\n",
    "*Link to RNA-seq reads (BioProject or SRA):* https://www.ncbi.nlm.nih.gov/sra?linkname=bioproject_sra_all&from_uid=400796\n",
    "\n",
    "## Results\n",
    "\n",
    "*Include figures that report RNA-seq results.*\n",
    "\n",
    "To include figures in Markdown, copy the image (JPG, PNG, etc.) to `1_paper_selection/assets/` and include the Markdown code: `![Image alt text](assets/{filename})`. The image will be rendered when the Markdown document is processed. If you want to adjust the size of the image, you have to use HTML tags. Examples can be found by lookinga the Markdown syntax in `1_paper.selection.ipynb`, i.e., `<img src=\"assets/journal.ppat.1012268.g001.PNG\" alt=\"experiment data\" width=\"600\">`\n",
    "\n",
    "- *Figure 1* \n",
    "\n",
    "<img src=\"https://journals.plos.org/plosgenetics/article/figure/image?size=medium&id=10.1371/journal.pgen.1007559.g001\" alt=\"experiment data\" width=\"500\">\n",
    "\n",
    "- *Figure 2*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosgenetics/article/figure/image?size=medium&id=10.1371/journal.pgen.1007559.g002\" width=\"500\">\n",
    "\n",
    "\n",
    "- *Figure 3*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosgenetics/article/figure/image?size=medium&id=10.1371/journal.pgen.1007559.g003\" width=\"500\">\n",
    "\n",
    "- *Figure 4*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosgenetics/article/figure/image?size=medium&id=10.1371/journal.pgen.1007559.g004\" width=\"500\">\n",
    "\n",
    "- *Figure 5*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosgenetics/article/figure/image?size=medium&id=10.1371/journal.pgen.1007559.g005\" width=\"500\">\n",
    "\n",
    "- *Figure 6*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosgenetics/article/figure/image?size=medium&id=10.1371/journal.pgen.1007559.g006\" width=\"500\">\n",
    "\n",
    "- *Figure 7*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosgenetics/article/figure/image?size=medium&id=10.1371/journal.pgen.1007559.g007\" width=\"500\">\n",
    "\n",
    "*Short description of each figure, including the results and conclusions derived from the figure:*\n",
    "\n",
    "- *Figure 1*\n",
    "\n",
    "Title: Sorting, RNA-seq, and analysis of isolated adult C. elegans tissues\n",
    "\n",
    "\n",
    "Description: In A, it shows what the adult worms looked like prior to cell disruption (photo included below adult). In B, shows PCA plot of all samples used in the study. In C, shows diagram of all expressed genes in the tissues + any overlap. In D, it shows a heatmap of the expressivity levels of the genes (white is low and blue is high). In E, it's a gene ontology analysis that shows ubiquitously expressed genes with larger bubbles representing more general GO terms. In F, shows diagram of embryonic, larval, and adult genes and the overlap between them in all 4 tissue samples. In G, it's a comparison of expressed genes with the adult proteomic data in 3 tissues; proteins were \"expressed\" if found in all 3.\n",
    "\n",
    "- *Figure 2*\n",
    "\n",
    "Title: Characterization of tissue-enriched gene sets\n",
    "\n",
    "\n",
    "Description: In A, shows gene ontology analysis of all 4 tissue samples in greater detail than in Fig. 1. In B, shows hypodermis GO analysis of uniquely expressed genes. In C, it's a comparison of the worm tissue to human tissue with red meaning highly significant enrichment and white meaning not significant.\n",
    "\n",
    "- *Figure 3*\n",
    "\n",
    "Title: Alternative splicing analysis\n",
    "\n",
    "\n",
    "Description: A-C shows a DEXseq analysis to identify significant splicing among the tissue samples. In D, shows a comparison of collagen expression in the adult tissue samples. In E, it shows a heatmap of those collagen expression values. In F-H, they show collagen expression levels among tissues and specific splicing patterns/characteristics.\n",
    "\n",
    "- *Figure 4*\n",
    "\n",
    "Title: Tissue specificity of gene expression\n",
    "\n",
    "\n",
    "Description: A shows multiple categories of genes expressed for each tissue type. B shows a heatmap of insulin expression in the adult tissue samples. C shows Class I (upregulated genes) and Class II (downregulated genes) diagrams that \"45–47% of previously identified targets are shared among all tissues\". \n",
    "\n",
    "- *Figure 5*\n",
    "\n",
    "Title: New prediction tool for sub-tissue/cellular gene expression\n",
    "\n",
    "\n",
    "Description: A shows their learning model of attaining training data and using gold standard data for genome-wide predictions on 76 tissues. They built a web server as well for users to utilize these prediction models. In B, it shows the prediction performance of their model. High precision is even found for tissues and types that have less data examples.\n",
    "\n",
    "- *Figure 6*\n",
    "\n",
    "Title: Tissue specificity of Insulin signaling/DAF-16 target expression\n",
    "\n",
    "\n",
    "Description: A (top) shows concordance between predicted tissue expression scores and experimental isolated RNA samples. Each square represents score of average tissue exprected expression. A (bottom) shows the contribution of the experimental samples on the tissue-gene prediction models. In B, it shows gene annotations from their model with predictions for each gene (red is highest; blue is lowest score)\n",
    "\n",
    "- *Figure 7*\n",
    "\n",
    "Title: CREB-dependent long-term memory (LTAM) genes (A) and CREB-dependent naïve (basal) target genes (B) from Day 1 adult worms [27] were analyzed for predicted tissue expression\n",
    "\n",
    "\n",
    "Description: Heatmaps that are plots from the NMF R package"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
