{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 1 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 9/20 by 11:59 pm***\n",
    "\n",
    "## Paper selection\n",
    "\n",
    "Link to paper: https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0299261#sec002 \n",
    "\n",
    "*Major research question:* Are certain genes in lemons involved in leaf abscission?\n",
    "\n",
    "*Short synopsis of paper (50-75 words):* The authors looked at the RNA of different lemon varieties, each at three different stages of drought stress. They also were looking at the hormone signal transduction pathway as well as the main metabolic pathways that are involved in leaf abscission. This research was done in order to improve screening of lemon defoliation genes for selection and breeding of high quality and highly resistant lemons.\n",
    "\n",
    "## RNA-seq\n",
    "\n",
    "*Description of the use of RNA-seq:* The authors looked at which genes where highly expressed during the initiation of abscission.\n",
    "\n",
    "### Experimental details\n",
    "\n",
    "*Species:* Allen Eureka and Yunning No. 1 types of lemons.\n",
    "\n",
    "*Tissue source:* Middle and lower mature petioles were collected from the abscission zone.\n",
    "\n",
    "*Variables (i.e., treatments, multiples tissue sources, time course, etc.):* soil water content/drought stress, 2 different varieties of lemon\n",
    "\n",
    "*Method of RNA extraction:* The RNAprep Pure Plant Kit was used for RNA extraction.\n",
    "\n",
    "*Method of library prep:* The mRNA was broken into short fragments with oligo (dT) magnetic beads.\n",
    "\n",
    "*Type of sequencer used:* The sequencer used was Illumina HiSeq.\n",
    "\n",
    "*Number and description of replicates:* There were three biological replicates of the two lemon varieties extracted at three different defolation stages.\n",
    "\n",
    "### Computational details\n",
    "\n",
    "*Link to reference genome:* https://www.citrusgenomedb.org/citrus_downloads/Citrus_limon/C.limon_EMF-UC_v1-Alternative_genome/assembly/Climon_v1_alternative-scaffolds.fa \n",
    "\n",
    "*Link to genome annotations (GTF):* https://github.com/UWEC-Bioinformatics/BIOL343/blob/Brandy/2_genome_exploration/lemons/lemon_annotations.gtf \n",
    "\n",
    "*Link to RNA-seq reads (BioProject or SRA):* https://www.ncbi.nlm.nih.gov/bioproject/PRJNA1033435  \n",
    "\n",
    "## Results\n",
    "\n",
    "*Include figures that report RNA-seq results.*\n",
    "\n",
    "To include figures in Markdown, copy the image (JPG, PNG, etc.) to `1_paper_selection/assets/` and include the Markdown code: `![Image alt text](assets/{filename})`. The image will be rendered when the Markdown document is processed. If you want to adjust the size of the image, you have to use HTML tags. Examples can be found by lookinga the Markdown syntax in `1_paper.selection.ipynb`, i.e., `<img src=\"assets/journal.ppat.1012268.g001.PNG\" alt=\"experiment data\" width=\"600\">`\n",
    "\n",
    "- *Figure 1*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosone/article/figure/image?size=large&id=10.1371/journal.pone.0299261.g003\" width=500px>\n",
    "\n",
    "- *Figure 2*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosone/article/figure/image?size=large&id=10.1371/journal.pone.0299261.g004\" width=500px>\n",
    "\n",
    "- *Figure 3*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosone/article/figure/image?size=large&id=10.1371/journal.pone.0299261.g005\" width=500px>\n",
    "\n",
    "- *Figure 4*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosone/article/figure/image?size=large&id=10.1371/journal.pone.0299261.g006\" width=500px>\n",
    "\n",
    "- *Figure 5*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosone/article/figure/image?size=large&id=10.1371/journal.pone.0299261.g007\" width=500px>\n",
    "\n",
    "- *Figure 6*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosone/article/figure/image?size=large&id=10.1371/journal.pone.0299261.g008\" width=500px>\n",
    "\n",
    "- *Figure 7*\n",
    "\n",
    "<img src=\"https://journals.plos.org/plosone/article/figure/image?size=large&id=10.1371/journal.pone.0299261.g012\" width=500px>\n",
    "\n",
    "\n",
    "*Short description of each figure, including the results and conclusions derived from the figure:*\n",
    "\n",
    "- *Figure 1* The PCA clustering of all 18 samples grouped all samples of the same variety and defoliation period together. The results show that there is good reproducabillity of all samples.\n",
    "\n",
    "- *Figure 2* Part A of this figure shows the difference of signifficantly expressed genes in Allen Eureka versus Yunning No. 1 lemons at each defoliation stage. The orange are upregulated genes and the blue are downregulated. Part B is a venn diagram showing the number of same and different signifficantly expressed genes between each defoliation stage. The yellow cirlce is predefoliation, the blue circle is middefoliation, and the purple circle is postdefoliation.\n",
    "\n",
    "- *Figure 3* This figure shows that the gene counts of molecular functions, cellular components, and biological processes were signifficantly different at each defoliation stage.\n",
    "\n",
    "- *Figure 4* The middefoliation stage had the highest number of differentially expressed genes and enriched related metabolic pathways. The postdefoliation stage was high, but not as high as the middefoliation stage. Predefoliation had few enriched metabolic pathways.\n",
    "\n",
    "- *Figure 5* The heat map shows differentially expressed genes that relate to hormone signal transduction pathway for both varietoes of lemon at each of the defoliation stages. Red is highly expressed and blue is lower expression.\n",
    "\n",
    "- *Figure 6* The four different heat maps represent genes in four metabolic pathways: phenylpropanoid biosynthesis, biosynthesis of amino acids, amino sugar and nucleotide sugar metabolism, and flavonoid biosynthesis. Horizontal coordinates are samples and vertical coordinates are differentially expressed genes.\n",
    "\n",
    "- *Figure 7* This figure shows the changes in fluorescence quantitative expression levels of six genes involved in plant hormone synthesis signaling and other processes related to abscission."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
