{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 1 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 9/19 by 11:59 pm***\n",
    "\n",
    "## Paper selection\n",
    "\n",
    "Link to paper: https://elifesciences.org/articles/30860#abstract\n",
    "\n",
    "*Major research question:* how does gene regulation play a role in the embryonic development of vertebraes?\n",
    "\n",
    "*Short synopsis of paper (50-75 words):* Gene regulation the fundamental way embryonic development happens and transcription a major way of how it occurs in the cell. This paper analyzed samples from zebrafish embryos during various stages of development and looked at the RNA expression at different stages to see how it is associated with the development.\n",
    "\n",
    "## RNA-seq\n",
    "\n",
    "*Description of the use of RNA-seq:* Was used to generate a comprehensive polyadenylated RNA expression profile of normal zebrafish development\n",
    "\n",
    "### Experimental details\n",
    "\n",
    "*Species:* Wild-type HLF strain zebrafish (Danio rerio)\n",
    "\n",
    "*Tissue source:* embryos: 18 different developmental stages from 1 cell to 5 days post-fertilization\n",
    "\n",
    "*Variables (i.e., treatments, multiples tissue sources, time course, etc.):* time: stage of embryonic development\n",
    "\n",
    "*Method of RNA extraction:* First mechanically broke the cells using beta mercaptoethanol. Then let RNA bind to beads (Agencourt RNAClean XP) before using a magnet to remove the liquid and then washed with 70% ethanol and air dried. The beads were then suspended in water and removed from the beads using a magnetic track.\n",
    "\n",
    "*Method of library prep:* Strand-specific RNA-seq libraries containing unique index sequences in the adapter were generated simultaneously following the dUTP method\n",
    "\n",
    "*Type of sequencer used:* Illumina HiSeq 2500 in 100 bp paired-end mode.\n",
    "\n",
    "*Number and description of replicates:* each time point had 5 replicates. Each replicate consisted of a pool of 12 embryos\n",
    "\n",
    "### Computational details\n",
    "\n",
    "*Link to reference genome:* https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000002035.5/\n",
    "\n",
    "*Link to genome annotations (GTF):* https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000002035.5/\n",
    "\n",
    "*Link to RNA-seq reads (BioProject or SRA):* Sequence data were deposited in ENA under accession ERP014517.\n",
    "\n",
    "## Results\n",
    "\n",
    "*Include figures that report RNA-seq results.*\n",
    "\n",
    "To include figures in Markdown, copy the image (JPG, PNG, etc.) to `1_paper_selection/assets/` and include the Markdown code: `![Image alt text](assets/{filename})`. The image will be rendered when the Markdown document is processed. If you want to adjust the size of the image, you have to use HTML tags. Examples can be found by lookinga the Markdown syntax in `1_paper.selection.ipynb`, i.e., `<img src=\"assets/journal.ppat.1012268.g001.PNG\" alt=\"experiment data\" width=\"600\">`\n",
    "\n",
    "- *Figure 1* ![Figure1a.png](assets/{filename})\n",
    "\n",
    "- *Figure 2* ![Figure1b.png](assets/{filename})\n",
    "\n",
    "- *Figure 3* ![Figure1d.png](assets/{filename})\n",
    "\n",
    "- *Figure 4* ![Figure3a.png](assets/{filename})\n",
    "\n",
    "- *Figure 5* ![Figure3b.png](assets/{filename})\n",
    "\n",
    "- *Figure 6* ![Figure4a.png](assets/{filename})\n",
    "\n",
    "*Short description of each figure, including the results and conclusions derived from the figure:*\n",
    "\n",
    "- *Figure 1*: Shows the developmental stages used in the study. Contains the ZFS stage ID, stage name, roughly how many hours post-fertilization, and the and the five developmental categories. Also shows that gene expression is very consistent in all replicates over the development cycle and that neighboring stages have the most similar gene expression.\n",
    "\n",
    "- *Figure 2*: Sample correlation matrix of Spearman correlation coefficients of every sample compared to every other sample. Shows that as devleopment goes on, the number of genes detected increases and a greater variety of genes are expressed.\n",
    "\n",
    "- *Figure 3*: 3D plot of first 3 PCs from PCA which explains 83.33% of the variance in data. Shows close clustering of replicates and smooth transition between stages as time goes on. \n",
    "\n",
    "- *Figure 4*: Heatmap of expression profiles with genes ordered based on which stage of development the maximum expression occurs at. The heatmap gives a visual of how at one stage, certain genes are specifically expressed in high amounts while others are expressed in relatively low amounts.\n",
    "\n",
    "- *Figure 5*: Table of the numbers of named and unnamed genes assigned to each stage. A notable increase in proportion of unnamed genes expressed during dome and gastrula stages could indicate that these stages are enriched by expression of genes unique to the species.\n",
    "\n",
    "- *Figure 6*: Heatmap of the expression profiles (log10[TPM]) of the ZnF genes from clusters 7, 23, 26 and 135. These are clusters with more chromosome 4 genes than expected. It shows there is a co-ordinated expression of these genes that happens at the onset of zygotic transcription since they all have high expression during dome to 75% epiboly and this begins at the 128- and 1000 cell stages. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
