{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 5 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 10/18 by 11:59 pm***"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTQ fetch, QC, and trimming/filtering\n",
    "\n",
    "Use the SRA-toolkit to fetch the relevant FASTQ files. You may use a hand-crafted `txt` file to provide the list of SRAs instead of manipulating an SRA Run Table with `cut`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting SRR26560370 from NCBI SRA\n",
      "2024-10-10 15:32:14,173 - SRR ids: ['SRR26560370']\n",
      "2024-10-10 15:32:14,173 - extra args: ['--gzip']\n",
      "2024-10-10 15:32:14,174 - tempdir: /local/scratch/job_98738/pfd_gd68ktt6\n",
      "2024-10-10 15:32:14,174 - CMD: sra-stat --meta --quick SRR26560370\n",
      "2024-10-10 15:32:17,093 - SRR26560370 spots: 23569582\n",
      "2024-10-10 15:32:17,093 - blocks: [[1, 1473098], [1473099, 2946196], [2946197, 4419294], [4419295, 5892392], [5892393, 7365490], [7365491, 8838588], [8838589, 10311686], [10311687, 11784784], [11784785, 13257882], [13257883, 14730980], [14730981, 16204078], [16204079, 17677176], [17677177, 19150274], [19150275, 20623372], [20623373, 22096470], [22096471, 23569582]]\n",
      "2024-10-10 15:32:17,093 - CMD: fastq-dump -N 1 -X 1473098 -O /local/scratch/job_98738/pfd_gd68ktt6/0 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,107 - CMD: fastq-dump -N 1473099 -X 2946196 -O /local/scratch/job_98738/pfd_gd68ktt6/1 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,108 - CMD: fastq-dump -N 2946197 -X 4419294 -O /local/scratch/job_98738/pfd_gd68ktt6/2 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,108 - CMD: fastq-dump -N 4419295 -X 5892392 -O /local/scratch/job_98738/pfd_gd68ktt6/3 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,108 - CMD: fastq-dump -N 5892393 -X 7365490 -O /local/scratch/job_98738/pfd_gd68ktt6/4 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,109 - CMD: fastq-dump -N 7365491 -X 8838588 -O /local/scratch/job_98738/pfd_gd68ktt6/5 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,110 - CMD: fastq-dump -N 8838589 -X 10311686 -O /local/scratch/job_98738/pfd_gd68ktt6/6 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,110 - CMD: fastq-dump -N 10311687 -X 11784784 -O /local/scratch/job_98738/pfd_gd68ktt6/7 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,110 - CMD: fastq-dump -N 11784785 -X 13257882 -O /local/scratch/job_98738/pfd_gd68ktt6/8 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,111 - CMD: fastq-dump -N 13257883 -X 14730980 -O /local/scratch/job_98738/pfd_gd68ktt6/9 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,111 - CMD: fastq-dump -N 14730981 -X 16204078 -O /local/scratch/job_98738/pfd_gd68ktt6/10 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,111 - CMD: fastq-dump -N 16204079 -X 17677176 -O /local/scratch/job_98738/pfd_gd68ktt6/11 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,112 - CMD: fastq-dump -N 17677177 -X 19150274 -O /local/scratch/job_98738/pfd_gd68ktt6/12 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,112 - CMD: fastq-dump -N 19150275 -X 20623372 -O /local/scratch/job_98738/pfd_gd68ktt6/13 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,112 - CMD: fastq-dump -N 20623373 -X 22096470 -O /local/scratch/job_98738/pfd_gd68ktt6/14 --gzip SRR26560370\n",
      "2024-10-10 15:32:17,112 - CMD: fastq-dump -N 22096471 -X 23569582 -O /local/scratch/job_98738/pfd_gd68ktt6/15 --gzip SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473112 spots for SRR26560370\n",
      "Written 1473112 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Read 1473098 spots for SRR26560370\n",
      "Written 1473098 spots for SRR26560370\n",
      "Getting SRR26560371 from NCBI SRA\n",
      "2024-10-10 15:36:21,252 - SRR ids: ['SRR26560371']\n",
      "2024-10-10 15:36:21,252 - extra args: ['--gzip']\n",
      "2024-10-10 15:36:21,252 - tempdir: /local/scratch/job_98738/pfd_u1dap012\n",
      "2024-10-10 15:36:21,252 - CMD: sra-stat --meta --quick SRR26560371\n",
      "2024-10-10 15:36:23,528 - SRR26560371 spots: 23087946\n",
      "2024-10-10 15:36:23,528 - blocks: [[1, 1442996], [1442997, 2885992], [2885993, 4328988], [4328989, 5771984], [5771985, 7214980], [7214981, 8657976], [8657977, 10100972], [10100973, 11543968], [11543969, 12986964], [12986965, 14429960], [14429961, 15872956], [15872957, 17315952], [17315953, 18758948], [18758949, 20201944], [20201945, 21644940], [21644941, 23087946]]\n",
      "2024-10-10 15:36:23,528 - CMD: fastq-dump -N 1 -X 1442996 -O /local/scratch/job_98738/pfd_u1dap012/0 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,529 - CMD: fastq-dump -N 1442997 -X 2885992 -O /local/scratch/job_98738/pfd_u1dap012/1 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,529 - CMD: fastq-dump -N 2885993 -X 4328988 -O /local/scratch/job_98738/pfd_u1dap012/2 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,530 - CMD: fastq-dump -N 4328989 -X 5771984 -O /local/scratch/job_98738/pfd_u1dap012/3 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,532 - CMD: fastq-dump -N 5771985 -X 7214980 -O /local/scratch/job_98738/pfd_u1dap012/4 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,533 - CMD: fastq-dump -N 7214981 -X 8657976 -O /local/scratch/job_98738/pfd_u1dap012/5 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,551 - CMD: fastq-dump -N 8657977 -X 10100972 -O /local/scratch/job_98738/pfd_u1dap012/6 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,554 - CMD: fastq-dump -N 10100973 -X 11543968 -O /local/scratch/job_98738/pfd_u1dap012/7 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,557 - CMD: fastq-dump -N 11543969 -X 12986964 -O /local/scratch/job_98738/pfd_u1dap012/8 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,564 - CMD: fastq-dump -N 12986965 -X 14429960 -O /local/scratch/job_98738/pfd_u1dap012/9 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,565 - CMD: fastq-dump -N 14429961 -X 15872956 -O /local/scratch/job_98738/pfd_u1dap012/10 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,566 - CMD: fastq-dump -N 15872957 -X 17315952 -O /local/scratch/job_98738/pfd_u1dap012/11 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,567 - CMD: fastq-dump -N 17315953 -X 18758948 -O /local/scratch/job_98738/pfd_u1dap012/12 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,568 - CMD: fastq-dump -N 18758949 -X 20201944 -O /local/scratch/job_98738/pfd_u1dap012/13 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,591 - CMD: fastq-dump -N 20201945 -X 21644940 -O /local/scratch/job_98738/pfd_u1dap012/14 --gzip SRR26560371\n",
      "2024-10-10 15:36:23,592 - CMD: fastq-dump -N 21644941 -X 23087946 -O /local/scratch/job_98738/pfd_u1dap012/15 --gzip SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1443006 spots for SRR26560371\n",
      "Written 1443006 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Read 1442996 spots for SRR26560371\n",
      "Written 1442996 spots for SRR26560371\n",
      "Getting SRR26560372 from NCBI SRA\n",
      "2024-10-10 15:40:14,123 - SRR ids: ['SRR26560372']\n",
      "2024-10-10 15:40:14,123 - extra args: ['--gzip']\n",
      "2024-10-10 15:40:14,123 - tempdir: /local/scratch/job_98738/pfd_oud7ft6n\n",
      "2024-10-10 15:40:14,123 - CMD: sra-stat --meta --quick SRR26560372\n",
      "2024-10-10 15:40:16,390 - SRR26560372 spots: 22434198\n",
      "2024-10-10 15:40:16,390 - blocks: [[1, 1402137], [1402138, 2804274], [2804275, 4206411], [4206412, 5608548], [5608549, 7010685], [7010686, 8412822], [8412823, 9814959], [9814960, 11217096], [11217097, 12619233], [12619234, 14021370], [14021371, 15423507], [15423508, 16825644], [16825645, 18227781], [18227782, 19629918], [19629919, 21032055], [21032056, 22434198]]\n",
      "2024-10-10 15:40:16,390 - CMD: fastq-dump -N 1 -X 1402137 -O /local/scratch/job_98738/pfd_oud7ft6n/0 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,391 - CMD: fastq-dump -N 1402138 -X 2804274 -O /local/scratch/job_98738/pfd_oud7ft6n/1 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,391 - CMD: fastq-dump -N 2804275 -X 4206411 -O /local/scratch/job_98738/pfd_oud7ft6n/2 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,392 - CMD: fastq-dump -N 4206412 -X 5608548 -O /local/scratch/job_98738/pfd_oud7ft6n/3 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,394 - CMD: fastq-dump -N 5608549 -X 7010685 -O /local/scratch/job_98738/pfd_oud7ft6n/4 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,395 - CMD: fastq-dump -N 7010686 -X 8412822 -O /local/scratch/job_98738/pfd_oud7ft6n/5 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,414 - CMD: fastq-dump -N 8412823 -X 9814959 -O /local/scratch/job_98738/pfd_oud7ft6n/6 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,415 - CMD: fastq-dump -N 9814960 -X 11217096 -O /local/scratch/job_98738/pfd_oud7ft6n/7 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,426 - CMD: fastq-dump -N 11217097 -X 12619233 -O /local/scratch/job_98738/pfd_oud7ft6n/8 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,427 - CMD: fastq-dump -N 12619234 -X 14021370 -O /local/scratch/job_98738/pfd_oud7ft6n/9 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,427 - CMD: fastq-dump -N 14021371 -X 15423507 -O /local/scratch/job_98738/pfd_oud7ft6n/10 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,428 - CMD: fastq-dump -N 15423508 -X 16825644 -O /local/scratch/job_98738/pfd_oud7ft6n/11 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,430 - CMD: fastq-dump -N 16825645 -X 18227781 -O /local/scratch/job_98738/pfd_oud7ft6n/12 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,452 - CMD: fastq-dump -N 18227782 -X 19629918 -O /local/scratch/job_98738/pfd_oud7ft6n/13 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,453 - CMD: fastq-dump -N 19629919 -X 21032055 -O /local/scratch/job_98738/pfd_oud7ft6n/14 --gzip SRR26560372\n",
      "2024-10-10 15:40:16,455 - CMD: fastq-dump -N 21032056 -X 22434198 -O /local/scratch/job_98738/pfd_oud7ft6n/15 --gzip SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402143 spots for SRR26560372\n",
      "Written 1402143 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Read 1402137 spots for SRR26560372\n",
      "Written 1402137 spots for SRR26560372\n",
      "Getting SRR26560373 from NCBI SRA\n",
      "2024-10-10 15:43:57,396 - SRR ids: ['SRR26560373']\n",
      "2024-10-10 15:43:57,397 - extra args: ['--gzip']\n",
      "2024-10-10 15:43:57,397 - tempdir: /local/scratch/job_98738/pfd_ik1gk7fb\n",
      "2024-10-10 15:43:57,397 - CMD: sra-stat --meta --quick SRR26560373\n",
      "2024-10-10 15:43:59,922 - SRR26560373 spots: 21865790\n",
      "2024-10-10 15:43:59,922 - blocks: [[1, 1366611], [1366612, 2733222], [2733223, 4099833], [4099834, 5466444], [5466445, 6833055], [6833056, 8199666], [8199667, 9566277], [9566278, 10932888], [10932889, 12299499], [12299500, 13666110], [13666111, 15032721], [15032722, 16399332], [16399333, 17765943], [17765944, 19132554], [19132555, 20499165], [20499166, 21865790]]\n",
      "2024-10-10 15:43:59,923 - CMD: fastq-dump -N 1 -X 1366611 -O /local/scratch/job_98738/pfd_ik1gk7fb/0 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,923 - CMD: fastq-dump -N 1366612 -X 2733222 -O /local/scratch/job_98738/pfd_ik1gk7fb/1 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,924 - CMD: fastq-dump -N 2733223 -X 4099833 -O /local/scratch/job_98738/pfd_ik1gk7fb/2 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,925 - CMD: fastq-dump -N 4099834 -X 5466444 -O /local/scratch/job_98738/pfd_ik1gk7fb/3 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,926 - CMD: fastq-dump -N 5466445 -X 6833055 -O /local/scratch/job_98738/pfd_ik1gk7fb/4 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,928 - CMD: fastq-dump -N 6833056 -X 8199666 -O /local/scratch/job_98738/pfd_ik1gk7fb/5 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,945 - CMD: fastq-dump -N 8199667 -X 9566277 -O /local/scratch/job_98738/pfd_ik1gk7fb/6 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,947 - CMD: fastq-dump -N 9566278 -X 10932888 -O /local/scratch/job_98738/pfd_ik1gk7fb/7 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,948 - CMD: fastq-dump -N 10932889 -X 12299499 -O /local/scratch/job_98738/pfd_ik1gk7fb/8 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,959 - CMD: fastq-dump -N 12299500 -X 13666110 -O /local/scratch/job_98738/pfd_ik1gk7fb/9 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,972 - CMD: fastq-dump -N 13666111 -X 15032721 -O /local/scratch/job_98738/pfd_ik1gk7fb/10 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,972 - CMD: fastq-dump -N 15032722 -X 16399332 -O /local/scratch/job_98738/pfd_ik1gk7fb/11 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,972 - CMD: fastq-dump -N 16399333 -X 17765943 -O /local/scratch/job_98738/pfd_ik1gk7fb/12 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,982 - CMD: fastq-dump -N 17765944 -X 19132554 -O /local/scratch/job_98738/pfd_ik1gk7fb/13 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,988 - CMD: fastq-dump -N 19132555 -X 20499165 -O /local/scratch/job_98738/pfd_ik1gk7fb/14 --gzip SRR26560373\n",
      "2024-10-10 15:43:59,997 - CMD: fastq-dump -N 20499166 -X 21865790 -O /local/scratch/job_98738/pfd_ik1gk7fb/15 --gzip SRR26560373\n"
     ]
    }
   ],
   "source": [
    "!mkdir fastq_lemons\n",
    "!cut -d ',' -f 1 SraRunTable_lemons.txt | tail -n +2 > sra_list_lemons.txt\n",
    "!while IFS= read -r line; do \\\n",
    "    echo \"Getting $line from NCBI SRA\"; \\\n",
    "    parallel-fastq-dump --sra-id $line --threads 16 --outdir fastq_lemons --gzip; \\\n",
    "    done < sra_list_lemons.txt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run QC on the FASTQ files and aggregate the results into a report."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir fastq_lemons/qc\n",
    "!fastqc -t 16 fastq_lemons/*.fastq.gz -o fastq_lemons/qc"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Edit this block with a short evaluation of each of the report sections. Each evaluation should include your important observations, reasons for potential QC failure, and an opinion on what should be done on the failure (i.e., trimming or filtering). Problem FASTQ files, if any, should be noted.\n",
    "\n",
    "*Per base sequence quality:*\n",
    "\n",
    "*Per tile sequence quality:*\n",
    "\n",
    "*Per sequence quality scores:*\n",
    "\n",
    "*Per base sequence content:*\n",
    "\n",
    "*Per sequence GC content:*\n",
    "\n",
    "*Per base N content:*\n",
    "\n",
    "*Sequence length distribution:*\n",
    "\n",
    "*Sequence duplication levels:*\n",
    "\n",
    "*Overrepresented sequences:*\n",
    "\n",
    "*Adapter content:*"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If necessary, use `cutadapt` to trim and filter your reads. Provide a justification for the `cutadapt` parameters you used:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%mkdir trimmed_lemons\n",
    "!for fastq in fastq_lemons/*.fastq.gz; do \\\n",
    "    base_name=$(basename \"$fastq\" .fastq.gz); \\\n",
    "    cutadapt -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed_lemons/$base_name.fastq.gz $fastq_lemons; \\\n",
    "done"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run QC on the trimmed/filtered reads and generate an aggregated report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir trimmed_lemons/qc\n",
    "!fastqc -t 16 trimmed_lemons/*.fastq.gz -o trimmed_lemons/qc\n",
    "!multiqc --force -d fastq_lemons/qc/ trimmed_lemons/qc/"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Report your observations on which QC metrics improved (or got worse):\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Snakemake Pipeline"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When running bioinformatics studies, it is wise to use notebooks (such as this one) to tie together code, decisions, and data visualization. The above task is something you might generate and report to a client or supervisor who might be interested in sequencing QC and what kind of pre-processing decisions you made prior to alignment. In addition, the preprocessing should be readily reproducible by all relevant stakeholders, so you should supplement notebooks with pipeline scripts. Throughout the semester, we will be creating snakemake pipelines to reproduce the published analysis that you've selected. Follow the below steps to begin producing this pipeline:\n",
    "\n",
    "1. Make a new `Snakefile` and `config.yaml` in `4_fastq`. \n",
    "2. In the `Snakefile`, produce the following rules:  \n",
    "    a. `fetch_fastq` - use `parallel-fastq-dump` to fetch every FASTQ file listed in `config.yaml`.  \n",
    "    b. `fastq_qc` - use `fastqc` to run QC on the raw FASTQ files.  \n",
    "    c. `trim_filter` - use `cutadapt` to trim and/or filter the FASTQ reads, using parameters justified in this notebook. ***Note:*** if you choose to trim/filter, it should be done on *all* of the FASTQ files; do not cherry pick.  \n",
    "    d. `trim_qc` - use `fastqc` to run QC on the trimmed FASTQ files.  \n",
    "    e. `report` - use `multiqc` to aggregate all QC data into a single report. . \n",
    "\n",
    "This `Snakefile` should not use any `wrappers` because we already have the relevant software installed in the `biol343` conda environment. The pipeline should run to completion when any instructor or classmate runs `snakemake --use-conda`. You may use the documentation found [here](https://multiqc.info/docs/usage/pipelines/#snakemake) or any other online documentation you may find. You may not use any AI tools to complete this homework."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biol343",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
