{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FASTQ download and QC\n",
    "\n",
    "## Fetching FASTQ data from NCIB SRA"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In previous weeks, we explored the genome and annotations. There were stored in FASTA and GTF files and will be used as the reference to which RNA sequencing reads are mapped/aligned. After alignment, we can count how many RNA-seq reads align to the annotations we care about (genes or transcripts). Now we need to get the FASTQ files that we want to align and analyze. To do so, we will be referencing an SRA read table, which is a list of all the FASTQ files associated with a given experiment uploaded to NCBI SRA. Here's the read table for the Winners vs. Losers experiment:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run,Assay Type,AvgSpotLen,Bases,BioProject,BioSample,BioSampleModel,Bytes,Center Name,Collection_Date,Consent,DATASTORE filetype,DATASTORE provider,DATASTORE region,dev_stage,Experiment,geo_loc_name_country,geo_loc_name_country_continent,geo_loc_name,Host,Instrument,isolate,isolation_source,Library Name,LibraryLayout,LibrarySelection,LibrarySource,name,Organism,Platform,ReleaseDate,create_date,version,Sample Name,SRA Study,tissue\n",
      "SRR26691082,RNA-Seq,69,692754411,PRJNA1036419,SAMN38137055,Invertebrate,504911946,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"run.zq,fastq,sra\",\"s3,ncbi,gs\",\"s3.us-east-1,gs.us-east1,ncbi.public\",egg,SRX22390854,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,LIV_ma1,SINGLE,PolyA,TRANSCRIPTOMIC,LIV_ma1,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:53:00Z,1,LIV_ma1,SRP470475,liver\n",
      "SRR26691083,RNA-Seq,69,607047855,PRJNA1036419,SAMN38137054,Invertebrate,440003332,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"run.zq,sra,fastq\",\"s3,gs,ncbi\",\"gs.us-east1,ncbi.public,s3.us-east-1\",egg,SRX22390853,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,LIV_im3,SINGLE,PolyA,TRANSCRIPTOMIC,LIV_im3,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:56:00Z,1,LIV_im3,SRP470475,liver\n",
      "SRR26691084,RNA-Seq,69,585361086,PRJNA1036419,SAMN38137053,Invertebrate,435877612,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"run.zq,sra,fastq\",\"gs,ncbi,s3\",\"ncbi.public,gs.us-east1,s3.us-east-1\",egg,SRX22390852,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,LIV_im2,SINGLE,PolyA,TRANSCRIPTOMIC,LIV_im2,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:57:00Z,1,LIV_im2,SRP470475,liver\n",
      "SRR26691085,RNA-Seq,69,570309771,PRJNA1036419,SAMN38137052,Invertebrate,424542018,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"fastq,sra,run.zq\",\"gs,s3,ncbi\",\"ncbi.public,s3.us-east-1,gs.us-east1\",egg,SRX22390851,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,LIV_im1,SINGLE,PolyA,TRANSCRIPTOMIC,LIV_im1,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:55:00Z,1,LIV_im1,SRP470475,liver\n",
      "SRR26691086,RNA-Seq,70,481072480,PRJNA1036419,SAMN38137050,Invertebrate,208923585,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"sra,run.zq,fastq\",\"ncbi,s3,gs\",\"s3.us-east-1,gs.us-east1,ncbi.public\",egg,SRX22390849,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,INT_ma2,SINGLE,PolyA,TRANSCRIPTOMIC,INT_ma2,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:53:00Z,1,INT_ma2,SRP470475,intestine\n",
      "SRR26691087,RNA-Seq,70,787070970,PRJNA1036419,SAMN38137049,Invertebrate,333680444,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"fastq,run.zq,sra\",\"gs,s3,ncbi\",\"s3.us-east-1,ncbi.public,gs.us-east1\",egg,SRX22390848,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,INT_ma1,SINGLE,PolyA,TRANSCRIPTOMIC,INT_ma1,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:55:00Z,1,INT_ma1,SRP470475,intestine\n",
      "SRR26691088,RNA-Seq,70,697767490,PRJNA1036419,SAMN38137048,Invertebrate,298328167,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"fastq,sra,run.zq\",\"gs,ncbi,s3\",\"ncbi.public,s3.us-east-1,gs.us-east1\",egg,SRX22390847,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,INT_im3,SINGLE,PolyA,TRANSCRIPTOMIC,INT_im3,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:55:00Z,1,INT_im3,SRP470475,intestine\n",
      "SRR26691089,RNA-Seq,70,786825270,PRJNA1036419,SAMN38137051,Invertebrate,335483342,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"run.zq,sra,fastq\",\"ncbi,s3,gs\",\"gs.us-east1,s3.us-east-1,ncbi.public\",egg,SRX22390850,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,INT_ma3,SINGLE,PolyA,TRANSCRIPTOMIC,INT_ma3,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:54:00Z,1,INT_ma3,SRP470475,intestine\n",
      "SRR26691090,RNA-Seq,69,706109844,PRJNA1036419,SAMN38137057,Invertebrate,514970874,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"sra,run.zq,fastq\",\"s3,gs,ncbi\",\"s3.us-east-1,gs.us-east1,ncbi.public\",egg,SRX22390846,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,LIV_ma3,SINGLE,PolyA,TRANSCRIPTOMIC,LIV_ma3,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:55:00Z,1,LIV_ma3,SRP470475,liver\n",
      "SRR26691091,RNA-Seq,69,667259118,PRJNA1036419,SAMN38137056,Invertebrate,484032264,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"fastq,sra,run.zq\",\"ncbi,gs,s3\",\"s3.us-east-1,gs.us-east1,ncbi.public\",egg,SRX22390845,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,LIV_ma2,SINGLE,PolyA,TRANSCRIPTOMIC,LIV_ma2,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:55:00Z,1,LIV_ma2,SRP470475,liver\n",
      "SRR26691092,RNA-Seq,70,871762010,PRJNA1036419,SAMN38137047,Invertebrate,373958956,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"sra,run.zq,fastq\",\"s3,ncbi,gs\",\"s3.us-east-1,ncbi.public,gs.us-east1\",egg,SRX22390844,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,INT_im2,SINGLE,PolyA,TRANSCRIPTOMIC,INT_im2,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:53:00Z,1,INT_im2,SRP470475,intestine\n",
      "SRR26691093,RNA-Seq,70,1015520100,PRJNA1036419,SAMN38137046,Invertebrate,441871985,\"CHARLES UNIVERSITY, FACULTY OF SCIENCE\",2022-11,public,\"sra,fastq,run.zq\",\"s3,gs,ncbi\",\"s3.us-east-1,ncbi.public,gs.us-east1\",egg,SRX22390843,Czech Republic,Europe,Czech Republic: Prague,Mus musculus,NextSeq 500,Belo Horizonte,laboratory,INT_im1,SINGLE,PolyA,TRANSCRIPTOMIC,INT_im1,Schistosoma mansoni,ILLUMINA,2023-12-31T00:00:00Z,2023-11-07T04:54:00Z,1,INT_im1,SRP470475,intestine\n",
      "4_homework.ipynb  5_fastq.ipynb  assets  sra_list.txt  SraRunTable.txt\n"
     ]
    }
   ],
   "source": [
    "!head -15 SraRunTable.txt\n",
    "!ls"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see that there are 12 different runs, each has an ID that looks something like `SRR26691082`. After the ID, quite a bit of run metadata is provided. You should spend some time learning about what each field denotes.\n",
    "\n",
    "As we know, the experiment had four different samples: \n",
    "- liver immature\n",
    "- liver mature\n",
    "- intestine immature\n",
    "- intestine mature\n",
    "\n",
    "Here's the data from Figure 1 of the paper:\n",
    "\n",
    "<img src=\"assets/journal.ppat.1012268.g001.PNG\" alt=\"experiment data\" width=\"600\">\n",
    "\n",
    "The 12 runs correspond to the 12 dots showin in Figure 1D.\n",
    "\n",
    "First we have to use `sra-tools` to download the FASTQ files from NCBI's SRA database. To do so, we'll use a `for` loop within bash. We can use `cut` (which we learned about in [week 2](../2_genome_exploration/2_genome_exploration.ipynb)) to get the first field, but because it's comma-delimited instead of tab-delimited, we have to tell the program with the `-d` option. This can be saved to a file, which then can be looped through line-by-line. The FASTQ files will be stored in a new directory called `fastq`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting SRR26691082 from NCBI SRA\n",
      "2024-10-08 16:01:04,926 - SRR ids: ['SRR26691082']\n",
      "2024-10-08 16:01:04,926 - extra args: ['--gzip']\n",
      "2024-10-08 16:01:04,927 - tempdir: /local/scratch/job_98296/pfd_c6xj1vjl\n",
      "2024-10-08 16:01:04,927 - CMD: sra-stat --meta --quick SRR26691082\n",
      "2024-10-08 16:01:07,163 - SRR26691082 spots: 10039919\n",
      "2024-10-08 16:01:07,163 - blocks: [[1, 627494], [627495, 1254988], [1254989, 1882482], [1882483, 2509976], [2509977, 3137470], [3137471, 3764964], [3764965, 4392458], [4392459, 5019952], [5019953, 5647446], [5647447, 6274940], [6274941, 6902434], [6902435, 7529928], [7529929, 8157422], [8157423, 8784916], [8784917, 9412410], [9412411, 10039919]]\n",
      "2024-10-08 16:01:07,163 - CMD: fastq-dump -N 1 -X 627494 -O /local/scratch/job_98296/pfd_c6xj1vjl/0 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,176 - CMD: fastq-dump -N 627495 -X 1254988 -O /local/scratch/job_98296/pfd_c6xj1vjl/1 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,177 - CMD: fastq-dump -N 1254989 -X 1882482 -O /local/scratch/job_98296/pfd_c6xj1vjl/2 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,177 - CMD: fastq-dump -N 1882483 -X 2509976 -O /local/scratch/job_98296/pfd_c6xj1vjl/3 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,177 - CMD: fastq-dump -N 2509977 -X 3137470 -O /local/scratch/job_98296/pfd_c6xj1vjl/4 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,178 - CMD: fastq-dump -N 3137471 -X 3764964 -O /local/scratch/job_98296/pfd_c6xj1vjl/5 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,178 - CMD: fastq-dump -N 3764965 -X 4392458 -O /local/scratch/job_98296/pfd_c6xj1vjl/6 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,178 - CMD: fastq-dump -N 4392459 -X 5019952 -O /local/scratch/job_98296/pfd_c6xj1vjl/7 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,179 - CMD: fastq-dump -N 5019953 -X 5647446 -O /local/scratch/job_98296/pfd_c6xj1vjl/8 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,179 - CMD: fastq-dump -N 5647447 -X 6274940 -O /local/scratch/job_98296/pfd_c6xj1vjl/9 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,179 - CMD: fastq-dump -N 6274941 -X 6902434 -O /local/scratch/job_98296/pfd_c6xj1vjl/10 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,180 - CMD: fastq-dump -N 6902435 -X 7529928 -O /local/scratch/job_98296/pfd_c6xj1vjl/11 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,180 - CMD: fastq-dump -N 7529929 -X 8157422 -O /local/scratch/job_98296/pfd_c6xj1vjl/12 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,180 - CMD: fastq-dump -N 8157423 -X 8784916 -O /local/scratch/job_98296/pfd_c6xj1vjl/13 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,180 - CMD: fastq-dump -N 8784917 -X 9412410 -O /local/scratch/job_98296/pfd_c6xj1vjl/14 --gzip SRR26691082\n",
      "2024-10-08 16:01:07,181 - CMD: fastq-dump -N 9412411 -X 10039919 -O /local/scratch/job_98296/pfd_c6xj1vjl/15 --gzip SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627509 spots for SRR26691082\n",
      "Written 627509 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Read 627494 spots for SRR26691082\n",
      "Written 627494 spots for SRR26691082\n",
      "Getting SRR26691083 from NCBI SRA\n",
      "2024-10-08 16:02:01,179 - SRR ids: ['SRR26691083']\n",
      "2024-10-08 16:02:01,179 - extra args: ['--gzip']\n",
      "2024-10-08 16:02:01,180 - tempdir: /local/scratch/job_98296/pfd_vdvgn240\n",
      "2024-10-08 16:02:01,180 - CMD: sra-stat --meta --quick SRR26691083\n",
      "2024-10-08 16:02:04,157 - SRR26691083 spots: 8797795\n",
      "2024-10-08 16:02:04,157 - blocks: [[1, 549862], [549863, 1099724], [1099725, 1649586], [1649587, 2199448], [2199449, 2749310], [2749311, 3299172], [3299173, 3849034], [3849035, 4398896], [4398897, 4948758], [4948759, 5498620], [5498621, 6048482], [6048483, 6598344], [6598345, 7148206], [7148207, 7698068], [7698069, 8247930], [8247931, 8797795]]\n",
      "2024-10-08 16:02:04,157 - CMD: fastq-dump -N 1 -X 549862 -O /local/scratch/job_98296/pfd_vdvgn240/0 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,158 - CMD: fastq-dump -N 549863 -X 1099724 -O /local/scratch/job_98296/pfd_vdvgn240/1 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,158 - CMD: fastq-dump -N 1099725 -X 1649586 -O /local/scratch/job_98296/pfd_vdvgn240/2 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,158 - CMD: fastq-dump -N 1649587 -X 2199448 -O /local/scratch/job_98296/pfd_vdvgn240/3 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,159 - CMD: fastq-dump -N 2199449 -X 2749310 -O /local/scratch/job_98296/pfd_vdvgn240/4 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,177 - CMD: fastq-dump -N 2749311 -X 3299172 -O /local/scratch/job_98296/pfd_vdvgn240/5 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,177 - CMD: fastq-dump -N 3299173 -X 3849034 -O /local/scratch/job_98296/pfd_vdvgn240/6 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,178 - CMD: fastq-dump -N 3849035 -X 4398896 -O /local/scratch/job_98296/pfd_vdvgn240/7 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,179 - CMD: fastq-dump -N 4398897 -X 4948758 -O /local/scratch/job_98296/pfd_vdvgn240/8 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,179 - CMD: fastq-dump -N 4948759 -X 5498620 -O /local/scratch/job_98296/pfd_vdvgn240/9 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,201 - CMD: fastq-dump -N 5498621 -X 6048482 -O /local/scratch/job_98296/pfd_vdvgn240/10 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,202 - CMD: fastq-dump -N 6048483 -X 6598344 -O /local/scratch/job_98296/pfd_vdvgn240/11 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,204 - CMD: fastq-dump -N 6598345 -X 7148206 -O /local/scratch/job_98296/pfd_vdvgn240/12 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,205 - CMD: fastq-dump -N 7148207 -X 7698068 -O /local/scratch/job_98296/pfd_vdvgn240/13 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,224 - CMD: fastq-dump -N 7698069 -X 8247930 -O /local/scratch/job_98296/pfd_vdvgn240/14 --gzip SRR26691083\n",
      "2024-10-08 16:02:04,226 - CMD: fastq-dump -N 8247931 -X 8797795 -O /local/scratch/job_98296/pfd_vdvgn240/15 --gzip SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549865 spots for SRR26691083\n",
      "Written 549865 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Read 549862 spots for SRR26691083\n",
      "Written 549862 spots for SRR26691083\n",
      "Getting SRR26691084 from NCBI SRA\n",
      "2024-10-08 16:02:47,128 - SRR ids: ['SRR26691084']\n",
      "2024-10-08 16:02:47,128 - extra args: ['--gzip']\n",
      "2024-10-08 16:02:47,128 - tempdir: /local/scratch/job_98296/pfd_bj6s_4pc\n",
      "2024-10-08 16:02:47,128 - CMD: sra-stat --meta --quick SRR26691084\n",
      "2024-10-08 16:02:50,053 - SRR26691084 spots: 8483494\n",
      "2024-10-08 16:02:50,053 - blocks: [[1, 530218], [530219, 1060436], [1060437, 1590654], [1590655, 2120872], [2120873, 2651090], [2651091, 3181308], [3181309, 3711526], [3711527, 4241744], [4241745, 4771962], [4771963, 5302180], [5302181, 5832398], [5832399, 6362616], [6362617, 6892834], [6892835, 7423052], [7423053, 7953270], [7953271, 8483494]]\n",
      "2024-10-08 16:02:50,053 - CMD: fastq-dump -N 1 -X 530218 -O /local/scratch/job_98296/pfd_bj6s_4pc/0 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,053 - CMD: fastq-dump -N 530219 -X 1060436 -O /local/scratch/job_98296/pfd_bj6s_4pc/1 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,054 - CMD: fastq-dump -N 1060437 -X 1590654 -O /local/scratch/job_98296/pfd_bj6s_4pc/2 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,054 - CMD: fastq-dump -N 1590655 -X 2120872 -O /local/scratch/job_98296/pfd_bj6s_4pc/3 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,055 - CMD: fastq-dump -N 2120873 -X 2651090 -O /local/scratch/job_98296/pfd_bj6s_4pc/4 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,056 - CMD: fastq-dump -N 2651091 -X 3181308 -O /local/scratch/job_98296/pfd_bj6s_4pc/5 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,058 - CMD: fastq-dump -N 3181309 -X 3711526 -O /local/scratch/job_98296/pfd_bj6s_4pc/6 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,074 - CMD: fastq-dump -N 3711527 -X 4241744 -O /local/scratch/job_98296/pfd_bj6s_4pc/7 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,080 - CMD: fastq-dump -N 4241745 -X 4771962 -O /local/scratch/job_98296/pfd_bj6s_4pc/8 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,103 - CMD: fastq-dump -N 4771963 -X 5302180 -O /local/scratch/job_98296/pfd_bj6s_4pc/9 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,104 - CMD: fastq-dump -N 5302181 -X 5832398 -O /local/scratch/job_98296/pfd_bj6s_4pc/10 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,105 - CMD: fastq-dump -N 5832399 -X 6362616 -O /local/scratch/job_98296/pfd_bj6s_4pc/11 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,106 - CMD: fastq-dump -N 6362617 -X 6892834 -O /local/scratch/job_98296/pfd_bj6s_4pc/12 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,122 - CMD: fastq-dump -N 6892835 -X 7423052 -O /local/scratch/job_98296/pfd_bj6s_4pc/13 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,123 - CMD: fastq-dump -N 7423053 -X 7953270 -O /local/scratch/job_98296/pfd_bj6s_4pc/14 --gzip SRR26691084\n",
      "2024-10-08 16:02:50,124 - CMD: fastq-dump -N 7953271 -X 8483494 -O /local/scratch/job_98296/pfd_bj6s_4pc/15 --gzip SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530218 spots for SRR26691084\n",
      "Written 530218 spots for SRR26691084\n",
      "Read 530224 spots for SRR26691084\n",
      "Written 530224 spots for SRR26691084\n",
      "Getting SRR26691085 from NCBI SRA\n",
      "2024-10-08 16:03:32,807 - SRR ids: ['SRR26691085']\n",
      "2024-10-08 16:03:32,807 - extra args: ['--gzip']\n",
      "2024-10-08 16:03:32,807 - tempdir: /local/scratch/job_98296/pfd_ryphmshz\n",
      "2024-10-08 16:03:32,807 - CMD: sra-stat --meta --quick SRR26691085\n",
      "2024-10-08 16:03:35,001 - SRR26691085 spots: 8265359\n",
      "2024-10-08 16:03:35,001 - blocks: [[1, 516584], [516585, 1033168], [1033169, 1549752], [1549753, 2066336], [2066337, 2582920], [2582921, 3099504], [3099505, 3616088], [3616089, 4132672], [4132673, 4649256], [4649257, 5165840], [5165841, 5682424], [5682425, 6199008], [6199009, 6715592], [6715593, 7232176], [7232177, 7748760], [7748761, 8265359]]\n",
      "2024-10-08 16:03:35,001 - CMD: fastq-dump -N 1 -X 516584 -O /local/scratch/job_98296/pfd_ryphmshz/0 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,002 - CMD: fastq-dump -N 516585 -X 1033168 -O /local/scratch/job_98296/pfd_ryphmshz/1 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,002 - CMD: fastq-dump -N 1033169 -X 1549752 -O /local/scratch/job_98296/pfd_ryphmshz/2 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,002 - CMD: fastq-dump -N 1549753 -X 2066336 -O /local/scratch/job_98296/pfd_ryphmshz/3 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,003 - CMD: fastq-dump -N 2066337 -X 2582920 -O /local/scratch/job_98296/pfd_ryphmshz/4 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,014 - CMD: fastq-dump -N 2582921 -X 3099504 -O /local/scratch/job_98296/pfd_ryphmshz/5 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,014 - CMD: fastq-dump -N 3099505 -X 3616088 -O /local/scratch/job_98296/pfd_ryphmshz/6 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,015 - CMD: fastq-dump -N 3616089 -X 4132672 -O /local/scratch/job_98296/pfd_ryphmshz/7 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,016 - CMD: fastq-dump -N 4132673 -X 4649256 -O /local/scratch/job_98296/pfd_ryphmshz/8 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,017 - CMD: fastq-dump -N 4649257 -X 5165840 -O /local/scratch/job_98296/pfd_ryphmshz/9 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,040 - CMD: fastq-dump -N 5165841 -X 5682424 -O /local/scratch/job_98296/pfd_ryphmshz/10 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,041 - CMD: fastq-dump -N 5682425 -X 6199008 -O /local/scratch/job_98296/pfd_ryphmshz/11 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,043 - CMD: fastq-dump -N 6199009 -X 6715592 -O /local/scratch/job_98296/pfd_ryphmshz/12 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,060 - CMD: fastq-dump -N 6715593 -X 7232176 -O /local/scratch/job_98296/pfd_ryphmshz/13 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,061 - CMD: fastq-dump -N 7232177 -X 7748760 -O /local/scratch/job_98296/pfd_ryphmshz/14 --gzip SRR26691085\n",
      "2024-10-08 16:03:35,062 - CMD: fastq-dump -N 7748761 -X 8265359 -O /local/scratch/job_98296/pfd_ryphmshz/15 --gzip SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516599 spots for SRR26691085\n",
      "Written 516599 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Read 516584 spots for SRR26691085\n",
      "Written 516584 spots for SRR26691085\n",
      "Getting SRR26691086 from NCBI SRA\n",
      "2024-10-08 16:04:24,099 - SRR ids: ['SRR26691086']\n",
      "2024-10-08 16:04:24,099 - extra args: ['--gzip']\n",
      "2024-10-08 16:04:24,099 - tempdir: /local/scratch/job_98296/pfd_h35pccll\n",
      "2024-10-08 16:04:24,099 - CMD: sra-stat --meta --quick SRR26691086\n",
      "2024-10-08 16:04:26,193 - SRR26691086 spots: 6872464\n",
      "2024-10-08 16:04:26,193 - blocks: [[1, 429529], [429530, 859058], [859059, 1288587], [1288588, 1718116], [1718117, 2147645], [2147646, 2577174], [2577175, 3006703], [3006704, 3436232], [3436233, 3865761], [3865762, 4295290], [4295291, 4724819], [4724820, 5154348], [5154349, 5583877], [5583878, 6013406], [6013407, 6442935], [6442936, 6872464]]\n",
      "2024-10-08 16:04:26,193 - CMD: fastq-dump -N 1 -X 429529 -O /local/scratch/job_98296/pfd_h35pccll/0 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,194 - CMD: fastq-dump -N 429530 -X 859058 -O /local/scratch/job_98296/pfd_h35pccll/1 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,194 - CMD: fastq-dump -N 859059 -X 1288587 -O /local/scratch/job_98296/pfd_h35pccll/2 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,195 - CMD: fastq-dump -N 1288588 -X 1718116 -O /local/scratch/job_98296/pfd_h35pccll/3 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,196 - CMD: fastq-dump -N 1718117 -X 2147645 -O /local/scratch/job_98296/pfd_h35pccll/4 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,205 - CMD: fastq-dump -N 2147646 -X 2577174 -O /local/scratch/job_98296/pfd_h35pccll/5 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,205 - CMD: fastq-dump -N 2577175 -X 3006703 -O /local/scratch/job_98296/pfd_h35pccll/6 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,206 - CMD: fastq-dump -N 3006704 -X 3436232 -O /local/scratch/job_98296/pfd_h35pccll/7 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,207 - CMD: fastq-dump -N 3436233 -X 3865761 -O /local/scratch/job_98296/pfd_h35pccll/8 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,227 - CMD: fastq-dump -N 3865762 -X 4295290 -O /local/scratch/job_98296/pfd_h35pccll/9 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,230 - CMD: fastq-dump -N 4295291 -X 4724819 -O /local/scratch/job_98296/pfd_h35pccll/10 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,230 - CMD: fastq-dump -N 4724820 -X 5154348 -O /local/scratch/job_98296/pfd_h35pccll/11 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,241 - CMD: fastq-dump -N 5154349 -X 5583877 -O /local/scratch/job_98296/pfd_h35pccll/12 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,253 - CMD: fastq-dump -N 5583878 -X 6013406 -O /local/scratch/job_98296/pfd_h35pccll/13 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,254 - CMD: fastq-dump -N 6013407 -X 6442935 -O /local/scratch/job_98296/pfd_h35pccll/14 --gzip SRR26691086\n",
      "2024-10-08 16:04:26,256 - CMD: fastq-dump -N 6442936 -X 6872464 -O /local/scratch/job_98296/pfd_h35pccll/15 --gzip SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Read 429529 spots for SRR26691086\n",
      "Written 429529 spots for SRR26691086\n",
      "Getting SRR26691087 from NCBI SRA\n",
      "2024-10-08 16:04:48,526 - SRR ids: ['SRR26691087']\n",
      "2024-10-08 16:04:48,526 - extra args: ['--gzip']\n",
      "2024-10-08 16:04:48,526 - tempdir: /local/scratch/job_98296/pfd_42w0hxr7\n",
      "2024-10-08 16:04:48,526 - CMD: sra-stat --meta --quick SRR26691087\n",
      "2024-10-08 16:04:50,560 - SRR26691087 spots: 11243871\n",
      "2024-10-08 16:04:50,560 - blocks: [[1, 702741], [702742, 1405482], [1405483, 2108223], [2108224, 2810964], [2810965, 3513705], [3513706, 4216446], [4216447, 4919187], [4919188, 5621928], [5621929, 6324669], [6324670, 7027410], [7027411, 7730151], [7730152, 8432892], [8432893, 9135633], [9135634, 9838374], [9838375, 10541115], [10541116, 11243871]]\n",
      "2024-10-08 16:04:50,560 - CMD: fastq-dump -N 1 -X 702741 -O /local/scratch/job_98296/pfd_42w0hxr7/0 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,560 - CMD: fastq-dump -N 702742 -X 1405482 -O /local/scratch/job_98296/pfd_42w0hxr7/1 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,561 - CMD: fastq-dump -N 1405483 -X 2108223 -O /local/scratch/job_98296/pfd_42w0hxr7/2 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,561 - CMD: fastq-dump -N 2108224 -X 2810964 -O /local/scratch/job_98296/pfd_42w0hxr7/3 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,562 - CMD: fastq-dump -N 2810965 -X 3513705 -O /local/scratch/job_98296/pfd_42w0hxr7/4 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,566 - CMD: fastq-dump -N 3513706 -X 4216446 -O /local/scratch/job_98296/pfd_42w0hxr7/5 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,567 - CMD: fastq-dump -N 4216447 -X 4919187 -O /local/scratch/job_98296/pfd_42w0hxr7/6 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,567 - CMD: fastq-dump -N 4919188 -X 5621928 -O /local/scratch/job_98296/pfd_42w0hxr7/7 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,568 - CMD: fastq-dump -N 5621929 -X 6324669 -O /local/scratch/job_98296/pfd_42w0hxr7/8 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,570 - CMD: fastq-dump -N 6324670 -X 7027410 -O /local/scratch/job_98296/pfd_42w0hxr7/9 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,591 - CMD: fastq-dump -N 7027411 -X 7730151 -O /local/scratch/job_98296/pfd_42w0hxr7/10 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,592 - CMD: fastq-dump -N 7730152 -X 8432892 -O /local/scratch/job_98296/pfd_42w0hxr7/11 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,593 - CMD: fastq-dump -N 8432893 -X 9135633 -O /local/scratch/job_98296/pfd_42w0hxr7/12 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,594 - CMD: fastq-dump -N 9135634 -X 9838374 -O /local/scratch/job_98296/pfd_42w0hxr7/13 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,617 - CMD: fastq-dump -N 9838375 -X 10541115 -O /local/scratch/job_98296/pfd_42w0hxr7/14 --gzip SRR26691087\n",
      "2024-10-08 16:04:50,630 - CMD: fastq-dump -N 10541116 -X 11243871 -O /local/scratch/job_98296/pfd_42w0hxr7/15 --gzip SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702756 spots for SRR26691087\n",
      "Written 702756 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Read 702741 spots for SRR26691087\n",
      "Written 702741 spots for SRR26691087\n",
      "Getting SRR26691088 from NCBI SRA\n",
      "2024-10-08 16:05:22,936 - SRR ids: ['SRR26691088']\n",
      "2024-10-08 16:05:22,936 - extra args: ['--gzip']\n",
      "2024-10-08 16:05:22,936 - tempdir: /local/scratch/job_98296/pfd_dy4lg9gg\n",
      "2024-10-08 16:05:22,936 - CMD: sra-stat --meta --quick SRR26691088\n",
      "2024-10-08 16:05:24,999 - SRR26691088 spots: 9968107\n",
      "2024-10-08 16:05:24,999 - blocks: [[1, 623006], [623007, 1246012], [1246013, 1869018], [1869019, 2492024], [2492025, 3115030], [3115031, 3738036], [3738037, 4361042], [4361043, 4984048], [4984049, 5607054], [5607055, 6230060], [6230061, 6853066], [6853067, 7476072], [7476073, 8099078], [8099079, 8722084], [8722085, 9345090], [9345091, 9968107]]\n",
      "2024-10-08 16:05:25,000 - CMD: fastq-dump -N 1 -X 623006 -O /local/scratch/job_98296/pfd_dy4lg9gg/0 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,000 - CMD: fastq-dump -N 623007 -X 1246012 -O /local/scratch/job_98296/pfd_dy4lg9gg/1 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,000 - CMD: fastq-dump -N 1246013 -X 1869018 -O /local/scratch/job_98296/pfd_dy4lg9gg/2 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,001 - CMD: fastq-dump -N 1869019 -X 2492024 -O /local/scratch/job_98296/pfd_dy4lg9gg/3 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,002 - CMD: fastq-dump -N 2492025 -X 3115030 -O /local/scratch/job_98296/pfd_dy4lg9gg/4 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,018 - CMD: fastq-dump -N 3115031 -X 3738036 -O /local/scratch/job_98296/pfd_dy4lg9gg/5 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,019 - CMD: fastq-dump -N 3738037 -X 4361042 -O /local/scratch/job_98296/pfd_dy4lg9gg/6 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,019 - CMD: fastq-dump -N 4361043 -X 4984048 -O /local/scratch/job_98296/pfd_dy4lg9gg/7 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,021 - CMD: fastq-dump -N 4984049 -X 5607054 -O /local/scratch/job_98296/pfd_dy4lg9gg/8 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,022 - CMD: fastq-dump -N 5607055 -X 6230060 -O /local/scratch/job_98296/pfd_dy4lg9gg/9 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,041 - CMD: fastq-dump -N 6230061 -X 6853066 -O /local/scratch/job_98296/pfd_dy4lg9gg/10 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,042 - CMD: fastq-dump -N 6853067 -X 7476072 -O /local/scratch/job_98296/pfd_dy4lg9gg/11 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,044 - CMD: fastq-dump -N 7476073 -X 8099078 -O /local/scratch/job_98296/pfd_dy4lg9gg/12 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,045 - CMD: fastq-dump -N 8099079 -X 8722084 -O /local/scratch/job_98296/pfd_dy4lg9gg/13 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,066 - CMD: fastq-dump -N 8722085 -X 9345090 -O /local/scratch/job_98296/pfd_dy4lg9gg/14 --gzip SRR26691088\n",
      "2024-10-08 16:05:25,080 - CMD: fastq-dump -N 9345091 -X 9968107 -O /local/scratch/job_98296/pfd_dy4lg9gg/15 --gzip SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623006 spots for SRR26691088\n",
      "Written 623006 spots for SRR26691088\n",
      "Read 623017 spots for SRR26691088\n",
      "Written 623017 spots for SRR26691088\n",
      "Getting SRR26691089 from NCBI SRA\n",
      "2024-10-08 16:05:52,614 - SRR ids: ['SRR26691089']\n",
      "2024-10-08 16:05:52,614 - extra args: ['--gzip']\n",
      "2024-10-08 16:05:52,615 - tempdir: /local/scratch/job_98296/pfd_69optcqj\n",
      "2024-10-08 16:05:52,615 - CMD: sra-stat --meta --quick SRR26691089\n",
      "2024-10-08 16:05:54,683 - SRR26691089 spots: 11240361\n",
      "2024-10-08 16:05:54,683 - blocks: [[1, 702522], [702523, 1405044], [1405045, 2107566], [2107567, 2810088], [2810089, 3512610], [3512611, 4215132], [4215133, 4917654], [4917655, 5620176], [5620177, 6322698], [6322699, 7025220], [7025221, 7727742], [7727743, 8430264], [8430265, 9132786], [9132787, 9835308], [9835309, 10537830], [10537831, 11240361]]\n",
      "2024-10-08 16:05:54,683 - CMD: fastq-dump -N 1 -X 702522 -O /local/scratch/job_98296/pfd_69optcqj/0 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,684 - CMD: fastq-dump -N 702523 -X 1405044 -O /local/scratch/job_98296/pfd_69optcqj/1 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,684 - CMD: fastq-dump -N 1405045 -X 2107566 -O /local/scratch/job_98296/pfd_69optcqj/2 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,685 - CMD: fastq-dump -N 2107567 -X 2810088 -O /local/scratch/job_98296/pfd_69optcqj/3 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,686 - CMD: fastq-dump -N 2810089 -X 3512610 -O /local/scratch/job_98296/pfd_69optcqj/4 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,686 - CMD: fastq-dump -N 3512611 -X 4215132 -O /local/scratch/job_98296/pfd_69optcqj/5 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,693 - CMD: fastq-dump -N 4215133 -X 4917654 -O /local/scratch/job_98296/pfd_69optcqj/6 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,694 - CMD: fastq-dump -N 4917655 -X 5620176 -O /local/scratch/job_98296/pfd_69optcqj/7 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,695 - CMD: fastq-dump -N 5620177 -X 6322698 -O /local/scratch/job_98296/pfd_69optcqj/8 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,697 - CMD: fastq-dump -N 6322699 -X 7025220 -O /local/scratch/job_98296/pfd_69optcqj/9 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,708 - CMD: fastq-dump -N 7025221 -X 7727742 -O /local/scratch/job_98296/pfd_69optcqj/10 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,716 - CMD: fastq-dump -N 7727743 -X 8430264 -O /local/scratch/job_98296/pfd_69optcqj/11 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,721 - CMD: fastq-dump -N 8430265 -X 9132786 -O /local/scratch/job_98296/pfd_69optcqj/12 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,740 - CMD: fastq-dump -N 9132787 -X 9835308 -O /local/scratch/job_98296/pfd_69optcqj/13 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,741 - CMD: fastq-dump -N 9835309 -X 10537830 -O /local/scratch/job_98296/pfd_69optcqj/14 --gzip SRR26691089\n",
      "2024-10-08 16:05:54,742 - CMD: fastq-dump -N 10537831 -X 11240361 -O /local/scratch/job_98296/pfd_69optcqj/15 --gzip SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702531 spots for SRR26691089\n",
      "Written 702531 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Read 702522 spots for SRR26691089\n",
      "Written 702522 spots for SRR26691089\n",
      "Getting SRR26691090 from NCBI SRA\n",
      "2024-10-08 16:06:29,975 - SRR ids: ['SRR26691090']\n",
      "2024-10-08 16:06:29,975 - extra args: ['--gzip']\n",
      "2024-10-08 16:06:29,976 - tempdir: /local/scratch/job_98296/pfd_6ltbd1i3\n",
      "2024-10-08 16:06:29,976 - CMD: sra-stat --meta --quick SRR26691090\n",
      "2024-10-08 16:06:32,044 - SRR26691090 spots: 10233476\n",
      "2024-10-08 16:06:32,044 - blocks: [[1, 639592], [639593, 1279184], [1279185, 1918776], [1918777, 2558368], [2558369, 3197960], [3197961, 3837552], [3837553, 4477144], [4477145, 5116736], [5116737, 5756328], [5756329, 6395920], [6395921, 7035512], [7035513, 7675104], [7675105, 8314696], [8314697, 8954288], [8954289, 9593880], [9593881, 10233476]]\n",
      "2024-10-08 16:06:32,044 - CMD: fastq-dump -N 1 -X 639592 -O /local/scratch/job_98296/pfd_6ltbd1i3/0 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,045 - CMD: fastq-dump -N 639593 -X 1279184 -O /local/scratch/job_98296/pfd_6ltbd1i3/1 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,045 - CMD: fastq-dump -N 1279185 -X 1918776 -O /local/scratch/job_98296/pfd_6ltbd1i3/2 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,046 - CMD: fastq-dump -N 1918777 -X 2558368 -O /local/scratch/job_98296/pfd_6ltbd1i3/3 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,047 - CMD: fastq-dump -N 2558369 -X 3197960 -O /local/scratch/job_98296/pfd_6ltbd1i3/4 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,065 - CMD: fastq-dump -N 3197961 -X 3837552 -O /local/scratch/job_98296/pfd_6ltbd1i3/5 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,071 - CMD: fastq-dump -N 3837553 -X 4477144 -O /local/scratch/job_98296/pfd_6ltbd1i3/6 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,072 - CMD: fastq-dump -N 4477145 -X 5116736 -O /local/scratch/job_98296/pfd_6ltbd1i3/7 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,077 - CMD: fastq-dump -N 5116737 -X 5756328 -O /local/scratch/job_98296/pfd_6ltbd1i3/8 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,095 - CMD: fastq-dump -N 5756329 -X 6395920 -O /local/scratch/job_98296/pfd_6ltbd1i3/9 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,098 - CMD: fastq-dump -N 6395921 -X 7035512 -O /local/scratch/job_98296/pfd_6ltbd1i3/10 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,101 - CMD: fastq-dump -N 7035513 -X 7675104 -O /local/scratch/job_98296/pfd_6ltbd1i3/11 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,103 - CMD: fastq-dump -N 7675105 -X 8314696 -O /local/scratch/job_98296/pfd_6ltbd1i3/12 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,119 - CMD: fastq-dump -N 8314697 -X 8954288 -O /local/scratch/job_98296/pfd_6ltbd1i3/13 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,122 - CMD: fastq-dump -N 8954289 -X 9593880 -O /local/scratch/job_98296/pfd_6ltbd1i3/14 --gzip SRR26691090\n",
      "2024-10-08 16:06:32,128 - CMD: fastq-dump -N 9593881 -X 10233476 -O /local/scratch/job_98296/pfd_6ltbd1i3/15 --gzip SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639596 spots for SRR26691090\n",
      "Written 639596 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Read 639592 spots for SRR26691090\n",
      "Written 639592 spots for SRR26691090\n",
      "Getting SRR26691091 from NCBI SRA\n",
      "2024-10-08 16:07:21,945 - SRR ids: ['SRR26691091']\n",
      "2024-10-08 16:07:21,945 - extra args: ['--gzip']\n",
      "2024-10-08 16:07:21,946 - tempdir: /local/scratch/job_98296/pfd_89ul7xr3\n",
      "2024-10-08 16:07:21,946 - CMD: sra-stat --meta --quick SRR26691091\n",
      "2024-10-08 16:07:24,091 - SRR26691091 spots: 9670422\n",
      "2024-10-08 16:07:24,091 - blocks: [[1, 604401], [604402, 1208802], [1208803, 1813203], [1813204, 2417604], [2417605, 3022005], [3022006, 3626406], [3626407, 4230807], [4230808, 4835208], [4835209, 5439609], [5439610, 6044010], [6044011, 6648411], [6648412, 7252812], [7252813, 7857213], [7857214, 8461614], [8461615, 9066015], [9066016, 9670422]]\n",
      "2024-10-08 16:07:24,091 - CMD: fastq-dump -N 1 -X 604401 -O /local/scratch/job_98296/pfd_89ul7xr3/0 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,092 - CMD: fastq-dump -N 604402 -X 1208802 -O /local/scratch/job_98296/pfd_89ul7xr3/1 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,092 - CMD: fastq-dump -N 1208803 -X 1813203 -O /local/scratch/job_98296/pfd_89ul7xr3/2 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,093 - CMD: fastq-dump -N 1813204 -X 2417604 -O /local/scratch/job_98296/pfd_89ul7xr3/3 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,094 - CMD: fastq-dump -N 2417605 -X 3022005 -O /local/scratch/job_98296/pfd_89ul7xr3/4 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,094 - CMD: fastq-dump -N 3022006 -X 3626406 -O /local/scratch/job_98296/pfd_89ul7xr3/5 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,113 - CMD: fastq-dump -N 3626407 -X 4230807 -O /local/scratch/job_98296/pfd_89ul7xr3/6 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,118 - CMD: fastq-dump -N 4230808 -X 4835208 -O /local/scratch/job_98296/pfd_89ul7xr3/7 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,119 - CMD: fastq-dump -N 4835209 -X 5439609 -O /local/scratch/job_98296/pfd_89ul7xr3/8 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,120 - CMD: fastq-dump -N 5439610 -X 6044010 -O /local/scratch/job_98296/pfd_89ul7xr3/9 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,131 - CMD: fastq-dump -N 6044011 -X 6648411 -O /local/scratch/job_98296/pfd_89ul7xr3/10 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,132 - CMD: fastq-dump -N 6648412 -X 7252812 -O /local/scratch/job_98296/pfd_89ul7xr3/11 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,133 - CMD: fastq-dump -N 7252813 -X 7857213 -O /local/scratch/job_98296/pfd_89ul7xr3/12 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,134 - CMD: fastq-dump -N 7857214 -X 8461614 -O /local/scratch/job_98296/pfd_89ul7xr3/13 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,142 - CMD: fastq-dump -N 8461615 -X 9066015 -O /local/scratch/job_98296/pfd_89ul7xr3/14 --gzip SRR26691091\n",
      "2024-10-08 16:07:24,155 - CMD: fastq-dump -N 9066016 -X 9670422 -O /local/scratch/job_98296/pfd_89ul7xr3/15 --gzip SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604407 spots for SRR26691091\n",
      "Written 604407 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Read 604401 spots for SRR26691091\n",
      "Written 604401 spots for SRR26691091\n",
      "Getting SRR26691092 from NCBI SRA\n",
      "2024-10-08 16:08:13,082 - SRR ids: ['SRR26691092']\n",
      "2024-10-08 16:08:13,082 - extra args: ['--gzip']\n",
      "2024-10-08 16:08:13,082 - tempdir: /local/scratch/job_98296/pfd__l9yi9y4\n",
      "2024-10-08 16:08:13,082 - CMD: sra-stat --meta --quick SRR26691092\n",
      "2024-10-08 16:08:15,224 - SRR26691092 spots: 12453743\n",
      "2024-10-08 16:08:15,224 - blocks: [[1, 778358], [778359, 1556716], [1556717, 2335074], [2335075, 3113432], [3113433, 3891790], [3891791, 4670148], [4670149, 5448506], [5448507, 6226864], [6226865, 7005222], [7005223, 7783580], [7783581, 8561938], [8561939, 9340296], [9340297, 10118654], [10118655, 10897012], [10897013, 11675370], [11675371, 12453743]]\n",
      "2024-10-08 16:08:15,224 - CMD: fastq-dump -N 1 -X 778358 -O /local/scratch/job_98296/pfd__l9yi9y4/0 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,225 - CMD: fastq-dump -N 778359 -X 1556716 -O /local/scratch/job_98296/pfd__l9yi9y4/1 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,225 - CMD: fastq-dump -N 1556717 -X 2335074 -O /local/scratch/job_98296/pfd__l9yi9y4/2 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,226 - CMD: fastq-dump -N 2335075 -X 3113432 -O /local/scratch/job_98296/pfd__l9yi9y4/3 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,227 - CMD: fastq-dump -N 3113433 -X 3891790 -O /local/scratch/job_98296/pfd__l9yi9y4/4 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,248 - CMD: fastq-dump -N 3891791 -X 4670148 -O /local/scratch/job_98296/pfd__l9yi9y4/5 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,248 - CMD: fastq-dump -N 4670149 -X 5448506 -O /local/scratch/job_98296/pfd__l9yi9y4/6 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,249 - CMD: fastq-dump -N 5448507 -X 6226864 -O /local/scratch/job_98296/pfd__l9yi9y4/7 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,250 - CMD: fastq-dump -N 6226865 -X 7005222 -O /local/scratch/job_98296/pfd__l9yi9y4/8 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,251 - CMD: fastq-dump -N 7005223 -X 7783580 -O /local/scratch/job_98296/pfd__l9yi9y4/9 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,273 - CMD: fastq-dump -N 7783581 -X 8561938 -O /local/scratch/job_98296/pfd__l9yi9y4/10 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,274 - CMD: fastq-dump -N 8561939 -X 9340296 -O /local/scratch/job_98296/pfd__l9yi9y4/11 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,286 - CMD: fastq-dump -N 9340297 -X 10118654 -O /local/scratch/job_98296/pfd__l9yi9y4/12 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,287 - CMD: fastq-dump -N 10118655 -X 10897012 -O /local/scratch/job_98296/pfd__l9yi9y4/13 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,302 - CMD: fastq-dump -N 10897013 -X 11675370 -O /local/scratch/job_98296/pfd__l9yi9y4/14 --gzip SRR26691092\n",
      "2024-10-08 16:08:15,303 - CMD: fastq-dump -N 11675371 -X 12453743 -O /local/scratch/job_98296/pfd__l9yi9y4/15 --gzip SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778373 spots for SRR26691092\n",
      "Written 778373 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Read 778358 spots for SRR26691092\n",
      "Written 778358 spots for SRR26691092\n",
      "Getting SRR26691093 from NCBI SRA\n",
      "2024-10-08 16:08:46,486 - SRR ids: ['SRR26691093']\n",
      "2024-10-08 16:08:46,486 - extra args: ['--gzip']\n",
      "2024-10-08 16:08:46,487 - tempdir: /local/scratch/job_98296/pfd_u_4fo0vz\n",
      "2024-10-08 16:08:46,487 - CMD: sra-stat --meta --quick SRR26691093\n",
      "2024-10-08 16:08:48,534 - SRR26691093 spots: 14507430\n",
      "2024-10-08 16:08:48,534 - blocks: [[1, 906714], [906715, 1813428], [1813429, 2720142], [2720143, 3626856], [3626857, 4533570], [4533571, 5440284], [5440285, 6346998], [6346999, 7253712], [7253713, 8160426], [8160427, 9067140], [9067141, 9973854], [9973855, 10880568], [10880569, 11787282], [11787283, 12693996], [12693997, 13600710], [13600711, 14507430]]\n",
      "2024-10-08 16:08:48,535 - CMD: fastq-dump -N 1 -X 906714 -O /local/scratch/job_98296/pfd_u_4fo0vz/0 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,535 - CMD: fastq-dump -N 906715 -X 1813428 -O /local/scratch/job_98296/pfd_u_4fo0vz/1 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,536 - CMD: fastq-dump -N 1813429 -X 2720142 -O /local/scratch/job_98296/pfd_u_4fo0vz/2 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,536 - CMD: fastq-dump -N 2720143 -X 3626856 -O /local/scratch/job_98296/pfd_u_4fo0vz/3 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,537 - CMD: fastq-dump -N 3626857 -X 4533570 -O /local/scratch/job_98296/pfd_u_4fo0vz/4 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,559 - CMD: fastq-dump -N 4533571 -X 5440284 -O /local/scratch/job_98296/pfd_u_4fo0vz/5 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,560 - CMD: fastq-dump -N 5440285 -X 6346998 -O /local/scratch/job_98296/pfd_u_4fo0vz/6 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,561 - CMD: fastq-dump -N 6346999 -X 7253712 -O /local/scratch/job_98296/pfd_u_4fo0vz/7 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,573 - CMD: fastq-dump -N 7253713 -X 8160426 -O /local/scratch/job_98296/pfd_u_4fo0vz/8 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,575 - CMD: fastq-dump -N 8160427 -X 9067140 -O /local/scratch/job_98296/pfd_u_4fo0vz/9 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,576 - CMD: fastq-dump -N 9067141 -X 9973854 -O /local/scratch/job_98296/pfd_u_4fo0vz/10 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,577 - CMD: fastq-dump -N 9973855 -X 10880568 -O /local/scratch/job_98296/pfd_u_4fo0vz/11 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,577 - CMD: fastq-dump -N 10880569 -X 11787282 -O /local/scratch/job_98296/pfd_u_4fo0vz/12 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,579 - CMD: fastq-dump -N 11787283 -X 12693996 -O /local/scratch/job_98296/pfd_u_4fo0vz/13 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,600 - CMD: fastq-dump -N 12693997 -X 13600710 -O /local/scratch/job_98296/pfd_u_4fo0vz/14 --gzip SRR26691093\n",
      "2024-10-08 16:08:48,600 - CMD: fastq-dump -N 13600711 -X 14507430 -O /local/scratch/job_98296/pfd_u_4fo0vz/15 --gzip SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906720 spots for SRR26691093\n",
      "Written 906720 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n",
      "Read 906714 spots for SRR26691093\n",
      "Written 906714 spots for SRR26691093\n"
     ]
    }
   ],
   "source": [
    "!mkdir fastq\n",
    "!cut -d ',' -f 1 SraRunTable.txt | tail -n +2 > sra_list.txt\n",
    "!while IFS= read -r line; do \\\n",
    "    echo \"Getting $line from NCBI SRA\"; \\\n",
    "    parallel-fastq-dump --sra-id $line --threads 16 --outdir fastq --gzip; \\\n",
    "    done < sra_list.txt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***8 minute time to completion***\n",
    "\n",
    "If we take another look at the metadata, there are a few things that are of interest to us and our analysis. First, these reads were generated with `PolyA` selection, which means reads should have many A's on their 3' end. Second, the reads are `ILLUMINA` reads generated on a `NextSeq 500` instrument, which means they may have Illumina adapters, depending on whether or not they were trimmed by the authors prior to uploading them to SRA. Third, from the paper (but not the run table), we know that these are single-end reads. This metadata will be important to us soon. You can get a bit more information about the data using the tool `seqkit`. For instance:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file                        format  type    num_seqs      sum_len  min_len  avg_len  max_len\n",
      "fastq/SRR26691082.fastq.gz  FASTQ   DNA   10,039,919  692,754,411       69       69       69\n"
     ]
    }
   ],
   "source": [
    "!seqkit stats fastq/SRR26691082.fastq.gz"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also simply use `grep` to count sequences. Because the files are gzipped, you have to decompress them first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10039919\n"
     ]
    }
   ],
   "source": [
    "!zcat -d fastq/SRR26691082.fastq.gz | grep -c '@'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FastQC\n",
    "\n",
    "Whenever we look at FASTQ files for the first time, we should perform quality control (QC). The primary tool used for read QC is called FastQC, which is installed in our environment. Let's take a look at the manual and then run QC on our reads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "            FastQC - A high throughput sequence QC analysis tool\n",
      "\n",
      "SYNOPSIS\n",
      "\n",
      "\tfastqc seqfile1 seqfile2 .. seqfileN\n",
      "\n",
      "    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] \n",
      "           [-c contaminant file] seqfile1 .. seqfileN\n",
      "\n",
      "DESCRIPTION\n",
      "\n",
      "    FastQC reads a set of sequence files and produces from each one a quality\n",
      "    control report consisting of a number of different modules, each one of \n",
      "    which will help to identify a different potential type of problem in your\n",
      "    data.\n",
      "    \n",
      "    If no files to process are specified on the command line then the program\n",
      "    will start as an interactive graphical application.  If files are provided\n",
      "    on the command line then the program will run with no user interaction\n",
      "    required.  In this mode it is suitable for inclusion into a standardised\n",
      "    analysis pipeline.\n",
      "    \n",
      "    The options for the program as as follows:\n",
      "    \n",
      "    -h --help       Print this help file and exit\n",
      "    \n",
      "    -v --version    Print the version of the program and exit\n",
      "    \n",
      "    -o --outdir     Create all output files in the specified output directory.\n",
      "                    Please note that this directory must exist as the program\n",
      "                    will not create it.  If this option is not set then the \n",
      "                    output file for each sequence file is created in the same\n",
      "                    directory as the sequence file which was processed.\n",
      "                    \n",
      "    --casava        Files come from raw casava output. Files in the same sample\n",
      "                    group (differing only by the group number) will be analysed\n",
      "                    as a set rather than individually. Sequences with the filter\n",
      "                    flag set in the header will be excluded from the analysis.\n",
      "                    Files must have the same names given to them by casava\n",
      "                    (including being gzipped and ending with .gz) otherwise they\n",
      "                    won't be grouped together correctly.\n",
      "                    \n",
      "    --nano          Files come from nanopore sequences and are in fast5 format. In\n",
      "                    this mode you can pass in directories to process and the program\n",
      "                    will take in all fast5 files within those directories and produce\n",
      "                    a single output file from the sequences found in all files.                    \n",
      "                    \n",
      "    --nofilter      If running with --casava then don't remove read flagged by\n",
      "                    casava as poor quality when performing the QC analysis.\n",
      "                   \n",
      "    --extract       If set then the zipped output file will be uncompressed in\n",
      "                    the same directory after it has been created. If --delete is \n",
      "                    also specified then the zip file will be removed after the \n",
      "                    contents are unzipped. \n",
      "                    \n",
      "    -j --java       Provides the full path to the java binary you want to use to\n",
      "                    launch fastqc. If not supplied then java is assumed to be in\n",
      "                    your path.\n",
      "                   \n",
      "    --noextract     Do not uncompress the output file after creating it.  You\n",
      "                    should set this option if you do not wish to uncompress\n",
      "                    the output when running in non-interactive mode.\n",
      "                    \n",
      "    --nogroup       Disable grouping of bases for reads >50bp. All reports will\n",
      "                    show data for every base in the read.  WARNING: Using this\n",
      "                    option will cause fastqc to crash and burn if you use it on\n",
      "                    really long reads, and your plots may end up a ridiculous size.\n",
      "                    You have been warned!\n",
      "                    \n",
      "    --min_length    Sets an artificial lower limit on the length of the sequence\n",
      "                    to be shown in the report.  As long as you set this to a value\n",
      "                    greater or equal to your longest read length then this will be\n",
      "                    the sequence length used to create your read groups.  This can\n",
      "                    be useful for making directly comaparable statistics from \n",
      "                    datasets with somewhat variable read lengths.\n",
      "\n",
      "    --dup_length    Sets a length to which the sequences will be truncated when \n",
      "                    defining them to be duplicates, affecting the duplication and\n",
      "                    overrepresented sequences plot.  This can be useful if you have\n",
      "                    long reads with higher levels of miscalls, or contamination with\n",
      "                    adapter dimers containing UMI sequences.\n",
      "\n",
      "                    \n",
      "    -f --format     Bypasses the normal sequence file format detection and\n",
      "                    forces the program to use the specified format.  Valid\n",
      "                    formats are bam,sam,bam_mapped,sam_mapped and fastq\n",
      "                    \n",
      "\n",
      "    --memory        Sets the base amount of memory, in Megabytes, used to process \n",
      "                    each file.  Defaults to 512MB.  You may need to increase this if\n",
      "                    you have a file with very long sequences in it.\n",
      "                \n",
      "    --svg           Save the graphs in the report in SVG format.\n",
      "\n",
      "    -t --threads    Specifies the number of files which can be processed\n",
      "                    simultaneously.  Each thread will be allocated 250MB of\n",
      "                    memory so you shouldn't run more threads than your\n",
      "                    available memory will cope with, and not more than\n",
      "                    6 threads on a 32 bit machine\n",
      "                  \n",
      "    -c              Specifies a non-default file which contains the list of\n",
      "    --contaminants  contaminants to screen overrepresented sequences against.\n",
      "                    The file must contain sets of named contaminants in the\n",
      "                    form name[tab]sequence.  Lines prefixed with a hash will\n",
      "                    be ignored.\n",
      "\n",
      "    -a              Specifies a non-default file which contains the list of\n",
      "    --adapters      adapter sequences which will be explicity searched against\n",
      "                    the library. The file must contain sets of named adapters\n",
      "                    in the form name[tab]sequence.  Lines prefixed with a hash\n",
      "                    will be ignored.\n",
      "                    \n",
      "    -l              Specifies a non-default file which contains a set of criteria\n",
      "    --limits        which will be used to determine the warn/error limits for the\n",
      "                    various modules.  This file can also be used to selectively \n",
      "                    remove some modules from the output all together.  The format\n",
      "                    needs to mirror the default limits.txt file found in the\n",
      "                    Configuration folder.\n",
      "                    \n",
      "   -k --kmers       Specifies the length of Kmer to look for in the Kmer content\n",
      "                    module. Specified Kmer length must be between 2 and 10. Default\n",
      "                    length is 7 if not specified.\n",
      "                    \n",
      "   -q --quiet       Suppress all progress messages on stdout and only report errors.\n",
      "   \n",
      "   -d --dir         Selects a directory to be used for temporary files written when\n",
      "                    generating report images. Defaults to system temp directory if\n",
      "                    not specified.\n",
      "                    \n",
      "BUGS\n",
      "\n",
      "    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk\n",
      "    or in www.bioinformatics.babraham.ac.uk/bugzilla/\n",
      "                   \n",
      "    \n"
     ]
    }
   ],
   "source": [
    "!fastqc -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR26691082.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR26691083.fastq.gz\n",
      "Started analysis of SRR26691084.fastq.gz\n",
      "Started analysis of SRR26691085.fastq.gz\n",
      "Started analysis of SRR26691086.fastq.gz\n",
      "Started analysis of SRR26691087.fastq.gz\n",
      "Approx 5% complete for SRR26691082.fastq.gz\n",
      "Approx 5% complete for SRR26691083.fastq.gz\n",
      "Started analysis of SRR26691088.fastq.gz\n",
      "Approx 5% complete for SRR26691084.fastq.gz\n",
      "Started analysis of SRR26691089.fastq.gz\n",
      "Approx 5% complete for SRR26691086.fastq.gz\n",
      "Approx 5% complete for SRR26691085.fastq.gz\n",
      "Started analysis of SRR26691090.fastq.gz\n",
      "Started analysis of SRR26691091.fastq.gz\n",
      "Started analysis of SRR26691092.fastq.gz\n",
      "Started analysis of SRR26691093.fastq.gz\n",
      "Approx 10% complete for SRR26691083.fastq.gz\n",
      "Approx 10% complete for SRR26691084.fastq.gz\n",
      "Approx 5% complete for SRR26691088.fastq.gz\n",
      "Approx 10% complete for SRR26691086.fastq.gz\n",
      "Approx 5% complete for SRR26691087.fastq.gz\n",
      "Approx 10% complete for SRR26691082.fastq.gz\n",
      "Approx 10% complete for SRR26691085.fastq.gz\n",
      "Approx 5% complete for SRR26691089.fastq.gz\n",
      "Approx 15% complete for SRR26691086.fastq.gz\n",
      "Approx 5% complete for SRR26691090.fastq.gz\n",
      "Approx 15% complete for SRR26691083.fastq.gz\n",
      "Approx 15% complete for SRR26691082.fastq.gz\n",
      "Approx 5% complete for SRR26691091.fastq.gz\n",
      "Approx 15% complete for SRR26691084.fastq.gz\n",
      "Approx 5% complete for SRR26691093.fastq.gz\n",
      "Approx 15% complete for SRR26691085.fastq.gz\n",
      "Approx 20% complete for SRR26691086.fastq.gz\n",
      "Approx 10% complete for SRR26691088.fastq.gz\n",
      "Approx 5% complete for SRR26691092.fastq.gz\n",
      "Approx 10% complete for SRR26691087.fastq.gz\n",
      "Approx 10% complete for SRR26691089.fastq.gz\n",
      "Approx 20% complete for SRR26691083.fastq.gz\n",
      "Approx 10% complete for SRR26691090.fastq.gz\n",
      "Approx 20% complete for SRR26691082.fastq.gz\n",
      "Approx 25% complete for SRR26691086.fastq.gz\n",
      "Approx 10% complete for SRR26691091.fastq.gz\n",
      "Approx 20% complete for SRR26691084.fastq.gz\n",
      "Approx 20% complete for SRR26691085.fastq.gz\n",
      "Approx 15% complete for SRR26691088.fastq.gz\n",
      "Approx 25% complete for SRR26691083.fastq.gz\n",
      "Approx 10% complete for SRR26691092.fastq.gz\n",
      "Approx 10% complete for SRR26691093.fastq.gz\n",
      "Approx 15% complete for SRR26691089.fastq.gz\n",
      "Approx 30% complete for SRR26691086.fastq.gz\n",
      "Approx 15% complete for SRR26691087.fastq.gz\n",
      "Approx 25% complete for SRR26691082.fastq.gz\n",
      "Approx 15% complete for SRR26691091.fastq.gz\n",
      "Approx 15% complete for SRR26691090.fastq.gz\n",
      "Approx 25% complete for SRR26691084.fastq.gz\n",
      "Approx 30% complete for SRR26691083.fastq.gz\n",
      "Approx 25% complete for SRR26691085.fastq.gz\n",
      "Approx 20% complete for SRR26691088.fastq.gz\n",
      "Approx 20% complete for SRR26691089.fastq.gz\n",
      "Approx 15% complete for SRR26691092.fastq.gz\n",
      "Approx 35% complete for SRR26691086.fastq.gz\n",
      "Approx 30% complete for SRR26691084.fastq.gz\n",
      "Approx 15% complete for SRR26691093.fastq.gz\n",
      "Approx 20% complete for SRR26691090.fastq.gz\n",
      "Approx 30% complete for SRR26691082.fastq.gz\n",
      "Approx 30% complete for SRR26691085.fastq.gz\n",
      "Approx 35% complete for SRR26691083.fastq.gz\n",
      "Approx 20% complete for SRR26691091.fastq.gz\n",
      "Approx 20% complete for SRR26691087.fastq.gz\n",
      "Approx 25% complete for SRR26691088.fastq.gz\n",
      "Approx 35% complete for SRR26691084.fastq.gz\n",
      "Approx 25% complete for SRR26691089.fastq.gz\n",
      "Approx 40% complete for SRR26691086.fastq.gz\n",
      "Approx 20% complete for SRR26691092.fastq.gz\n",
      "Approx 35% complete for SRR26691085.fastq.gz\n",
      "Approx 25% complete for SRR26691090.fastq.gz\n",
      "Approx 35% complete for SRR26691082.fastq.gz\n",
      "Approx 20% complete for SRR26691093.fastq.gz\n",
      "Approx 40% complete for SRR26691083.fastq.gz\n",
      "Approx 30% complete for SRR26691088.fastq.gz\n",
      "Approx 30% complete for SRR26691089.fastq.gz\n",
      "Approx 40% complete for SRR26691084.fastq.gz\n",
      "Approx 45% complete for SRR26691086.fastq.gz\n",
      "Approx 25% complete for SRR26691091.fastq.gz\n",
      "Approx 40% complete for SRR26691085.fastq.gz\n",
      "Approx 25% complete for SRR26691087.fastq.gz\n",
      "Approx 30% complete for SRR26691090.fastq.gz\n",
      "Approx 40% complete for SRR26691082.fastq.gz\n",
      "Approx 25% complete for SRR26691092.fastq.gz\n",
      "Approx 35% complete for SRR26691089.fastq.gz\n",
      "Approx 45% complete for SRR26691085.fastq.gz\n",
      "Approx 45% complete for SRR26691083.fastq.gz\n",
      "Approx 50% complete for SRR26691086.fastq.gz\n",
      "Approx 35% complete for SRR26691090.fastq.gz\n",
      "Approx 45% complete for SRR26691082.fastq.gz\n",
      "Approx 45% complete for SRR26691084.fastq.gz\n",
      "Approx 50% complete for SRR26691085.fastq.gz\n",
      "Approx 35% complete for SRR26691088.fastq.gz\n",
      "Approx 40% complete for SRR26691089.fastq.gz\n",
      "Approx 30% complete for SRR26691091.fastq.gz\n",
      "Approx 30% complete for SRR26691087.fastq.gz\n",
      "Approx 55% complete for SRR26691086.fastq.gz\n",
      "Approx 25% complete for SRR26691093.fastq.gz\n",
      "Approx 30% complete for SRR26691092.fastq.gz\n",
      "Approx 40% complete for SRR26691090.fastq.gz\n",
      "Approx 50% complete for SRR26691082.fastq.gz\n",
      "Approx 45% complete for SRR26691089.fastq.gz\n",
      "Approx 50% complete for SRR26691083.fastq.gz\n",
      "Approx 55% complete for SRR26691085.fastq.gz\n",
      "Approx 60% complete for SRR26691086.fastq.gz\n",
      "Approx 50% complete for SRR26691084.fastq.gz\n",
      "Approx 40% complete for SRR26691088.fastq.gz\n",
      "Approx 45% complete for SRR26691090.fastq.gz\n",
      "Approx 35% complete for SRR26691092.fastq.gz\n",
      "Approx 55% complete for SRR26691082.fastq.gz\n",
      "Approx 35% complete for SRR26691091.fastq.gz\n",
      "Approx 65% complete for SRR26691086.fastq.gz\n",
      "Approx 35% complete for SRR26691087.fastq.gz\n",
      "Approx 50% complete for SRR26691089.fastq.gz\n",
      "Approx 55% complete for SRR26691083.fastq.gz\n",
      "Approx 50% complete for SRR26691090.fastq.gz\n",
      "Approx 55% complete for SRR26691084.fastq.gz\n",
      "Approx 30% complete for SRR26691093.fastq.gz\n",
      "Approx 60% complete for SRR26691085.fastq.gz\n",
      "Approx 45% complete for SRR26691088.fastq.gz\n",
      "Approx 40% complete for SRR26691092.fastq.gz\n",
      "Approx 70% complete for SRR26691086.fastq.gz\n",
      "Approx 60% complete for SRR26691082.fastq.gz\n",
      "Approx 55% complete for SRR26691089.fastq.gz\n",
      "Approx 60% complete for SRR26691084.fastq.gz\n",
      "Approx 40% complete for SRR26691091.fastq.gz\n",
      "Approx 55% complete for SRR26691090.fastq.gz\n",
      "Approx 40% complete for SRR26691087.fastq.gz\n",
      "Approx 60% complete for SRR26691083.fastq.gz\n",
      "Approx 65% complete for SRR26691085.fastq.gz\n",
      "Approx 35% complete for SRR26691093.fastq.gz\n",
      "Approx 45% complete for SRR26691092.fastq.gz\n",
      "Approx 75% complete for SRR26691086.fastq.gz\n",
      "Approx 50% complete for SRR26691088.fastq.gz\n",
      "Approx 60% complete for SRR26691090.fastq.gz\n",
      "Approx 70% complete for SRR26691085.fastq.gz\n",
      "Approx 65% complete for SRR26691084.fastq.gz\n",
      "Approx 65% complete for SRR26691082.fastq.gz\n",
      "Approx 60% complete for SRR26691089.fastq.gz\n",
      "Approx 50% complete for SRR26691092.fastq.gz\n",
      "Approx 40% complete for SRR26691093.fastq.gz\n",
      "Approx 65% complete for SRR26691083.fastq.gz\n",
      "Approx 45% complete for SRR26691091.fastq.gz\n",
      "Approx 80% complete for SRR26691086.fastq.gz\n",
      "Approx 65% complete for SRR26691090.fastq.gz\n",
      "Approx 45% complete for SRR26691087.fastq.gz\n",
      "Approx 75% complete for SRR26691085.fastq.gz\n",
      "Approx 85% complete for SRR26691086.fastq.gz\n",
      "Approx 55% complete for SRR26691092.fastq.gz\n",
      "Approx 70% complete for SRR26691084.fastq.gz\n",
      "Approx 55% complete for SRR26691088.fastq.gz\n",
      "Approx 70% complete for SRR26691082.fastq.gz\n",
      "Approx 70% complete for SRR26691090.fastq.gz\n",
      "Approx 90% complete for SRR26691086.fastq.gz\n",
      "Approx 50% complete for SRR26691091.fastq.gz\n",
      "Approx 65% complete for SRR26691089.fastq.gz\n",
      "Approx 70% complete for SRR26691083.fastq.gz\n",
      "Approx 45% complete for SRR26691093.fastq.gz\n",
      "Approx 80% complete for SRR26691085.fastq.gz\n",
      "Approx 75% complete for SRR26691084.fastq.gz\n",
      "Approx 95% complete for SRR26691086.fastq.gz\n",
      "Approx 75% complete for SRR26691090.fastq.gz\n",
      "Approx 55% complete for SRR26691091.fastq.gz\n",
      "Approx 50% complete for SRR26691087.fastq.gz\n",
      "Approx 60% complete for SRR26691088.fastq.gz\n",
      "Analysis complete for SRR26691086.fastq.gz\n",
      "Approx 80% complete for SRR26691084.fastq.gz\n",
      "Approx 75% complete for SRR26691082.fastq.gz\n",
      "Approx 75% complete for SRR26691083.fastq.gz\n",
      "Approx 60% complete for SRR26691092.fastq.gz\n",
      "Approx 85% complete for SRR26691085.fastq.gz\n",
      "Approx 70% complete for SRR26691089.fastq.gz\n",
      "Approx 80% complete for SRR26691090.fastq.gz\n",
      "Approx 60% complete for SRR26691091.fastq.gz\n",
      "Approx 85% complete for SRR26691084.fastq.gz\n",
      "Approx 50% complete for SRR26691093.fastq.gz\n",
      "Approx 65% complete for SRR26691088.fastq.gz\n",
      "Approx 80% complete for SRR26691083.fastq.gz\n",
      "Approx 55% complete for SRR26691087.fastq.gz\n",
      "Approx 80% complete for SRR26691082.fastq.gz\n",
      "Approx 90% complete for SRR26691085.fastq.gz\n",
      "Approx 65% complete for SRR26691092.fastq.gz\n",
      "Approx 75% complete for SRR26691089.fastq.gz\n",
      "Approx 85% complete for SRR26691083.fastq.gz\n",
      "Approx 90% complete for SRR26691084.fastq.gz\n",
      "Approx 70% complete for SRR26691088.fastq.gz\n",
      "Approx 85% complete for SRR26691090.fastq.gz\n",
      "Approx 60% complete for SRR26691087.fastq.gz\n",
      "Approx 65% complete for SRR26691091.fastq.gz\n",
      "Approx 55% complete for SRR26691093.fastq.gz\n",
      "Approx 95% complete for SRR26691085.fastq.gz\n",
      "Approx 85% complete for SRR26691082.fastq.gz\n",
      "Approx 95% complete for SRR26691084.fastq.gz\n",
      "Approx 90% complete for SRR26691083.fastq.gz\n",
      "Approx 75% complete for SRR26691088.fastq.gz\n",
      "Approx 65% complete for SRR26691087.fastq.gz\n",
      "Approx 80% complete for SRR26691089.fastq.gz\n",
      "Approx 90% complete for SRR26691090.fastq.gz\n",
      "Approx 70% complete for SRR26691091.fastq.gz\n",
      "Approx 70% complete for SRR26691092.fastq.gz\n",
      "Analysis complete for SRR26691085.fastq.gz\n",
      "Analysis complete for SRR26691084.fastq.gz\n",
      "Approx 95% complete for SRR26691083.fastq.gz\n",
      "Approx 90% complete for SRR26691082.fastq.gz\n",
      "Approx 60% complete for SRR26691093.fastq.gz\n",
      "Approx 80% complete for SRR26691088.fastq.gz\n",
      "Approx 70% complete for SRR26691087.fastq.gz\n",
      "Approx 85% complete for SRR26691089.fastq.gz\n",
      "Approx 75% complete for SRR26691091.fastq.gz\n",
      "Approx 95% complete for SRR26691090.fastq.gz\n",
      "Approx 95% complete for SRR26691082.fastq.gz\n",
      "Approx 75% complete for SRR26691092.fastq.gz\n",
      "Analysis complete for SRR26691083.fastq.gz\n",
      "Approx 85% complete for SRR26691088.fastq.gz\n",
      "Approx 65% complete for SRR26691093.fastq.gz\n",
      "Approx 80% complete for SRR26691091.fastq.gz\n",
      "Approx 90% complete for SRR26691089.fastq.gz\n",
      "Approx 75% complete for SRR26691087.fastq.gz\n",
      "Analysis complete for SRR26691090.fastq.gz\n",
      "Analysis complete for SRR26691082.fastq.gz\n",
      "Approx 80% complete for SRR26691092.fastq.gz\n",
      "Approx 90% complete for SRR26691088.fastq.gz\n",
      "Approx 85% complete for SRR26691091.fastq.gz\n",
      "Approx 80% complete for SRR26691087.fastq.gz\n",
      "Approx 95% complete for SRR26691089.fastq.gz\n",
      "Approx 70% complete for SRR26691093.fastq.gz\n",
      "Approx 85% complete for SRR26691092.fastq.gz\n",
      "Approx 95% complete for SRR26691088.fastq.gz\n",
      "Approx 90% complete for SRR26691091.fastq.gz\n",
      "Approx 85% complete for SRR26691087.fastq.gz\n",
      "Analysis complete for SRR26691089.fastq.gz\n",
      "Analysis complete for SRR26691088.fastq.gz\n",
      "Approx 75% complete for SRR26691093.fastq.gz\n",
      "Approx 90% complete for SRR26691092.fastq.gz\n",
      "Approx 95% complete for SRR26691091.fastq.gz\n",
      "Approx 90% complete for SRR26691087.fastq.gz\n",
      "Approx 95% complete for SRR26691092.fastq.gz\n",
      "Analysis complete for SRR26691091.fastq.gz\n",
      "Approx 80% complete for SRR26691093.fastq.gz\n",
      "Approx 95% complete for SRR26691087.fastq.gz\n",
      "Analysis complete for SRR26691092.fastq.gz\n",
      "Analysis complete for SRR26691087.fastq.gz\n",
      "Approx 85% complete for SRR26691093.fastq.gz\n",
      "Approx 90% complete for SRR26691093.fastq.gz\n",
      "Approx 95% complete for SRR26691093.fastq.gz\n",
      "Analysis complete for SRR26691093.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "!mkdir fastq/qc\n",
    "!fastqc -t 16 fastq/*.fastq.gz -o fastq/qc"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***1 minute time to completion***\n",
    "\n",
    "Let's go through each section one-by-one.\n",
    "\n",
    "### Per base sequence quality\n",
    "\n",
    "Base quality scores are represented in Phred+33 encoding. Remember that Q20 is a popular threshold, as the likelihood of an error substantially increases as the Phred quality gets less than 20. It is typical to see high base quality at the 5' end of reads and lower quality at the 3' ends. For this reason, sometimes the 3' end of reads may be trimmed to remove low quality bases.\n",
    "\n",
    "### Per tile sequence quality\n",
    "\n",
    "Tiles are physical locations on the flow cell. Warnings in this section can be associated with flow cells that have been overloaded.\n",
    "\n",
    "### Per sequence quality scores\n",
    "\n",
    "This distribution shows that average Phred quality for entire reads. It helps you make decisions about removing entire reads based on their average quality.\n",
    "\n",
    "### Per base sequence content\n",
    "\n",
    "This distribution shows the bases most likely to be found at any given position in the reads. In general, the likelihood of finding any of A, T, G, or C at a given position should be the same across the entire sequence. Deviations from this expectation may be due to the presence of barcodes, adaptors, UMIs, UMI linkers, or PolyA tails. Here's an example of a dataset that has the UMI linker on the 5' end.\n",
    "\n",
    "<img src=\"assets/linker.png\" alt=\"UMI linker\" width=\"450\">\n",
    "\n",
    "### Per sequence GC content\n",
    "\n",
    "Since our reads are derived from mRNA fragments, we would expect that the percentage of Gs and Cs would be normally distributed in a way that mimics the underlying transcriptome of the organism from which the mRNA was extracted. If the distribution isn't normal (that is, if there are sharp peaks), that is an indication of some sort of contamination. Here is an example of a sharp peak:\n",
    "\n",
    "<img src=\"assets/gc_peak.png\" alt=\"GC\" width=\"450\">\n",
    "\n",
    "This peak is indicative of something in the library that has a higher GC% than the rest of the mRNA. Sharp peaks are often associated with adaptor dimers. These are reads generated from two adaptors that were ligated together during library prep - they have the sequencing primer binding site, but they don't have an insert. Here's the adaptor sequence used for this library, with Gs and Cs annotated:\n",
    "\n",
    "```\n",
    "AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC = 34\n",
    "x  xxx  x xx x xx x x  x xx x x x = 18\n",
    "18 / 34 = 53%\n",
    "```\n",
    "\n",
    "So, it would make sense that if many reads are associated with adaptor dimers, these reads would have a significantly higher GC% than average.\n",
    "\n",
    "### Per base N content\n",
    "\n",
    "Ns are called when the base caller can't make a decision for that cycle. N calls are usually associated (and correlated) with per base sequence quality.\n",
    "\n",
    "### Sequence length distribution\n",
    "\n",
    "Some sequencing libraries will have non-uniform read lengths, in which case we'd see a distribution of different lenghts. For runs that have a pre-defined read lenght, we should see a sharp peak at that length.\n",
    "\n",
    "### Sequence duplication levels\n",
    "\n",
    "For this module, FastQC will scan the first 50 bp of the first 100,000 reads and tally duplicates. The best libraries will have almost no, if any, duplicates. High levels of duplication may be indicative of PCR over amplification. Sometimes over amplification is a desired outcome; in this case, highly expressed transcripts will be highly duplicated, but this is a tradeoff for recovering more lowly experssed transcripts. FASTQs can be deduplicated before or after the alignment step.\n",
    "\n",
    "### Overrepresented sequences\n",
    "\n",
    "If you have many duplicate reads, it's likely there are overrepresented sequences. You can also use this list of sequences to help evalulate if you have adaptor dimers. A simple way to check for problems is to search a sequence database for the overrepresented sequence. If it hits to a sequences deriving from your organism of interest, then you don't have a problem and this overrepresentation may be indicative of true underlying biology. The most useful search tool is BLAST. You can use [this website](https://blast.ncbi.nlm.nih.gov/Blast.cgi) to BLAST the sequences against the NCBI's entire nucleotide database.\n",
    "\n",
    "### Adapter content\n",
    "\n",
    "This module searches reads for popular adaptors and PolyA stretches. If you performed mRNA enrichment, you may see higher PolyA instances at either end of the reads. \n",
    "\n",
    "## FastQC evaluation\n",
    "\n",
    "To properly evaluate these QC data, we need to consider how the library was prepared. This consideration will give us an idea of what (if any) adaptors we may expect to see, how much sequence duplication might be present, and if PolyA tails are expected. From the [paper](https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1012268#sec010), we know that they used the QuantSeq 3' mRNA FWD V2 library prep kit with UMI Second Strand Synthesis and 6 nt Unique Dual Indexing (UDI). You can read a bit about the kit [here](https://www.lexogen.com/store/quantseq-3-mrna-seq-v2-fwd-with-udi/). This kit is optimized for a few things:\n",
    "\n",
    "- Degraded mRNA - it will only get sequences from the 3' end (and PolyA tails)\n",
    "- Low input RNA - which is often the case for small eukaryotic parasites, especially when working with eggs, which are difficult to get RNA out of\n",
    "\n",
    "Remember how library prep works. When doing PolyA selection, an oligo(dT) primer will be used for first-strand cDNA synthesis and a random primer will be used for second strand synthesis. The library is then amplified, during which barcodes and P5/P7 are added. Here's the schematic from Lexogen:\n",
    "\n",
    "<img src=\"assets/library_prep.png\" alt=\"library prep\" width=\"600\">\n",
    "\n",
    "The QuantSeq 3' mRNA FWD V2 library also allows for incorporation of unique molecular identifiers (UMIs) during second strand synthesis. This UMI will be present during the amplification step, which means that each read then has the UMI tag that informs which *original fragment* the read was derived from. The UMI is 6 bp long and has TATA immediately after it, prior to the insert. This tells us how many duplicated reads are present in the data. Here's what that looks like:\n",
    "\n",
    "<img src=\"assets/umi.png\" alt=\"UMI\" width=\"450\">\n",
    "\n",
    "Using this information, we can make the following inferences:\n",
    "\n",
    "- We are likely to have PolyA tails\n",
    "- We may see UMIs and UMI linkers (TATA)\n",
    "- Because the RNA was low-input, some duplicated sequences are to be expected.\n",
    "\n",
    "Take some time to look at all the FastQC files and see if there are any other problems that should be noted."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MultiQC\n",
    "\n",
    "When you have many FASTQ files, all the relevant QC can be difficult to manage. MultiQC is designed to coalesce QC metrics for many different bioinformatics datatypes.\n",
    "\n",
    "Check the manual first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!multiqc -h"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MultiQC is very easy to use. All you have to do is run the command followed by the path to the working directory. It will discover all relevant QC files and wrap them up into a very helpful report. Run it now, then take a look at `multiqc_report.html` (you'll have to download it and open it in your browser)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=796612;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.17\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/corwinbm5021/BIOL343/5_fastq\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m45/45\u001b[0m  2m./fastq/qc/SRR26691089_fastqc.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 12 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "!multiqc ."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FASTQ trimming and filtering\n",
    "\n",
    "QC metrics will inform whether or not reads should be trimmed (i.e., PolyA tails, low quality bases, or adaptors removed) or filtered (low quality reads removed). Typically, manufacturers will provide some guidance on how to trim and filter sequences produced from their kit. For Lexogen (the maker of the QuantSeq kit), this information can be found in the [FAQs](https://faqs.lexogen.com/faq/what-sequences-should-be-trimmed). This site shows that `AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC` is the adapter sequence. Is that sequence (or one like it) in the \"Overrepresented sequences\" section? If so, we need to trim the adapter. If not, we just need to trim the PolyA tail.\n",
    "\n",
    "Here's what the authors of the paper said about read QC:\n",
    "\n",
    "> Quality control of raw reads was performed in FastQC (v0.11.9). Reads were mapped to the S. mansoni genome (Wormbase ParaSite v9) using STAR aligner (version 2.7.10a) with options adjusted to the QuantSeq FWD 3mRNA Library Prep Kit. The UMIs were deduplicated using open source UMI-tools software package (version 1.1.2).\n",
    "\n",
    "Although they did look at QC, they say nothing about trimming or filtering. They do mention that they deduplicated reads using UMI-tools. We won't be able to reproduce this because the reads that we fetched from NCBI SRA don't all have UMIs, some only have the UMI linker (TATA).\n",
    "\n",
    "For QuantSeq libraries, each read may include a 6 nt UMI, a 4 nt (TATA) linker, the insert, the PolyA tail, and then the adapter. Based on our QC data, it looks like the adapters have been trimmed (`AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC` isn't overrepresented). The PolyA tail *may* have been trimmed, but it's not obvious - we still see up to 40% PolyA at position 59 in SRR26691093. On the other side of the read, we see TATA in all 4 files, and SRR26691093 and SRR26691087 contain the 6 nt UMI while SRR26691085 and SRR26691082 do not.\n",
    "\n",
    "We definitely want to trim PolyA tails, but trimming UDIs and linkers actually depends on the type of aligner being used. STAR (the aligner we'll be using) can soft-clip the ends of reads that have a high mismatch rate. This also means that we don't need to trim low-quality bases, because they can also be soft-clipped. What about the duplicates? Again, those can be marked during alignment, so we don't need to remove them either.\n",
    "\n",
    "In the end, we'll just trim PolyA tails and do some trimming that's particular to the NextSeq instrument that we're not going to get into. We will use the tool [cutadapt](https://cutadapt.readthedocs.io/en/stable/) to perform the trimming. First, check out the manual:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cutadapt version 4.9\n",
      "\n",
      "Copyright (C) 2010 Marcel Martin <marcel.martin@scilifelab.se> and contributors\n",
      "\n",
      "Cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. All reads from input.fastq will be written to\n",
      "output.fastq with the adapter sequence removed. Adapter matching is\n",
      "error-tolerant. Multiple adapter sequences can be given (use further -a\n",
      "options), but only the best-matching adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Run \"cutadapt --help\" to see all command-line options.\n",
      "See https://cutadapt.readthedocs.io/ for full documentation.\n",
      "\n",
      "Options:\n",
      "  -h, --help            Show this help message and exit\n",
      "  --version             Show version number and exit\n",
      "  --debug               Print debug log. Use twice to also print DP matrices\n",
      "  -j CORES, --cores CORES\n",
      "                        Number of CPU cores to use. Use 0 to auto-detect.\n",
      "                        Default: 1\n",
      "\n",
      "Finding adapters:\n",
      "  Parameters -a, -g, -b specify adapters to be removed from each read (or from\n",
      "  R1 if data is paired-end. If specified multiple times, only the best\n",
      "  matching adapter is trimmed (but see the --times option). Use notation\n",
      "  'file:FILE' to read adapter sequences from a FASTA file.\n",
      "\n",
      "  -a ADAPTER, --adapter ADAPTER\n",
      "                        Sequence of an adapter ligated to the 3' end (paired\n",
      "                        data: of the first read). The adapter and subsequent\n",
      "                        bases are trimmed. If a '$' character is appended\n",
      "                        ('anchoring'), the adapter is only found if it is a\n",
      "                        suffix of the read.\n",
      "  -g ADAPTER, --front ADAPTER\n",
      "                        Sequence of an adapter ligated to the 5' end (paired\n",
      "                        data: of the first read). The adapter and any preceding\n",
      "                        bases are trimmed. Partial matches at the 5' end are\n",
      "                        allowed. If a '^' character is prepended ('anchoring'),\n",
      "                        the adapter is only found if it is a prefix of the read.\n",
      "  -b ADAPTER, --anywhere ADAPTER\n",
      "                        Sequence of an adapter that may be ligated to the 5' or\n",
      "                        3' end (paired data: of the first read). Both types of\n",
      "                        matches as described under -a and -g are allowed. If the\n",
      "                        first base of the read is part of the match, the\n",
      "                        behavior is as with -g, otherwise as with -a. This\n",
      "                        option is mostly for rescuing failed library\n",
      "                        preparations - do not use if you know which end your\n",
      "                        adapter was ligated to!\n",
      "  -e E, --error-rate E, --errors E\n",
      "                        Maximum allowed error rate (if 0 <= E < 1), or absolute\n",
      "                        number of errors for full-length adapter match (if E is\n",
      "                        an integer >= 1). Error rate = no. of errors divided by\n",
      "                        length of matching region. Default: 0.1 (10%)\n",
      "  --no-indels           Allow only mismatches in alignments. Default: allow both\n",
      "                        mismatches and indels\n",
      "  -n COUNT, --times COUNT\n",
      "                        Remove up to COUNT adapters from each read. Default: 1\n",
      "  -O MINLENGTH, --overlap MINLENGTH\n",
      "                        Require MINLENGTH overlap between read and adapter for\n",
      "                        an adapter to be found. Default: 3\n",
      "  --match-read-wildcards\n",
      "                        Interpret IUPAC wildcards in reads. Default: False\n",
      "  -N, --no-match-adapter-wildcards\n",
      "                        Do not interpret IUPAC wildcards in adapters.\n",
      "  --action {trim,retain,mask,lowercase,crop,none}\n",
      "                        What to do if a match was found. trim: trim adapter and\n",
      "                        up- or downstream sequence; retain: trim, but retain\n",
      "                        adapter; mask: replace with 'N' characters; lowercase:\n",
      "                        convert to lowercase; crop: trim up and downstream\n",
      "                        sequence; none: leave unchanged. Default: trim\n",
      "  --rc, --revcomp       Check both the read and its reverse complement for\n",
      "                        adapter matches. If match is on reverse-complemented\n",
      "                        version, output that one. Default: check only read\n",
      "\n",
      "Additional read modifications:\n",
      "  -u LEN, --cut LEN     Remove LEN bases from each read (or R1 if paired; use -U\n",
      "                        option for R2). If LEN is positive, remove bases from\n",
      "                        the beginning. If LEN is negative, remove bases from the\n",
      "                        end. Can be used twice if LENs have different signs.\n",
      "                        Applied *before* adapter trimming.\n",
      "  --nextseq-trim 3'CUTOFF\n",
      "                        NextSeq-specific quality trimming (each read). Trims\n",
      "                        also dark cycles appearing as high-quality G bases.\n",
      "  -q [5'CUTOFF,]3'CUTOFF, --quality-cutoff [5'CUTOFF,]3'CUTOFF\n",
      "                        Trim low-quality bases from 5' and/or 3' ends of each\n",
      "                        read before adapter removal. Applied to both reads if\n",
      "                        data is paired. If one value is given, only the 3' end\n",
      "                        is trimmed. If two comma-separated cutoffs are given,\n",
      "                        the 5' end is trimmed with the first cutoff, the 3' end\n",
      "                        with the second.\n",
      "  --quality-base N      Assume that quality values in FASTQ are encoded as\n",
      "                        ascii(quality + N). This needs to be set to 64 for some\n",
      "                        old Illumina FASTQ files. Default: 33\n",
      "  --poly-a              Trim poly-A tails\n",
      "  --length LENGTH, -l LENGTH\n",
      "                        Shorten reads to LENGTH. Positive values remove bases at\n",
      "                        the end while negative ones remove bases at the\n",
      "                        beginning. This and the following modifications are\n",
      "                        applied after adapter trimming.\n",
      "  --trim-n              Trim N's on ends of reads.\n",
      "  --length-tag TAG      Search for TAG followed by a decimal number in the\n",
      "                        description field of the read. Replace the decimal\n",
      "                        number with the correct length of the trimmed read. For\n",
      "                        example, use --length-tag 'length=' to correct fields\n",
      "                        like 'length=123'.\n",
      "  --strip-suffix STRIP_SUFFIX\n",
      "                        Remove this suffix from read names if present. Can be\n",
      "                        given multiple times.\n",
      "  -x PREFIX, --prefix PREFIX\n",
      "                        Add this prefix to read names. Use {name} to insert the\n",
      "                        name of the matching adapter.\n",
      "  -y SUFFIX, --suffix SUFFIX\n",
      "                        Add this suffix to read names; can also include {name}\n",
      "  --rename TEMPLATE     Rename reads using TEMPLATE containing variables such as\n",
      "                        {id}, {adapter_name} etc. (see documentation)\n",
      "  --zero-cap, -z        Change negative quality values to zero.\n",
      "\n",
      "Filtering of processed reads:\n",
      "  Filters are applied after above read modifications. Paired-end reads are\n",
      "  always discarded pairwise (see also --pair-filter).\n",
      "\n",
      "  -m LEN[:LEN2], --minimum-length LEN[:LEN2]\n",
      "                        Discard reads shorter than LEN. Default: 0\n",
      "  -M LEN[:LEN2], --maximum-length LEN[:LEN2]\n",
      "                        Discard reads longer than LEN. Default: no limit\n",
      "  --max-n COUNT         Discard reads with more than COUNT 'N' bases. If COUNT\n",
      "                        is a number between 0 and 1, it is interpreted as a\n",
      "                        fraction of the read length.\n",
      "  --max-expected-errors ERRORS, --max-ee ERRORS\n",
      "                        Discard reads whose expected number of errors (computed\n",
      "                        from quality values) exceeds ERRORS.\n",
      "  --max-average-error-rate ERROR_RATE, --max-aer ERROR_RATE\n",
      "                        as --max-expected-errors (see above), but divided by\n",
      "                        length to account for reads of varying length.\n",
      "  --discard-trimmed, --discard\n",
      "                        Discard reads that contain an adapter. Use also -O to\n",
      "                        avoid discarding too many randomly matching reads.\n",
      "  --discard-untrimmed, --trimmed-only\n",
      "                        Discard reads that do not contain an adapter.\n",
      "  --discard-casava      Discard reads that did not pass CASAVA filtering (header\n",
      "                        has :Y:).\n",
      "\n",
      "Output:\n",
      "  --quiet               Print only error messages.\n",
      "  --report {full,minimal}\n",
      "                        Which type of report to print: 'full' or 'minimal'.\n",
      "                        Default: full\n",
      "  --json FILE           Dump report in JSON format to FILE\n",
      "  -o FILE, --output FILE\n",
      "                        Write trimmed reads to FILE. FASTQ or FASTA format is\n",
      "                        chosen depending on input. Summary report is sent to\n",
      "                        standard output. Use '{name}' for demultiplexing (see\n",
      "                        docs). Default: write to standard output\n",
      "  --fasta               Output FASTA to standard output even on FASTQ input.\n",
      "  -Z                    Use compression level 1 for gzipped output files\n",
      "                        (faster, but uses more space)\n",
      "  --info-file FILE      Write information about each read and its adapter\n",
      "                        matches into FILE. See the documentation for the file\n",
      "                        format.\n",
      "  -r FILE, --rest-file FILE\n",
      "                        When the adapter matches in the middle of a read, write\n",
      "                        the rest (after the adapter) to FILE.\n",
      "  --wildcard-file FILE  When the adapter has N wildcard bases, write adapter\n",
      "                        bases matching wildcard positions to FILE. (Inaccurate\n",
      "                        with indels.)\n",
      "  --too-short-output FILE\n",
      "                        Write reads that are too short (according to length\n",
      "                        specified by -m) to FILE. Default: discard reads\n",
      "  --too-long-output FILE\n",
      "                        Write reads that are too long (according to length\n",
      "                        specified by -M) to FILE. Default: discard reads\n",
      "  --untrimmed-output FILE\n",
      "                        Write reads that do not contain any adapter to FILE.\n",
      "                        Default: output to same file as trimmed reads\n",
      "\n",
      "Paired-end options:\n",
      "  The -A/-G/-B/-U/-Q options work like their lowercase counterparts, but are\n",
      "  applied to R2 (second read in pair)\n",
      "\n",
      "  -A ADAPTER            3' adapter to be removed from R2\n",
      "  -G ADAPTER            5' adapter to be removed from R2\n",
      "  -B ADAPTER            5'/3 adapter to be removed from R2\n",
      "  -U LENGTH             Remove LENGTH bases from R2\n",
      "  -Q [5'CUTOFF,]3'CUTOFF\n",
      "                        Quality-trimming cutoff for R2. Default: same as for R1\n",
      "  -L LENGTH             Shorten R2 to LENGTH. Default: same as for R1\n",
      "  -p FILE, --paired-output FILE\n",
      "                        Write R2 to FILE.\n",
      "  --pair-adapters       Treat adapters given with -a/-A etc. as pairs. Either\n",
      "                        both or none are removed from each read pair.\n",
      "  --pair-filter {any,both,first}\n",
      "                        Which of the reads in a paired-end read have to match\n",
      "                        the filtering criterion in order for the pair to be\n",
      "                        filtered. Default: any\n",
      "  --interleaved         Read and/or write interleaved paired-end reads.\n",
      "  --untrimmed-paired-output FILE\n",
      "                        Write second read in a pair to this FILE when no adapter\n",
      "                        was found. Use with --untrimmed-output. Default: output\n",
      "                        to same file as trimmed reads\n",
      "  --too-short-paired-output FILE\n",
      "                        Write second read in a pair to this file if pair is too\n",
      "                        short.\n",
      "  --too-long-paired-output FILE\n",
      "                        Write second read in a pair to this file if pair is too\n",
      "                        long.\n"
     ]
    }
   ],
   "source": [
    "!cutadapt --help"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Edit this block and define the result of each of flags used in the above commands.\n",
    "\n",
    "`-j`: \n",
    "\n",
    "`-m`: \n",
    "\n",
    "`-O`:\n",
    "\n",
    "`-a`:\n",
    "\n",
    "`-n`:"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's trim and run QC on these new files to see how they look. We'll use a bash loop again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691082.fastq.gz fastq/SRR26691082.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:12    10,039,919 reads @   1.2 s/read;  49.11 M reads/minute\n",
      "Finished in 12.337 s (1.229 s/read; 48.83 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              10,039,919\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:              60,645 (0.6%)\n",
      "Reads written (passing filters):     9,979,274 (99.4%)\n",
      "\n",
      "Total basepairs processed:   692,754,411 bp\n",
      "Quality-trimmed:               1,674,822 bp (0.2%)\n",
      "Poly-A-trimmed:               27,670,085 bp (4.0%)\n",
      "Total written (filtered):    662,590,840 bp (95.6%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t7854189\n",
      "3\t308134\n",
      "4\t155349\n",
      "5\t215901\n",
      "6\t171737\n",
      "7\t124653\n",
      "8\t98316\n",
      "9\t89926\n",
      "10\t85678\n",
      "11\t67989\n",
      "12\t86254\n",
      "13\t66030\n",
      "14\t50981\n",
      "15\t51620\n",
      "16\t42716\n",
      "17\t40238\n",
      "18\t39007\n",
      "19\t37771\n",
      "20\t36185\n",
      "21\t46458\n",
      "22\t69529\n",
      "23\t36987\n",
      "24\t18668\n",
      "25\t18581\n",
      "26\t24193\n",
      "27\t24867\n",
      "28\t18301\n",
      "29\t11383\n",
      "30\t11335\n",
      "31\t7783\n",
      "32\t5493\n",
      "33\t5292\n",
      "34\t6469\n",
      "35\t4616\n",
      "36\t4139\n",
      "37\t3205\n",
      "38\t3210\n",
      "39\t2958\n",
      "40\t2729\n",
      "41\t2925\n",
      "42\t2442\n",
      "43\t2659\n",
      "44\t2547\n",
      "45\t2949\n",
      "46\t3145\n",
      "47\t4189\n",
      "48\t5012\n",
      "49\t6952\n",
      "50\t7813\n",
      "51\t6215\n",
      "52\t7560\n",
      "53\t6576\n",
      "54\t5049\n",
      "55\t4123\n",
      "56\t2977\n",
      "57\t3001\n",
      "58\t2181\n",
      "59\t2471\n",
      "60\t1909\n",
      "61\t1802\n",
      "62\t1294\n",
      "63\t1226\n",
      "64\t843\n",
      "65\t301\n",
      "66\t2510\n",
      "67\t57\n",
      "68\t168\n",
      "69\t153\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691083.fastq.gz fastq/SRR26691083.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:10     8,797,795 reads @   1.2 s/read;  48.35 M reads/minute\n",
      "Finished in 10.995 s (1.250 s/read; 48.01 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               8,797,795\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:              78,257 (0.9%)\n",
      "Reads written (passing filters):     8,719,538 (99.1%)\n",
      "\n",
      "Total basepairs processed:   607,047,855 bp\n",
      "Quality-trimmed:               1,837,069 bp (0.3%)\n",
      "Poly-A-trimmed:               19,267,321 bp (3.2%)\n",
      "Total written (filtered):    584,908,127 bp (96.4%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t7433179\n",
      "3\t210395\n",
      "4\t95919\n",
      "5\t141164\n",
      "6\t83701\n",
      "7\t57554\n",
      "8\t58383\n",
      "9\t83804\n",
      "10\t59391\n",
      "11\t40167\n",
      "12\t47395\n",
      "13\t31645\n",
      "14\t26349\n",
      "15\t33350\n",
      "16\t22319\n",
      "17\t22257\n",
      "18\t23049\n",
      "19\t22523\n",
      "20\t19225\n",
      "21\t19729\n",
      "22\t20860\n",
      "23\t18589\n",
      "24\t10177\n",
      "25\t10566\n",
      "26\t13891\n",
      "27\t13929\n",
      "28\t9175\n",
      "29\t7783\n",
      "30\t7731\n",
      "31\t6476\n",
      "32\t4294\n",
      "33\t4601\n",
      "34\t5092\n",
      "35\t3902\n",
      "36\t3977\n",
      "37\t3475\n",
      "38\t3834\n",
      "39\t2903\n",
      "40\t2679\n",
      "41\t3101\n",
      "42\t2696\n",
      "43\t2851\n",
      "44\t2651\n",
      "45\t3384\n",
      "46\t3578\n",
      "47\t5402\n",
      "48\t5985\n",
      "49\t6917\n",
      "50\t9317\n",
      "51\t8213\n",
      "52\t9428\n",
      "53\t7684\n",
      "54\t5904\n",
      "55\t4927\n",
      "56\t3784\n",
      "57\t4143\n",
      "58\t3203\n",
      "59\t3599\n",
      "60\t3319\n",
      "61\t2956\n",
      "62\t2166\n",
      "63\t1655\n",
      "64\t1253\n",
      "65\t465\n",
      "66\t3228\n",
      "67\t98\n",
      "68\t300\n",
      "69\t156\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691084.fastq.gz fastq/SRR26691084.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:10     8,483,494 reads @   1.2 s/read;  49.52 M reads/minute\n",
      "Finished in 10.359 s (1.221 s/read; 49.13 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               8,483,494\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:              45,423 (0.5%)\n",
      "Reads written (passing filters):     8,438,071 (99.5%)\n",
      "\n",
      "Total basepairs processed:   585,361,086 bp\n",
      "Quality-trimmed:               1,829,736 bp (0.3%)\n",
      "Poly-A-trimmed:               15,725,362 bp (2.7%)\n",
      "Total written (filtered):    567,217,784 bp (96.9%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t7235314\n",
      "3\t205984\n",
      "4\t92733\n",
      "5\t128701\n",
      "6\t81991\n",
      "7\t56234\n",
      "8\t60192\n",
      "9\t80898\n",
      "10\t53643\n",
      "11\t36826\n",
      "12\t35274\n",
      "13\t29725\n",
      "14\t25434\n",
      "15\t34482\n",
      "16\t22152\n",
      "17\t21219\n",
      "18\t22577\n",
      "19\t21599\n",
      "20\t19015\n",
      "21\t20375\n",
      "22\t21520\n",
      "23\t19035\n",
      "24\t9357\n",
      "25\t9088\n",
      "26\t12232\n",
      "27\t12406\n",
      "28\t7679\n",
      "29\t6478\n",
      "30\t6292\n",
      "31\t5247\n",
      "32\t3421\n",
      "33\t3554\n",
      "34\t3802\n",
      "35\t3016\n",
      "36\t2673\n",
      "37\t2169\n",
      "38\t2116\n",
      "39\t1997\n",
      "40\t1900\n",
      "41\t2033\n",
      "42\t1689\n",
      "43\t1799\n",
      "44\t1608\n",
      "45\t2124\n",
      "46\t2094\n",
      "47\t3263\n",
      "48\t3578\n",
      "49\t4233\n",
      "50\t5709\n",
      "51\t4380\n",
      "52\t5093\n",
      "53\t4682\n",
      "54\t3529\n",
      "55\t2811\n",
      "56\t2090\n",
      "57\t2230\n",
      "58\t1675\n",
      "59\t1871\n",
      "60\t1700\n",
      "61\t1508\n",
      "62\t1118\n",
      "63\t970\n",
      "64\t678\n",
      "65\t230\n",
      "66\t2164\n",
      "67\t42\n",
      "68\t172\n",
      "69\t71\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691085.fastq.gz fastq/SRR26691085.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:10     8,265,359 reads @   1.2 s/read;  48.04 M reads/minute\n",
      "Finished in 10.399 s (1.258 s/read; 47.69 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               8,265,359\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:               6,882 (0.1%)\n",
      "Reads written (passing filters):     8,258,477 (99.9%)\n",
      "\n",
      "Total basepairs processed:   570,309,771 bp\n",
      "Quality-trimmed:               1,909,807 bp (0.3%)\n",
      "Poly-A-trimmed:                8,410,771 bp (1.5%)\n",
      "Total written (filtered):    559,923,242 bp (98.2%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t7378454\n",
      "3\t198826\n",
      "4\t83431\n",
      "5\t112265\n",
      "6\t66255\n",
      "7\t40861\n",
      "8\t35864\n",
      "9\t45627\n",
      "10\t36648\n",
      "11\t24316\n",
      "12\t30456\n",
      "13\t17261\n",
      "14\t14834\n",
      "15\t20142\n",
      "16\t12811\n",
      "17\t12725\n",
      "18\t13324\n",
      "19\t12511\n",
      "20\t10825\n",
      "21\t14260\n",
      "22\t15296\n",
      "23\t11583\n",
      "24\t5014\n",
      "25\t5210\n",
      "26\t8087\n",
      "27\t8484\n",
      "28\t4829\n",
      "29\t3306\n",
      "30\t3543\n",
      "31\t2493\n",
      "32\t1340\n",
      "33\t1514\n",
      "34\t1399\n",
      "35\t1018\n",
      "36\t713\n",
      "37\t530\n",
      "38\t453\n",
      "39\t360\n",
      "40\t401\n",
      "41\t506\n",
      "42\t282\n",
      "43\t316\n",
      "44\t289\n",
      "45\t298\n",
      "46\t330\n",
      "47\t421\n",
      "48\t432\n",
      "49\t533\n",
      "50\t788\n",
      "51\t497\n",
      "52\t573\n",
      "53\t389\n",
      "54\t305\n",
      "55\t241\n",
      "56\t229\n",
      "57\t229\n",
      "58\t146\n",
      "59\t190\n",
      "60\t183\n",
      "61\t220\n",
      "62\t127\n",
      "63\t125\n",
      "64\t100\n",
      "65\t68\n",
      "66\t252\n",
      "67\t2\n",
      "68\t13\n",
      "69\t6\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691086.fastq.gz fastq/SRR26691086.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:07     6,872,464 reads @   1.1 s/read;  52.57 M reads/minute\n",
      "Finished in 7.918 s (1.152 s/read; 52.08 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               6,872,464\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:               6,163 (0.1%)\n",
      "Reads written (passing filters):     6,866,301 (99.9%)\n",
      "\n",
      "Total basepairs processed:   481,072,480 bp\n",
      "Quality-trimmed:               2,801,751 bp (0.6%)\n",
      "Poly-A-trimmed:               16,004,817 bp (3.3%)\n",
      "Total written (filtered):    462,211,334 bp (96.1%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t5626589\n",
      "3\t136790\n",
      "4\t129856\n",
      "5\t94248\n",
      "6\t65376\n",
      "7\t66694\n",
      "8\t40164\n",
      "9\t41709\n",
      "10\t55962\n",
      "11\t32577\n",
      "12\t33301\n",
      "13\t42417\n",
      "14\t32119\n",
      "15\t28012\n",
      "16\t30337\n",
      "17\t28592\n",
      "18\t27308\n",
      "19\t31752\n",
      "20\t27517\n",
      "21\t40734\n",
      "22\t41556\n",
      "23\t50902\n",
      "24\t23456\n",
      "25\t31403\n",
      "26\t40821\n",
      "27\t23702\n",
      "28\t14496\n",
      "29\t6527\n",
      "30\t3285\n",
      "31\t1765\n",
      "32\t1197\n",
      "33\t1048\n",
      "34\t957\n",
      "35\t1284\n",
      "36\t1381\n",
      "37\t894\n",
      "38\t1061\n",
      "39\t741\n",
      "40\t761\n",
      "41\t664\n",
      "42\t1141\n",
      "43\t812\n",
      "44\t1136\n",
      "45\t2290\n",
      "46\t1002\n",
      "47\t932\n",
      "48\t955\n",
      "49\t823\n",
      "50\t648\n",
      "51\t508\n",
      "52\t576\n",
      "53\t390\n",
      "54\t293\n",
      "55\t298\n",
      "56\t184\n",
      "57\t134\n",
      "58\t92\n",
      "59\t70\n",
      "60\t49\n",
      "61\t143\n",
      "62\t6\n",
      "63\t11\n",
      "65\t1\n",
      "67\t4\n",
      "69\t2\n",
      "70\t9\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691087.fastq.gz fastq/SRR26691087.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:13    11,243,871 reads @   1.2 s/read;  50.54 M reads/minute\n",
      "Finished in 13.411 s (1.193 s/read; 50.30 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              11,243,871\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:               8,994 (0.1%)\n",
      "Reads written (passing filters):    11,234,877 (99.9%)\n",
      "\n",
      "Total basepairs processed:   787,070,970 bp\n",
      "Quality-trimmed:               3,672,248 bp (0.5%)\n",
      "Poly-A-trimmed:               27,545,287 bp (3.5%)\n",
      "Total written (filtered):    755,772,076 bp (96.0%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t9000207\n",
      "3\t281562\n",
      "4\t207048\n",
      "5\t176305\n",
      "6\t129387\n",
      "7\t129829\n",
      "8\t77621\n",
      "9\t72672\n",
      "10\t101284\n",
      "11\t62078\n",
      "12\t60145\n",
      "13\t68062\n",
      "14\t59097\n",
      "15\t53896\n",
      "16\t62038\n",
      "17\t60846\n",
      "18\t60930\n",
      "19\t68538\n",
      "20\t56352\n",
      "21\t76500\n",
      "22\t72898\n",
      "23\t86719\n",
      "24\t34344\n",
      "25\t39677\n",
      "26\t47266\n",
      "27\t28940\n",
      "28\t15529\n",
      "29\t7482\n",
      "30\t5341\n",
      "31\t3410\n",
      "32\t2438\n",
      "33\t2097\n",
      "34\t1856\n",
      "35\t2260\n",
      "36\t2130\n",
      "37\t1447\n",
      "38\t1442\n",
      "39\t1153\n",
      "40\t1409\n",
      "41\t1303\n",
      "42\t1672\n",
      "43\t1536\n",
      "44\t1856\n",
      "45\t3033\n",
      "46\t1879\n",
      "47\t1825\n",
      "48\t1947\n",
      "49\t1476\n",
      "50\t1322\n",
      "51\t954\n",
      "52\t858\n",
      "53\t480\n",
      "54\t372\n",
      "55\t396\n",
      "56\t202\n",
      "57\t142\n",
      "58\t110\n",
      "59\t75\n",
      "60\t22\n",
      "61\t122\n",
      "62\t15\n",
      "63\t17\n",
      "64\t8\n",
      "65\t3\n",
      "66\t1\n",
      "67\t1\n",
      "68\t2\n",
      "69\t5\n",
      "70\t2\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691088.fastq.gz fastq/SRR26691088.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:11     9,968,107 reads @   1.2 s/read;  51.33 M reads/minute\n",
      "Finished in 11.722 s (1.176 s/read; 51.02 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               9,968,107\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:              17,077 (0.2%)\n",
      "Reads written (passing filters):     9,951,030 (99.8%)\n",
      "\n",
      "Total basepairs processed:   697,767,490 bp\n",
      "Quality-trimmed:               4,049,453 bp (0.6%)\n",
      "Poly-A-trimmed:               25,207,517 bp (3.6%)\n",
      "Total written (filtered):    668,297,037 bp (95.8%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t8011883\n",
      "3\t238966\n",
      "4\t165418\n",
      "5\t154786\n",
      "6\t116704\n",
      "7\t122902\n",
      "8\t67643\n",
      "9\t61558\n",
      "10\t86701\n",
      "11\t54868\n",
      "12\t54871\n",
      "13\t61324\n",
      "14\t53653\n",
      "15\t50222\n",
      "16\t52157\n",
      "17\t57109\n",
      "18\t53303\n",
      "19\t56639\n",
      "20\t45544\n",
      "21\t57114\n",
      "22\t54741\n",
      "23\t60908\n",
      "24\t27441\n",
      "25\t30781\n",
      "26\t37367\n",
      "27\t22470\n",
      "28\t12682\n",
      "29\t7528\n",
      "30\t6050\n",
      "31\t3899\n",
      "32\t3270\n",
      "33\t2847\n",
      "34\t2725\n",
      "35\t3433\n",
      "36\t3549\n",
      "37\t2421\n",
      "38\t2755\n",
      "39\t2348\n",
      "40\t2668\n",
      "41\t2705\n",
      "42\t4022\n",
      "43\t3649\n",
      "44\t4505\n",
      "45\t8567\n",
      "46\t4496\n",
      "47\t4420\n",
      "48\t4408\n",
      "49\t3508\n",
      "50\t2695\n",
      "51\t2183\n",
      "52\t2384\n",
      "53\t1627\n",
      "54\t1305\n",
      "55\t1059\n",
      "56\t817\n",
      "57\t548\n",
      "58\t482\n",
      "59\t324\n",
      "60\t234\n",
      "61\t739\n",
      "62\t42\n",
      "63\t67\n",
      "64\t14\n",
      "65\t21\n",
      "66\t8\n",
      "67\t4\n",
      "68\t7\n",
      "69\t5\n",
      "70\t14\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691089.fastq.gz fastq/SRR26691089.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:12    11,240,361 reads @   1.1 s/read;  52.74 M reads/minute\n",
      "Finished in 12.859 s (1.144 s/read; 52.45 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              11,240,361\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:              12,141 (0.1%)\n",
      "Reads written (passing filters):    11,228,220 (99.9%)\n",
      "\n",
      "Total basepairs processed:   786,825,270 bp\n",
      "Quality-trimmed:               3,574,237 bp (0.5%)\n",
      "Poly-A-trimmed:               24,311,522 bp (3.1%)\n",
      "Total written (filtered):    758,815,011 bp (96.4%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t9378423\n",
      "3\t162265\n",
      "4\t192581\n",
      "5\t111320\n",
      "6\t94681\n",
      "7\t152404\n",
      "8\t61685\n",
      "9\t60560\n",
      "10\t89281\n",
      "11\t51968\n",
      "12\t51532\n",
      "13\t68602\n",
      "14\t56369\n",
      "15\t41048\n",
      "16\t48614\n",
      "17\t49167\n",
      "18\t45314\n",
      "19\t53285\n",
      "20\t47715\n",
      "21\t66262\n",
      "22\t62961\n",
      "23\t74800\n",
      "24\t31512\n",
      "25\t40775\n",
      "26\t47056\n",
      "27\t25061\n",
      "28\t14118\n",
      "29\t5711\n",
      "30\t4263\n",
      "31\t2674\n",
      "32\t2190\n",
      "33\t1761\n",
      "34\t1575\n",
      "35\t2225\n",
      "36\t2446\n",
      "37\t1394\n",
      "38\t1471\n",
      "39\t1184\n",
      "40\t1506\n",
      "41\t1338\n",
      "42\t2248\n",
      "43\t2033\n",
      "44\t2453\n",
      "45\t5553\n",
      "46\t2648\n",
      "47\t2789\n",
      "48\t2887\n",
      "49\t2216\n",
      "50\t1914\n",
      "51\t1475\n",
      "52\t1572\n",
      "53\t888\n",
      "54\t593\n",
      "55\t465\n",
      "56\t307\n",
      "57\t270\n",
      "58\t192\n",
      "59\t129\n",
      "60\t86\n",
      "61\t319\n",
      "62\t42\n",
      "63\t72\n",
      "64\t8\n",
      "65\t9\n",
      "66\t6\n",
      "67\t10\n",
      "68\t10\n",
      "69\t20\n",
      "70\t50\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691090.fastq.gz fastq/SRR26691090.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:12    10,233,476 reads @   1.2 s/read;  49.33 M reads/minute\n",
      "Finished in 12.520 s (1.223 s/read; 49.04 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              10,233,476\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:             385,836 (3.8%)\n",
      "Reads written (passing filters):     9,847,640 (96.2%)\n",
      "\n",
      "Total basepairs processed:   706,109,844 bp\n",
      "Quality-trimmed:               2,816,605 bp (0.4%)\n",
      "Poly-A-trimmed:               65,051,855 bp (9.2%)\n",
      "Total written (filtered):    633,534,424 bp (89.7%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t7046287\n",
      "3\t327396\n",
      "4\t151352\n",
      "5\t199264\n",
      "6\t141393\n",
      "7\t114469\n",
      "8\t102808\n",
      "9\t94362\n",
      "10\t93777\n",
      "11\t85203\n",
      "12\t88246\n",
      "13\t85377\n",
      "14\t75028\n",
      "15\t76346\n",
      "16\t67801\n",
      "17\t67654\n",
      "18\t70560\n",
      "19\t71792\n",
      "20\t65635\n",
      "21\t76902\n",
      "22\t81220\n",
      "23\t72547\n",
      "24\t38591\n",
      "25\t41451\n",
      "26\t55327\n",
      "27\t57124\n",
      "28\t40798\n",
      "29\t28904\n",
      "30\t29294\n",
      "31\t21539\n",
      "32\t17550\n",
      "33\t16236\n",
      "34\t20140\n",
      "35\t16095\n",
      "36\t15333\n",
      "37\t13782\n",
      "38\t14402\n",
      "39\t12427\n",
      "40\t11758\n",
      "41\t12727\n",
      "42\t11296\n",
      "43\t11495\n",
      "44\t11363\n",
      "45\t14228\n",
      "46\t15266\n",
      "47\t20658\n",
      "48\t23358\n",
      "49\t26391\n",
      "50\t33338\n",
      "51\t33082\n",
      "52\t32048\n",
      "53\t31204\n",
      "54\t27634\n",
      "55\t25225\n",
      "56\t22319\n",
      "57\t23561\n",
      "58\t20951\n",
      "59\t23338\n",
      "60\t18464\n",
      "61\t17226\n",
      "62\t13458\n",
      "63\t12225\n",
      "64\t8921\n",
      "65\t3458\n",
      "66\t29219\n",
      "67\t755\n",
      "68\t2944\n",
      "69\t1154\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691091.fastq.gz fastq/SRR26691091.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:11     9,670,422 reads @   1.2 s/read;  48.59 M reads/minute\n",
      "Finished in 12.015 s (1.242 s/read; 48.29 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               9,670,422\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:              21,326 (0.2%)\n",
      "Reads written (passing filters):     9,649,096 (99.8%)\n",
      "\n",
      "Total basepairs processed:   667,259,118 bp\n",
      "Quality-trimmed:               2,077,890 bp (0.3%)\n",
      "Poly-A-trimmed:               12,984,082 bp (1.9%)\n",
      "Total written (filtered):    651,917,838 bp (97.7%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t8424209\n",
      "3\t264068\n",
      "4\t106083\n",
      "5\t156048\n",
      "6\t95765\n",
      "7\t63203\n",
      "8\t51375\n",
      "9\t49269\n",
      "10\t49142\n",
      "11\t34883\n",
      "12\t44721\n",
      "13\t28965\n",
      "14\t22961\n",
      "15\t27162\n",
      "16\t19332\n",
      "17\t18878\n",
      "18\t18547\n",
      "19\t17728\n",
      "20\t15687\n",
      "21\t19269\n",
      "22\t19405\n",
      "23\t16587\n",
      "24\t7829\n",
      "25\t7735\n",
      "26\t11486\n",
      "27\t11463\n",
      "28\t7378\n",
      "29\t4872\n",
      "30\t5135\n",
      "31\t3897\n",
      "32\t2492\n",
      "33\t2561\n",
      "34\t2686\n",
      "35\t1982\n",
      "36\t1770\n",
      "37\t1310\n",
      "38\t1342\n",
      "39\t1167\n",
      "40\t1059\n",
      "41\t1212\n",
      "42\t955\n",
      "43\t963\n",
      "44\t848\n",
      "45\t1125\n",
      "46\t1213\n",
      "47\t1531\n",
      "48\t1796\n",
      "49\t2286\n",
      "50\t2802\n",
      "51\t2304\n",
      "52\t2261\n",
      "53\t2287\n",
      "54\t1835\n",
      "55\t1514\n",
      "56\t1059\n",
      "57\t1080\n",
      "58\t726\n",
      "59\t805\n",
      "60\t595\n",
      "61\t424\n",
      "62\t290\n",
      "63\t243\n",
      "64\t184\n",
      "65\t70\n",
      "66\t505\n",
      "67\t11\n",
      "68\t31\n",
      "69\t16\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691092.fastq.gz fastq/SRR26691092.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:14    12,453,743 reads @   1.2 s/read;  51.71 M reads/minute\n",
      "Finished in 14.523 s (1.166 s/read; 51.45 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              12,453,743\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:              30,165 (0.2%)\n",
      "Reads written (passing filters):    12,423,578 (99.8%)\n",
      "\n",
      "Total basepairs processed:   871,762,010 bp\n",
      "Quality-trimmed:               5,300,083 bp (0.6%)\n",
      "Poly-A-trimmed:               47,865,188 bp (5.5%)\n",
      "Total written (filtered):    818,156,527 bp (93.9%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t9109946\n",
      "3\t474059\n",
      "4\t222205\n",
      "5\t166532\n",
      "6\t140128\n",
      "7\t142475\n",
      "8\t96445\n",
      "9\t96972\n",
      "10\t129687\n",
      "11\t90399\n",
      "12\t86327\n",
      "13\t101907\n",
      "14\t109416\n",
      "15\t92581\n",
      "16\t101200\n",
      "17\t109709\n",
      "18\t113973\n",
      "19\t127242\n",
      "20\t95549\n",
      "21\t118762\n",
      "22\t120221\n",
      "23\t131161\n",
      "24\t56226\n",
      "25\t57746\n",
      "26\t79356\n",
      "27\t49971\n",
      "28\t23913\n",
      "29\t12032\n",
      "30\t11336\n",
      "31\t8263\n",
      "32\t6555\n",
      "33\t5322\n",
      "34\t4754\n",
      "35\t6069\n",
      "36\t5990\n",
      "37\t4353\n",
      "38\t5155\n",
      "39\t4751\n",
      "40\t5217\n",
      "41\t5627\n",
      "42\t9811\n",
      "43\t8364\n",
      "44\t11079\n",
      "45\t19990\n",
      "46\t11550\n",
      "47\t12208\n",
      "48\t11690\n",
      "49\t9291\n",
      "50\t7479\n",
      "51\t5781\n",
      "52\t5850\n",
      "53\t3457\n",
      "54\t2287\n",
      "55\t1428\n",
      "56\t984\n",
      "57\t835\n",
      "58\t475\n",
      "59\t350\n",
      "60\t252\n",
      "61\t558\n",
      "62\t134\n",
      "63\t235\n",
      "64\t37\n",
      "65\t36\n",
      "66\t12\n",
      "67\t6\n",
      "68\t6\n",
      "69\t6\n",
      "70\t20\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/SRR26691093.fastq.gz fastq/SRR26691093.fastq.gz\n",
      "Processing single-end reads on 16 cores ...\n",
      "Done           00:00:16    14,507,430 reads @   1.2 s/read;  51.77 M reads/minute\n",
      "Finished in 16.877 s (1.163 s/read; 51.58 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              14,507,430\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:              47,882 (0.3%)\n",
      "Reads written (passing filters):    14,459,548 (99.7%)\n",
      "\n",
      "Total basepairs processed: 1,015,520,100 bp\n",
      "Quality-trimmed:              11,082,919 bp (1.1%)\n",
      "Poly-A-trimmed:               59,339,312 bp (5.8%)\n",
      "Total written (filtered):    944,371,188 bp (93.0%)\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t10762119\n",
      "3\t499896\n",
      "4\t215252\n",
      "5\t171663\n",
      "6\t140769\n",
      "7\t131573\n",
      "8\t105857\n",
      "9\t102584\n",
      "10\t128092\n",
      "11\t97008\n",
      "12\t92151\n",
      "13\t106673\n",
      "14\t113016\n",
      "15\t103523\n",
      "16\t111088\n",
      "17\t121208\n",
      "18\t123267\n",
      "19\t136127\n",
      "20\t109494\n",
      "21\t139924\n",
      "22\t146202\n",
      "23\t154820\n",
      "24\t67803\n",
      "25\t72124\n",
      "26\t91775\n",
      "27\t62040\n",
      "28\t31084\n",
      "29\t17787\n",
      "30\t14986\n",
      "31\t11747\n",
      "32\t9982\n",
      "33\t8607\n",
      "34\t7479\n",
      "35\t8975\n",
      "36\t10256\n",
      "37\t7386\n",
      "38\t8694\n",
      "39\t8224\n",
      "40\t9882\n",
      "41\t11234\n",
      "42\t18898\n",
      "43\t17636\n",
      "44\t22763\n",
      "45\t35010\n",
      "46\t25131\n",
      "47\t25777\n",
      "48\t22624\n",
      "49\t17161\n",
      "50\t13174\n",
      "51\t9966\n",
      "52\t9619\n",
      "53\t5357\n",
      "54\t3849\n",
      "55\t2393\n",
      "56\t1534\n",
      "57\t1280\n",
      "58\t671\n",
      "59\t532\n",
      "60\t328\n",
      "61\t649\n",
      "62\t220\n",
      "63\t319\n",
      "64\t54\n",
      "65\t35\n",
      "66\t14\n",
      "67\t14\n",
      "68\t10\n",
      "69\t12\n",
      "70\t29\n"
     ]
    }
   ],
   "source": [
    "%mkdir trimmed\n",
    "!for fastq in fastq/*.fastq.gz; do \\\n",
    "    base_name=$(basename \"$fastq\" .fastq.gz); \\\n",
    "    cutadapt -j 16 -m 20 --poly-a --nextseq-trim=10 -o ./trimmed/$base_name.fastq.gz $fastq; \\\n",
    "done"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***2 minute time to completion***\n",
    "\n",
    "Re-run FastQC so we compare the metrics before and after trimming. Generate an overall report with MultiQC. This time when running, we use the `--force` option to overwrite previous reports and the `-d` option to direct the tool to specific directories."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=935423;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.17\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Prepending directory to sample names\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/corwinbm5021/BIOL343/5_fastq/fastq/qc\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/corwinbm5021/BIOL343/5_fastq/trimmed/qc\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m48/48\u001b[0m  stqc.html\u001b[0m87_fastqc.html\u001b[0m.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 24 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc_report.html   (overwritten)\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc_data   (overwritten)\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "#!mkdir trimmed/qc\n",
    "#!fastqc -t 16 trimmed/*.fastq.gz -o trimmed/qc\n",
    "!multiqc --force -d fastq/qc/ trimmed/qc/"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***2 minute time to completion***\n",
    "\n",
    "Check out the new MultiQC report to see what changed. Not a whole lot, but in some samples (i.e., SRR26691093) the PolyA tail was substantially trimmed. Our trimming did not deal with UMIs or linkers - we'll talk more about those when we get to alignment next week. You can open the individual QC files to inspect them further.\n",
    "\n",
    "You'll notice that the \"Sequence Length Distribution\" sections often went from  to . That's because the trimmed PolyA tail was different for each read, so now we have a broad distribution of lengths rather than all of them being 69 nt. \n",
    "\n",
    "Even though we don't have adapters, we still have have overrepresented sequences. What are they? BLAST them against a nucleotide database and see if they are concerning or not."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
