{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FASTQ download and QC\n",
    "\n",
    "## Fetching FASTQ data from NCBI SRA"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In previous weeks, we explored the genome and annotations. They were stored in FASTA and GTF files and will be used as the reference to which RNA sequencing reads are mapped/aligned. After alignment, we can count how many RNA-seq reads align to the annotations we care about (genes or transcripts).\n",
    "\n",
    "Now, we need to get the FASTQ files that we want to align and analyze. To do so, we will be referencing an SRA read table, which is a list of all the FASTQ files associated with a given experiment uploaded to NCBI SRA. Here's the read table for the *in vivo* PZQ treatment paper:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run,AGE,Assay Type,AvgSpotLen,Bases,BioProject,BioSample,BioSampleModel,Bytes,Center Name,Consent,DATASTORE filetype,DATASTORE provider,DATASTORE region,dev_stage,Experiment,Instrument,Library Name,LibraryLayout,LibrarySelection,LibrarySource,Organism,Platform,ReleaseDate,replicate,create_date,version,Sample Name,sex,SRA Study,strain,tissue,treatment\n",
      "SRR10776762,7 weeks,RNA-Seq,102,2570036880,PRJNA597909,SAMN13691180,Model organism or animal,1127285500,MEDICAL COLLEGE OF WISCONSIN,public,\"fastq,run.zq,sra\",\"gs,ncbi,s3\",\"gs.us-east1,ncbi.public,s3.us-east-1\",sexually mature,SRX7450683,Illumina HiSeq 2500,ADULT_PZQ_E,PAIRED,RT-PCR,TRANSCRIPTOMIC,Schistosoma mansoni,ILLUMINA,2020-01-21T00:00:00Z,Biological replicate 5,2019-12-27T20:33:00Z,1,ADULT_PZQ_biological_replicate_5,pooled male and female,SRP238922,NMRI,whole animal,PZQ\n",
      "SRR10776763,7 weeks,RNA-Seq,102,2444119104,PRJNA597909,SAMN13691179,Model organism or animal,1065352234,MEDICAL COLLEGE OF WISCONSIN,public,\"fastq,run.zq,sra\",\"gs,ncbi,s3\",\"gs.us-east1,ncbi.public,s3.us-east-1\",sexually mature,SRX7450682,Illumina HiSeq 2500,ADULT_PZQ_D,PAIRED,RT-PCR,TRANSCRIPTOMIC,Schistosoma mansoni,ILLUMINA,2020-01-21T00:00:00Z,Biological replicate 4,2019-12-27T20:32:00Z,1,ADULT_PZQ_biological_replicate_4,pooled male and female,SRP238922,NMRI,whole animal,PZQ\n",
      "SRR10776764,7 weeks,RNA-Seq,102,2205190734,PRJNA597909,SAMN13691178,Model organism or animal,974816565,MEDICAL COLLEGE OF WISCONSIN,public,\"fastq,run.zq,sra\",\"gs,ncbi,s3\",\"gs.us-east1,ncbi.public,s3.us-east-1\",sexually mature,SRX7450681,Illumina HiSeq 2500,ADULT_PZQ_C,PAIRED,RT-PCR,TRANSCRIPTOMIC,Schistosoma mansoni,ILLUMINA,2020-01-21T00:00:00Z,Biological replicate 3,2019-12-27T20:32:00Z,1,ADULT_PZQ_biological_replicate_3,pooled male and female,SRP238922,NMRI,whole animal,PZQ\n",
      "SRR10776765,7 weeks,RNA-Seq,102,2439166902,PRJNA597909,SAMN13691177,Model organism or animal,1087248278,MEDICAL COLLEGE OF WISCONSIN,public,\"fastq,run.zq,sra\",\"gs,ncbi,s3\",\"gs.us-east1,ncbi.public,s3.us-east-1\",sexually mature,SRX7450680,Illumina HiSeq 2500,ADULT_PZQ_B,PAIRED,RT-PCR,TRANSCRIPTOMIC,Schistosoma mansoni,ILLUMINA,2020-01-21T00:00:00Z,Biological replicate 2,2019-12-27T20:33:00Z,1,ADULT_PZQ_biological_replicate_2,pooled male and female,SRP238922,NMRI,whole animal,PZQ\n",
      "SRR10776766,7 weeks,RNA-Seq,102,2221488090,PRJNA597909,SAMN13691176,Model organism or animal,966712364,MEDICAL COLLEGE OF WISCONSIN,public,\"fastq,run.zq,sra\",\"gs,ncbi,s3\",\"gs.us-east1,ncbi.public,s3.us-east-1\",sexually mature,SRX7450679,Illumina HiSeq 2500,ADULT_PZQ_A,PAIRED,RT-PCR,TRANSCRIPTOMIC,Schistosoma mansoni,ILLUMINA,2020-01-21T00:00:00Z,Biological replicate 1,2019-12-27T20:33:00Z,1,ADULT_PZQ_biological_replicate_1,pooled male and female,SRP238922,NMRI,whole animal,PZQ\n",
      "SRR10776767,7 weeks,RNA-Seq,102,2255055372,PRJNA597909,SAMN13691175,Model organism or animal,996224047,MEDICAL COLLEGE OF WISCONSIN,public,\"fastq,run.zq,sra\",\"gs,ncbi,s3\",\"gs.us-east1,ncbi.public,s3.us-east-1\",sexually mature,SRX7450678,Illumina HiSeq 2500,ADULT_CTRL_E,PAIRED,RT-PCR,TRANSCRIPTOMIC,Schistosoma mansoni,ILLUMINA,2020-01-21T00:00:00Z,Biological replicate 5,2019-12-27T20:33:00Z,1,ADULT_CTRL_biological_replicate_5,pooled male and female,SRP238922,NMRI,whole animal,CTRL\n",
      "SRR10776768,7 weeks,RNA-Seq,102,2383524780,PRJNA597909,SAMN13691174,Model organism or animal,1060446984,MEDICAL COLLEGE OF WISCONSIN,public,\"fastq,run.zq,sra\",\"gs,ncbi,s3\",\"gs.us-east1,ncbi.public,s3.us-east-1\",sexually mature,SRX7450677,Illumina HiSeq 2500,ADULT_CTRL_D,PAIRED,RT-PCR,TRANSCRIPTOMIC,Schistosoma mansoni,ILLUMINA,2020-01-21T00:00:00Z,Biological replicate 4,2019-12-27T20:34:00Z,1,ADULT_CTRL_biological_replicate_4,pooled male and female,SRP238922,NMRI,whole animal,CTRL\n",
      "SRR10776769,7 weeks,RNA-Seq,102,2449494300,PRJNA597909,SAMN13691173,Model organism or animal,1080572631,MEDICAL COLLEGE OF WISCONSIN,public,\"fastq,run.zq,sra\",\"gs,ncbi,s3\",\"gs.us-east1,ncbi.public,s3.us-east-1\",sexually mature,SRX7450676,Illumina HiSeq 2500,ADULT_CTRL_C,PAIRED,RT-PCR,TRANSCRIPTOMIC,Schistosoma mansoni,ILLUMINA,2020-01-21T00:00:00Z,Biological replicate 3,2019-12-27T20:33:00Z,1,ADULT_CTRL_biological_replicate_3,pooled male and female,SRP238922,NMRI,whole animal,CTRL\n",
      "SRR10776770,7 weeks,RNA-Seq,102,2433967350,PRJNA597909,SAMN13691172,Model organism or animal,1069772374,MEDICAL COLLEGE OF WISCONSIN,public,\"fastq,run.zq,sra\",\"gs,ncbi,s3\",\"gs.us-east1,ncbi.public,s3.us-east-1\",sexually mature,SRX7450675,Illumina HiSeq 2500,ADULT_CTRL_B,PAIRED,RT-PCR,TRANSCRIPTOMIC,Schistosoma mansoni,ILLUMINA,2020-01-21T00:00:00Z,Biological replicate 2,2019-12-27T20:34:00Z,1,ADULT_CTRL_biological_replicate_2,pooled male and female,SRP238922,NMRI,whole animal,CTRL\n",
      "SRR10776771,7 weeks,RNA-Seq,102,2332436448,PRJNA597909,SAMN13691171,Model organism or animal,1023093905,MEDICAL COLLEGE OF WISCONSIN,public,\"fastq,run.zq,sra\",\"gs,ncbi,s3\",\"gs.us-east1,ncbi.public,s3.us-east-1\",sexually mature,SRX7450674,Illumina HiSeq 2500,ADULT_CTRL_A,PAIRED,RT-PCR,TRANSCRIPTOMIC,Schistosoma mansoni,ILLUMINA,2020-01-21T00:00:00Z,Biological replicate 1,2019-12-27T20:35:00Z,1,ADULT_CTRL_biological_replicate_1,pooled male and female,SRP238922,NMRI,whole animal,CTRL\n"
     ]
    }
   ],
   "source": [
    "!head -11 SraRunTable.txt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see that there are 10 different runs, each has an ID that looks something like `SRR10776762`. After the ID, quite a bit of run metadata is provided. You should spend some time learning about what each field means.\n",
    "\n",
    "The experiment that we're reanalyzing had 2 different samples: parasites that were harvested from untreated mice and parasites that were harvested from mice treated with 100 mg/kg PZQ. Each sample had 5 replicates, totaling 10 RNA-seq runs.\n",
    "\n",
    "Most RNA-seq data in the USA is stored on NCBI's SRA database. We have to use `sra-tools` to download the FASTQ files from SRA. To do so, we'll use a `for` loop within bash. We can use `cut` (which we learned about in [week 2](../2_genome_exploration/2_genome_exploration.ipynb)) to get the first field, but because it's comma-delimited instead of tab-delimited, we have to use the `-d` option to tell what \"delimiter\" to use. The output can be saved to a file, which then can be looped through line-by-line. The FASTQ files will be stored in a new directory called `fastq`. Here's the code to get the list of runs we want to download. It will create a new file called `sra_list.txt`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir fastq\n",
    "!cut -d ',' -f 1 SraRunTable.txt | tail -n +2 > sra_list.txt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's the code to download the FASTQ files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting SRR10776762 from NCBI SRA\n",
      "2025-10-14 15:46:38,134 - SRR ids: ['SRR10776762']\n",
      "2025-10-14 15:46:38,134 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 15:46:38,202 - tempdir: /local/scratch/job_322423/pfd_l58tusbj\n",
      "2025-10-14 15:46:38,202 - CMD: sra-stat --meta --quick SRR10776762\n",
      "2025-10-14 15:46:40,792 - SRR10776762 spots: 25196440\n",
      "2025-10-14 15:46:40,792 - blocks: [[1, 1574777], [1574778, 3149554], [3149555, 4724331], [4724332, 6299108], [6299109, 7873885], [7873886, 9448662], [9448663, 11023439], [11023440, 12598216], [12598217, 14172993], [14172994, 15747770], [15747771, 17322547], [17322548, 18897324], [18897325, 20472101], [20472102, 22046878], [22046879, 23621655], [23621656, 25196440]]\n",
      "2025-10-14 15:46:40,792 - CMD: fastq-dump -N 1 -X 1574777 -O /local/scratch/job_322423/pfd_l58tusbj/0 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,976 - CMD: fastq-dump -N 1574778 -X 3149554 -O /local/scratch/job_322423/pfd_l58tusbj/1 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,976 - CMD: fastq-dump -N 3149555 -X 4724331 -O /local/scratch/job_322423/pfd_l58tusbj/2 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,977 - CMD: fastq-dump -N 4724332 -X 6299108 -O /local/scratch/job_322423/pfd_l58tusbj/3 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,977 - CMD: fastq-dump -N 6299109 -X 7873885 -O /local/scratch/job_322423/pfd_l58tusbj/4 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,977 - CMD: fastq-dump -N 7873886 -X 9448662 -O /local/scratch/job_322423/pfd_l58tusbj/5 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,978 - CMD: fastq-dump -N 9448663 -X 11023439 -O /local/scratch/job_322423/pfd_l58tusbj/6 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,978 - CMD: fastq-dump -N 11023440 -X 12598216 -O /local/scratch/job_322423/pfd_l58tusbj/7 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,978 - CMD: fastq-dump -N 12598217 -X 14172993 -O /local/scratch/job_322423/pfd_l58tusbj/8 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,979 - CMD: fastq-dump -N 14172994 -X 15747770 -O /local/scratch/job_322423/pfd_l58tusbj/9 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,979 - CMD: fastq-dump -N 15747771 -X 17322547 -O /local/scratch/job_322423/pfd_l58tusbj/10 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,979 - CMD: fastq-dump -N 17322548 -X 18897324 -O /local/scratch/job_322423/pfd_l58tusbj/11 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,979 - CMD: fastq-dump -N 18897325 -X 20472101 -O /local/scratch/job_322423/pfd_l58tusbj/12 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,980 - CMD: fastq-dump -N 20472102 -X 22046878 -O /local/scratch/job_322423/pfd_l58tusbj/13 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,980 - CMD: fastq-dump -N 22046879 -X 23621655 -O /local/scratch/job_322423/pfd_l58tusbj/14 --split-files --gzip SRR10776762\n",
      "2025-10-14 15:46:40,980 - CMD: fastq-dump -N 23621656 -X 25196440 -O /local/scratch/job_322423/pfd_l58tusbj/15 --split-files --gzip SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574785 spots for SRR10776762\n",
      "Written 1574785 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Read 1574777 spots for SRR10776762\n",
      "Written 1574777 spots for SRR10776762\n",
      "Getting SRR10776763 from NCBI SRA\n",
      "2025-10-14 15:48:11,324 - SRR ids: ['SRR10776763']\n",
      "2025-10-14 15:48:11,324 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 15:48:11,326 - tempdir: /local/scratch/job_322423/pfd_3xajk3ft\n",
      "2025-10-14 15:48:11,326 - CMD: sra-stat --meta --quick SRR10776763\n",
      "2025-10-14 15:48:13,166 - SRR10776763 spots: 23961952\n",
      "2025-10-14 15:48:13,166 - blocks: [[1, 1497622], [1497623, 2995244], [2995245, 4492866], [4492867, 5990488], [5990489, 7488110], [7488111, 8985732], [8985733, 10483354], [10483355, 11980976], [11980977, 13478598], [13478599, 14976220], [14976221, 16473842], [16473843, 17971464], [17971465, 19469086], [19469087, 20966708], [20966709, 22464330], [22464331, 23961952]]\n",
      "2025-10-14 15:48:13,166 - CMD: fastq-dump -N 1 -X 1497622 -O /local/scratch/job_322423/pfd_3xajk3ft/0 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,217 - CMD: fastq-dump -N 1497623 -X 2995244 -O /local/scratch/job_322423/pfd_3xajk3ft/1 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,218 - CMD: fastq-dump -N 2995245 -X 4492866 -O /local/scratch/job_322423/pfd_3xajk3ft/2 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,218 - CMD: fastq-dump -N 4492867 -X 5990488 -O /local/scratch/job_322423/pfd_3xajk3ft/3 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,219 - CMD: fastq-dump -N 5990489 -X 7488110 -O /local/scratch/job_322423/pfd_3xajk3ft/4 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,219 - CMD: fastq-dump -N 7488111 -X 8985732 -O /local/scratch/job_322423/pfd_3xajk3ft/5 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,219 - CMD: fastq-dump -N 8985733 -X 10483354 -O /local/scratch/job_322423/pfd_3xajk3ft/6 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,220 - CMD: fastq-dump -N 10483355 -X 11980976 -O /local/scratch/job_322423/pfd_3xajk3ft/7 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,220 - CMD: fastq-dump -N 11980977 -X 13478598 -O /local/scratch/job_322423/pfd_3xajk3ft/8 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,220 - CMD: fastq-dump -N 13478599 -X 14976220 -O /local/scratch/job_322423/pfd_3xajk3ft/9 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,221 - CMD: fastq-dump -N 14976221 -X 16473842 -O /local/scratch/job_322423/pfd_3xajk3ft/10 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,221 - CMD: fastq-dump -N 16473843 -X 17971464 -O /local/scratch/job_322423/pfd_3xajk3ft/11 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,221 - CMD: fastq-dump -N 17971465 -X 19469086 -O /local/scratch/job_322423/pfd_3xajk3ft/12 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,221 - CMD: fastq-dump -N 19469087 -X 20966708 -O /local/scratch/job_322423/pfd_3xajk3ft/13 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,222 - CMD: fastq-dump -N 20966709 -X 22464330 -O /local/scratch/job_322423/pfd_3xajk3ft/14 --split-files --gzip SRR10776763\n",
      "2025-10-14 15:48:13,222 - CMD: fastq-dump -N 22464331 -X 23961952 -O /local/scratch/job_322423/pfd_3xajk3ft/15 --split-files --gzip SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Read 1497622 spots for SRR10776763\n",
      "Written 1497622 spots for SRR10776763\n",
      "Getting SRR10776764 from NCBI SRA\n",
      "2025-10-14 15:49:36,649 - SRR ids: ['SRR10776764']\n",
      "2025-10-14 15:49:36,649 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 15:49:36,651 - tempdir: /local/scratch/job_322423/pfd_a_tmbqfb\n",
      "2025-10-14 15:49:36,651 - CMD: sra-stat --meta --quick SRR10776764\n",
      "2025-10-14 15:49:38,283 - SRR10776764 spots: 21619517\n",
      "2025-10-14 15:49:38,283 - blocks: [[1, 1351219], [1351220, 2702438], [2702439, 4053657], [4053658, 5404876], [5404877, 6756095], [6756096, 8107314], [8107315, 9458533], [9458534, 10809752], [10809753, 12160971], [12160972, 13512190], [13512191, 14863409], [14863410, 16214628], [16214629, 17565847], [17565848, 18917066], [18917067, 20268285], [20268286, 21619517]]\n",
      "2025-10-14 15:49:38,283 - CMD: fastq-dump -N 1 -X 1351219 -O /local/scratch/job_322423/pfd_a_tmbqfb/0 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,294 - CMD: fastq-dump -N 1351220 -X 2702438 -O /local/scratch/job_322423/pfd_a_tmbqfb/1 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,294 - CMD: fastq-dump -N 2702439 -X 4053657 -O /local/scratch/job_322423/pfd_a_tmbqfb/2 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,295 - CMD: fastq-dump -N 4053658 -X 5404876 -O /local/scratch/job_322423/pfd_a_tmbqfb/3 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,295 - CMD: fastq-dump -N 5404877 -X 6756095 -O /local/scratch/job_322423/pfd_a_tmbqfb/4 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,295 - CMD: fastq-dump -N 6756096 -X 8107314 -O /local/scratch/job_322423/pfd_a_tmbqfb/5 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,296 - CMD: fastq-dump -N 8107315 -X 9458533 -O /local/scratch/job_322423/pfd_a_tmbqfb/6 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,296 - CMD: fastq-dump -N 9458534 -X 10809752 -O /local/scratch/job_322423/pfd_a_tmbqfb/7 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,296 - CMD: fastq-dump -N 10809753 -X 12160971 -O /local/scratch/job_322423/pfd_a_tmbqfb/8 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,297 - CMD: fastq-dump -N 12160972 -X 13512190 -O /local/scratch/job_322423/pfd_a_tmbqfb/9 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,297 - CMD: fastq-dump -N 13512191 -X 14863409 -O /local/scratch/job_322423/pfd_a_tmbqfb/10 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,297 - CMD: fastq-dump -N 14863410 -X 16214628 -O /local/scratch/job_322423/pfd_a_tmbqfb/11 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,298 - CMD: fastq-dump -N 16214629 -X 17565847 -O /local/scratch/job_322423/pfd_a_tmbqfb/12 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,298 - CMD: fastq-dump -N 17565848 -X 18917066 -O /local/scratch/job_322423/pfd_a_tmbqfb/13 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,298 - CMD: fastq-dump -N 18917067 -X 20268285 -O /local/scratch/job_322423/pfd_a_tmbqfb/14 --split-files --gzip SRR10776764\n",
      "2025-10-14 15:49:38,298 - CMD: fastq-dump -N 20268286 -X 21619517 -O /local/scratch/job_322423/pfd_a_tmbqfb/15 --split-files --gzip SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351232 spots for SRR10776764\n",
      "Written 1351232 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Read 1351219 spots for SRR10776764\n",
      "Written 1351219 spots for SRR10776764\n",
      "Getting SRR10776765 from NCBI SRA\n",
      "2025-10-14 15:50:55,788 - SRR ids: ['SRR10776765']\n",
      "2025-10-14 15:50:55,788 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 15:50:55,788 - tempdir: /local/scratch/job_322423/pfd_q_rsz094\n",
      "2025-10-14 15:50:55,788 - CMD: sra-stat --meta --quick SRR10776765\n",
      "2025-10-14 15:50:57,766 - SRR10776765 spots: 23913401\n",
      "2025-10-14 15:50:57,766 - blocks: [[1, 1494587], [1494588, 2989174], [2989175, 4483761], [4483762, 5978348], [5978349, 7472935], [7472936, 8967522], [8967523, 10462109], [10462110, 11956696], [11956697, 13451283], [13451284, 14945870], [14945871, 16440457], [16440458, 17935044], [17935045, 19429631], [19429632, 20924218], [20924219, 22418805], [22418806, 23913401]]\n",
      "2025-10-14 15:50:57,766 - CMD: fastq-dump -N 1 -X 1494587 -O /local/scratch/job_322423/pfd_q_rsz094/0 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,793 - CMD: fastq-dump -N 1494588 -X 2989174 -O /local/scratch/job_322423/pfd_q_rsz094/1 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,793 - CMD: fastq-dump -N 2989175 -X 4483761 -O /local/scratch/job_322423/pfd_q_rsz094/2 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,794 - CMD: fastq-dump -N 4483762 -X 5978348 -O /local/scratch/job_322423/pfd_q_rsz094/3 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,794 - CMD: fastq-dump -N 5978349 -X 7472935 -O /local/scratch/job_322423/pfd_q_rsz094/4 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,795 - CMD: fastq-dump -N 7472936 -X 8967522 -O /local/scratch/job_322423/pfd_q_rsz094/5 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,795 - CMD: fastq-dump -N 8967523 -X 10462109 -O /local/scratch/job_322423/pfd_q_rsz094/6 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,795 - CMD: fastq-dump -N 10462110 -X 11956696 -O /local/scratch/job_322423/pfd_q_rsz094/7 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,796 - CMD: fastq-dump -N 11956697 -X 13451283 -O /local/scratch/job_322423/pfd_q_rsz094/8 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,796 - CMD: fastq-dump -N 13451284 -X 14945870 -O /local/scratch/job_322423/pfd_q_rsz094/9 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,797 - CMD: fastq-dump -N 14945871 -X 16440457 -O /local/scratch/job_322423/pfd_q_rsz094/10 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,797 - CMD: fastq-dump -N 16440458 -X 17935044 -O /local/scratch/job_322423/pfd_q_rsz094/11 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,797 - CMD: fastq-dump -N 17935045 -X 19429631 -O /local/scratch/job_322423/pfd_q_rsz094/12 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,798 - CMD: fastq-dump -N 19429632 -X 20924218 -O /local/scratch/job_322423/pfd_q_rsz094/13 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,798 - CMD: fastq-dump -N 20924219 -X 22418805 -O /local/scratch/job_322423/pfd_q_rsz094/14 --split-files --gzip SRR10776765\n",
      "2025-10-14 15:50:57,798 - CMD: fastq-dump -N 22418806 -X 23913401 -O /local/scratch/job_322423/pfd_q_rsz094/15 --split-files --gzip SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494596 spots for SRR10776765\n",
      "Written 1494596 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Read 1494587 spots for SRR10776765\n",
      "Written 1494587 spots for SRR10776765\n",
      "Getting SRR10776766 from NCBI SRA\n",
      "2025-10-14 15:52:18,968 - SRR ids: ['SRR10776766']\n",
      "2025-10-14 15:52:18,968 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 15:52:18,968 - tempdir: /local/scratch/job_322423/pfd_m5t7a2nf\n",
      "2025-10-14 15:52:18,968 - CMD: sra-stat --meta --quick SRR10776766\n",
      "2025-10-14 15:52:20,605 - SRR10776766 spots: 21779295\n",
      "2025-10-14 15:52:20,605 - blocks: [[1, 1361205], [1361206, 2722410], [2722411, 4083615], [4083616, 5444820], [5444821, 6806025], [6806026, 8167230], [8167231, 9528435], [9528436, 10889640], [10889641, 12250845], [12250846, 13612050], [13612051, 14973255], [14973256, 16334460], [16334461, 17695665], [17695666, 19056870], [19056871, 20418075], [20418076, 21779295]]\n",
      "2025-10-14 15:52:20,605 - CMD: fastq-dump -N 1 -X 1361205 -O /local/scratch/job_322423/pfd_m5t7a2nf/0 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,643 - CMD: fastq-dump -N 1361206 -X 2722410 -O /local/scratch/job_322423/pfd_m5t7a2nf/1 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,643 - CMD: fastq-dump -N 2722411 -X 4083615 -O /local/scratch/job_322423/pfd_m5t7a2nf/2 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,644 - CMD: fastq-dump -N 4083616 -X 5444820 -O /local/scratch/job_322423/pfd_m5t7a2nf/3 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,644 - CMD: fastq-dump -N 5444821 -X 6806025 -O /local/scratch/job_322423/pfd_m5t7a2nf/4 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,645 - CMD: fastq-dump -N 6806026 -X 8167230 -O /local/scratch/job_322423/pfd_m5t7a2nf/5 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,645 - CMD: fastq-dump -N 8167231 -X 9528435 -O /local/scratch/job_322423/pfd_m5t7a2nf/6 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,645 - CMD: fastq-dump -N 9528436 -X 10889640 -O /local/scratch/job_322423/pfd_m5t7a2nf/7 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,646 - CMD: fastq-dump -N 10889641 -X 12250845 -O /local/scratch/job_322423/pfd_m5t7a2nf/8 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,646 - CMD: fastq-dump -N 12250846 -X 13612050 -O /local/scratch/job_322423/pfd_m5t7a2nf/9 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,646 - CMD: fastq-dump -N 13612051 -X 14973255 -O /local/scratch/job_322423/pfd_m5t7a2nf/10 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,646 - CMD: fastq-dump -N 14973256 -X 16334460 -O /local/scratch/job_322423/pfd_m5t7a2nf/11 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,647 - CMD: fastq-dump -N 16334461 -X 17695665 -O /local/scratch/job_322423/pfd_m5t7a2nf/12 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,647 - CMD: fastq-dump -N 17695666 -X 19056870 -O /local/scratch/job_322423/pfd_m5t7a2nf/13 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,647 - CMD: fastq-dump -N 19056871 -X 20418075 -O /local/scratch/job_322423/pfd_m5t7a2nf/14 --split-files --gzip SRR10776766\n",
      "2025-10-14 15:52:20,648 - CMD: fastq-dump -N 20418076 -X 21779295 -O /local/scratch/job_322423/pfd_m5t7a2nf/15 --split-files --gzip SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361205 spots for SRR10776766\n",
      "Written 1361205 spots for SRR10776766\n",
      "Read 1361220 spots for SRR10776766\n",
      "Written 1361220 spots for SRR10776766\n",
      "Getting SRR10776767 from NCBI SRA\n",
      "2025-10-14 15:53:48,474 - SRR ids: ['SRR10776767']\n",
      "2025-10-14 15:53:48,475 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 15:53:48,475 - tempdir: /local/scratch/job_322423/pfd_2zdkxpln\n",
      "2025-10-14 15:53:48,475 - CMD: sra-stat --meta --quick SRR10776767\n",
      "2025-10-14 15:53:50,071 - SRR10776767 spots: 22108386\n",
      "2025-10-14 15:53:50,071 - blocks: [[1, 1381774], [1381775, 2763548], [2763549, 4145322], [4145323, 5527096], [5527097, 6908870], [6908871, 8290644], [8290645, 9672418], [9672419, 11054192], [11054193, 12435966], [12435967, 13817740], [13817741, 15199514], [15199515, 16581288], [16581289, 17963062], [17963063, 19344836], [19344837, 20726610], [20726611, 22108386]]\n",
      "2025-10-14 15:53:50,071 - CMD: fastq-dump -N 1 -X 1381774 -O /local/scratch/job_322423/pfd_2zdkxpln/0 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,095 - CMD: fastq-dump -N 1381775 -X 2763548 -O /local/scratch/job_322423/pfd_2zdkxpln/1 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,096 - CMD: fastq-dump -N 2763549 -X 4145322 -O /local/scratch/job_322423/pfd_2zdkxpln/2 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,096 - CMD: fastq-dump -N 4145323 -X 5527096 -O /local/scratch/job_322423/pfd_2zdkxpln/3 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,097 - CMD: fastq-dump -N 5527097 -X 6908870 -O /local/scratch/job_322423/pfd_2zdkxpln/4 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,097 - CMD: fastq-dump -N 6908871 -X 8290644 -O /local/scratch/job_322423/pfd_2zdkxpln/5 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,098 - CMD: fastq-dump -N 8290645 -X 9672418 -O /local/scratch/job_322423/pfd_2zdkxpln/6 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,098 - CMD: fastq-dump -N 9672419 -X 11054192 -O /local/scratch/job_322423/pfd_2zdkxpln/7 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,098 - CMD: fastq-dump -N 11054193 -X 12435966 -O /local/scratch/job_322423/pfd_2zdkxpln/8 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,098 - CMD: fastq-dump -N 12435967 -X 13817740 -O /local/scratch/job_322423/pfd_2zdkxpln/9 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,099 - CMD: fastq-dump -N 13817741 -X 15199514 -O /local/scratch/job_322423/pfd_2zdkxpln/10 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,099 - CMD: fastq-dump -N 15199515 -X 16581288 -O /local/scratch/job_322423/pfd_2zdkxpln/11 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,100 - CMD: fastq-dump -N 16581289 -X 17963062 -O /local/scratch/job_322423/pfd_2zdkxpln/12 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,100 - CMD: fastq-dump -N 17963063 -X 19344836 -O /local/scratch/job_322423/pfd_2zdkxpln/13 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,100 - CMD: fastq-dump -N 19344837 -X 20726610 -O /local/scratch/job_322423/pfd_2zdkxpln/14 --split-files --gzip SRR10776767\n",
      "2025-10-14 15:53:50,101 - CMD: fastq-dump -N 20726611 -X 22108386 -O /local/scratch/job_322423/pfd_2zdkxpln/15 --split-files --gzip SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381776 spots for SRR10776767\n",
      "Written 1381776 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Read 1381774 spots for SRR10776767\n",
      "Written 1381774 spots for SRR10776767\n",
      "Getting SRR10776768 from NCBI SRA\n",
      "2025-10-14 15:55:27,192 - SRR ids: ['SRR10776768']\n",
      "2025-10-14 15:55:27,192 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 15:55:27,193 - tempdir: /local/scratch/job_322423/pfd_yg8mehqe\n",
      "2025-10-14 15:55:27,193 - CMD: sra-stat --meta --quick SRR10776768\n",
      "2025-10-14 15:55:29,001 - SRR10776768 spots: 23367890\n",
      "2025-10-14 15:55:29,001 - blocks: [[1, 1460493], [1460494, 2920986], [2920987, 4381479], [4381480, 5841972], [5841973, 7302465], [7302466, 8762958], [8762959, 10223451], [10223452, 11683944], [11683945, 13144437], [13144438, 14604930], [14604931, 16065423], [16065424, 17525916], [17525917, 18986409], [18986410, 20446902], [20446903, 21907395], [21907396, 23367890]]\n",
      "2025-10-14 15:55:29,001 - CMD: fastq-dump -N 1 -X 1460493 -O /local/scratch/job_322423/pfd_yg8mehqe/0 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,144 - CMD: fastq-dump -N 1460494 -X 2920986 -O /local/scratch/job_322423/pfd_yg8mehqe/1 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,144 - CMD: fastq-dump -N 2920987 -X 4381479 -O /local/scratch/job_322423/pfd_yg8mehqe/2 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,145 - CMD: fastq-dump -N 4381480 -X 5841972 -O /local/scratch/job_322423/pfd_yg8mehqe/3 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,145 - CMD: fastq-dump -N 5841973 -X 7302465 -O /local/scratch/job_322423/pfd_yg8mehqe/4 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,145 - CMD: fastq-dump -N 7302466 -X 8762958 -O /local/scratch/job_322423/pfd_yg8mehqe/5 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,146 - CMD: fastq-dump -N 8762959 -X 10223451 -O /local/scratch/job_322423/pfd_yg8mehqe/6 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,146 - CMD: fastq-dump -N 10223452 -X 11683944 -O /local/scratch/job_322423/pfd_yg8mehqe/7 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,146 - CMD: fastq-dump -N 11683945 -X 13144437 -O /local/scratch/job_322423/pfd_yg8mehqe/8 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,147 - CMD: fastq-dump -N 13144438 -X 14604930 -O /local/scratch/job_322423/pfd_yg8mehqe/9 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,147 - CMD: fastq-dump -N 14604931 -X 16065423 -O /local/scratch/job_322423/pfd_yg8mehqe/10 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,147 - CMD: fastq-dump -N 16065424 -X 17525916 -O /local/scratch/job_322423/pfd_yg8mehqe/11 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,148 - CMD: fastq-dump -N 17525917 -X 18986409 -O /local/scratch/job_322423/pfd_yg8mehqe/12 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,148 - CMD: fastq-dump -N 18986410 -X 20446902 -O /local/scratch/job_322423/pfd_yg8mehqe/13 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,148 - CMD: fastq-dump -N 20446903 -X 21907395 -O /local/scratch/job_322423/pfd_yg8mehqe/14 --split-files --gzip SRR10776768\n",
      "2025-10-14 15:55:29,149 - CMD: fastq-dump -N 21907396 -X 23367890 -O /local/scratch/job_322423/pfd_yg8mehqe/15 --split-files --gzip SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460495 spots for SRR10776768\n",
      "Written 1460495 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Read 1460493 spots for SRR10776768\n",
      "Written 1460493 spots for SRR10776768\n",
      "Getting SRR10776769 from NCBI SRA\n",
      "2025-10-14 15:56:46,391 - SRR ids: ['SRR10776769']\n",
      "2025-10-14 15:56:46,391 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 15:56:46,428 - tempdir: /local/scratch/job_322423/pfd_tc3r5e7j\n",
      "2025-10-14 15:56:46,428 - CMD: sra-stat --meta --quick SRR10776769\n",
      "2025-10-14 15:56:48,021 - SRR10776769 spots: 24014650\n",
      "2025-10-14 15:56:48,021 - blocks: [[1, 1500915], [1500916, 3001830], [3001831, 4502745], [4502746, 6003660], [6003661, 7504575], [7504576, 9005490], [9005491, 10506405], [10506406, 12007320], [12007321, 13508235], [13508236, 15009150], [15009151, 16510065], [16510066, 18010980], [18010981, 19511895], [19511896, 21012810], [21012811, 22513725], [22513726, 24014650]]\n",
      "2025-10-14 15:56:48,021 - CMD: fastq-dump -N 1 -X 1500915 -O /local/scratch/job_322423/pfd_tc3r5e7j/0 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,035 - CMD: fastq-dump -N 1500916 -X 3001830 -O /local/scratch/job_322423/pfd_tc3r5e7j/1 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,035 - CMD: fastq-dump -N 3001831 -X 4502745 -O /local/scratch/job_322423/pfd_tc3r5e7j/2 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,036 - CMD: fastq-dump -N 4502746 -X 6003660 -O /local/scratch/job_322423/pfd_tc3r5e7j/3 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,036 - CMD: fastq-dump -N 6003661 -X 7504575 -O /local/scratch/job_322423/pfd_tc3r5e7j/4 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,036 - CMD: fastq-dump -N 7504576 -X 9005490 -O /local/scratch/job_322423/pfd_tc3r5e7j/5 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,037 - CMD: fastq-dump -N 9005491 -X 10506405 -O /local/scratch/job_322423/pfd_tc3r5e7j/6 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,037 - CMD: fastq-dump -N 10506406 -X 12007320 -O /local/scratch/job_322423/pfd_tc3r5e7j/7 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,038 - CMD: fastq-dump -N 12007321 -X 13508235 -O /local/scratch/job_322423/pfd_tc3r5e7j/8 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,038 - CMD: fastq-dump -N 13508236 -X 15009150 -O /local/scratch/job_322423/pfd_tc3r5e7j/9 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,038 - CMD: fastq-dump -N 15009151 -X 16510065 -O /local/scratch/job_322423/pfd_tc3r5e7j/10 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,039 - CMD: fastq-dump -N 16510066 -X 18010980 -O /local/scratch/job_322423/pfd_tc3r5e7j/11 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,039 - CMD: fastq-dump -N 18010981 -X 19511895 -O /local/scratch/job_322423/pfd_tc3r5e7j/12 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,039 - CMD: fastq-dump -N 19511896 -X 21012810 -O /local/scratch/job_322423/pfd_tc3r5e7j/13 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,040 - CMD: fastq-dump -N 21012811 -X 22513725 -O /local/scratch/job_322423/pfd_tc3r5e7j/14 --split-files --gzip SRR10776769\n",
      "2025-10-14 15:56:48,040 - CMD: fastq-dump -N 22513726 -X 24014650 -O /local/scratch/job_322423/pfd_tc3r5e7j/15 --split-files --gzip SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Read 1500925 spots for SRR10776769\n",
      "Written 1500925 spots for SRR10776769\n",
      "Read 1500915 spots for SRR10776769\n",
      "Written 1500915 spots for SRR10776769\n",
      "Getting SRR10776770 from NCBI SRA\n",
      "2025-10-14 15:58:18,796 - SRR ids: ['SRR10776770']\n",
      "2025-10-14 15:58:18,796 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 15:58:18,798 - tempdir: /local/scratch/job_322423/pfd_vox57ye5\n",
      "2025-10-14 15:58:18,798 - CMD: sra-stat --meta --quick SRR10776770\n",
      "2025-10-14 15:58:20,751 - SRR10776770 spots: 23862425\n",
      "2025-10-14 15:58:20,751 - blocks: [[1, 1491401], [1491402, 2982802], [2982803, 4474203], [4474204, 5965604], [5965605, 7457005], [7457006, 8948406], [8948407, 10439807], [10439808, 11931208], [11931209, 13422609], [13422610, 14914010], [14914011, 16405411], [16405412, 17896812], [17896813, 19388213], [19388214, 20879614], [20879615, 22371015], [22371016, 23862425]]\n",
      "2025-10-14 15:58:20,751 - CMD: fastq-dump -N 1 -X 1491401 -O /local/scratch/job_322423/pfd_vox57ye5/0 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,784 - CMD: fastq-dump -N 1491402 -X 2982802 -O /local/scratch/job_322423/pfd_vox57ye5/1 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,785 - CMD: fastq-dump -N 2982803 -X 4474203 -O /local/scratch/job_322423/pfd_vox57ye5/2 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,785 - CMD: fastq-dump -N 4474204 -X 5965604 -O /local/scratch/job_322423/pfd_vox57ye5/3 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,786 - CMD: fastq-dump -N 5965605 -X 7457005 -O /local/scratch/job_322423/pfd_vox57ye5/4 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,786 - CMD: fastq-dump -N 7457006 -X 8948406 -O /local/scratch/job_322423/pfd_vox57ye5/5 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,786 - CMD: fastq-dump -N 8948407 -X 10439807 -O /local/scratch/job_322423/pfd_vox57ye5/6 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,787 - CMD: fastq-dump -N 10439808 -X 11931208 -O /local/scratch/job_322423/pfd_vox57ye5/7 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,787 - CMD: fastq-dump -N 11931209 -X 13422609 -O /local/scratch/job_322423/pfd_vox57ye5/8 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,787 - CMD: fastq-dump -N 13422610 -X 14914010 -O /local/scratch/job_322423/pfd_vox57ye5/9 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,788 - CMD: fastq-dump -N 14914011 -X 16405411 -O /local/scratch/job_322423/pfd_vox57ye5/10 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,788 - CMD: fastq-dump -N 16405412 -X 17896812 -O /local/scratch/job_322423/pfd_vox57ye5/11 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,788 - CMD: fastq-dump -N 17896813 -X 19388213 -O /local/scratch/job_322423/pfd_vox57ye5/12 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,788 - CMD: fastq-dump -N 19388214 -X 20879614 -O /local/scratch/job_322423/pfd_vox57ye5/13 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,789 - CMD: fastq-dump -N 20879615 -X 22371015 -O /local/scratch/job_322423/pfd_vox57ye5/14 --split-files --gzip SRR10776770\n",
      "2025-10-14 15:58:20,789 - CMD: fastq-dump -N 22371016 -X 23862425 -O /local/scratch/job_322423/pfd_vox57ye5/15 --split-files --gzip SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491410 spots for SRR10776770\n",
      "Written 1491410 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Read 1491401 spots for SRR10776770\n",
      "Written 1491401 spots for SRR10776770\n",
      "Getting SRR10776771 from NCBI SRA\n",
      "2025-10-14 16:00:02,964 - SRR ids: ['SRR10776771']\n",
      "2025-10-14 16:00:02,964 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 16:00:03,030 - tempdir: /local/scratch/job_322423/pfd_8hnhzdsb\n",
      "2025-10-14 16:00:03,030 - CMD: sra-stat --meta --quick SRR10776771\n",
      "2025-10-14 16:00:05,628 - SRR10776771 spots: 22867024\n",
      "2025-10-14 16:00:05,628 - blocks: [[1, 1429189], [1429190, 2858378], [2858379, 4287567], [4287568, 5716756], [5716757, 7145945], [7145946, 8575134], [8575135, 10004323], [10004324, 11433512], [11433513, 12862701], [12862702, 14291890], [14291891, 15721079], [15721080, 17150268], [17150269, 18579457], [18579458, 20008646], [20008647, 21437835], [21437836, 22867024]]\n",
      "2025-10-14 16:00:05,628 - CMD: fastq-dump -N 1 -X 1429189 -O /local/scratch/job_322423/pfd_8hnhzdsb/0 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,872 - CMD: fastq-dump -N 1429190 -X 2858378 -O /local/scratch/job_322423/pfd_8hnhzdsb/1 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,873 - CMD: fastq-dump -N 2858379 -X 4287567 -O /local/scratch/job_322423/pfd_8hnhzdsb/2 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,873 - CMD: fastq-dump -N 4287568 -X 5716756 -O /local/scratch/job_322423/pfd_8hnhzdsb/3 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,873 - CMD: fastq-dump -N 5716757 -X 7145945 -O /local/scratch/job_322423/pfd_8hnhzdsb/4 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,874 - CMD: fastq-dump -N 7145946 -X 8575134 -O /local/scratch/job_322423/pfd_8hnhzdsb/5 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,874 - CMD: fastq-dump -N 8575135 -X 10004323 -O /local/scratch/job_322423/pfd_8hnhzdsb/6 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,875 - CMD: fastq-dump -N 10004324 -X 11433512 -O /local/scratch/job_322423/pfd_8hnhzdsb/7 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,875 - CMD: fastq-dump -N 11433513 -X 12862701 -O /local/scratch/job_322423/pfd_8hnhzdsb/8 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,875 - CMD: fastq-dump -N 12862702 -X 14291890 -O /local/scratch/job_322423/pfd_8hnhzdsb/9 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,875 - CMD: fastq-dump -N 14291891 -X 15721079 -O /local/scratch/job_322423/pfd_8hnhzdsb/10 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,876 - CMD: fastq-dump -N 15721080 -X 17150268 -O /local/scratch/job_322423/pfd_8hnhzdsb/11 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,876 - CMD: fastq-dump -N 17150269 -X 18579457 -O /local/scratch/job_322423/pfd_8hnhzdsb/12 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,876 - CMD: fastq-dump -N 18579458 -X 20008646 -O /local/scratch/job_322423/pfd_8hnhzdsb/13 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,877 - CMD: fastq-dump -N 20008647 -X 21437835 -O /local/scratch/job_322423/pfd_8hnhzdsb/14 --split-files --gzip SRR10776771\n",
      "2025-10-14 16:00:05,877 - CMD: fastq-dump -N 21437836 -X 22867024 -O /local/scratch/job_322423/pfd_8hnhzdsb/15 --split-files --gzip SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n",
      "Read 1429189 spots for SRR10776771\n",
      "Written 1429189 spots for SRR10776771\n"
     ]
    }
   ],
   "source": [
    "!while IFS= read -r line; do \\\n",
    "    echo \"Getting $line from NCBI SRA\"; \\\n",
    "    parallel-fastq-dump --split-files --sra-id $line --threads 16 --outdir fastq --gzip; \\\n",
    "    done < sra_list.txt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***14 minute time to completion***\n",
    "\n",
    "If we take another look at the metadata, there are a few things that are of interest to us and our analysis. First, the reads are `ILLUMINA` reads generated on a `HiSeq 2500` instrument, which means they may have Illumina adapters, depending on whether or not they were trimmed by the authors prior to uploading them to SRA. Secpmd, we can see  that these are paired-end reads (`PAIRED`). This information will be important to us soon. You can get a bit more information about the data using the tool `seqkit`. For instance:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!seqkit stats fastq/SRR10776762_1.fastq.gz"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***1 minute time to completion***"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FastQC\n",
    "\n",
    "Whenever we look at FASTQ files for the first time, we should perform quality control (QC). The primary tool used for read QC is called FastQC, which is installed in our environment. Let's take a look at the manual and then run QC on our reads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!fastqc -h"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following code uses the `*` regular expression, which will expand to any file ending in `.fastq.gz`. This line will run QC on all 10 files and use 32 cores (threads):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.009s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /dev/cpuset.\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR10776762.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR10776763.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR10776764.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR10776765.fastq.gz\n",
      "application/gzip\n",
      "Started analysis of SRR10776766.fastq.gz\n",
      "Started analysis of SRR10776767.fastq.gz\n",
      "Started analysis of SRR10776768.fastq.gz\n",
      "Started analysis of SRR10776769.fastq.gz\n",
      "Started analysis of SRR10776770.fastq.gz\n",
      "Started analysis of SRR10776771.fastq.gz\n",
      "Approx 5% complete for SRR10776764.fastq.gz\n",
      "Approx 5% complete for SRR10776762.fastq.gz\n",
      "Approx 5% complete for SRR10776763.fastq.gz\n",
      "Approx 5% complete for SRR10776766.fastq.gz\n",
      "Approx 5% complete for SRR10776767.fastq.gz\n",
      "Approx 5% complete for SRR10776765.fastq.gz\n",
      "Approx 5% complete for SRR10776768.fastq.gz\n",
      "Approx 5% complete for SRR10776771.fastq.gz\n",
      "Approx 5% complete for SRR10776769.fastq.gz\n",
      "Approx 5% complete for SRR10776770.fastq.gz\n",
      "Approx 10% complete for SRR10776764.fastq.gz\n",
      "Approx 10% complete for SRR10776766.fastq.gz\n",
      "Approx 10% complete for SRR10776767.fastq.gz\n",
      "Approx 10% complete for SRR10776763.fastq.gz\n",
      "Approx 10% complete for SRR10776762.fastq.gz\n",
      "Approx 10% complete for SRR10776765.fastq.gz\n",
      "Approx 10% complete for SRR10776771.fastq.gz\n",
      "Approx 10% complete for SRR10776768.fastq.gz\n",
      "Approx 10% complete for SRR10776769.fastq.gz\n",
      "Approx 10% complete for SRR10776770.fastq.gz\n",
      "Approx 15% complete for SRR10776764.fastq.gz\n",
      "Approx 15% complete for SRR10776766.fastq.gz\n",
      "Approx 15% complete for SRR10776767.fastq.gz\n",
      "Approx 15% complete for SRR10776771.fastq.gz\n",
      "Approx 15% complete for SRR10776763.fastq.gz\n",
      "Approx 15% complete for SRR10776762.fastq.gz\n",
      "Approx 15% complete for SRR10776765.fastq.gz\n",
      "Approx 15% complete for SRR10776768.fastq.gz\n",
      "Approx 15% complete for SRR10776769.fastq.gz\n",
      "Approx 15% complete for SRR10776770.fastq.gz\n",
      "Approx 20% complete for SRR10776764.fastq.gz\n",
      "Approx 20% complete for SRR10776766.fastq.gz\n",
      "Approx 20% complete for SRR10776767.fastq.gz\n",
      "Approx 20% complete for SRR10776771.fastq.gz\n",
      "Approx 20% complete for SRR10776763.fastq.gz\n",
      "Approx 20% complete for SRR10776762.fastq.gz\n",
      "Approx 20% complete for SRR10776765.fastq.gz\n",
      "Approx 20% complete for SRR10776768.fastq.gz\n",
      "Approx 20% complete for SRR10776769.fastq.gz\n",
      "Approx 20% complete for SRR10776770.fastq.gz\n",
      "Approx 25% complete for SRR10776764.fastq.gz\n",
      "Approx 25% complete for SRR10776766.fastq.gz\n",
      "Approx 25% complete for SRR10776767.fastq.gz\n",
      "Approx 25% complete for SRR10776771.fastq.gz\n",
      "Approx 25% complete for SRR10776763.fastq.gz\n",
      "Approx 25% complete for SRR10776768.fastq.gz\n",
      "Approx 25% complete for SRR10776765.fastq.gz\n",
      "Approx 25% complete for SRR10776762.fastq.gz\n",
      "Approx 25% complete for SRR10776769.fastq.gz\n",
      "Approx 25% complete for SRR10776770.fastq.gz\n",
      "Approx 30% complete for SRR10776764.fastq.gz\n",
      "Approx 30% complete for SRR10776766.fastq.gz\n",
      "Approx 30% complete for SRR10776767.fastq.gz\n",
      "Approx 30% complete for SRR10776771.fastq.gz\n",
      "Approx 30% complete for SRR10776763.fastq.gz\n",
      "Approx 30% complete for SRR10776768.fastq.gz\n",
      "Approx 30% complete for SRR10776765.fastq.gz\n",
      "Approx 30% complete for SRR10776762.fastq.gz\n",
      "Approx 30% complete for SRR10776769.fastq.gz\n",
      "Approx 35% complete for SRR10776764.fastq.gz\n",
      "Approx 30% complete for SRR10776770.fastq.gz\n",
      "Approx 35% complete for SRR10776766.fastq.gz\n",
      "Approx 35% complete for SRR10776767.fastq.gz\n",
      "Approx 35% complete for SRR10776771.fastq.gz\n",
      "Approx 35% complete for SRR10776768.fastq.gz\n",
      "Approx 35% complete for SRR10776763.fastq.gz\n",
      "Approx 35% complete for SRR10776765.fastq.gz\n",
      "Approx 40% complete for SRR10776764.fastq.gz\n",
      "Approx 35% complete for SRR10776769.fastq.gz\n",
      "Approx 35% complete for SRR10776762.fastq.gz\n",
      "Approx 40% complete for SRR10776766.fastq.gz\n",
      "Approx 35% complete for SRR10776770.fastq.gz\n",
      "Approx 40% complete for SRR10776767.fastq.gz\n",
      "Approx 40% complete for SRR10776771.fastq.gz\n",
      "Approx 40% complete for SRR10776768.fastq.gz\n",
      "Approx 45% complete for SRR10776764.fastq.gz\n",
      "Approx 40% complete for SRR10776763.fastq.gz\n",
      "Approx 40% complete for SRR10776765.fastq.gz\n",
      "Approx 40% complete for SRR10776769.fastq.gz\n",
      "Approx 45% complete for SRR10776766.fastq.gz\n",
      "Approx 40% complete for SRR10776762.fastq.gz\n",
      "Approx 40% complete for SRR10776770.fastq.gz\n",
      "Approx 45% complete for SRR10776767.fastq.gz\n",
      "Approx 45% complete for SRR10776771.fastq.gz\n",
      "Approx 50% complete for SRR10776764.fastq.gz\n",
      "Approx 45% complete for SRR10776768.fastq.gz\n",
      "Approx 45% complete for SRR10776763.fastq.gz\n",
      "Approx 50% complete for SRR10776766.fastq.gz\n",
      "Approx 45% complete for SRR10776765.fastq.gz\n",
      "Approx 45% complete for SRR10776769.fastq.gz\n",
      "Approx 45% complete for SRR10776762.fastq.gz\n",
      "Approx 45% complete for SRR10776770.fastq.gz\n",
      "Approx 50% complete for SRR10776767.fastq.gz\n",
      "Approx 50% complete for SRR10776771.fastq.gz\n",
      "Approx 55% complete for SRR10776764.fastq.gz\n",
      "Approx 50% complete for SRR10776768.fastq.gz\n",
      "Approx 55% complete for SRR10776766.fastq.gz\n",
      "Approx 50% complete for SRR10776763.fastq.gz\n",
      "Approx 50% complete for SRR10776765.fastq.gz\n",
      "Approx 50% complete for SRR10776769.fastq.gz\n",
      "Approx 55% complete for SRR10776767.fastq.gz\n",
      "Approx 55% complete for SRR10776771.fastq.gz\n",
      "Approx 50% complete for SRR10776770.fastq.gz\n",
      "Approx 50% complete for SRR10776762.fastq.gz\n",
      "Approx 60% complete for SRR10776764.fastq.gz\n",
      "Approx 60% complete for SRR10776766.fastq.gz\n",
      "Approx 55% complete for SRR10776768.fastq.gz\n",
      "Approx 55% complete for SRR10776763.fastq.gz\n",
      "Approx 55% complete for SRR10776765.fastq.gz\n",
      "Approx 55% complete for SRR10776769.fastq.gz\n",
      "Approx 60% complete for SRR10776767.fastq.gz\n",
      "Approx 60% complete for SRR10776771.fastq.gz\n",
      "Approx 55% complete for SRR10776770.fastq.gz\n",
      "Approx 55% complete for SRR10776762.fastq.gz\n",
      "Approx 65% complete for SRR10776764.fastq.gz\n",
      "Approx 65% complete for SRR10776766.fastq.gz\n",
      "Approx 60% complete for SRR10776768.fastq.gz\n",
      "Approx 60% complete for SRR10776763.fastq.gz\n",
      "Approx 60% complete for SRR10776765.fastq.gz\n",
      "Approx 65% complete for SRR10776767.fastq.gz\n",
      "Approx 65% complete for SRR10776771.fastq.gz\n",
      "Approx 60% complete for SRR10776769.fastq.gz\n",
      "Approx 70% complete for SRR10776764.fastq.gz\n",
      "Approx 60% complete for SRR10776770.fastq.gz\n",
      "Approx 60% complete for SRR10776762.fastq.gz\n",
      "Approx 70% complete for SRR10776766.fastq.gz\n",
      "Approx 65% complete for SRR10776768.fastq.gz\n",
      "Approx 65% complete for SRR10776763.fastq.gz\n",
      "Approx 70% complete for SRR10776767.fastq.gz\n",
      "Approx 70% complete for SRR10776771.fastq.gz\n",
      "Approx 65% complete for SRR10776765.fastq.gz\n",
      "Approx 65% complete for SRR10776769.fastq.gz\n",
      "Approx 75% complete for SRR10776764.fastq.gz\n",
      "Approx 65% complete for SRR10776770.fastq.gz\n",
      "Approx 75% complete for SRR10776766.fastq.gz\n",
      "Approx 65% complete for SRR10776762.fastq.gz\n",
      "Approx 70% complete for SRR10776768.fastq.gz\n",
      "Approx 70% complete for SRR10776763.fastq.gz\n",
      "Approx 75% complete for SRR10776767.fastq.gz\n",
      "Approx 75% complete for SRR10776771.fastq.gz\n",
      "Approx 70% complete for SRR10776765.fastq.gz\n",
      "Approx 80% complete for SRR10776764.fastq.gz\n",
      "Approx 70% complete for SRR10776769.fastq.gz\n",
      "Approx 80% complete for SRR10776766.fastq.gz\n",
      "Approx 70% complete for SRR10776770.fastq.gz\n",
      "Approx 70% complete for SRR10776762.fastq.gz\n",
      "Approx 75% complete for SRR10776768.fastq.gz\n",
      "Approx 80% complete for SRR10776767.fastq.gz\n",
      "Approx 80% complete for SRR10776771.fastq.gz\n",
      "Approx 75% complete for SRR10776763.fastq.gz\n",
      "Approx 85% complete for SRR10776764.fastq.gz\n",
      "Approx 75% complete for SRR10776765.fastq.gz\n",
      "Approx 75% complete for SRR10776769.fastq.gz\n",
      "Approx 85% complete for SRR10776766.fastq.gz\n",
      "Approx 75% complete for SRR10776770.fastq.gz\n",
      "Approx 75% complete for SRR10776762.fastq.gz\n",
      "Approx 80% complete for SRR10776768.fastq.gz\n",
      "Approx 85% complete for SRR10776767.fastq.gz\n",
      "Approx 85% complete for SRR10776771.fastq.gz\n",
      "Approx 90% complete for SRR10776764.fastq.gz\n",
      "Approx 80% complete for SRR10776763.fastq.gz\n",
      "Approx 80% complete for SRR10776765.fastq.gz\n",
      "Approx 90% complete for SRR10776766.fastq.gz\n",
      "Approx 80% complete for SRR10776769.fastq.gz\n",
      "Approx 80% complete for SRR10776770.fastq.gz\n",
      "Approx 85% complete for SRR10776768.fastq.gz\n",
      "Approx 90% complete for SRR10776767.fastq.gz\n",
      "Approx 80% complete for SRR10776762.fastq.gz\n",
      "Approx 95% complete for SRR10776764.fastq.gz\n",
      "Approx 90% complete for SRR10776771.fastq.gz\n",
      "Approx 85% complete for SRR10776763.fastq.gz\n",
      "Approx 85% complete for SRR10776765.fastq.gz\n",
      "Approx 95% complete for SRR10776766.fastq.gz\n",
      "Approx 85% complete for SRR10776769.fastq.gz\n",
      "Approx 85% complete for SRR10776770.fastq.gz\n",
      "Approx 95% complete for SRR10776767.fastq.gz\n",
      "Approx 90% complete for SRR10776768.fastq.gz\n",
      "Analysis complete for SRR10776764.fastq.gz\n",
      "Approx 95% complete for SRR10776771.fastq.gz\n",
      "Approx 85% complete for SRR10776762.fastq.gz\n",
      "Approx 90% complete for SRR10776763.fastq.gz\n",
      "Analysis complete for SRR10776766.fastq.gz\n",
      "Approx 90% complete for SRR10776765.fastq.gz\n",
      "Approx 90% complete for SRR10776769.fastq.gz\n",
      "Approx 90% complete for SRR10776770.fastq.gz\n",
      "Analysis complete for SRR10776767.fastq.gz\n",
      "Analysis complete for SRR10776771.fastq.gz\n",
      "Approx 95% complete for SRR10776768.fastq.gz\n",
      "Approx 90% complete for SRR10776762.fastq.gz\n",
      "Approx 95% complete for SRR10776763.fastq.gz\n",
      "Approx 95% complete for SRR10776765.fastq.gz\n",
      "Approx 95% complete for SRR10776769.fastq.gz\n",
      "Approx 95% complete for SRR10776770.fastq.gz\n",
      "Analysis complete for SRR10776768.fastq.gz\n",
      "Approx 95% complete for SRR10776762.fastq.gz\n",
      "Analysis complete for SRR10776763.fastq.gz\n",
      "Analysis complete for SRR10776765.fastq.gz\n",
      "Analysis complete for SRR10776769.fastq.gz\n",
      "Analysis complete for SRR10776770.fastq.gz\n",
      "Analysis complete for SRR10776762.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "!mkdir fastq/qc\n",
    "!fastqc -t 32 fastq/*.fastq.gz -o fastq/qc"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***3 minute time to completion***\n",
    "\n",
    "Let's go through each section one-by-one.\n",
    "\n",
    "### Per base sequence quality\n",
    "\n",
    "Base quality scores are represented in Phred+33 encoding. Remember that Q20 is a popular threshold, as the likelihood of an error substantially increases as the Phred quality gets less than 20. It is typical to see high base quality at the 5' end of reads and lower quality at the 3' ends. For this reason, sometimes the 3' end of reads may be trimmed to remove low quality bases.\n",
    "\n",
    "### Per tile sequence quality\n",
    "\n",
    "Tiles are physical locations on the flow cell. Warnings in this section can be associated with flow cells that have been overloaded.\n",
    "\n",
    "### Per sequence quality scores\n",
    "\n",
    "This distribution shows that average Phred quality for entire reads. It helps you make decisions about removing entire reads based on their average quality.\n",
    "\n",
    "### Per base sequence content\n",
    "\n",
    "This distribution shows the bases most likely to be found at any given position in the reads. In general, the likelihood of finding any of A, T, G, or C at a given position should be the similar across the entire sequence. Deviations from this expectation may be due to the presence of barcodes or adaptors. Here's an example of a dataset that has the adapter on the 5' end.\n",
    "\n",
    "<img src=\"assets/linker.png\" alt=\"linker\" width=\"450\">\n",
    "\n",
    "### Per sequence GC content\n",
    "\n",
    "Since our reads are derived from mRNA fragments, we would expect that the percentage of Gs and Cs would be normally distributed in a way that mimics the underlying transcriptome of the organism from which the mRNA was extracted. If the distribution isn't normal (that is, if there are sharp peaks), that is an indication of some sort of contamination. Here is an example of a sharp peak:\n",
    "\n",
    "<img src=\"assets/gc_peak.png\" alt=\"GC\" width=\"450\">\n",
    "\n",
    "This peak is indicative of something in the library that has a higher GC% than the rest of the mRNA. Sharp peaks are often associated with adapter dimers. These are reads generated from two adapters that were ligated together during library prep - they have the sequencing primer binding site, but they don't have an insert. Here's the adaptor sequence used for this library, with Gs and Cs annotated:\n",
    "\n",
    "```\n",
    "AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC = 34\n",
    " x  xxx  x xx x xx x x  x xx x x x = 18\n",
    "18 / 34 = 53%\n",
    "```\n",
    "\n",
    "So, it would make sense that if many reads are associated with adaptor dimers, these reads would have a significantly higher GC% than average.\n",
    "\n",
    "### Per base N content\n",
    "\n",
    "Ns are called when the base caller can't make a decision for that cycle. N calls are usually associated (and correlated) with per base sequence quality.\n",
    "\n",
    "### Sequence length distribution\n",
    "\n",
    "Some sequencing libraries will have non-uniform read lengths, in which case we'd see a distribution of different lenghts. For runs that have a pre-defined read lenght, we should see a sharp peak at that length.\n",
    "\n",
    "### Sequence duplication levels\n",
    "\n",
    "For this module, FastQC will scan the first 50 bp of the first 100,000 reads and tally duplicates. The best libraries will have almost no, if any, duplicates. High levels of duplication may be indicative of PCR over amplification. Sometimes over amplification is a desired outcome; in this case, highly expressed transcripts will be highly duplicated, but this is a tradeoff for recovering more lowly expressed transcripts. FASTQs can be deduplicated before or after the alignment step.\n",
    "\n",
    "### Overrepresented sequences\n",
    "\n",
    "If you have many duplicate reads, it's likely there are overrepresented sequences. You can also use this list of sequences to help evalulate if you have adapter dimers. A simple way to check for problems is to search a sequence database for the overrepresented sequence. If it hits to a sequences deriving from your organism of interest, then you don't have a problem and this overrepresentation may be indicative of true underlying biology. The most useful search tool is BLAST. You can use [this website](https://blast.ncbi.nlm.nih.gov/Blast.cgi) to BLAST the sequences against the NCBI's entire nucleotide database.\n",
    "\n",
    "### Adapter content\n",
    "\n",
    "This module searches reads for popular adapters and PolyA stretches. If you performed mRNA enrichment, you may see higher PolyA instances at either end of the reads. "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MultiQC\n",
    "\n",
    "When you have many FASTQ files, all the relevant QC can be difficult to manage. MultiQC is designed to coalesce QC metrics for many different bioinformatics datatypes.\n",
    "\n",
    "Check the manual first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!multiqc -h"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MultiQC is very easy to use. All you have to do is run the command followed by the path to the working directory. It will discover all relevant QC files and wrap them up into a very helpful report. Run it now, then take a look at `multiqc_report.html` (you'll have to download it and open it in your browser)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=184737;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.17\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/heky1803/BIOL343/5_fastq\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m39/39\u001b[0m  /qc/SRR10776770_fastqc.html\u001b[0m.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 10 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "!multiqc ."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FASTQ trimming and filtering\n",
    "\n",
    "QC metrics will inform whether or not reads should be trimmed (i.e., PolyA tails, low quality bases, or adapters removed) or filtered (low quality reads removed). Typically, manufacturers will provide some guidance on how to trim and filter sequences produced from their kit. For Illumina (the maker of the TruSeq kit used in the paper), this information can be found in the [FAQs](https://support-docs.illumina.com/SHARE/AdapterSequences/Content/SHARE/AdapterSeq/TruSeq/UDIndexes.htm). This site shows that `AGATCGGAAGAGCACACGTCTGAACTCCAGTCA` is the adapter sequence for read 1 and `AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT` is the adapter sequence for read 2. However, our FastQC eval for these data show that there are no adapters or overrepresented sequences, which means they've likely already been trimmed.\n",
    "\n",
    "Here's what the authors of the paper said about read QC:\n",
    "\n",
    "> Trimmed reads were mapped to the Schistosoma mansoni genome (v7.2)...\n",
    "\n",
    "Although they likely looked at QC, they say nothing about how they trimmed or filtered.\n",
    "\n",
    "In the end, we'll just trim low quality sequences from the end (using Q = 20 as a threshold). We will use the tool [cutadapt](https://cutadapt.readthedocs.io/en/stable/) to perform the trimming. First, check out the manual:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cutadapt version 4.9\n",
      "\n",
      "Copyright (C) 2010 Marcel Martin <marcel.martin@scilifelab.se> and contributors\n",
      "\n",
      "Cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. All reads from input.fastq will be written to\n",
      "output.fastq with the adapter sequence removed. Adapter matching is\n",
      "error-tolerant. Multiple adapter sequences can be given (use further -a\n",
      "options), but only the best-matching adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Run \"cutadapt --help\" to see all command-line options.\n",
      "See https://cutadapt.readthedocs.io/ for full documentation.\n",
      "\n",
      "Options:\n",
      "  -h, --help            Show this help message and exit\n",
      "  --version             Show version number and exit\n",
      "  --debug               Print debug log. Use twice to also print DP matrices\n",
      "  -j CORES, --cores CORES\n",
      "                        Number of CPU cores to use. Use 0 to auto-detect.\n",
      "                        Default: 1\n",
      "\n",
      "Finding adapters:\n",
      "  Parameters -a, -g, -b specify adapters to be removed from each read (or from\n",
      "  R1 if data is paired-end. If specified multiple times, only the best\n",
      "  matching adapter is trimmed (but see the --times option). Use notation\n",
      "  'file:FILE' to read adapter sequences from a FASTA file.\n",
      "\n",
      "  -a ADAPTER, --adapter ADAPTER\n",
      "                        Sequence of an adapter ligated to the 3' end (paired\n",
      "                        data: of the first read). The adapter and subsequent\n",
      "                        bases are trimmed. If a '$' character is appended\n",
      "                        ('anchoring'), the adapter is only found if it is a\n",
      "                        suffix of the read.\n",
      "  -g ADAPTER, --front ADAPTER\n",
      "                        Sequence of an adapter ligated to the 5' end (paired\n",
      "                        data: of the first read). The adapter and any preceding\n",
      "                        bases are trimmed. Partial matches at the 5' end are\n",
      "                        allowed. If a '^' character is prepended ('anchoring'),\n",
      "                        the adapter is only found if it is a prefix of the read.\n",
      "  -b ADAPTER, --anywhere ADAPTER\n",
      "                        Sequence of an adapter that may be ligated to the 5' or\n",
      "                        3' end (paired data: of the first read). Both types of\n",
      "                        matches as described under -a and -g are allowed. If the\n",
      "                        first base of the read is part of the match, the\n",
      "                        behavior is as with -g, otherwise as with -a. This\n",
      "                        option is mostly for rescuing failed library\n",
      "                        preparations - do not use if you know which end your\n",
      "                        adapter was ligated to!\n",
      "  -e E, --error-rate E, --errors E\n",
      "                        Maximum allowed error rate (if 0 <= E < 1), or absolute\n",
      "                        number of errors for full-length adapter match (if E is\n",
      "                        an integer >= 1). Error rate = no. of errors divided by\n",
      "                        length of matching region. Default: 0.1 (10%)\n",
      "  --no-indels           Allow only mismatches in alignments. Default: allow both\n",
      "                        mismatches and indels\n",
      "  -n COUNT, --times COUNT\n",
      "                        Remove up to COUNT adapters from each read. Default: 1\n",
      "  -O MINLENGTH, --overlap MINLENGTH\n",
      "                        Require MINLENGTH overlap between read and adapter for\n",
      "                        an adapter to be found. Default: 3\n",
      "  --match-read-wildcards\n",
      "                        Interpret IUPAC wildcards in reads. Default: False\n",
      "  -N, --no-match-adapter-wildcards\n",
      "                        Do not interpret IUPAC wildcards in adapters.\n",
      "  --action {trim,retain,mask,lowercase,crop,none}\n",
      "                        What to do if a match was found. trim: trim adapter and\n",
      "                        up- or downstream sequence; retain: trim, but retain\n",
      "                        adapter; mask: replace with 'N' characters; lowercase:\n",
      "                        convert to lowercase; crop: trim up and downstream\n",
      "                        sequence; none: leave unchanged. Default: trim\n",
      "  --rc, --revcomp       Check both the read and its reverse complement for\n",
      "                        adapter matches. If match is on reverse-complemented\n",
      "                        version, output that one. Default: check only read\n",
      "\n",
      "Additional read modifications:\n",
      "  -u LEN, --cut LEN     Remove LEN bases from each read (or R1 if paired; use -U\n",
      "                        option for R2). If LEN is positive, remove bases from\n",
      "                        the beginning. If LEN is negative, remove bases from the\n",
      "                        end. Can be used twice if LENs have different signs.\n",
      "                        Applied *before* adapter trimming.\n",
      "  --nextseq-trim 3'CUTOFF\n",
      "                        NextSeq-specific quality trimming (each read). Trims\n",
      "                        also dark cycles appearing as high-quality G bases.\n",
      "  -q [5'CUTOFF,]3'CUTOFF, --quality-cutoff [5'CUTOFF,]3'CUTOFF\n",
      "                        Trim low-quality bases from 5' and/or 3' ends of each\n",
      "                        read before adapter removal. Applied to both reads if\n",
      "                        data is paired. If one value is given, only the 3' end\n",
      "                        is trimmed. If two comma-separated cutoffs are given,\n",
      "                        the 5' end is trimmed with the first cutoff, the 3' end\n",
      "                        with the second.\n",
      "  --quality-base N      Assume that quality values in FASTQ are encoded as\n",
      "                        ascii(quality + N). This needs to be set to 64 for some\n",
      "                        old Illumina FASTQ files. Default: 33\n",
      "  --poly-a              Trim poly-A tails\n",
      "  --length LENGTH, -l LENGTH\n",
      "                        Shorten reads to LENGTH. Positive values remove bases at\n",
      "                        the end while negative ones remove bases at the\n",
      "                        beginning. This and the following modifications are\n",
      "                        applied after adapter trimming.\n",
      "  --trim-n              Trim N's on ends of reads.\n",
      "  --length-tag TAG      Search for TAG followed by a decimal number in the\n",
      "                        description field of the read. Replace the decimal\n",
      "                        number with the correct length of the trimmed read. For\n",
      "                        example, use --length-tag 'length=' to correct fields\n",
      "                        like 'length=123'.\n",
      "  --strip-suffix STRIP_SUFFIX\n",
      "                        Remove this suffix from read names if present. Can be\n",
      "                        given multiple times.\n",
      "  -x PREFIX, --prefix PREFIX\n",
      "                        Add this prefix to read names. Use {name} to insert the\n",
      "                        name of the matching adapter.\n",
      "  -y SUFFIX, --suffix SUFFIX\n",
      "                        Add this suffix to read names; can also include {name}\n",
      "  --rename TEMPLATE     Rename reads using TEMPLATE containing variables such as\n",
      "                        {id}, {adapter_name} etc. (see documentation)\n",
      "  --zero-cap, -z        Change negative quality values to zero.\n",
      "\n",
      "Filtering of processed reads:\n",
      "  Filters are applied after above read modifications. Paired-end reads are\n",
      "  always discarded pairwise (see also --pair-filter).\n",
      "\n",
      "  -m LEN[:LEN2], --minimum-length LEN[:LEN2]\n",
      "                        Discard reads shorter than LEN. Default: 0\n",
      "  -M LEN[:LEN2], --maximum-length LEN[:LEN2]\n",
      "                        Discard reads longer than LEN. Default: no limit\n",
      "  --max-n COUNT         Discard reads with more than COUNT 'N' bases. If COUNT\n",
      "                        is a number between 0 and 1, it is interpreted as a\n",
      "                        fraction of the read length.\n",
      "  --max-expected-errors ERRORS, --max-ee ERRORS\n",
      "                        Discard reads whose expected number of errors (computed\n",
      "                        from quality values) exceeds ERRORS.\n",
      "  --max-average-error-rate ERROR_RATE, --max-aer ERROR_RATE\n",
      "                        as --max-expected-errors (see above), but divided by\n",
      "                        length to account for reads of varying length.\n",
      "  --discard-trimmed, --discard\n",
      "                        Discard reads that contain an adapter. Use also -O to\n",
      "                        avoid discarding too many randomly matching reads.\n",
      "  --discard-untrimmed, --trimmed-only\n",
      "                        Discard reads that do not contain an adapter.\n",
      "  --discard-casava      Discard reads that did not pass CASAVA filtering (header\n",
      "                        has :Y:).\n",
      "\n",
      "Output:\n",
      "  --quiet               Print only error messages.\n",
      "  --report {full,minimal}\n",
      "                        Which type of report to print: 'full' or 'minimal'.\n",
      "                        Default: full\n",
      "  --json FILE           Dump report in JSON format to FILE\n",
      "  -o FILE, --output FILE\n",
      "                        Write trimmed reads to FILE. FASTQ or FASTA format is\n",
      "                        chosen depending on input. Summary report is sent to\n",
      "                        standard output. Use '{name}' for demultiplexing (see\n",
      "                        docs). Default: write to standard output\n",
      "  --fasta               Output FASTA to standard output even on FASTQ input.\n",
      "  -Z                    Use compression level 1 for gzipped output files\n",
      "                        (faster, but uses more space)\n",
      "  --info-file FILE      Write information about each read and its adapter\n",
      "                        matches into FILE. See the documentation for the file\n",
      "                        format.\n",
      "  -r FILE, --rest-file FILE\n",
      "                        When the adapter matches in the middle of a read, write\n",
      "                        the rest (after the adapter) to FILE.\n",
      "  --wildcard-file FILE  When the adapter has N wildcard bases, write adapter\n",
      "                        bases matching wildcard positions to FILE. (Inaccurate\n",
      "                        with indels.)\n",
      "  --too-short-output FILE\n",
      "                        Write reads that are too short (according to length\n",
      "                        specified by -m) to FILE. Default: discard reads\n",
      "  --too-long-output FILE\n",
      "                        Write reads that are too long (according to length\n",
      "                        specified by -M) to FILE. Default: discard reads\n",
      "  --untrimmed-output FILE\n",
      "                        Write reads that do not contain any adapter to FILE.\n",
      "                        Default: output to same file as trimmed reads\n",
      "\n",
      "Paired-end options:\n",
      "  The -A/-G/-B/-U/-Q options work like their lowercase counterparts, but are\n",
      "  applied to R2 (second read in pair)\n",
      "\n",
      "  -A ADAPTER            3' adapter to be removed from R2\n",
      "  -G ADAPTER            5' adapter to be removed from R2\n",
      "  -B ADAPTER            5'/3 adapter to be removed from R2\n",
      "  -U LENGTH             Remove LENGTH bases from R2\n",
      "  -Q [5'CUTOFF,]3'CUTOFF\n",
      "                        Quality-trimming cutoff for R2. Default: same as for R1\n",
      "  -L LENGTH             Shorten R2 to LENGTH. Default: same as for R1\n",
      "  -p FILE, --paired-output FILE\n",
      "                        Write R2 to FILE.\n",
      "  --pair-adapters       Treat adapters given with -a/-A etc. as pairs. Either\n",
      "                        both or none are removed from each read pair.\n",
      "  --pair-filter {any,both,first}\n",
      "                        Which of the reads in a paired-end read have to match\n",
      "                        the filtering criterion in order for the pair to be\n",
      "                        filtered. Default: any\n",
      "  --interleaved         Read and/or write interleaved paired-end reads.\n",
      "  --untrimmed-paired-output FILE\n",
      "                        Write second read in a pair to this FILE when no adapter\n",
      "                        was found. Use with --untrimmed-output. Default: output\n",
      "                        to same file as trimmed reads\n",
      "  --too-short-paired-output FILE\n",
      "                        Write second read in a pair to this file if pair is too\n",
      "                        short.\n",
      "  --too-long-paired-output FILE\n",
      "                        Write second read in a pair to this file if pair is too\n",
      "                        long.\n"
     ]
    }
   ],
   "source": [
    "!cutadapt --help"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Edit this block and define the result of each of flags used in the above commands.\n",
    "\n",
    "`-j`: \n",
    "\n",
    "`-q`: \n",
    "\n",
    "`-m`:\n",
    "\n",
    "`-o`:"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's trim and run QC on these new files to see how they look. We'll use a bash loop again. Note that though our data was originally paired-end, the authors have already joined the read pairs so we're treating it like single-end (one read per fragment cluster)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -q 20 -m 100 -o ./trimmed/SRR10776762.fastq.gz fastq/SRR10776762.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:41    25,196,440 reads @   1.7 s/read;  36.05 M reads/minute\n",
      "Finished in 42.116 s (1.672 s/read; 35.90 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              25,196,440\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:             123,383 (0.5%)\n",
      "Reads written (passing filters):    25,073,057 (99.5%)\n",
      "\n",
      "Total basepairs processed: 2,570,036,880 bp\n",
      "Quality-trimmed:               2,429,454 bp (0.1%)\n",
      "Total written (filtered):  2,556,825,458 bp (99.5%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -q 20 -m 100 -o ./trimmed/SRR10776763.fastq.gz fastq/SRR10776763.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:39    23,961,952 reads @   1.6 s/read;  36.59 M reads/minute\n",
      "Finished in 39.372 s (1.643 s/read; 36.52 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              23,961,952\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:             120,539 (0.5%)\n",
      "Reads written (passing filters):    23,841,413 (99.5%)\n",
      "\n",
      "Total basepairs processed: 2,444,119,104 bp\n",
      "Quality-trimmed:               2,433,801 bp (0.1%)\n",
      "Total written (filtered):  2,431,274,864 bp (99.5%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -q 20 -m 100 -o ./trimmed/SRR10776764.fastq.gz fastq/SRR10776764.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:35    21,619,517 reads @   1.6 s/read;  36.69 M reads/minute\n",
      "Finished in 35.447 s (1.640 s/read; 36.59 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              21,619,517\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:             108,322 (0.5%)\n",
      "Reads written (passing filters):    21,511,195 (99.5%)\n",
      "\n",
      "Total basepairs processed: 2,205,190,734 bp\n",
      "Quality-trimmed:               2,210,582 bp (0.1%)\n",
      "Total written (filtered):  2,193,584,446 bp (99.5%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -q 20 -m 100 -o ./trimmed/SRR10776765.fastq.gz fastq/SRR10776765.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:39    23,913,401 reads @   1.6 s/read;  36.47 M reads/minute\n",
      "Finished in 39.432 s (1.649 s/read; 36.39 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              23,913,401\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:             103,814 (0.4%)\n",
      "Reads written (passing filters):    23,809,587 (99.6%)\n",
      "\n",
      "Total basepairs processed: 2,439,166,902 bp\n",
      "Quality-trimmed:               1,965,981 bp (0.1%)\n",
      "Total written (filtered):  2,427,905,542 bp (99.5%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -q 20 -m 100 -o ./trimmed/SRR10776766.fastq.gz fastq/SRR10776766.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:35    21,779,295 reads @   1.6 s/read;  36.40 M reads/minute\n",
      "Finished in 36.022 s (1.654 s/read; 36.28 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              21,779,295\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:              64,383 (0.3%)\n",
      "Reads written (passing filters):    21,714,912 (99.7%)\n",
      "\n",
      "Total basepairs processed: 2,221,488,090 bp\n",
      "Quality-trimmed:               1,268,299 bp (0.1%)\n",
      "Total written (filtered):  2,214,442,332 bp (99.7%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -q 20 -m 100 -o ./trimmed/SRR10776767.fastq.gz fastq/SRR10776767.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:37    22,108,386 reads @   1.7 s/read;  35.40 M reads/minute\n",
      "Finished in 37.554 s (1.699 s/read; 35.32 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              22,108,386\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:              90,735 (0.4%)\n",
      "Reads written (passing filters):    22,017,651 (99.6%)\n",
      "\n",
      "Total basepairs processed: 2,255,055,372 bp\n",
      "Quality-trimmed:               1,791,413 bp (0.1%)\n",
      "Total written (filtered):  2,245,249,544 bp (99.6%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -q 20 -m 100 -o ./trimmed/SRR10776768.fastq.gz fastq/SRR10776768.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:40    23,367,890 reads @   1.7 s/read;  34.97 M reads/minute\n",
      "Finished in 40.180 s (1.719 s/read; 34.89 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              23,367,890\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:             109,195 (0.5%)\n",
      "Reads written (passing filters):    23,258,695 (99.5%)\n",
      "\n",
      "Total basepairs processed: 2,383,524,780 bp\n",
      "Quality-trimmed:               2,153,995 bp (0.1%)\n",
      "Total written (filtered):  2,371,735,195 bp (99.5%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -q 20 -m 100 -o ./trimmed/SRR10776769.fastq.gz fastq/SRR10776769.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:40    24,014,650 reads @   1.7 s/read;  35.81 M reads/minute\n",
      "Finished in 40.320 s (1.679 s/read; 35.74 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              24,014,650\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:             142,356 (0.6%)\n",
      "Reads written (passing filters):    23,872,294 (99.4%)\n",
      "\n",
      "Total basepairs processed: 2,449,494,300 bp\n",
      "Quality-trimmed:               2,818,988 bp (0.1%)\n",
      "Total written (filtered):  2,434,320,622 bp (99.4%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -q 20 -m 100 -o ./trimmed/SRR10776770.fastq.gz fastq/SRR10776770.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:38    23,862,425 reads @   1.6 s/read;  37.01 M reads/minute\n",
      "Finished in 38.893 s (1.630 s/read; 36.81 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              23,862,425\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:             104,529 (0.4%)\n",
      "Reads written (passing filters):    23,757,896 (99.6%)\n",
      "\n",
      "Total basepairs processed: 2,433,967,350 bp\n",
      "Quality-trimmed:               2,066,804 bp (0.1%)\n",
      "Total written (filtered):  2,422,699,909 bp (99.5%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -q 20 -m 100 -o ./trimmed/SRR10776771.fastq.gz fastq/SRR10776771.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:37    22,867,024 reads @   1.6 s/read;  36.63 M reads/minute\n",
      "Finished in 37.556 s (1.642 s/read; 36.53 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              22,867,024\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:             125,731 (0.5%)\n",
      "Reads written (passing filters):    22,741,293 (99.5%)\n",
      "\n",
      "Total basepairs processed: 2,332,436,448 bp\n",
      "Quality-trimmed:               2,488,134 bp (0.1%)\n",
      "Total written (filtered):  2,319,022,364 bp (99.4%)\n"
     ]
    }
   ],
   "source": [
    "%mkdir trimmed\n",
    "!for fastq in fastq/*.fastq.gz; do \\\n",
    "    base_name=$(basename \"$fastq\" .fastq.gz); \\\n",
    "    cutadapt -j 32 -q 20 -m 20 -o ./trimmed/$base_name.fastq.gz $fastq; \\\n",
    "done"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***6 minute time to completion***\n",
    "\n",
    "You can see from the cutadapt logs that very few reads were trimmed or filtered, and that's because the authors already did these steps before uploading to SRA. Not all authors do this, so we must always do QC when downloading new data to re-analyze. For our own new data, we are responsible for QC, trimming, and filtering.\n",
    "\n",
    "Re-run FastQC so we compare the metrics before and after trimming. Generate an overall report with MultiQC. This time when running, we use the `--force` option to overwrite previous reports and the `-d` option to direct the tool to specific directories."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.042s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /dev/cpuset.\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR10776762.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR10776763.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR10776764.fastq.gz\n",
      "application/gzip\n",
      "Started analysis of SRR10776765.fastq.gz\n",
      "Started analysis of SRR10776766.fastq.gz\n",
      "Started analysis of SRR10776767.fastq.gz\n",
      "Started analysis of SRR10776768.fastq.gz\n",
      "Started analysis of SRR10776769.fastq.gz\n",
      "Started analysis of SRR10776770.fastq.gz\n",
      "Started analysis of SRR10776771.fastq.gz\n",
      "Approx 5% complete for SRR10776764.fastq.gz\n",
      "Approx 5% complete for SRR10776762.fastq.gz\n",
      "Approx 5% complete for SRR10776763.fastq.gz\n",
      "Approx 5% complete for SRR10776766.fastq.gz\n",
      "Approx 5% complete for SRR10776765.fastq.gz\n",
      "Approx 5% complete for SRR10776767.fastq.gz\n",
      "Approx 5% complete for SRR10776768.fastq.gz\n",
      "Approx 5% complete for SRR10776771.fastq.gz\n",
      "Approx 5% complete for SRR10776769.fastq.gz\n",
      "Approx 5% complete for SRR10776770.fastq.gz\n",
      "Approx 10% complete for SRR10776764.fastq.gz\n",
      "Approx 10% complete for SRR10776763.fastq.gz\n",
      "Approx 10% complete for SRR10776762.fastq.gz\n",
      "Approx 10% complete for SRR10776766.fastq.gz\n",
      "Approx 10% complete for SRR10776767.fastq.gz\n",
      "Approx 10% complete for SRR10776765.fastq.gz\n",
      "Approx 10% complete for SRR10776771.fastq.gz\n",
      "Approx 10% complete for SRR10776768.fastq.gz\n",
      "Approx 10% complete for SRR10776769.fastq.gz\n",
      "Approx 10% complete for SRR10776770.fastq.gz\n",
      "Approx 15% complete for SRR10776764.fastq.gz\n",
      "Approx 15% complete for SRR10776766.fastq.gz\n",
      "Approx 15% complete for SRR10776767.fastq.gz\n",
      "Approx 15% complete for SRR10776762.fastq.gz\n",
      "Approx 15% complete for SRR10776763.fastq.gz\n",
      "Approx 15% complete for SRR10776771.fastq.gz\n",
      "Approx 15% complete for SRR10776765.fastq.gz\n",
      "Approx 15% complete for SRR10776768.fastq.gz\n",
      "Approx 15% complete for SRR10776769.fastq.gz\n",
      "Approx 15% complete for SRR10776770.fastq.gz\n",
      "Approx 20% complete for SRR10776764.fastq.gz\n",
      "Approx 20% complete for SRR10776766.fastq.gz\n",
      "Approx 20% complete for SRR10776767.fastq.gz\n",
      "Approx 20% complete for SRR10776763.fastq.gz\n",
      "Approx 20% complete for SRR10776762.fastq.gz\n",
      "Approx 20% complete for SRR10776771.fastq.gz\n",
      "Approx 20% complete for SRR10776768.fastq.gz\n",
      "Approx 20% complete for SRR10776765.fastq.gz\n",
      "Approx 20% complete for SRR10776769.fastq.gz\n",
      "Approx 20% complete for SRR10776770.fastq.gz\n",
      "Approx 25% complete for SRR10776764.fastq.gz\n",
      "Approx 25% complete for SRR10776766.fastq.gz\n",
      "Approx 25% complete for SRR10776767.fastq.gz\n",
      "Approx 25% complete for SRR10776771.fastq.gz\n",
      "Approx 25% complete for SRR10776762.fastq.gz\n",
      "Approx 25% complete for SRR10776763.fastq.gz\n",
      "Approx 25% complete for SRR10776768.fastq.gz\n",
      "Approx 25% complete for SRR10776765.fastq.gz\n",
      "Approx 25% complete for SRR10776769.fastq.gz\n",
      "Approx 30% complete for SRR10776764.fastq.gz\n",
      "Approx 25% complete for SRR10776770.fastq.gz\n",
      "Approx 30% complete for SRR10776766.fastq.gz\n",
      "Approx 30% complete for SRR10776767.fastq.gz\n",
      "Approx 30% complete for SRR10776771.fastq.gz\n",
      "Approx 30% complete for SRR10776762.fastq.gz\n",
      "Approx 30% complete for SRR10776763.fastq.gz\n",
      "Approx 30% complete for SRR10776768.fastq.gz\n",
      "Approx 30% complete for SRR10776765.fastq.gz\n",
      "Approx 35% complete for SRR10776764.fastq.gz\n",
      "Approx 30% complete for SRR10776769.fastq.gz\n",
      "Approx 30% complete for SRR10776770.fastq.gz\n",
      "Approx 35% complete for SRR10776766.fastq.gz\n",
      "Approx 35% complete for SRR10776767.fastq.gz\n",
      "Approx 35% complete for SRR10776771.fastq.gz\n",
      "Approx 35% complete for SRR10776762.fastq.gz\n",
      "Approx 35% complete for SRR10776763.fastq.gz\n",
      "Approx 35% complete for SRR10776768.fastq.gz\n",
      "Approx 35% complete for SRR10776765.fastq.gz\n",
      "Approx 40% complete for SRR10776764.fastq.gz\n",
      "Approx 35% complete for SRR10776769.fastq.gz\n",
      "Approx 40% complete for SRR10776766.fastq.gz\n",
      "Approx 35% complete for SRR10776770.fastq.gz\n",
      "Approx 40% complete for SRR10776767.fastq.gz\n",
      "Approx 40% complete for SRR10776771.fastq.gz\n",
      "Approx 40% complete for SRR10776762.fastq.gz\n",
      "Approx 40% complete for SRR10776763.fastq.gz\n",
      "Approx 40% complete for SRR10776768.fastq.gz\n",
      "Approx 45% complete for SRR10776764.fastq.gz\n",
      "Approx 40% complete for SRR10776765.fastq.gz\n",
      "Approx 45% complete for SRR10776766.fastq.gz\n",
      "Approx 40% complete for SRR10776769.fastq.gz\n",
      "Approx 45% complete for SRR10776767.fastq.gz\n",
      "Approx 40% complete for SRR10776770.fastq.gz\n",
      "Approx 45% complete for SRR10776771.fastq.gz\n",
      "Approx 45% complete for SRR10776762.fastq.gz\n",
      "Approx 45% complete for SRR10776763.fastq.gz\n",
      "Approx 50% complete for SRR10776764.fastq.gz\n",
      "Approx 45% complete for SRR10776768.fastq.gz\n",
      "Approx 45% complete for SRR10776765.fastq.gz\n",
      "Approx 50% complete for SRR10776766.fastq.gz\n",
      "Approx 50% complete for SRR10776767.fastq.gz\n",
      "Approx 50% complete for SRR10776771.fastq.gz\n",
      "Approx 45% complete for SRR10776769.fastq.gz\n",
      "Approx 45% complete for SRR10776770.fastq.gz\n",
      "Approx 50% complete for SRR10776762.fastq.gz\n",
      "Approx 55% complete for SRR10776764.fastq.gz\n",
      "Approx 50% complete for SRR10776763.fastq.gz\n",
      "Approx 50% complete for SRR10776768.fastq.gz\n",
      "Approx 55% complete for SRR10776766.fastq.gz\n",
      "Approx 50% complete for SRR10776765.fastq.gz\n",
      "Approx 55% complete for SRR10776767.fastq.gz\n",
      "Approx 55% complete for SRR10776771.fastq.gz\n",
      "Approx 50% complete for SRR10776769.fastq.gz\n",
      "Approx 50% complete for SRR10776770.fastq.gz\n",
      "Approx 60% complete for SRR10776764.fastq.gz\n",
      "Approx 55% complete for SRR10776762.fastq.gz\n",
      "Approx 55% complete for SRR10776763.fastq.gz\n",
      "Approx 55% complete for SRR10776768.fastq.gz\n",
      "Approx 60% complete for SRR10776766.fastq.gz\n",
      "Approx 60% complete for SRR10776767.fastq.gz\n",
      "Approx 60% complete for SRR10776771.fastq.gz\n",
      "Approx 55% complete for SRR10776765.fastq.gz\n",
      "Approx 55% complete for SRR10776769.fastq.gz\n",
      "Approx 65% complete for SRR10776764.fastq.gz\n",
      "Approx 60% complete for SRR10776762.fastq.gz\n",
      "Approx 55% complete for SRR10776770.fastq.gz\n",
      "Approx 60% complete for SRR10776763.fastq.gz\n",
      "Approx 60% complete for SRR10776768.fastq.gz\n",
      "Approx 65% complete for SRR10776766.fastq.gz\n",
      "Approx 65% complete for SRR10776767.fastq.gz\n",
      "Approx 65% complete for SRR10776771.fastq.gz\n",
      "Approx 60% complete for SRR10776765.fastq.gz\n",
      "Approx 70% complete for SRR10776764.fastq.gz\n",
      "Approx 60% complete for SRR10776769.fastq.gz\n",
      "Approx 65% complete for SRR10776762.fastq.gz\n",
      "Approx 65% complete for SRR10776763.fastq.gz\n",
      "Approx 60% complete for SRR10776770.fastq.gz\n",
      "Approx 70% complete for SRR10776766.fastq.gz\n",
      "Approx 65% complete for SRR10776768.fastq.gz\n",
      "Approx 70% complete for SRR10776771.fastq.gz\n",
      "Approx 70% complete for SRR10776767.fastq.gz\n",
      "Approx 65% complete for SRR10776765.fastq.gz\n",
      "Approx 75% complete for SRR10776764.fastq.gz\n",
      "Approx 65% complete for SRR10776769.fastq.gz\n",
      "Approx 70% complete for SRR10776762.fastq.gz\n",
      "Approx 70% complete for SRR10776763.fastq.gz\n",
      "Approx 75% complete for SRR10776766.fastq.gz\n",
      "Approx 70% complete for SRR10776768.fastq.gz\n",
      "Approx 65% complete for SRR10776770.fastq.gz\n",
      "Approx 75% complete for SRR10776771.fastq.gz\n",
      "Approx 75% complete for SRR10776767.fastq.gz\n",
      "Approx 80% complete for SRR10776764.fastq.gz\n",
      "Approx 70% complete for SRR10776765.fastq.gz\n",
      "Approx 70% complete for SRR10776769.fastq.gz\n",
      "Approx 75% complete for SRR10776762.fastq.gz\n",
      "Approx 80% complete for SRR10776766.fastq.gz\n",
      "Approx 75% complete for SRR10776763.fastq.gz\n",
      "Approx 75% complete for SRR10776768.fastq.gz\n",
      "Approx 80% complete for SRR10776771.fastq.gz\n",
      "Approx 80% complete for SRR10776767.fastq.gz\n",
      "Approx 70% complete for SRR10776770.fastq.gz\n",
      "Approx 85% complete for SRR10776764.fastq.gz\n",
      "Approx 75% complete for SRR10776765.fastq.gz\n",
      "Approx 85% complete for SRR10776766.fastq.gz\n",
      "Approx 75% complete for SRR10776769.fastq.gz\n",
      "Approx 80% complete for SRR10776762.fastq.gz\n",
      "Approx 80% complete for SRR10776763.fastq.gz\n",
      "Approx 85% complete for SRR10776771.fastq.gz\n",
      "Approx 80% complete for SRR10776768.fastq.gz\n",
      "Approx 85% complete for SRR10776767.fastq.gz\n",
      "Approx 75% complete for SRR10776770.fastq.gz\n",
      "Approx 90% complete for SRR10776764.fastq.gz\n",
      "Approx 80% complete for SRR10776765.fastq.gz\n",
      "Approx 90% complete for SRR10776766.fastq.gz\n",
      "Approx 85% complete for SRR10776762.fastq.gz\n",
      "Approx 80% complete for SRR10776769.fastq.gz\n",
      "Approx 85% complete for SRR10776763.fastq.gz\n",
      "Approx 90% complete for SRR10776771.fastq.gz\n",
      "Approx 85% complete for SRR10776768.fastq.gz\n",
      "Approx 90% complete for SRR10776767.fastq.gz\n",
      "Approx 95% complete for SRR10776764.fastq.gz\n",
      "Approx 80% complete for SRR10776770.fastq.gz\n",
      "Approx 95% complete for SRR10776766.fastq.gz\n",
      "Approx 85% complete for SRR10776765.fastq.gz\n",
      "Approx 90% complete for SRR10776762.fastq.gz\n",
      "Approx 85% complete for SRR10776769.fastq.gz\n",
      "Approx 95% complete for SRR10776771.fastq.gz\n",
      "Approx 90% complete for SRR10776763.fastq.gz\n",
      "Approx 90% complete for SRR10776768.fastq.gz\n",
      "Approx 95% complete for SRR10776767.fastq.gz\n",
      "Analysis complete for SRR10776764.fastq.gz\n",
      "Approx 85% complete for SRR10776770.fastq.gz\n",
      "Analysis complete for SRR10776766.fastq.gz\n",
      "Approx 95% complete for SRR10776762.fastq.gz\n",
      "Approx 95% complete for SRR10776763.fastq.gz\n",
      "Approx 90% complete for SRR10776765.fastq.gz\n",
      "Analysis complete for SRR10776771.fastq.gz\n",
      "Approx 90% complete for SRR10776769.fastq.gz\n",
      "Analysis complete for SRR10776767.fastq.gz\n",
      "Approx 95% complete for SRR10776768.fastq.gz\n",
      "Approx 90% complete for SRR10776770.fastq.gz\n",
      "Analysis complete for SRR10776762.fastq.gz\n",
      "Analysis complete for SRR10776763.fastq.gz\n",
      "Approx 95% complete for SRR10776765.fastq.gz\n",
      "Approx 95% complete for SRR10776769.fastq.gz\n",
      "Analysis complete for SRR10776768.fastq.gz\n",
      "Approx 95% complete for SRR10776770.fastq.gz\n",
      "Analysis complete for SRR10776765.fastq.gz\n",
      "Analysis complete for SRR10776769.fastq.gz\n",
      "Analysis complete for SRR10776770.fastq.gz\n",
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=197387;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.17\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Prepending directory to sample names\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/heky1803/BIOL343/5_fastq/fastq/qc\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/heky1803/BIOL343/5_fastq/trimmed/qc\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m40/40\u001b[0m  /qc/SRR10776765_fastqc.html\u001b[0m.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 20 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc_report.html   (overwritten)\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc_data   (overwritten)\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "!mkdir trimmed/qc\n",
    "!fastqc -t 32 trimmed/*.fastq.gz -o trimmed/qc\n",
    "!multiqc --force -d fastq/qc/ trimmed/qc/"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***3 minute time to completion***\n",
    "\n",
    "Check out the new MultiQC report to see what changed. Not a whole lot, but in some samples (i.e., SRR26691093) the PolyA tail was substantially trimmed. Our trimming did not deal with UMIs or linkers - we'll talk more about those when we get to alignment next week. You can open the individual QC files to inspect them further.\n",
    "\n",
    "You'll notice that the \"Sequence Length Distribution\" sections often went from  to . That's because the trimmed PolyA tail was different for each read, so now we have a broad distribution of lengths rather than all of them being 69 nt. \n",
    "\n",
    "Even though we don't have adapters, we still have have overrepresented sequences. What are they? BLAST them against a nucleotide database and see if they are concerning or not."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
