{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 5 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 10/17 by 11:59 pm***"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTQ fetch, QC, and trimming/filtering\n",
    "\n",
    "Use the SRA-toolkit to fetch the relevant FASTQ files. You may use a hand-crafted `txt` file to provide the list of SRAs instead of manipulating an SRA Run Table with `cut`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting ERR1442631 from NCBI SRA\n",
      "2025-10-14 21:21:31,360 - SRR ids: ['ERR1442631']\n",
      "2025-10-14 21:21:31,360 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 21:21:31,364 - tempdir: /local/scratch/job_322458/pfd_tnla3xhm\n",
      "2025-10-14 21:21:31,364 - CMD: sra-stat --meta --quick ERR1442631\n",
      "2025-10-14 21:21:38,782 - ERR1442631 spots: 1785080\n",
      "2025-10-14 21:21:38,782 - blocks: [[1, 55783], [55784, 111566], [111567, 167349], [167350, 223132], [223133, 278915], [278916, 334698], [334699, 390481], [390482, 446264], [446265, 502047], [502048, 557830], [557831, 613613], [613614, 669396], [669397, 725179], [725180, 780962], [780963, 836745], [836746, 892528], [892529, 948311], [948312, 1004094], [1004095, 1059877], [1059878, 1115660], [1115661, 1171443], [1171444, 1227226], [1227227, 1283009], [1283010, 1338792], [1338793, 1394575], [1394576, 1450358], [1450359, 1506141], [1506142, 1561924], [1561925, 1617707], [1617708, 1673490], [1673491, 1729273], [1729274, 1785080]]\n",
      "2025-10-14 21:21:38,783 - CMD: fastq-dump -N 1 -X 55783 -O /local/scratch/job_322458/pfd_tnla3xhm/0 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,853 - CMD: fastq-dump -N 55784 -X 111566 -O /local/scratch/job_322458/pfd_tnla3xhm/1 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,854 - CMD: fastq-dump -N 111567 -X 167349 -O /local/scratch/job_322458/pfd_tnla3xhm/2 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,854 - CMD: fastq-dump -N 167350 -X 223132 -O /local/scratch/job_322458/pfd_tnla3xhm/3 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,855 - CMD: fastq-dump -N 223133 -X 278915 -O /local/scratch/job_322458/pfd_tnla3xhm/4 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,855 - CMD: fastq-dump -N 278916 -X 334698 -O /local/scratch/job_322458/pfd_tnla3xhm/5 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,855 - CMD: fastq-dump -N 334699 -X 390481 -O /local/scratch/job_322458/pfd_tnla3xhm/6 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,855 - CMD: fastq-dump -N 390482 -X 446264 -O /local/scratch/job_322458/pfd_tnla3xhm/7 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,856 - CMD: fastq-dump -N 446265 -X 502047 -O /local/scratch/job_322458/pfd_tnla3xhm/8 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,856 - CMD: fastq-dump -N 502048 -X 557830 -O /local/scratch/job_322458/pfd_tnla3xhm/9 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,856 - CMD: fastq-dump -N 557831 -X 613613 -O /local/scratch/job_322458/pfd_tnla3xhm/10 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,857 - CMD: fastq-dump -N 613614 -X 669396 -O /local/scratch/job_322458/pfd_tnla3xhm/11 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,857 - CMD: fastq-dump -N 669397 -X 725179 -O /local/scratch/job_322458/pfd_tnla3xhm/12 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,857 - CMD: fastq-dump -N 725180 -X 780962 -O /local/scratch/job_322458/pfd_tnla3xhm/13 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,857 - CMD: fastq-dump -N 780963 -X 836745 -O /local/scratch/job_322458/pfd_tnla3xhm/14 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,858 - CMD: fastq-dump -N 836746 -X 892528 -O /local/scratch/job_322458/pfd_tnla3xhm/15 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,858 - CMD: fastq-dump -N 892529 -X 948311 -O /local/scratch/job_322458/pfd_tnla3xhm/16 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,858 - CMD: fastq-dump -N 948312 -X 1004094 -O /local/scratch/job_322458/pfd_tnla3xhm/17 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,859 - CMD: fastq-dump -N 1004095 -X 1059877 -O /local/scratch/job_322458/pfd_tnla3xhm/18 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,859 - CMD: fastq-dump -N 1059878 -X 1115660 -O /local/scratch/job_322458/pfd_tnla3xhm/19 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,859 - CMD: fastq-dump -N 1115661 -X 1171443 -O /local/scratch/job_322458/pfd_tnla3xhm/20 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,859 - CMD: fastq-dump -N 1171444 -X 1227226 -O /local/scratch/job_322458/pfd_tnla3xhm/21 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,860 - CMD: fastq-dump -N 1227227 -X 1283009 -O /local/scratch/job_322458/pfd_tnla3xhm/22 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,860 - CMD: fastq-dump -N 1283010 -X 1338792 -O /local/scratch/job_322458/pfd_tnla3xhm/23 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,860 - CMD: fastq-dump -N 1338793 -X 1394575 -O /local/scratch/job_322458/pfd_tnla3xhm/24 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,860 - CMD: fastq-dump -N 1394576 -X 1450358 -O /local/scratch/job_322458/pfd_tnla3xhm/25 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,861 - CMD: fastq-dump -N 1450359 -X 1506141 -O /local/scratch/job_322458/pfd_tnla3xhm/26 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,861 - CMD: fastq-dump -N 1506142 -X 1561924 -O /local/scratch/job_322458/pfd_tnla3xhm/27 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,861 - CMD: fastq-dump -N 1561925 -X 1617707 -O /local/scratch/job_322458/pfd_tnla3xhm/28 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,861 - CMD: fastq-dump -N 1617708 -X 1673490 -O /local/scratch/job_322458/pfd_tnla3xhm/29 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,862 - CMD: fastq-dump -N 1673491 -X 1729273 -O /local/scratch/job_322458/pfd_tnla3xhm/30 --split-files --gzip ERR1442631\n",
      "2025-10-14 21:21:38,862 - CMD: fastq-dump -N 1729274 -X 1785080 -O /local/scratch/job_322458/pfd_tnla3xhm/31 --split-files --gzip ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55807 spots for ERR1442631\n",
      "Written 55807 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Read 55783 spots for ERR1442631\n",
      "Written 55783 spots for ERR1442631\n",
      "Getting ERR1442632 from NCBI SRA\n",
      "2025-10-14 21:22:40,445 - SRR ids: ['ERR1442632']\n",
      "2025-10-14 21:22:40,445 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 21:22:40,445 - tempdir: /local/scratch/job_322458/pfd_8kk7vldx\n",
      "2025-10-14 21:22:40,445 - CMD: sra-stat --meta --quick ERR1442632\n",
      "2025-10-14 21:22:45,580 - ERR1442632 spots: 1601181\n",
      "2025-10-14 21:22:45,580 - blocks: [[1, 50036], [50037, 100072], [100073, 150108], [150109, 200144], [200145, 250180], [250181, 300216], [300217, 350252], [350253, 400288], [400289, 450324], [450325, 500360], [500361, 550396], [550397, 600432], [600433, 650468], [650469, 700504], [700505, 750540], [750541, 800576], [800577, 850612], [850613, 900648], [900649, 950684], [950685, 1000720], [1000721, 1050756], [1050757, 1100792], [1100793, 1150828], [1150829, 1200864], [1200865, 1250900], [1250901, 1300936], [1300937, 1350972], [1350973, 1401008], [1401009, 1451044], [1451045, 1501080], [1501081, 1551116], [1551117, 1601181]]\n",
      "2025-10-14 21:22:45,580 - CMD: fastq-dump -N 1 -X 50036 -O /local/scratch/job_322458/pfd_8kk7vldx/0 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,684 - CMD: fastq-dump -N 50037 -X 100072 -O /local/scratch/job_322458/pfd_8kk7vldx/1 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,685 - CMD: fastq-dump -N 100073 -X 150108 -O /local/scratch/job_322458/pfd_8kk7vldx/2 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,685 - CMD: fastq-dump -N 150109 -X 200144 -O /local/scratch/job_322458/pfd_8kk7vldx/3 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,685 - CMD: fastq-dump -N 200145 -X 250180 -O /local/scratch/job_322458/pfd_8kk7vldx/4 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,686 - CMD: fastq-dump -N 250181 -X 300216 -O /local/scratch/job_322458/pfd_8kk7vldx/5 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,686 - CMD: fastq-dump -N 300217 -X 350252 -O /local/scratch/job_322458/pfd_8kk7vldx/6 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,686 - CMD: fastq-dump -N 350253 -X 400288 -O /local/scratch/job_322458/pfd_8kk7vldx/7 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,687 - CMD: fastq-dump -N 400289 -X 450324 -O /local/scratch/job_322458/pfd_8kk7vldx/8 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,687 - CMD: fastq-dump -N 450325 -X 500360 -O /local/scratch/job_322458/pfd_8kk7vldx/9 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,687 - CMD: fastq-dump -N 500361 -X 550396 -O /local/scratch/job_322458/pfd_8kk7vldx/10 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,688 - CMD: fastq-dump -N 550397 -X 600432 -O /local/scratch/job_322458/pfd_8kk7vldx/11 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,688 - CMD: fastq-dump -N 600433 -X 650468 -O /local/scratch/job_322458/pfd_8kk7vldx/12 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,688 - CMD: fastq-dump -N 650469 -X 700504 -O /local/scratch/job_322458/pfd_8kk7vldx/13 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,689 - CMD: fastq-dump -N 700505 -X 750540 -O /local/scratch/job_322458/pfd_8kk7vldx/14 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,689 - CMD: fastq-dump -N 750541 -X 800576 -O /local/scratch/job_322458/pfd_8kk7vldx/15 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,689 - CMD: fastq-dump -N 800577 -X 850612 -O /local/scratch/job_322458/pfd_8kk7vldx/16 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,690 - CMD: fastq-dump -N 850613 -X 900648 -O /local/scratch/job_322458/pfd_8kk7vldx/17 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,690 - CMD: fastq-dump -N 900649 -X 950684 -O /local/scratch/job_322458/pfd_8kk7vldx/18 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,690 - CMD: fastq-dump -N 950685 -X 1000720 -O /local/scratch/job_322458/pfd_8kk7vldx/19 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,690 - CMD: fastq-dump -N 1000721 -X 1050756 -O /local/scratch/job_322458/pfd_8kk7vldx/20 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,691 - CMD: fastq-dump -N 1050757 -X 1100792 -O /local/scratch/job_322458/pfd_8kk7vldx/21 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,691 - CMD: fastq-dump -N 1100793 -X 1150828 -O /local/scratch/job_322458/pfd_8kk7vldx/22 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,691 - CMD: fastq-dump -N 1150829 -X 1200864 -O /local/scratch/job_322458/pfd_8kk7vldx/23 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,692 - CMD: fastq-dump -N 1200865 -X 1250900 -O /local/scratch/job_322458/pfd_8kk7vldx/24 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,692 - CMD: fastq-dump -N 1250901 -X 1300936 -O /local/scratch/job_322458/pfd_8kk7vldx/25 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,692 - CMD: fastq-dump -N 1300937 -X 1350972 -O /local/scratch/job_322458/pfd_8kk7vldx/26 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,692 - CMD: fastq-dump -N 1350973 -X 1401008 -O /local/scratch/job_322458/pfd_8kk7vldx/27 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,693 - CMD: fastq-dump -N 1401009 -X 1451044 -O /local/scratch/job_322458/pfd_8kk7vldx/28 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,693 - CMD: fastq-dump -N 1451045 -X 1501080 -O /local/scratch/job_322458/pfd_8kk7vldx/29 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,693 - CMD: fastq-dump -N 1501081 -X 1551116 -O /local/scratch/job_322458/pfd_8kk7vldx/30 --split-files --gzip ERR1442632\n",
      "2025-10-14 21:22:45,694 - CMD: fastq-dump -N 1551117 -X 1601181 -O /local/scratch/job_322458/pfd_8kk7vldx/31 --split-files --gzip ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50065 spots for ERR1442632\n",
      "Written 50065 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Read 50036 spots for ERR1442632\n",
      "Written 50036 spots for ERR1442632\n",
      "Getting ERR1442633 from NCBI SRA\n",
      "2025-10-14 21:23:44,419 - SRR ids: ['ERR1442633']\n",
      "2025-10-14 21:23:44,420 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 21:23:44,435 - tempdir: /local/scratch/job_322458/pfd_ebtmgnb1\n",
      "2025-10-14 21:23:44,435 - CMD: sra-stat --meta --quick ERR1442633\n",
      "2025-10-14 21:23:50,047 - ERR1442633 spots: 1912345\n",
      "2025-10-14 21:23:50,047 - blocks: [[1, 59760], [59761, 119520], [119521, 179280], [179281, 239040], [239041, 298800], [298801, 358560], [358561, 418320], [418321, 478080], [478081, 537840], [537841, 597600], [597601, 657360], [657361, 717120], [717121, 776880], [776881, 836640], [836641, 896400], [896401, 956160], [956161, 1015920], [1015921, 1075680], [1075681, 1135440], [1135441, 1195200], [1195201, 1254960], [1254961, 1314720], [1314721, 1374480], [1374481, 1434240], [1434241, 1494000], [1494001, 1553760], [1553761, 1613520], [1613521, 1673280], [1673281, 1733040], [1733041, 1792800], [1792801, 1852560], [1852561, 1912345]]\n",
      "2025-10-14 21:23:50,047 - CMD: fastq-dump -N 1 -X 59760 -O /local/scratch/job_322458/pfd_ebtmgnb1/0 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,054 - CMD: fastq-dump -N 59761 -X 119520 -O /local/scratch/job_322458/pfd_ebtmgnb1/1 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,054 - CMD: fastq-dump -N 119521 -X 179280 -O /local/scratch/job_322458/pfd_ebtmgnb1/2 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,054 - CMD: fastq-dump -N 179281 -X 239040 -O /local/scratch/job_322458/pfd_ebtmgnb1/3 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,055 - CMD: fastq-dump -N 239041 -X 298800 -O /local/scratch/job_322458/pfd_ebtmgnb1/4 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,055 - CMD: fastq-dump -N 298801 -X 358560 -O /local/scratch/job_322458/pfd_ebtmgnb1/5 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,056 - CMD: fastq-dump -N 358561 -X 418320 -O /local/scratch/job_322458/pfd_ebtmgnb1/6 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,056 - CMD: fastq-dump -N 418321 -X 478080 -O /local/scratch/job_322458/pfd_ebtmgnb1/7 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,057 - CMD: fastq-dump -N 478081 -X 537840 -O /local/scratch/job_322458/pfd_ebtmgnb1/8 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,057 - CMD: fastq-dump -N 537841 -X 597600 -O /local/scratch/job_322458/pfd_ebtmgnb1/9 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,057 - CMD: fastq-dump -N 597601 -X 657360 -O /local/scratch/job_322458/pfd_ebtmgnb1/10 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,057 - CMD: fastq-dump -N 657361 -X 717120 -O /local/scratch/job_322458/pfd_ebtmgnb1/11 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,058 - CMD: fastq-dump -N 717121 -X 776880 -O /local/scratch/job_322458/pfd_ebtmgnb1/12 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,058 - CMD: fastq-dump -N 776881 -X 836640 -O /local/scratch/job_322458/pfd_ebtmgnb1/13 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,058 - CMD: fastq-dump -N 836641 -X 896400 -O /local/scratch/job_322458/pfd_ebtmgnb1/14 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,059 - CMD: fastq-dump -N 896401 -X 956160 -O /local/scratch/job_322458/pfd_ebtmgnb1/15 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,059 - CMD: fastq-dump -N 956161 -X 1015920 -O /local/scratch/job_322458/pfd_ebtmgnb1/16 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,059 - CMD: fastq-dump -N 1015921 -X 1075680 -O /local/scratch/job_322458/pfd_ebtmgnb1/17 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,059 - CMD: fastq-dump -N 1075681 -X 1135440 -O /local/scratch/job_322458/pfd_ebtmgnb1/18 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,060 - CMD: fastq-dump -N 1135441 -X 1195200 -O /local/scratch/job_322458/pfd_ebtmgnb1/19 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,060 - CMD: fastq-dump -N 1195201 -X 1254960 -O /local/scratch/job_322458/pfd_ebtmgnb1/20 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,060 - CMD: fastq-dump -N 1254961 -X 1314720 -O /local/scratch/job_322458/pfd_ebtmgnb1/21 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,061 - CMD: fastq-dump -N 1314721 -X 1374480 -O /local/scratch/job_322458/pfd_ebtmgnb1/22 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,061 - CMD: fastq-dump -N 1374481 -X 1434240 -O /local/scratch/job_322458/pfd_ebtmgnb1/23 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,061 - CMD: fastq-dump -N 1434241 -X 1494000 -O /local/scratch/job_322458/pfd_ebtmgnb1/24 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,062 - CMD: fastq-dump -N 1494001 -X 1553760 -O /local/scratch/job_322458/pfd_ebtmgnb1/25 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,062 - CMD: fastq-dump -N 1553761 -X 1613520 -O /local/scratch/job_322458/pfd_ebtmgnb1/26 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,062 - CMD: fastq-dump -N 1613521 -X 1673280 -O /local/scratch/job_322458/pfd_ebtmgnb1/27 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,062 - CMD: fastq-dump -N 1673281 -X 1733040 -O /local/scratch/job_322458/pfd_ebtmgnb1/28 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,063 - CMD: fastq-dump -N 1733041 -X 1792800 -O /local/scratch/job_322458/pfd_ebtmgnb1/29 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,063 - CMD: fastq-dump -N 1792801 -X 1852560 -O /local/scratch/job_322458/pfd_ebtmgnb1/30 --split-files --gzip ERR1442633\n",
      "2025-10-14 21:23:50,063 - CMD: fastq-dump -N 1852561 -X 1912345 -O /local/scratch/job_322458/pfd_ebtmgnb1/31 --split-files --gzip ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59785 spots for ERR1442633\n",
      "Written 59785 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Read 59760 spots for ERR1442633\n",
      "Written 59760 spots for ERR1442633\n",
      "Getting ERR1442641 from NCBI SRA\n",
      "2025-10-14 21:24:53,628 - SRR ids: ['ERR1442641']\n",
      "2025-10-14 21:24:53,628 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 21:24:53,628 - tempdir: /local/scratch/job_322458/pfd_yh3btgbh\n",
      "2025-10-14 21:24:53,628 - CMD: sra-stat --meta --quick ERR1442641\n",
      "2025-10-14 21:24:58,646 - ERR1442641 spots: 2116966\n",
      "2025-10-14 21:24:58,646 - blocks: [[1, 66155], [66156, 132310], [132311, 198465], [198466, 264620], [264621, 330775], [330776, 396930], [396931, 463085], [463086, 529240], [529241, 595395], [595396, 661550], [661551, 727705], [727706, 793860], [793861, 860015], [860016, 926170], [926171, 992325], [992326, 1058480], [1058481, 1124635], [1124636, 1190790], [1190791, 1256945], [1256946, 1323100], [1323101, 1389255], [1389256, 1455410], [1455411, 1521565], [1521566, 1587720], [1587721, 1653875], [1653876, 1720030], [1720031, 1786185], [1786186, 1852340], [1852341, 1918495], [1918496, 1984650], [1984651, 2050805], [2050806, 2116966]]\n",
      "2025-10-14 21:24:58,646 - CMD: fastq-dump -N 1 -X 66155 -O /local/scratch/job_322458/pfd_yh3btgbh/0 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,648 - CMD: fastq-dump -N 66156 -X 132310 -O /local/scratch/job_322458/pfd_yh3btgbh/1 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,649 - CMD: fastq-dump -N 132311 -X 198465 -O /local/scratch/job_322458/pfd_yh3btgbh/2 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,649 - CMD: fastq-dump -N 198466 -X 264620 -O /local/scratch/job_322458/pfd_yh3btgbh/3 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,649 - CMD: fastq-dump -N 264621 -X 330775 -O /local/scratch/job_322458/pfd_yh3btgbh/4 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,650 - CMD: fastq-dump -N 330776 -X 396930 -O /local/scratch/job_322458/pfd_yh3btgbh/5 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,650 - CMD: fastq-dump -N 396931 -X 463085 -O /local/scratch/job_322458/pfd_yh3btgbh/6 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,650 - CMD: fastq-dump -N 463086 -X 529240 -O /local/scratch/job_322458/pfd_yh3btgbh/7 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,651 - CMD: fastq-dump -N 529241 -X 595395 -O /local/scratch/job_322458/pfd_yh3btgbh/8 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,651 - CMD: fastq-dump -N 595396 -X 661550 -O /local/scratch/job_322458/pfd_yh3btgbh/9 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,651 - CMD: fastq-dump -N 661551 -X 727705 -O /local/scratch/job_322458/pfd_yh3btgbh/10 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,651 - CMD: fastq-dump -N 727706 -X 793860 -O /local/scratch/job_322458/pfd_yh3btgbh/11 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,652 - CMD: fastq-dump -N 793861 -X 860015 -O /local/scratch/job_322458/pfd_yh3btgbh/12 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,652 - CMD: fastq-dump -N 860016 -X 926170 -O /local/scratch/job_322458/pfd_yh3btgbh/13 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,652 - CMD: fastq-dump -N 926171 -X 992325 -O /local/scratch/job_322458/pfd_yh3btgbh/14 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,653 - CMD: fastq-dump -N 992326 -X 1058480 -O /local/scratch/job_322458/pfd_yh3btgbh/15 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,653 - CMD: fastq-dump -N 1058481 -X 1124635 -O /local/scratch/job_322458/pfd_yh3btgbh/16 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,653 - CMD: fastq-dump -N 1124636 -X 1190790 -O /local/scratch/job_322458/pfd_yh3btgbh/17 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,654 - CMD: fastq-dump -N 1190791 -X 1256945 -O /local/scratch/job_322458/pfd_yh3btgbh/18 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,654 - CMD: fastq-dump -N 1256946 -X 1323100 -O /local/scratch/job_322458/pfd_yh3btgbh/19 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,654 - CMD: fastq-dump -N 1323101 -X 1389255 -O /local/scratch/job_322458/pfd_yh3btgbh/20 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,654 - CMD: fastq-dump -N 1389256 -X 1455410 -O /local/scratch/job_322458/pfd_yh3btgbh/21 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,655 - CMD: fastq-dump -N 1455411 -X 1521565 -O /local/scratch/job_322458/pfd_yh3btgbh/22 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,655 - CMD: fastq-dump -N 1521566 -X 1587720 -O /local/scratch/job_322458/pfd_yh3btgbh/23 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,655 - CMD: fastq-dump -N 1587721 -X 1653875 -O /local/scratch/job_322458/pfd_yh3btgbh/24 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,656 - CMD: fastq-dump -N 1653876 -X 1720030 -O /local/scratch/job_322458/pfd_yh3btgbh/25 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,656 - CMD: fastq-dump -N 1720031 -X 1786185 -O /local/scratch/job_322458/pfd_yh3btgbh/26 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,656 - CMD: fastq-dump -N 1786186 -X 1852340 -O /local/scratch/job_322458/pfd_yh3btgbh/27 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,656 - CMD: fastq-dump -N 1852341 -X 1918495 -O /local/scratch/job_322458/pfd_yh3btgbh/28 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,657 - CMD: fastq-dump -N 1918496 -X 1984650 -O /local/scratch/job_322458/pfd_yh3btgbh/29 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,657 - CMD: fastq-dump -N 1984651 -X 2050805 -O /local/scratch/job_322458/pfd_yh3btgbh/30 --split-files --gzip ERR1442641\n",
      "2025-10-14 21:24:58,657 - CMD: fastq-dump -N 2050806 -X 2116966 -O /local/scratch/job_322458/pfd_yh3btgbh/31 --split-files --gzip ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66161 spots for ERR1442641\n",
      "Written 66161 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Read 66155 spots for ERR1442641\n",
      "Written 66155 spots for ERR1442641\n",
      "Getting ERR1442642 from NCBI SRA\n",
      "2025-10-14 21:26:01,633 - SRR ids: ['ERR1442642']\n",
      "2025-10-14 21:26:01,633 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 21:26:01,633 - tempdir: /local/scratch/job_322458/pfd_y_rd6z37\n",
      "2025-10-14 21:26:01,633 - CMD: sra-stat --meta --quick ERR1442642\n",
      "2025-10-14 21:26:05,995 - ERR1442642 spots: 2090906\n",
      "2025-10-14 21:26:05,995 - blocks: [[1, 65340], [65341, 130680], [130681, 196020], [196021, 261360], [261361, 326700], [326701, 392040], [392041, 457380], [457381, 522720], [522721, 588060], [588061, 653400], [653401, 718740], [718741, 784080], [784081, 849420], [849421, 914760], [914761, 980100], [980101, 1045440], [1045441, 1110780], [1110781, 1176120], [1176121, 1241460], [1241461, 1306800], [1306801, 1372140], [1372141, 1437480], [1437481, 1502820], [1502821, 1568160], [1568161, 1633500], [1633501, 1698840], [1698841, 1764180], [1764181, 1829520], [1829521, 1894860], [1894861, 1960200], [1960201, 2025540], [2025541, 2090906]]\n",
      "2025-10-14 21:26:05,995 - CMD: fastq-dump -N 1 -X 65340 -O /local/scratch/job_322458/pfd_y_rd6z37/0 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,003 - CMD: fastq-dump -N 65341 -X 130680 -O /local/scratch/job_322458/pfd_y_rd6z37/1 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,003 - CMD: fastq-dump -N 130681 -X 196020 -O /local/scratch/job_322458/pfd_y_rd6z37/2 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,004 - CMD: fastq-dump -N 196021 -X 261360 -O /local/scratch/job_322458/pfd_y_rd6z37/3 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,004 - CMD: fastq-dump -N 261361 -X 326700 -O /local/scratch/job_322458/pfd_y_rd6z37/4 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,004 - CMD: fastq-dump -N 326701 -X 392040 -O /local/scratch/job_322458/pfd_y_rd6z37/5 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,005 - CMD: fastq-dump -N 392041 -X 457380 -O /local/scratch/job_322458/pfd_y_rd6z37/6 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,005 - CMD: fastq-dump -N 457381 -X 522720 -O /local/scratch/job_322458/pfd_y_rd6z37/7 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,006 - CMD: fastq-dump -N 522721 -X 588060 -O /local/scratch/job_322458/pfd_y_rd6z37/8 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,006 - CMD: fastq-dump -N 588061 -X 653400 -O /local/scratch/job_322458/pfd_y_rd6z37/9 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,006 - CMD: fastq-dump -N 653401 -X 718740 -O /local/scratch/job_322458/pfd_y_rd6z37/10 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,007 - CMD: fastq-dump -N 718741 -X 784080 -O /local/scratch/job_322458/pfd_y_rd6z37/11 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,007 - CMD: fastq-dump -N 784081 -X 849420 -O /local/scratch/job_322458/pfd_y_rd6z37/12 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,007 - CMD: fastq-dump -N 849421 -X 914760 -O /local/scratch/job_322458/pfd_y_rd6z37/13 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,007 - CMD: fastq-dump -N 914761 -X 980100 -O /local/scratch/job_322458/pfd_y_rd6z37/14 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,008 - CMD: fastq-dump -N 980101 -X 1045440 -O /local/scratch/job_322458/pfd_y_rd6z37/15 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,008 - CMD: fastq-dump -N 1045441 -X 1110780 -O /local/scratch/job_322458/pfd_y_rd6z37/16 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,008 - CMD: fastq-dump -N 1110781 -X 1176120 -O /local/scratch/job_322458/pfd_y_rd6z37/17 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,009 - CMD: fastq-dump -N 1176121 -X 1241460 -O /local/scratch/job_322458/pfd_y_rd6z37/18 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,009 - CMD: fastq-dump -N 1241461 -X 1306800 -O /local/scratch/job_322458/pfd_y_rd6z37/19 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,009 - CMD: fastq-dump -N 1306801 -X 1372140 -O /local/scratch/job_322458/pfd_y_rd6z37/20 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,009 - CMD: fastq-dump -N 1372141 -X 1437480 -O /local/scratch/job_322458/pfd_y_rd6z37/21 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,010 - CMD: fastq-dump -N 1437481 -X 1502820 -O /local/scratch/job_322458/pfd_y_rd6z37/22 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,010 - CMD: fastq-dump -N 1502821 -X 1568160 -O /local/scratch/job_322458/pfd_y_rd6z37/23 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,010 - CMD: fastq-dump -N 1568161 -X 1633500 -O /local/scratch/job_322458/pfd_y_rd6z37/24 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,011 - CMD: fastq-dump -N 1633501 -X 1698840 -O /local/scratch/job_322458/pfd_y_rd6z37/25 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,011 - CMD: fastq-dump -N 1698841 -X 1764180 -O /local/scratch/job_322458/pfd_y_rd6z37/26 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,011 - CMD: fastq-dump -N 1764181 -X 1829520 -O /local/scratch/job_322458/pfd_y_rd6z37/27 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,012 - CMD: fastq-dump -N 1829521 -X 1894860 -O /local/scratch/job_322458/pfd_y_rd6z37/28 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,012 - CMD: fastq-dump -N 1894861 -X 1960200 -O /local/scratch/job_322458/pfd_y_rd6z37/29 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,012 - CMD: fastq-dump -N 1960201 -X 2025540 -O /local/scratch/job_322458/pfd_y_rd6z37/30 --split-files --gzip ERR1442642\n",
      "2025-10-14 21:26:06,012 - CMD: fastq-dump -N 2025541 -X 2090906 -O /local/scratch/job_322458/pfd_y_rd6z37/31 --split-files --gzip ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65366 spots for ERR1442642\n",
      "Written 65366 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Read 65340 spots for ERR1442642\n",
      "Written 65340 spots for ERR1442642\n",
      "Getting ERR1442643 from NCBI SRA\n",
      "2025-10-14 21:27:05,121 - SRR ids: ['ERR1442643']\n",
      "2025-10-14 21:27:05,121 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 21:27:05,122 - tempdir: /local/scratch/job_322458/pfd_140xhgtd\n",
      "2025-10-14 21:27:05,122 - CMD: sra-stat --meta --quick ERR1442643\n",
      "2025-10-14 21:27:09,287 - ERR1442643 spots: 1563602\n",
      "2025-10-14 21:27:09,287 - blocks: [[1, 48862], [48863, 97724], [97725, 146586], [146587, 195448], [195449, 244310], [244311, 293172], [293173, 342034], [342035, 390896], [390897, 439758], [439759, 488620], [488621, 537482], [537483, 586344], [586345, 635206], [635207, 684068], [684069, 732930], [732931, 781792], [781793, 830654], [830655, 879516], [879517, 928378], [928379, 977240], [977241, 1026102], [1026103, 1074964], [1074965, 1123826], [1123827, 1172688], [1172689, 1221550], [1221551, 1270412], [1270413, 1319274], [1319275, 1368136], [1368137, 1416998], [1416999, 1465860], [1465861, 1514722], [1514723, 1563602]]\n",
      "2025-10-14 21:27:09,287 - CMD: fastq-dump -N 1 -X 48862 -O /local/scratch/job_322458/pfd_140xhgtd/0 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,288 - CMD: fastq-dump -N 48863 -X 97724 -O /local/scratch/job_322458/pfd_140xhgtd/1 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,289 - CMD: fastq-dump -N 97725 -X 146586 -O /local/scratch/job_322458/pfd_140xhgtd/2 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,289 - CMD: fastq-dump -N 146587 -X 195448 -O /local/scratch/job_322458/pfd_140xhgtd/3 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,290 - CMD: fastq-dump -N 195449 -X 244310 -O /local/scratch/job_322458/pfd_140xhgtd/4 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,290 - CMD: fastq-dump -N 244311 -X 293172 -O /local/scratch/job_322458/pfd_140xhgtd/5 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,290 - CMD: fastq-dump -N 293173 -X 342034 -O /local/scratch/job_322458/pfd_140xhgtd/6 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,291 - CMD: fastq-dump -N 342035 -X 390896 -O /local/scratch/job_322458/pfd_140xhgtd/7 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,291 - CMD: fastq-dump -N 390897 -X 439758 -O /local/scratch/job_322458/pfd_140xhgtd/8 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,291 - CMD: fastq-dump -N 439759 -X 488620 -O /local/scratch/job_322458/pfd_140xhgtd/9 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,291 - CMD: fastq-dump -N 488621 -X 537482 -O /local/scratch/job_322458/pfd_140xhgtd/10 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,292 - CMD: fastq-dump -N 537483 -X 586344 -O /local/scratch/job_322458/pfd_140xhgtd/11 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,292 - CMD: fastq-dump -N 586345 -X 635206 -O /local/scratch/job_322458/pfd_140xhgtd/12 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,292 - CMD: fastq-dump -N 635207 -X 684068 -O /local/scratch/job_322458/pfd_140xhgtd/13 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,293 - CMD: fastq-dump -N 684069 -X 732930 -O /local/scratch/job_322458/pfd_140xhgtd/14 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,293 - CMD: fastq-dump -N 732931 -X 781792 -O /local/scratch/job_322458/pfd_140xhgtd/15 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,293 - CMD: fastq-dump -N 781793 -X 830654 -O /local/scratch/job_322458/pfd_140xhgtd/16 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,294 - CMD: fastq-dump -N 830655 -X 879516 -O /local/scratch/job_322458/pfd_140xhgtd/17 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,294 - CMD: fastq-dump -N 879517 -X 928378 -O /local/scratch/job_322458/pfd_140xhgtd/18 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,294 - CMD: fastq-dump -N 928379 -X 977240 -O /local/scratch/job_322458/pfd_140xhgtd/19 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,294 - CMD: fastq-dump -N 977241 -X 1026102 -O /local/scratch/job_322458/pfd_140xhgtd/20 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,295 - CMD: fastq-dump -N 1026103 -X 1074964 -O /local/scratch/job_322458/pfd_140xhgtd/21 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,295 - CMD: fastq-dump -N 1074965 -X 1123826 -O /local/scratch/job_322458/pfd_140xhgtd/22 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,295 - CMD: fastq-dump -N 1123827 -X 1172688 -O /local/scratch/job_322458/pfd_140xhgtd/23 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,296 - CMD: fastq-dump -N 1172689 -X 1221550 -O /local/scratch/job_322458/pfd_140xhgtd/24 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,296 - CMD: fastq-dump -N 1221551 -X 1270412 -O /local/scratch/job_322458/pfd_140xhgtd/25 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,296 - CMD: fastq-dump -N 1270413 -X 1319274 -O /local/scratch/job_322458/pfd_140xhgtd/26 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,296 - CMD: fastq-dump -N 1319275 -X 1368136 -O /local/scratch/job_322458/pfd_140xhgtd/27 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,297 - CMD: fastq-dump -N 1368137 -X 1416998 -O /local/scratch/job_322458/pfd_140xhgtd/28 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,297 - CMD: fastq-dump -N 1416999 -X 1465860 -O /local/scratch/job_322458/pfd_140xhgtd/29 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,297 - CMD: fastq-dump -N 1465861 -X 1514722 -O /local/scratch/job_322458/pfd_140xhgtd/30 --split-files --gzip ERR1442643\n",
      "2025-10-14 21:27:09,298 - CMD: fastq-dump -N 1514723 -X 1563602 -O /local/scratch/job_322458/pfd_140xhgtd/31 --split-files --gzip ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48880 spots for ERR1442643\n",
      "Written 48880 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Read 48862 spots for ERR1442643\n",
      "Written 48862 spots for ERR1442643\n",
      "Getting ERR1442621 from NCBI SRA\n",
      "2025-10-14 21:28:04,855 - SRR ids: ['ERR1442621']\n",
      "2025-10-14 21:28:04,855 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 21:28:04,866 - tempdir: /local/scratch/job_322458/pfd_qkxdv0l4\n",
      "2025-10-14 21:28:04,866 - CMD: sra-stat --meta --quick ERR1442621\n",
      "2025-10-14 21:28:10,390 - ERR1442621 spots: 1706411\n",
      "2025-10-14 21:28:10,390 - blocks: [[1, 53325], [53326, 106650], [106651, 159975], [159976, 213300], [213301, 266625], [266626, 319950], [319951, 373275], [373276, 426600], [426601, 479925], [479926, 533250], [533251, 586575], [586576, 639900], [639901, 693225], [693226, 746550], [746551, 799875], [799876, 853200], [853201, 906525], [906526, 959850], [959851, 1013175], [1013176, 1066500], [1066501, 1119825], [1119826, 1173150], [1173151, 1226475], [1226476, 1279800], [1279801, 1333125], [1333126, 1386450], [1386451, 1439775], [1439776, 1493100], [1493101, 1546425], [1546426, 1599750], [1599751, 1653075], [1653076, 1706411]]\n",
      "2025-10-14 21:28:10,390 - CMD: fastq-dump -N 1 -X 53325 -O /local/scratch/job_322458/pfd_qkxdv0l4/0 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,400 - CMD: fastq-dump -N 53326 -X 106650 -O /local/scratch/job_322458/pfd_qkxdv0l4/1 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,401 - CMD: fastq-dump -N 106651 -X 159975 -O /local/scratch/job_322458/pfd_qkxdv0l4/2 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,401 - CMD: fastq-dump -N 159976 -X 213300 -O /local/scratch/job_322458/pfd_qkxdv0l4/3 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,401 - CMD: fastq-dump -N 213301 -X 266625 -O /local/scratch/job_322458/pfd_qkxdv0l4/4 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,402 - CMD: fastq-dump -N 266626 -X 319950 -O /local/scratch/job_322458/pfd_qkxdv0l4/5 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,402 - CMD: fastq-dump -N 319951 -X 373275 -O /local/scratch/job_322458/pfd_qkxdv0l4/6 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,402 - CMD: fastq-dump -N 373276 -X 426600 -O /local/scratch/job_322458/pfd_qkxdv0l4/7 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,403 - CMD: fastq-dump -N 426601 -X 479925 -O /local/scratch/job_322458/pfd_qkxdv0l4/8 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,403 - CMD: fastq-dump -N 479926 -X 533250 -O /local/scratch/job_322458/pfd_qkxdv0l4/9 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,403 - CMD: fastq-dump -N 533251 -X 586575 -O /local/scratch/job_322458/pfd_qkxdv0l4/10 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,404 - CMD: fastq-dump -N 586576 -X 639900 -O /local/scratch/job_322458/pfd_qkxdv0l4/11 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,404 - CMD: fastq-dump -N 639901 -X 693225 -O /local/scratch/job_322458/pfd_qkxdv0l4/12 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,404 - CMD: fastq-dump -N 693226 -X 746550 -O /local/scratch/job_322458/pfd_qkxdv0l4/13 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,405 - CMD: fastq-dump -N 746551 -X 799875 -O /local/scratch/job_322458/pfd_qkxdv0l4/14 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,405 - CMD: fastq-dump -N 799876 -X 853200 -O /local/scratch/job_322458/pfd_qkxdv0l4/15 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,405 - CMD: fastq-dump -N 853201 -X 906525 -O /local/scratch/job_322458/pfd_qkxdv0l4/16 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,406 - CMD: fastq-dump -N 906526 -X 959850 -O /local/scratch/job_322458/pfd_qkxdv0l4/17 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,406 - CMD: fastq-dump -N 959851 -X 1013175 -O /local/scratch/job_322458/pfd_qkxdv0l4/18 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,406 - CMD: fastq-dump -N 1013176 -X 1066500 -O /local/scratch/job_322458/pfd_qkxdv0l4/19 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,407 - CMD: fastq-dump -N 1066501 -X 1119825 -O /local/scratch/job_322458/pfd_qkxdv0l4/20 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,407 - CMD: fastq-dump -N 1119826 -X 1173150 -O /local/scratch/job_322458/pfd_qkxdv0l4/21 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,407 - CMD: fastq-dump -N 1173151 -X 1226475 -O /local/scratch/job_322458/pfd_qkxdv0l4/22 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,408 - CMD: fastq-dump -N 1226476 -X 1279800 -O /local/scratch/job_322458/pfd_qkxdv0l4/23 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,408 - CMD: fastq-dump -N 1279801 -X 1333125 -O /local/scratch/job_322458/pfd_qkxdv0l4/24 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,408 - CMD: fastq-dump -N 1333126 -X 1386450 -O /local/scratch/job_322458/pfd_qkxdv0l4/25 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,408 - CMD: fastq-dump -N 1386451 -X 1439775 -O /local/scratch/job_322458/pfd_qkxdv0l4/26 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,409 - CMD: fastq-dump -N 1439776 -X 1493100 -O /local/scratch/job_322458/pfd_qkxdv0l4/27 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,409 - CMD: fastq-dump -N 1493101 -X 1546425 -O /local/scratch/job_322458/pfd_qkxdv0l4/28 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,409 - CMD: fastq-dump -N 1546426 -X 1599750 -O /local/scratch/job_322458/pfd_qkxdv0l4/29 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,410 - CMD: fastq-dump -N 1599751 -X 1653075 -O /local/scratch/job_322458/pfd_qkxdv0l4/30 --split-files --gzip ERR1442621\n",
      "2025-10-14 21:28:10,410 - CMD: fastq-dump -N 1653076 -X 1706411 -O /local/scratch/job_322458/pfd_qkxdv0l4/31 --split-files --gzip ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53336 spots for ERR1442621\n",
      "Written 53336 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Read 53325 spots for ERR1442621\n",
      "Written 53325 spots for ERR1442621\n",
      "Getting ERR1442622 from NCBI SRA\n",
      "2025-10-14 21:29:18,550 - SRR ids: ['ERR1442622']\n",
      "2025-10-14 21:29:18,550 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 21:29:18,560 - tempdir: /local/scratch/job_322458/pfd_hj1kvurt\n",
      "2025-10-14 21:29:18,560 - CMD: sra-stat --meta --quick ERR1442622\n",
      "2025-10-14 21:29:24,372 - ERR1442622 spots: 1853271\n",
      "2025-10-14 21:29:24,372 - blocks: [[1, 57914], [57915, 115828], [115829, 173742], [173743, 231656], [231657, 289570], [289571, 347484], [347485, 405398], [405399, 463312], [463313, 521226], [521227, 579140], [579141, 637054], [637055, 694968], [694969, 752882], [752883, 810796], [810797, 868710], [868711, 926624], [926625, 984538], [984539, 1042452], [1042453, 1100366], [1100367, 1158280], [1158281, 1216194], [1216195, 1274108], [1274109, 1332022], [1332023, 1389936], [1389937, 1447850], [1447851, 1505764], [1505765, 1563678], [1563679, 1621592], [1621593, 1679506], [1679507, 1737420], [1737421, 1795334], [1795335, 1853271]]\n",
      "2025-10-14 21:29:24,372 - CMD: fastq-dump -N 1 -X 57914 -O /local/scratch/job_322458/pfd_hj1kvurt/0 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,374 - CMD: fastq-dump -N 57915 -X 115828 -O /local/scratch/job_322458/pfd_hj1kvurt/1 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,374 - CMD: fastq-dump -N 115829 -X 173742 -O /local/scratch/job_322458/pfd_hj1kvurt/2 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,374 - CMD: fastq-dump -N 173743 -X 231656 -O /local/scratch/job_322458/pfd_hj1kvurt/3 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,375 - CMD: fastq-dump -N 231657 -X 289570 -O /local/scratch/job_322458/pfd_hj1kvurt/4 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,375 - CMD: fastq-dump -N 289571 -X 347484 -O /local/scratch/job_322458/pfd_hj1kvurt/5 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,375 - CMD: fastq-dump -N 347485 -X 405398 -O /local/scratch/job_322458/pfd_hj1kvurt/6 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,376 - CMD: fastq-dump -N 405399 -X 463312 -O /local/scratch/job_322458/pfd_hj1kvurt/7 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,376 - CMD: fastq-dump -N 463313 -X 521226 -O /local/scratch/job_322458/pfd_hj1kvurt/8 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,376 - CMD: fastq-dump -N 521227 -X 579140 -O /local/scratch/job_322458/pfd_hj1kvurt/9 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,377 - CMD: fastq-dump -N 579141 -X 637054 -O /local/scratch/job_322458/pfd_hj1kvurt/10 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,377 - CMD: fastq-dump -N 637055 -X 694968 -O /local/scratch/job_322458/pfd_hj1kvurt/11 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,377 - CMD: fastq-dump -N 694969 -X 752882 -O /local/scratch/job_322458/pfd_hj1kvurt/12 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,378 - CMD: fastq-dump -N 752883 -X 810796 -O /local/scratch/job_322458/pfd_hj1kvurt/13 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,378 - CMD: fastq-dump -N 810797 -X 868710 -O /local/scratch/job_322458/pfd_hj1kvurt/14 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,378 - CMD: fastq-dump -N 868711 -X 926624 -O /local/scratch/job_322458/pfd_hj1kvurt/15 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,379 - CMD: fastq-dump -N 926625 -X 984538 -O /local/scratch/job_322458/pfd_hj1kvurt/16 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,379 - CMD: fastq-dump -N 984539 -X 1042452 -O /local/scratch/job_322458/pfd_hj1kvurt/17 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,379 - CMD: fastq-dump -N 1042453 -X 1100366 -O /local/scratch/job_322458/pfd_hj1kvurt/18 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,379 - CMD: fastq-dump -N 1100367 -X 1158280 -O /local/scratch/job_322458/pfd_hj1kvurt/19 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,380 - CMD: fastq-dump -N 1158281 -X 1216194 -O /local/scratch/job_322458/pfd_hj1kvurt/20 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,380 - CMD: fastq-dump -N 1216195 -X 1274108 -O /local/scratch/job_322458/pfd_hj1kvurt/21 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,380 - CMD: fastq-dump -N 1274109 -X 1332022 -O /local/scratch/job_322458/pfd_hj1kvurt/22 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,381 - CMD: fastq-dump -N 1332023 -X 1389936 -O /local/scratch/job_322458/pfd_hj1kvurt/23 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,381 - CMD: fastq-dump -N 1389937 -X 1447850 -O /local/scratch/job_322458/pfd_hj1kvurt/24 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,381 - CMD: fastq-dump -N 1447851 -X 1505764 -O /local/scratch/job_322458/pfd_hj1kvurt/25 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,382 - CMD: fastq-dump -N 1505765 -X 1563678 -O /local/scratch/job_322458/pfd_hj1kvurt/26 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,382 - CMD: fastq-dump -N 1563679 -X 1621592 -O /local/scratch/job_322458/pfd_hj1kvurt/27 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,382 - CMD: fastq-dump -N 1621593 -X 1679506 -O /local/scratch/job_322458/pfd_hj1kvurt/28 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,382 - CMD: fastq-dump -N 1679507 -X 1737420 -O /local/scratch/job_322458/pfd_hj1kvurt/29 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,383 - CMD: fastq-dump -N 1737421 -X 1795334 -O /local/scratch/job_322458/pfd_hj1kvurt/30 --split-files --gzip ERR1442622\n",
      "2025-10-14 21:29:24,383 - CMD: fastq-dump -N 1795335 -X 1853271 -O /local/scratch/job_322458/pfd_hj1kvurt/31 --split-files --gzip ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57937 spots for ERR1442622\n",
      "Written 57937 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Read 57914 spots for ERR1442622\n",
      "Written 57914 spots for ERR1442622\n",
      "Getting ERR1442623 from NCBI SRA\n",
      "2025-10-14 21:30:26,556 - SRR ids: ['ERR1442623']\n",
      "2025-10-14 21:30:26,556 - extra args: ['--split-files', '--gzip']\n",
      "2025-10-14 21:30:26,569 - tempdir: /local/scratch/job_322458/pfd_8c51r__j\n",
      "2025-10-14 21:30:26,569 - CMD: sra-stat --meta --quick ERR1442623\n",
      "2025-10-14 21:30:31,229 - ERR1442623 spots: 1803705\n",
      "2025-10-14 21:30:31,229 - blocks: [[1, 56365], [56366, 112730], [112731, 169095], [169096, 225460], [225461, 281825], [281826, 338190], [338191, 394555], [394556, 450920], [450921, 507285], [507286, 563650], [563651, 620015], [620016, 676380], [676381, 732745], [732746, 789110], [789111, 845475], [845476, 901840], [901841, 958205], [958206, 1014570], [1014571, 1070935], [1070936, 1127300], [1127301, 1183665], [1183666, 1240030], [1240031, 1296395], [1296396, 1352760], [1352761, 1409125], [1409126, 1465490], [1465491, 1521855], [1521856, 1578220], [1578221, 1634585], [1634586, 1690950], [1690951, 1747315], [1747316, 1803705]]\n",
      "2025-10-14 21:30:31,229 - CMD: fastq-dump -N 1 -X 56365 -O /local/scratch/job_322458/pfd_8c51r__j/0 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,245 - CMD: fastq-dump -N 56366 -X 112730 -O /local/scratch/job_322458/pfd_8c51r__j/1 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,246 - CMD: fastq-dump -N 112731 -X 169095 -O /local/scratch/job_322458/pfd_8c51r__j/2 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,246 - CMD: fastq-dump -N 169096 -X 225460 -O /local/scratch/job_322458/pfd_8c51r__j/3 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,246 - CMD: fastq-dump -N 225461 -X 281825 -O /local/scratch/job_322458/pfd_8c51r__j/4 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,247 - CMD: fastq-dump -N 281826 -X 338190 -O /local/scratch/job_322458/pfd_8c51r__j/5 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,247 - CMD: fastq-dump -N 338191 -X 394555 -O /local/scratch/job_322458/pfd_8c51r__j/6 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,247 - CMD: fastq-dump -N 394556 -X 450920 -O /local/scratch/job_322458/pfd_8c51r__j/7 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,248 - CMD: fastq-dump -N 450921 -X 507285 -O /local/scratch/job_322458/pfd_8c51r__j/8 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,248 - CMD: fastq-dump -N 507286 -X 563650 -O /local/scratch/job_322458/pfd_8c51r__j/9 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,248 - CMD: fastq-dump -N 563651 -X 620015 -O /local/scratch/job_322458/pfd_8c51r__j/10 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,249 - CMD: fastq-dump -N 620016 -X 676380 -O /local/scratch/job_322458/pfd_8c51r__j/11 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,249 - CMD: fastq-dump -N 676381 -X 732745 -O /local/scratch/job_322458/pfd_8c51r__j/12 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,249 - CMD: fastq-dump -N 732746 -X 789110 -O /local/scratch/job_322458/pfd_8c51r__j/13 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,250 - CMD: fastq-dump -N 789111 -X 845475 -O /local/scratch/job_322458/pfd_8c51r__j/14 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,250 - CMD: fastq-dump -N 845476 -X 901840 -O /local/scratch/job_322458/pfd_8c51r__j/15 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,250 - CMD: fastq-dump -N 901841 -X 958205 -O /local/scratch/job_322458/pfd_8c51r__j/16 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,250 - CMD: fastq-dump -N 958206 -X 1014570 -O /local/scratch/job_322458/pfd_8c51r__j/17 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,251 - CMD: fastq-dump -N 1014571 -X 1070935 -O /local/scratch/job_322458/pfd_8c51r__j/18 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,251 - CMD: fastq-dump -N 1070936 -X 1127300 -O /local/scratch/job_322458/pfd_8c51r__j/19 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,251 - CMD: fastq-dump -N 1127301 -X 1183665 -O /local/scratch/job_322458/pfd_8c51r__j/20 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,252 - CMD: fastq-dump -N 1183666 -X 1240030 -O /local/scratch/job_322458/pfd_8c51r__j/21 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,252 - CMD: fastq-dump -N 1240031 -X 1296395 -O /local/scratch/job_322458/pfd_8c51r__j/22 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,252 - CMD: fastq-dump -N 1296396 -X 1352760 -O /local/scratch/job_322458/pfd_8c51r__j/23 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,252 - CMD: fastq-dump -N 1352761 -X 1409125 -O /local/scratch/job_322458/pfd_8c51r__j/24 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,253 - CMD: fastq-dump -N 1409126 -X 1465490 -O /local/scratch/job_322458/pfd_8c51r__j/25 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,253 - CMD: fastq-dump -N 1465491 -X 1521855 -O /local/scratch/job_322458/pfd_8c51r__j/26 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,253 - CMD: fastq-dump -N 1521856 -X 1578220 -O /local/scratch/job_322458/pfd_8c51r__j/27 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,254 - CMD: fastq-dump -N 1578221 -X 1634585 -O /local/scratch/job_322458/pfd_8c51r__j/28 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,254 - CMD: fastq-dump -N 1634586 -X 1690950 -O /local/scratch/job_322458/pfd_8c51r__j/29 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,254 - CMD: fastq-dump -N 1690951 -X 1747315 -O /local/scratch/job_322458/pfd_8c51r__j/30 --split-files --gzip ERR1442623\n",
      "2025-10-14 21:30:31,254 - CMD: fastq-dump -N 1747316 -X 1803705 -O /local/scratch/job_322458/pfd_8c51r__j/31 --split-files --gzip ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56390 spots for ERR1442623\n",
      "Written 56390 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n",
      "Read 56365 spots for ERR1442623\n",
      "Written 56365 spots for ERR1442623\n"
     ]
    }
   ],
   "source": [
    "!while IFS= read -r line; do \\\n",
    "    echo \"Getting $line from NCBI SRA\"; \\\n",
    "    parallel-fastq-dump --split-files --sra-id $line --threads 32 --outdir hw_fastq --gzip; \\\n",
    "    done < hwsra_list.txt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run QC on the FASTQ files and aggregate the results into a report."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.003s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /dev/cpuset.\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of ERR1442621_1.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of ERR1442621_2.fastq.gz\n",
      "Approx 5% complete for ERR1442621_1.fastq.gz\n",
      "Started analysis of ERR1442622_1.fastq.gz\n",
      "Approx 5% complete for ERR1442621_2.fastq.gz\n",
      "Approx 10% complete for ERR1442621_1.fastq.gz\n",
      "Approx 10% complete for ERR1442621_2.fastq.gz\n",
      "Started analysis of ERR1442622_2.fastq.gz\n",
      "Approx 15% complete for ERR1442621_1.fastq.gz\n",
      "Approx 5% complete for ERR1442622_1.fastq.gz\n",
      "Approx 15% complete for ERR1442621_2.fastq.gz\n",
      "Approx 20% complete for ERR1442621_1.fastq.gz\n",
      "Started analysis of ERR1442623_1.fastq.gz\n",
      "Approx 5% complete for ERR1442622_2.fastq.gz\n",
      "Approx 10% complete for ERR1442622_1.fastq.gz\n",
      "Approx 20% complete for ERR1442621_2.fastq.gz\n",
      "Approx 25% complete for ERR1442621_1.fastq.gz\n",
      "Approx 5% complete for ERR1442623_1.fastq.gz\n",
      "Started analysis of ERR1442623_2.fastq.gz\n",
      "Approx 15% complete for ERR1442622_1.fastq.gz\n",
      "Approx 10% complete for ERR1442622_2.fastq.gz\n",
      "Approx 25% complete for ERR1442621_2.fastq.gz\n",
      "Approx 30% complete for ERR1442621_1.fastq.gz\n",
      "Approx 10% complete for ERR1442623_1.fastq.gz\n",
      "Started analysis of ERR1442631_1.fastq.gz\n",
      "Approx 20% complete for ERR1442622_1.fastq.gz\n",
      "Approx 5% complete for ERR1442623_2.fastq.gz\n",
      "Approx 15% complete for ERR1442622_2.fastq.gz\n",
      "Approx 35% complete for ERR1442621_1.fastq.gz\n",
      "Approx 30% complete for ERR1442621_2.fastq.gz\n",
      "Started analysis of ERR1442631_2.fastq.gz\n",
      "Approx 15% complete for ERR1442623_1.fastq.gz\n",
      "Approx 5% complete for ERR1442631_1.fastq.gz\n",
      "Approx 25% complete for ERR1442622_1.fastq.gz\n",
      "Approx 10% complete for ERR1442623_2.fastq.gz\n",
      "Approx 20% complete for ERR1442622_2.fastq.gz\n",
      "Approx 40% complete for ERR1442621_1.fastq.gz\n",
      "Approx 35% complete for ERR1442621_2.fastq.gz\n",
      "Started analysis of ERR1442632_1.fastq.gz\n",
      "Approx 5% complete for ERR1442631_2.fastq.gz\n",
      "Approx 20% complete for ERR1442623_1.fastq.gz\n",
      "Approx 10% complete for ERR1442631_1.fastq.gz\n",
      "Approx 30% complete for ERR1442622_1.fastq.gz\n",
      "Approx 15% complete for ERR1442623_2.fastq.gz\n",
      "Approx 45% complete for ERR1442621_1.fastq.gz\n",
      "Approx 25% complete for ERR1442622_2.fastq.gz\n",
      "Started analysis of ERR1442632_2.fastq.gz\n",
      "Approx 40% complete for ERR1442621_2.fastq.gz\n",
      "Approx 5% complete for ERR1442632_1.fastq.gz\n",
      "Approx 25% complete for ERR1442623_1.fastq.gz\n",
      "Approx 10% complete for ERR1442631_2.fastq.gz\n",
      "Approx 50% complete for ERR1442621_1.fastq.gz\n",
      "Approx 15% complete for ERR1442631_1.fastq.gz\n",
      "Approx 35% complete for ERR1442622_1.fastq.gz\n",
      "Approx 20% complete for ERR1442623_2.fastq.gz\n",
      "Approx 5% complete for ERR1442632_2.fastq.gz\n",
      "Approx 45% complete for ERR1442621_2.fastq.gz\n",
      "Approx 30% complete for ERR1442622_2.fastq.gz\n",
      "Started analysis of ERR1442633_1.fastq.gz\n",
      "Approx 10% complete for ERR1442632_1.fastq.gz\n",
      "Approx 55% complete for ERR1442621_1.fastq.gz\n",
      "Approx 30% complete for ERR1442623_1.fastq.gz\n",
      "Approx 15% complete for ERR1442631_2.fastq.gz\n",
      "Approx 40% complete for ERR1442622_1.fastq.gz\n",
      "Approx 20% complete for ERR1442631_1.fastq.gz\n",
      "Approx 50% complete for ERR1442621_2.fastq.gz\n",
      "Approx 25% complete for ERR1442623_2.fastq.gz\n",
      "Approx 10% complete for ERR1442632_2.fastq.gz\n",
      "Started analysis of ERR1442633_2.fastq.gz\n",
      "Approx 15% complete for ERR1442632_1.fastq.gz\n",
      "Approx 35% complete for ERR1442622_2.fastq.gz\n",
      "Approx 5% complete for ERR1442633_1.fastq.gz\n",
      "Approx 60% complete for ERR1442621_1.fastq.gz\n",
      "Approx 35% complete for ERR1442623_1.fastq.gz\n",
      "Approx 25% complete for ERR1442631_1.fastq.gz\n",
      "Approx 55% complete for ERR1442621_2.fastq.gz\n",
      "Approx 45% complete for ERR1442622_1.fastq.gz\n",
      "Approx 20% complete for ERR1442631_2.fastq.gz\n",
      "Approx 30% complete for ERR1442623_2.fastq.gz\n",
      "Approx 15% complete for ERR1442632_2.fastq.gz\n",
      "Started analysis of ERR1442641_1.fastq.gz\n",
      "Approx 20% complete for ERR1442632_1.fastq.gz\n",
      "Approx 65% complete for ERR1442621_1.fastq.gz\n",
      "Approx 40% complete for ERR1442622_2.fastq.gz\n",
      "Approx 5% complete for ERR1442633_2.fastq.gz\n",
      "Approx 10% complete for ERR1442633_1.fastq.gz\n",
      "Approx 40% complete for ERR1442623_1.fastq.gz\n",
      "Started analysis of ERR1442641_2.fastq.gz\n",
      "Approx 60% complete for ERR1442621_2.fastq.gz\n",
      "Approx 50% complete for ERR1442622_1.fastq.gz\n",
      "Approx 30% complete for ERR1442631_1.fastq.gz\n",
      "Approx 25% complete for ERR1442631_2.fastq.gz\n",
      "Approx 20% complete for ERR1442632_2.fastq.gz\n",
      "Approx 70% complete for ERR1442621_1.fastq.gz\n",
      "Approx 35% complete for ERR1442623_2.fastq.gz\n",
      "Approx 25% complete for ERR1442632_1.fastq.gz\n",
      "Approx 5% complete for ERR1442641_1.fastq.gz\n",
      "Approx 45% complete for ERR1442622_2.fastq.gz\n",
      "Approx 10% complete for ERR1442633_2.fastq.gz\n",
      "Approx 45% complete for ERR1442623_1.fastq.gz\n",
      "Started analysis of ERR1442642_1.fastq.gz\n",
      "Approx 65% complete for ERR1442621_2.fastq.gz\n",
      "Approx 15% complete for ERR1442633_1.fastq.gz\n",
      "Approx 75% complete for ERR1442621_1.fastq.gz\n",
      "Approx 25% complete for ERR1442632_2.fastq.gz\n",
      "Approx 35% complete for ERR1442631_1.fastq.gz\n",
      "Approx 55% complete for ERR1442622_1.fastq.gz\n",
      "Approx 30% complete for ERR1442631_2.fastq.gz\n",
      "Approx 5% complete for ERR1442641_2.fastq.gz\n",
      "Approx 30% complete for ERR1442632_1.fastq.gz\n",
      "Approx 40% complete for ERR1442623_2.fastq.gz\n",
      "Started analysis of ERR1442642_2.fastq.gz\n",
      "Approx 50% complete for ERR1442622_2.fastq.gz\n",
      "Approx 50% complete for ERR1442623_1.fastq.gz\n",
      "Approx 10% complete for ERR1442641_1.fastq.gz\n",
      "Approx 70% complete for ERR1442621_2.fastq.gz\n",
      "Approx 80% complete for ERR1442621_1.fastq.gz\n",
      "Approx 15% complete for ERR1442633_2.fastq.gz\n",
      "Approx 30% complete for ERR1442632_2.fastq.gz\n",
      "Approx 5% complete for ERR1442642_1.fastq.gz\n",
      "Approx 20% complete for ERR1442633_1.fastq.gz\n",
      "Approx 35% complete for ERR1442632_1.fastq.gz\n",
      "Approx 60% complete for ERR1442622_1.fastq.gz\n",
      "Approx 40% complete for ERR1442631_1.fastq.gz\n",
      "Approx 35% complete for ERR1442631_2.fastq.gz\n",
      "Approx 45% complete for ERR1442623_2.fastq.gz\n",
      "Started analysis of ERR1442643_1.fastq.gz\n",
      "Approx 10% complete for ERR1442641_2.fastq.gz\n",
      "Approx 85% complete for ERR1442621_1.fastq.gz\n",
      "Approx 75% complete for ERR1442621_2.fastq.gz\n",
      "Approx 55% complete for ERR1442623_1.fastq.gz\n",
      "Approx 55% complete for ERR1442622_2.fastq.gz\n",
      "Approx 5% complete for ERR1442642_2.fastq.gz\n",
      "Approx 15% complete for ERR1442641_1.fastq.gz\n",
      "Approx 35% complete for ERR1442632_2.fastq.gz\n",
      "Approx 20% complete for ERR1442633_2.fastq.gz\n",
      "Approx 65% complete for ERR1442622_1.fastq.gz\n",
      "Approx 40% complete for ERR1442632_1.fastq.gz\n",
      "Started analysis of ERR1442643_2.fastq.gz\n",
      "Approx 50% complete for ERR1442623_2.fastq.gz\n",
      "Approx 5% complete for ERR1442643_1.fastq.gz\n",
      "Approx 45% complete for ERR1442631_1.fastq.gz\n",
      "Approx 40% complete for ERR1442631_2.fastq.gz\n",
      "Approx 10% complete for ERR1442642_1.fastq.gz\n",
      "Approx 25% complete for ERR1442633_1.fastq.gz\n",
      "Approx 90% complete for ERR1442621_1.fastq.gz\n",
      "Approx 80% complete for ERR1442621_2.fastq.gz\n",
      "Approx 60% complete for ERR1442623_1.fastq.gz\n",
      "Approx 15% complete for ERR1442641_2.fastq.gz\n",
      "Approx 60% complete for ERR1442622_2.fastq.gz\n",
      "Approx 40% complete for ERR1442632_2.fastq.gz\n",
      "Approx 70% complete for ERR1442622_1.fastq.gz\n",
      "Approx 45% complete for ERR1442632_1.fastq.gz\n",
      "Approx 5% complete for ERR1442643_2.fastq.gz\n",
      "Approx 20% complete for ERR1442641_1.fastq.gz\n",
      "Approx 10% complete for ERR1442642_2.fastq.gz\n",
      "Approx 10% complete for ERR1442643_1.fastq.gz\n",
      "Approx 50% complete for ERR1442631_1.fastq.gz\n",
      "Approx 25% complete for ERR1442633_2.fastq.gz\n",
      "Approx 95% complete for ERR1442621_1.fastq.gz\n",
      "Approx 55% complete for ERR1442623_2.fastq.gz\n",
      "Approx 85% complete for ERR1442621_2.fastq.gz\n",
      "Approx 45% complete for ERR1442631_2.fastq.gz\n",
      "Approx 65% complete for ERR1442623_1.fastq.gz\n",
      "Approx 30% complete for ERR1442633_1.fastq.gz\n",
      "Approx 15% complete for ERR1442642_1.fastq.gz\n",
      "Approx 45% complete for ERR1442632_2.fastq.gz\n",
      "Approx 65% complete for ERR1442622_2.fastq.gz\n",
      "Approx 50% complete for ERR1442632_1.fastq.gz\n",
      "Approx 75% complete for ERR1442622_1.fastq.gz\n",
      "Approx 20% complete for ERR1442641_2.fastq.gz\n",
      "Approx 10% complete for ERR1442643_2.fastq.gz\n",
      "Analysis complete for ERR1442621_1.fastq.gz\n",
      "Approx 15% complete for ERR1442643_1.fastq.gz\n",
      "Approx 90% complete for ERR1442621_2.fastq.gz\n",
      "Approx 50% complete for ERR1442631_2.fastq.gz\n",
      "Approx 55% complete for ERR1442631_1.fastq.gz\n",
      "Approx 60% complete for ERR1442623_2.fastq.gz\n",
      "Approx 15% complete for ERR1442642_2.fastq.gz\n",
      "Approx 30% complete for ERR1442633_2.fastq.gz\n",
      "Approx 25% complete for ERR1442641_1.fastq.gz\n",
      "Approx 70% complete for ERR1442623_1.fastq.gz\n",
      "Approx 50% complete for ERR1442632_2.fastq.gz\n",
      "Approx 35% complete for ERR1442633_1.fastq.gz\n",
      "Approx 70% complete for ERR1442622_2.fastq.gz\n",
      "Approx 20% complete for ERR1442642_1.fastq.gz\n",
      "Approx 55% complete for ERR1442632_1.fastq.gz\n",
      "Approx 80% complete for ERR1442622_1.fastq.gz\n",
      "Approx 15% complete for ERR1442643_2.fastq.gz\n",
      "Approx 20% complete for ERR1442643_1.fastq.gz\n",
      "Approx 95% complete for ERR1442621_2.fastq.gz\n",
      "Approx 25% complete for ERR1442641_2.fastq.gz\n",
      "Approx 65% complete for ERR1442623_2.fastq.gz\n",
      "Approx 60% complete for ERR1442631_1.fastq.gz\n",
      "Approx 75% complete for ERR1442623_1.fastq.gz\n",
      "Approx 55% complete for ERR1442631_2.fastq.gz\n",
      "Approx 35% complete for ERR1442633_2.fastq.gz\n",
      "Approx 55% complete for ERR1442632_2.fastq.gz\n",
      "Approx 20% complete for ERR1442642_2.fastq.gz\n",
      "Approx 30% complete for ERR1442641_1.fastq.gz\n",
      "Approx 75% complete for ERR1442622_2.fastq.gz\n",
      "Approx 60% complete for ERR1442632_1.fastq.gz\n",
      "Approx 85% complete for ERR1442622_1.fastq.gz\n",
      "Approx 20% complete for ERR1442643_2.fastq.gz\n",
      "Approx 25% complete for ERR1442643_1.fastq.gz\n",
      "Approx 40% complete for ERR1442633_1.fastq.gz\n",
      "Analysis complete for ERR1442621_2.fastq.gz\n",
      "Approx 25% complete for ERR1442642_1.fastq.gz\n",
      "Approx 80% complete for ERR1442623_1.fastq.gz\n",
      "Approx 70% complete for ERR1442623_2.fastq.gz\n",
      "Approx 65% complete for ERR1442631_1.fastq.gz\n",
      "Approx 30% complete for ERR1442641_2.fastq.gz\n",
      "Approx 60% complete for ERR1442632_2.fastq.gz\n",
      "Approx 60% complete for ERR1442631_2.fastq.gz\n",
      "Approx 40% complete for ERR1442633_2.fastq.gz\n",
      "Approx 80% complete for ERR1442622_2.fastq.gz\n",
      "Approx 90% complete for ERR1442622_1.fastq.gz\n",
      "Approx 65% complete for ERR1442632_1.fastq.gz\n",
      "Approx 25% complete for ERR1442643_2.fastq.gz\n",
      "Approx 30% complete for ERR1442643_1.fastq.gz\n",
      "Approx 25% complete for ERR1442642_2.fastq.gz\n",
      "Approx 35% complete for ERR1442641_1.fastq.gz\n",
      "Approx 85% complete for ERR1442623_1.fastq.gz\n",
      "Approx 75% complete for ERR1442623_2.fastq.gz\n",
      "Approx 45% complete for ERR1442633_1.fastq.gz\n",
      "Approx 70% complete for ERR1442631_1.fastq.gz\n",
      "Approx 65% complete for ERR1442632_2.fastq.gz\n",
      "Approx 30% complete for ERR1442642_1.fastq.gz\n",
      "Approx 65% complete for ERR1442631_2.fastq.gz\n",
      "Approx 35% complete for ERR1442641_2.fastq.gz\n",
      "Approx 70% complete for ERR1442632_1.fastq.gz\n",
      "Approx 95% complete for ERR1442622_1.fastq.gz\n",
      "Approx 35% complete for ERR1442643_1.fastq.gz\n",
      "Approx 30% complete for ERR1442643_2.fastq.gz\n",
      "Approx 45% complete for ERR1442633_2.fastq.gz\n",
      "Approx 85% complete for ERR1442622_2.fastq.gz\n",
      "Approx 90% complete for ERR1442623_1.fastq.gz\n",
      "Approx 80% complete for ERR1442623_2.fastq.gz\n",
      "Approx 40% complete for ERR1442641_1.fastq.gz\n",
      "Approx 30% complete for ERR1442642_2.fastq.gz\n",
      "Approx 75% complete for ERR1442631_1.fastq.gz\n",
      "Approx 70% complete for ERR1442632_2.fastq.gz\n",
      "Approx 50% complete for ERR1442633_1.fastq.gz\n",
      "Approx 75% complete for ERR1442632_1.fastq.gz\n",
      "Analysis complete for ERR1442622_1.fastq.gz\n",
      "Approx 70% complete for ERR1442631_2.fastq.gz\n",
      "Approx 40% complete for ERR1442643_1.fastq.gz\n",
      "Approx 35% complete for ERR1442643_2.fastq.gz\n",
      "Approx 50% complete for ERR1442633_2.fastq.gz\n",
      "Approx 35% complete for ERR1442642_1.fastq.gz\n",
      "Approx 90% complete for ERR1442622_2.fastq.gz\n",
      "Approx 40% complete for ERR1442641_2.fastq.gz\n",
      "Approx 95% complete for ERR1442623_1.fastq.gz\n",
      "Approx 85% complete for ERR1442623_2.fastq.gz\n",
      "Approx 75% complete for ERR1442632_2.fastq.gz\n",
      "Approx 80% complete for ERR1442631_1.fastq.gz\n",
      "Approx 80% complete for ERR1442632_1.fastq.gz\n",
      "Approx 45% complete for ERR1442641_1.fastq.gz\n",
      "Approx 35% complete for ERR1442642_2.fastq.gz\n",
      "Approx 45% complete for ERR1442643_1.fastq.gz\n",
      "Approx 40% complete for ERR1442643_2.fastq.gz\n",
      "Approx 55% complete for ERR1442633_1.fastq.gz\n",
      "Approx 75% complete for ERR1442631_2.fastq.gz\n",
      "Approx 95% complete for ERR1442622_2.fastq.gz\n",
      "Analysis complete for ERR1442623_1.fastq.gz\n",
      "Approx 55% complete for ERR1442633_2.fastq.gz\n",
      "Approx 45% complete for ERR1442641_2.fastq.gz\n",
      "Approx 80% complete for ERR1442632_2.fastq.gz\n",
      "Approx 40% complete for ERR1442642_1.fastq.gz\n",
      "Approx 90% complete for ERR1442623_2.fastq.gz\n",
      "Approx 85% complete for ERR1442631_1.fastq.gz\n",
      "Approx 85% complete for ERR1442632_1.fastq.gz\n",
      "Approx 50% complete for ERR1442643_1.fastq.gz\n",
      "Approx 45% complete for ERR1442643_2.fastq.gz\n",
      "Analysis complete for ERR1442622_2.fastq.gz\n",
      "Approx 50% complete for ERR1442641_1.fastq.gz\n",
      "Approx 80% complete for ERR1442631_2.fastq.gz\n",
      "Approx 40% complete for ERR1442642_2.fastq.gz\n",
      "Approx 60% complete for ERR1442633_1.fastq.gz\n",
      "Approx 85% complete for ERR1442632_2.fastq.gz\n",
      "Approx 60% complete for ERR1442633_2.fastq.gz\n",
      "Approx 95% complete for ERR1442623_2.fastq.gz\n",
      "Approx 50% complete for ERR1442641_2.fastq.gz\n",
      "Approx 90% complete for ERR1442632_1.fastq.gz\n",
      "Approx 90% complete for ERR1442631_1.fastq.gz\n",
      "Approx 55% complete for ERR1442643_1.fastq.gz\n",
      "Approx 50% complete for ERR1442643_2.fastq.gz\n",
      "Approx 45% complete for ERR1442642_1.fastq.gz\n",
      "Approx 85% complete for ERR1442631_2.fastq.gz\n",
      "Approx 90% complete for ERR1442632_2.fastq.gz\n",
      "Approx 55% complete for ERR1442641_1.fastq.gz\n",
      "Analysis complete for ERR1442623_2.fastq.gz\n",
      "Approx 65% complete for ERR1442633_1.fastq.gz\n",
      "Approx 45% complete for ERR1442642_2.fastq.gz\n",
      "Approx 65% complete for ERR1442633_2.fastq.gz\n",
      "Approx 95% complete for ERR1442632_1.fastq.gz\n",
      "Approx 60% complete for ERR1442643_1.fastq.gz\n",
      "Approx 55% complete for ERR1442643_2.fastq.gz\n",
      "Approx 95% complete for ERR1442631_1.fastq.gz\n",
      "Approx 55% complete for ERR1442641_2.fastq.gz\n",
      "Approx 50% complete for ERR1442642_1.fastq.gz\n",
      "Approx 95% complete for ERR1442632_2.fastq.gz\n",
      "Approx 90% complete for ERR1442631_2.fastq.gz\n",
      "Analysis complete for ERR1442632_1.fastq.gz\n",
      "Approx 70% complete for ERR1442633_1.fastq.gz\n",
      "Approx 65% complete for ERR1442643_1.fastq.gz\n",
      "Approx 60% complete for ERR1442643_2.fastq.gz\n",
      "Approx 50% complete for ERR1442642_2.fastq.gz\n",
      "Analysis complete for ERR1442631_1.fastq.gz\n",
      "Approx 70% complete for ERR1442633_2.fastq.gz\n",
      "Approx 60% complete for ERR1442641_1.fastq.gz\n",
      "Analysis complete for ERR1442632_2.fastq.gz\n",
      "Approx 60% complete for ERR1442641_2.fastq.gz\n",
      "Approx 95% complete for ERR1442631_2.fastq.gz\n",
      "Approx 70% complete for ERR1442643_1.fastq.gz\n",
      "Approx 55% complete for ERR1442642_1.fastq.gz\n",
      "Approx 75% complete for ERR1442633_1.fastq.gz\n",
      "Approx 65% complete for ERR1442643_2.fastq.gz\n",
      "Approx 75% complete for ERR1442633_2.fastq.gz\n",
      "Approx 65% complete for ERR1442641_1.fastq.gz\n",
      "Approx 55% complete for ERR1442642_2.fastq.gz\n",
      "Analysis complete for ERR1442631_2.fastq.gz\n",
      "Approx 65% complete for ERR1442641_2.fastq.gz\n",
      "Approx 75% complete for ERR1442643_1.fastq.gz\n",
      "Approx 70% complete for ERR1442643_2.fastq.gz\n",
      "Approx 80% complete for ERR1442633_1.fastq.gz\n",
      "Approx 60% complete for ERR1442642_1.fastq.gz\n",
      "Approx 80% complete for ERR1442633_2.fastq.gz\n",
      "Approx 70% complete for ERR1442641_1.fastq.gz\n",
      "Approx 80% complete for ERR1442643_1.fastq.gz\n",
      "Approx 75% complete for ERR1442643_2.fastq.gz\n",
      "Approx 60% complete for ERR1442642_2.fastq.gz\n",
      "Approx 70% complete for ERR1442641_2.fastq.gz\n",
      "Approx 85% complete for ERR1442633_1.fastq.gz\n",
      "Approx 85% complete for ERR1442633_2.fastq.gz\n",
      "Approx 85% complete for ERR1442643_1.fastq.gz\n",
      "Approx 65% complete for ERR1442642_1.fastq.gz\n",
      "Approx 80% complete for ERR1442643_2.fastq.gz\n",
      "Approx 75% complete for ERR1442641_1.fastq.gz\n",
      "Approx 90% complete for ERR1442633_1.fastq.gz\n",
      "Approx 75% complete for ERR1442641_2.fastq.gz\n",
      "Approx 65% complete for ERR1442642_2.fastq.gz\n",
      "Approx 90% complete for ERR1442643_1.fastq.gz\n",
      "Approx 85% complete for ERR1442643_2.fastq.gz\n",
      "Approx 90% complete for ERR1442633_2.fastq.gz\n",
      "Approx 70% complete for ERR1442642_1.fastq.gz\n",
      "Approx 80% complete for ERR1442641_1.fastq.gz\n",
      "Approx 95% complete for ERR1442633_1.fastq.gz\n",
      "Approx 95% complete for ERR1442643_1.fastq.gz\n",
      "Approx 80% complete for ERR1442641_2.fastq.gz\n",
      "Approx 70% complete for ERR1442642_2.fastq.gz\n",
      "Approx 90% complete for ERR1442643_2.fastq.gz\n",
      "Approx 95% complete for ERR1442633_2.fastq.gz\n",
      "Approx 85% complete for ERR1442641_1.fastq.gz\n",
      "Approx 75% complete for ERR1442642_1.fastq.gz\n",
      "Analysis complete for ERR1442643_1.fastq.gz\n",
      "Analysis complete for ERR1442633_1.fastq.gz\n",
      "Approx 95% complete for ERR1442643_2.fastq.gz\n",
      "Approx 85% complete for ERR1442641_2.fastq.gz\n",
      "Approx 75% complete for ERR1442642_2.fastq.gz\n",
      "Analysis complete for ERR1442633_2.fastq.gz\n",
      "Analysis complete for ERR1442643_2.fastq.gz\n",
      "Approx 90% complete for ERR1442641_1.fastq.gz\n",
      "Approx 80% complete for ERR1442642_1.fastq.gz\n",
      "Approx 90% complete for ERR1442641_2.fastq.gz\n",
      "Approx 80% complete for ERR1442642_2.fastq.gz\n",
      "Approx 95% complete for ERR1442641_1.fastq.gz\n",
      "Approx 85% complete for ERR1442642_1.fastq.gz\n",
      "Approx 95% complete for ERR1442641_2.fastq.gz\n",
      "Approx 85% complete for ERR1442642_2.fastq.gz\n",
      "Analysis complete for ERR1442641_1.fastq.gz\n",
      "Approx 90% complete for ERR1442642_1.fastq.gz\n",
      "Analysis complete for ERR1442641_2.fastq.gz\n",
      "Approx 90% complete for ERR1442642_2.fastq.gz\n",
      "Approx 95% complete for ERR1442642_1.fastq.gz\n",
      "Approx 95% complete for ERR1442642_2.fastq.gz\n",
      "Analysis complete for ERR1442642_1.fastq.gz\n",
      "Analysis complete for ERR1442642_2.fastq.gz\n",
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=23359;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.17\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/heky1803/BIOL343/5_fastq/hw_fastq/qc\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m36/36\u001b[0m  qc/ERR1442642_2_fastqc.html\u001b[0m.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 18 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "!fastqc -t 32 hw_fastq/*.fastq.gz -o hw_fastq/qc\n",
    "!multiqc hw_fastq/qc"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Edit this block with a short evaluation of each of the report sections. Each evaluation should include your important observations, reasons for potential QC failure, and an opinion on what should be done on the failure (i.e., trimming or filtering). Problem FASTQ files, if any, should be noted.\n",
    "\n",
    "*Per base sequence quality:* These are good\n",
    "\n",
    "*Per tile sequence quality:*\n",
    "\n",
    "*Per sequence quality scores:* These are good\n",
    "\n",
    "*Per base sequence content:* This is bad. There is a big mess at one end of the reads. There are probably adapters remaining on that side. This will need trimming.\n",
    "\n",
    "*Per sequence GC content:* There are 4 that seems to be a big problem (red) while others are minor problems (yellow).\n",
    "\n",
    "*Per base N content:* This is good\n",
    "\n",
    "*Sequence length distribution:* All samples are the same length so it is good\n",
    "\n",
    "*Sequence duplication levels:* Most of the library is not duplicate which is good\n",
    "\n",
    "*Overrepresented sequences:* 7 samples have somewhat overrepresented sequences (yellow).\n",
    "\n",
    "*Adapter content:* this is good as well."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If necessary, use `cutadapt` to trim and filter your reads. Provide a justification for the `cutadapt` parameters you used:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442621_1.fastq.gz hw_fastq/ERR1442621_1.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,706,411 reads @   1.1 s/read;  56.81 M reads/minute\n",
      "Finished in 2.087 s (1.223 s/read; 49.07 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,706,411\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,706,411 (100.0%)\n",
      "\n",
      "Total basepairs processed:   170,641,100 bp\n",
      "Total written (filtered):    145,044,935 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442621_2.fastq.gz hw_fastq/ERR1442621_2.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,706,411 reads @   0.9 s/read;  64.91 M reads/minute\n",
      "Finished in 1.689 s (0.990 s/read; 60.61 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,706,411\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,706,411 (100.0%)\n",
      "\n",
      "Total basepairs processed:   170,641,100 bp\n",
      "Total written (filtered):    145,044,935 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442622_1.fastq.gz hw_fastq/ERR1442622_1.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,853,271 reads @   0.9 s/read;  66.36 M reads/minute\n",
      "Finished in 1.777 s (0.959 s/read; 62.58 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,853,271\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,853,271 (100.0%)\n",
      "\n",
      "Total basepairs processed:   185,327,100 bp\n",
      "Total written (filtered):    157,528,035 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442622_2.fastq.gz hw_fastq/ERR1442622_2.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,853,271 reads @   0.9 s/read;  65.35 M reads/minute\n",
      "Finished in 1.815 s (0.979 s/read; 61.27 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,853,271\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,853,271 (100.0%)\n",
      "\n",
      "Total basepairs processed:   185,327,100 bp\n",
      "Total written (filtered):    157,528,035 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442623_1.fastq.gz hw_fastq/ERR1442623_1.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,803,705 reads @   0.9 s/read;  69.61 M reads/minute\n",
      "Finished in 1.657 s (0.919 s/read; 65.32 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,803,705\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,803,705 (100.0%)\n",
      "\n",
      "Total basepairs processed:   180,370,500 bp\n",
      "Total written (filtered):    153,314,925 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442623_2.fastq.gz hw_fastq/ERR1442623_2.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,803,705 reads @   1.0 s/read;  60.87 M reads/minute\n",
      "Finished in 1.870 s (1.037 s/read; 57.89 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,803,705\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,803,705 (100.0%)\n",
      "\n",
      "Total basepairs processed:   180,370,500 bp\n",
      "Total written (filtered):    153,314,925 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442631_1.fastq.gz hw_fastq/ERR1442631_1.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,785,080 reads @   0.9 s/read;  64.48 M reads/minute\n",
      "Finished in 1.778 s (0.996 s/read; 60.22 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,785,080\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,785,080 (100.0%)\n",
      "\n",
      "Total basepairs processed:   178,508,000 bp\n",
      "Total written (filtered):    151,731,800 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442631_2.fastq.gz hw_fastq/ERR1442631_2.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,785,080 reads @   1.1 s/read;  55.61 M reads/minute\n",
      "Finished in 2.033 s (1.139 s/read; 52.67 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,785,080\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,785,080 (100.0%)\n",
      "\n",
      "Total basepairs processed:   178,508,000 bp\n",
      "Total written (filtered):    151,731,800 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442632_1.fastq.gz hw_fastq/ERR1442632_1.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,601,181 reads @   1.0 s/read;  62.19 M reads/minute\n",
      "Finished in 1.635 s (1.021 s/read; 58.74 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,601,181\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,601,181 (100.0%)\n",
      "\n",
      "Total basepairs processed:   160,118,100 bp\n",
      "Total written (filtered):    136,100,385 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442632_2.fastq.gz hw_fastq/ERR1442632_2.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,601,181 reads @   1.0 s/read;  58.68 M reads/minute\n",
      "Finished in 1.735 s (1.084 s/read; 55.36 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,601,181\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,601,181 (100.0%)\n",
      "\n",
      "Total basepairs processed:   160,118,100 bp\n",
      "Total written (filtered):    136,100,385 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442633_1.fastq.gz hw_fastq/ERR1442633_1.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,912,345 reads @   0.9 s/read;  64.35 M reads/minute\n",
      "Finished in 1.891 s (0.989 s/read; 60.69 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,912,345\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,912,345 (100.0%)\n",
      "\n",
      "Total basepairs processed:   191,234,500 bp\n",
      "Total written (filtered):    162,549,325 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442633_2.fastq.gz hw_fastq/ERR1442633_2.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,912,345 reads @   1.0 s/read;  58.71 M reads/minute\n",
      "Finished in 2.019 s (1.056 s/read; 56.84 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,912,345\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,912,345 (100.0%)\n",
      "\n",
      "Total basepairs processed:   191,234,500 bp\n",
      "Total written (filtered):    162,549,325 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442641_1.fastq.gz hw_fastq/ERR1442641_1.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:02     2,116,966 reads @   1.1 s/read;  55.80 M reads/minute\n",
      "Finished in 2.370 s (1.120 s/read; 53.59 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               2,116,966\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     2,116,966 (100.0%)\n",
      "\n",
      "Total basepairs processed:   211,696,600 bp\n",
      "Total written (filtered):    179,942,110 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442641_2.fastq.gz hw_fastq/ERR1442641_2.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:02     2,116,966 reads @   1.1 s/read;  54.26 M reads/minute\n",
      "Finished in 2.683 s (1.267 s/read; 47.34 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               2,116,966\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     2,116,966 (100.0%)\n",
      "\n",
      "Total basepairs processed:   211,696,600 bp\n",
      "Total written (filtered):    179,942,110 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442642_1.fastq.gz hw_fastq/ERR1442642_1.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:02     2,090,906 reads @   1.1 s/read;  55.47 M reads/minute\n",
      "Finished in 2.348 s (1.123 s/read; 53.43 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               2,090,906\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     2,090,906 (100.0%)\n",
      "\n",
      "Total basepairs processed:   209,090,600 bp\n",
      "Total written (filtered):    177,727,010 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442642_2.fastq.gz hw_fastq/ERR1442642_2.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:02     2,090,906 reads @   1.3 s/read;  47.98 M reads/minute\n",
      "Finished in 2.843 s (1.360 s/read; 44.13 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               2,090,906\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     2,090,906 (100.0%)\n",
      "\n",
      "Total basepairs processed:   209,090,600 bp\n",
      "Total written (filtered):    177,727,010 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442643_1.fastq.gz hw_fastq/ERR1442643_1.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,563,602 reads @   1.2 s/read;  51.29 M reads/minute\n",
      "Finished in 1.918 s (1.226 s/read; 48.92 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,563,602\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,563,602 (100.0%)\n",
      "\n",
      "Total basepairs processed:   156,360,200 bp\n",
      "Total written (filtered):    132,906,170 bp (85.0%)\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -j 32 -u 15 -m 85 -o ./hwtrimmed/ERR1442643_2.fastq.gz hw_fastq/ERR1442643_2.fastq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Done           00:00:01     1,563,602 reads @   1.2 s/read;  49.47 M reads/minute\n",
      "Finished in 1.988 s (1.271 s/read; 47.20 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,563,602\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                   0 (0.0%)\n",
      "Reads written (passing filters):     1,563,602 (100.0%)\n",
      "\n",
      "Total basepairs processed:   156,360,200 bp\n",
      "Total written (filtered):    132,906,170 bp (85.0%)\n"
     ]
    }
   ],
   "source": [
    "!for fastq in hw_fastq/*.fastq.gz; do \\\n",
    "    base_name=$(basename \"$fastq\" .fastq.gz); \\\n",
    "    cutadapt -j 32 -u 15 -m 85 -o ./hwtrimmed/$base_name.fastq.gz $fastq; \\\n",
    "done"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run QC on the trimmed/filtered reads and generate an aggregated report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.001s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /dev/cpuset.\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of ERR1442621_1.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of ERR1442621_2.fastq.gz\n",
      "Approx 5% complete for ERR1442621_1.fastq.gz\n",
      "Approx 5% complete for ERR1442621_2.fastq.gz\n",
      "Started analysis of ERR1442622_1.fastq.gz\n",
      "Approx 10% complete for ERR1442621_1.fastq.gz\n",
      "Approx 10% complete for ERR1442621_2.fastq.gz\n",
      "Approx 15% complete for ERR1442621_1.fastq.gz\n",
      "Approx 5% complete for ERR1442622_1.fastq.gz\n",
      "Started analysis of ERR1442622_2.fastq.gz\n",
      "Approx 15% complete for ERR1442621_2.fastq.gz\n",
      "Approx 20% complete for ERR1442621_1.fastq.gz\n",
      "Approx 10% complete for ERR1442622_1.fastq.gz\n",
      "Started analysis of ERR1442623_1.fastq.gz\n",
      "Approx 5% complete for ERR1442622_2.fastq.gz\n",
      "Approx 20% complete for ERR1442621_2.fastq.gz\n",
      "Approx 25% complete for ERR1442621_1.fastq.gz\n",
      "Approx 15% complete for ERR1442622_1.fastq.gz\n",
      "Approx 5% complete for ERR1442623_1.fastq.gz\n",
      "Started analysis of ERR1442623_2.fastq.gz\n",
      "Approx 10% complete for ERR1442622_2.fastq.gz\n",
      "Approx 25% complete for ERR1442621_2.fastq.gz\n",
      "Approx 30% complete for ERR1442621_1.fastq.gz\n",
      "Approx 20% complete for ERR1442622_1.fastq.gz\n",
      "Started analysis of ERR1442631_1.fastq.gz\n",
      "Approx 10% complete for ERR1442623_1.fastq.gz\n",
      "Approx 5% complete for ERR1442623_2.fastq.gz\n",
      "Approx 30% complete for ERR1442621_2.fastq.gz\n",
      "Approx 15% complete for ERR1442622_2.fastq.gz\n",
      "Approx 35% complete for ERR1442621_1.fastq.gz\n",
      "Approx 25% complete for ERR1442622_1.fastq.gz\n",
      "Started analysis of ERR1442631_2.fastq.gz\n",
      "Approx 15% complete for ERR1442623_1.fastq.gz\n",
      "Approx 10% complete for ERR1442623_2.fastq.gz\n",
      "Approx 35% complete for ERR1442621_2.fastq.gz\n",
      "Approx 5% complete for ERR1442631_1.fastq.gz\n",
      "Approx 40% complete for ERR1442621_1.fastq.gz\n",
      "Approx 20% complete for ERR1442622_2.fastq.gz\n",
      "Approx 30% complete for ERR1442622_1.fastq.gz\n",
      "Started analysis of ERR1442632_1.fastq.gz\n",
      "Approx 20% complete for ERR1442623_1.fastq.gz\n",
      "Approx 15% complete for ERR1442623_2.fastq.gz\n",
      "Approx 5% complete for ERR1442631_2.fastq.gz\n",
      "Approx 45% complete for ERR1442621_1.fastq.gz\n",
      "Approx 40% complete for ERR1442621_2.fastq.gz\n",
      "Approx 10% complete for ERR1442631_1.fastq.gz\n",
      "Approx 25% complete for ERR1442622_2.fastq.gz\n",
      "Approx 35% complete for ERR1442622_1.fastq.gz\n",
      "Approx 25% complete for ERR1442623_1.fastq.gz\n",
      "Approx 50% complete for ERR1442621_1.fastq.gz\n",
      "Started analysis of ERR1442632_2.fastq.gz\n",
      "Approx 20% complete for ERR1442623_2.fastq.gz\n",
      "Approx 45% complete for ERR1442621_2.fastq.gz\n",
      "Approx 10% complete for ERR1442631_2.fastq.gz\n",
      "Approx 5% complete for ERR1442632_1.fastq.gz\n",
      "Approx 15% complete for ERR1442631_1.fastq.gz\n",
      "Approx 30% complete for ERR1442622_2.fastq.gz\n",
      "Approx 40% complete for ERR1442622_1.fastq.gz\n",
      "Approx 30% complete for ERR1442623_1.fastq.gz\n",
      "Started analysis of ERR1442633_1.fastq.gz\n",
      "Approx 50% complete for ERR1442621_2.fastq.gz\n",
      "Approx 55% complete for ERR1442621_1.fastq.gz\n",
      "Approx 5% complete for ERR1442632_2.fastq.gz\n",
      "Approx 25% complete for ERR1442623_2.fastq.gz\n",
      "Approx 10% complete for ERR1442632_1.fastq.gz\n",
      "Approx 15% complete for ERR1442631_2.fastq.gz\n",
      "Approx 20% complete for ERR1442631_1.fastq.gz\n",
      "Approx 35% complete for ERR1442622_2.fastq.gz\n",
      "Approx 45% complete for ERR1442622_1.fastq.gz\n",
      "Approx 60% complete for ERR1442621_1.fastq.gz\n",
      "Approx 55% complete for ERR1442621_2.fastq.gz\n",
      "Approx 35% complete for ERR1442623_1.fastq.gz\n",
      "Started analysis of ERR1442633_2.fastq.gz\n",
      "Approx 10% complete for ERR1442632_2.fastq.gz\n",
      "Approx 30% complete for ERR1442623_2.fastq.gz\n",
      "Approx 15% complete for ERR1442632_1.fastq.gz\n",
      "Approx 25% complete for ERR1442631_1.fastq.gz\n",
      "Approx 20% complete for ERR1442631_2.fastq.gz\n",
      "Approx 65% complete for ERR1442621_1.fastq.gz\n",
      "Approx 40% complete for ERR1442622_2.fastq.gz\n",
      "Approx 50% complete for ERR1442622_1.fastq.gz\n",
      "Approx 60% complete for ERR1442621_2.fastq.gz\n",
      "Approx 40% complete for ERR1442623_1.fastq.gz\n",
      "Approx 15% complete for ERR1442632_2.fastq.gz\n",
      "Started analysis of ERR1442641_1.fastq.gz\n",
      "Approx 35% complete for ERR1442623_2.fastq.gz\n",
      "Approx 20% complete for ERR1442632_1.fastq.gz\n",
      "Approx 5% complete for ERR1442633_1.fastq.gz\n",
      "Approx 5% complete for ERR1442633_2.fastq.gz\n",
      "Approx 30% complete for ERR1442631_1.fastq.gz\n",
      "Approx 70% complete for ERR1442621_1.fastq.gz\n",
      "Approx 25% complete for ERR1442631_2.fastq.gz\n",
      "Approx 65% complete for ERR1442621_2.fastq.gz\n",
      "Started analysis of ERR1442641_2.fastq.gz\n",
      "Approx 55% complete for ERR1442622_1.fastq.gz\n",
      "Approx 45% complete for ERR1442622_2.fastq.gz\n",
      "Approx 45% complete for ERR1442623_1.fastq.gz\n",
      "Approx 20% complete for ERR1442632_2.fastq.gz\n",
      "Approx 25% complete for ERR1442632_1.fastq.gz\n",
      "Approx 40% complete for ERR1442623_2.fastq.gz\n",
      "Approx 5% complete for ERR1442641_1.fastq.gz\n",
      "Approx 75% complete for ERR1442621_1.fastq.gz\n",
      "Approx 35% complete for ERR1442631_1.fastq.gz\n",
      "Approx 10% complete for ERR1442633_2.fastq.gz\n",
      "Approx 70% complete for ERR1442621_2.fastq.gz\n",
      "Approx 10% complete for ERR1442633_1.fastq.gz\n",
      "Approx 30% complete for ERR1442631_2.fastq.gz\n",
      "Approx 60% complete for ERR1442622_1.fastq.gz\n",
      "Approx 50% complete for ERR1442623_1.fastq.gz\n",
      "Approx 50% complete for ERR1442622_2.fastq.gz\n",
      "Approx 30% complete for ERR1442632_1.fastq.gz\n",
      "Approx 5% complete for ERR1442641_2.fastq.gz\n",
      "Approx 25% complete for ERR1442632_2.fastq.gz\n",
      "Started analysis of ERR1442642_1.fastq.gz\n",
      "Approx 45% complete for ERR1442623_2.fastq.gz\n",
      "Approx 80% complete for ERR1442621_1.fastq.gz\n",
      "Approx 75% complete for ERR1442621_2.fastq.gz\n",
      "Approx 40% complete for ERR1442631_1.fastq.gz\n",
      "Approx 15% complete for ERR1442633_2.fastq.gz\n",
      "Approx 65% complete for ERR1442622_1.fastq.gz\n",
      "Approx 15% complete for ERR1442633_1.fastq.gz\n",
      "Approx 55% complete for ERR1442623_1.fastq.gz\n",
      "Approx 35% complete for ERR1442631_2.fastq.gz\n",
      "Approx 35% complete for ERR1442632_1.fastq.gz\n",
      "Approx 55% complete for ERR1442622_2.fastq.gz\n",
      "Approx 85% complete for ERR1442621_1.fastq.gz\n",
      "Approx 30% complete for ERR1442632_2.fastq.gz\n",
      "Approx 10% complete for ERR1442641_1.fastq.gz\n",
      "Approx 50% complete for ERR1442623_2.fastq.gz\n",
      "Approx 80% complete for ERR1442621_2.fastq.gz\n",
      "Approx 10% complete for ERR1442641_2.fastq.gz\n",
      "Approx 5% complete for ERR1442642_1.fastq.gz\n",
      "Started analysis of ERR1442642_2.fastq.gz\n",
      "Approx 70% complete for ERR1442622_1.fastq.gz\n",
      "Approx 45% complete for ERR1442631_1.fastq.gz\n",
      "Approx 60% complete for ERR1442623_1.fastq.gz\n",
      "Approx 20% complete for ERR1442633_2.fastq.gz\n",
      "Approx 40% complete for ERR1442631_2.fastq.gz\n",
      "Approx 40% complete for ERR1442632_1.fastq.gz\n",
      "Approx 20% complete for ERR1442633_1.fastq.gz\n",
      "Approx 90% complete for ERR1442621_1.fastq.gz\n",
      "Approx 60% complete for ERR1442622_2.fastq.gz\n",
      "Approx 35% complete for ERR1442632_2.fastq.gz\n",
      "Approx 55% complete for ERR1442623_2.fastq.gz\n",
      "Approx 85% complete for ERR1442621_2.fastq.gz\n",
      "Started analysis of ERR1442643_1.fastq.gz\n",
      "Approx 15% complete for ERR1442641_1.fastq.gz\n",
      "Approx 75% complete for ERR1442622_1.fastq.gz\n",
      "Approx 50% complete for ERR1442631_1.fastq.gz\n",
      "Approx 15% complete for ERR1442641_2.fastq.gz\n",
      "Approx 65% complete for ERR1442623_1.fastq.gz\n",
      "Approx 5% complete for ERR1442642_2.fastq.gz\n",
      "Approx 10% complete for ERR1442642_1.fastq.gz\n",
      "Approx 95% complete for ERR1442621_1.fastq.gz\n",
      "Approx 45% complete for ERR1442632_1.fastq.gz\n",
      "Approx 45% complete for ERR1442631_2.fastq.gz\n",
      "Approx 25% complete for ERR1442633_2.fastq.gz\n",
      "Approx 25% complete for ERR1442633_1.fastq.gz\n",
      "Approx 90% complete for ERR1442621_2.fastq.gz\n",
      "Approx 65% complete for ERR1442622_2.fastq.gz\n",
      "Approx 40% complete for ERR1442632_2.fastq.gz\n",
      "Approx 5% complete for ERR1442643_1.fastq.gz\n",
      "Approx 60% complete for ERR1442623_2.fastq.gz\n",
      "Approx 80% complete for ERR1442622_1.fastq.gz\n",
      "Approx 20% complete for ERR1442641_1.fastq.gz\n",
      "Approx 70% complete for ERR1442623_1.fastq.gz\n",
      "Approx 55% complete for ERR1442631_1.fastq.gz\n",
      "Analysis complete for ERR1442621_1.fastq.gz\n",
      "Started analysis of ERR1442643_2.fastq.gz\n",
      "Approx 50% complete for ERR1442632_1.fastq.gz\n",
      "Approx 50% complete for ERR1442631_2.fastq.gz\n",
      "Approx 20% complete for ERR1442641_2.fastq.gz\n",
      "Approx 95% complete for ERR1442621_2.fastq.gz\n",
      "Approx 10% complete for ERR1442642_2.fastq.gz\n",
      "Approx 15% complete for ERR1442642_1.fastq.gz\n",
      "Approx 30% complete for ERR1442633_2.fastq.gz\n",
      "Approx 10% complete for ERR1442643_1.fastq.gz\n",
      "Approx 70% complete for ERR1442622_2.fastq.gz\n",
      "Approx 65% complete for ERR1442623_2.fastq.gz\n",
      "Approx 30% complete for ERR1442633_1.fastq.gz\n",
      "Approx 45% complete for ERR1442632_2.fastq.gz\n",
      "Approx 85% complete for ERR1442622_1.fastq.gz\n",
      "Approx 75% complete for ERR1442623_1.fastq.gz\n",
      "Approx 5% complete for ERR1442643_2.fastq.gz\n",
      "Approx 55% complete for ERR1442632_1.fastq.gz\n",
      "Approx 60% complete for ERR1442631_1.fastq.gz\n",
      "Approx 25% complete for ERR1442641_1.fastq.gz\n",
      "Approx 55% complete for ERR1442631_2.fastq.gz\n",
      "Analysis complete for ERR1442621_2.fastq.gz\n",
      "Approx 15% complete for ERR1442643_1.fastq.gz\n",
      "Approx 70% complete for ERR1442623_2.fastq.gz\n",
      "Approx 50% complete for ERR1442632_2.fastq.gz\n",
      "Approx 75% complete for ERR1442622_2.fastq.gz\n",
      "Approx 35% complete for ERR1442633_2.fastq.gz\n",
      "Approx 20% complete for ERR1442642_1.fastq.gz\n",
      "Approx 90% complete for ERR1442622_1.fastq.gz\n",
      "Approx 15% complete for ERR1442642_2.fastq.gz\n",
      "Approx 35% complete for ERR1442633_1.fastq.gz\n",
      "Approx 80% complete for ERR1442623_1.fastq.gz\n",
      "Approx 10% complete for ERR1442643_2.fastq.gz\n",
      "Approx 60% complete for ERR1442632_1.fastq.gz\n",
      "Approx 25% complete for ERR1442641_2.fastq.gz\n",
      "Approx 65% complete for ERR1442631_1.fastq.gz\n",
      "Approx 30% complete for ERR1442641_1.fastq.gz\n",
      "Approx 20% complete for ERR1442643_1.fastq.gz\n",
      "Approx 75% complete for ERR1442623_2.fastq.gz\n",
      "Approx 55% complete for ERR1442632_2.fastq.gz\n",
      "Approx 80% complete for ERR1442622_2.fastq.gz\n",
      "Approx 60% complete for ERR1442631_2.fastq.gz\n",
      "Approx 95% complete for ERR1442622_1.fastq.gz\n",
      "Approx 40% complete for ERR1442633_2.fastq.gz\n",
      "Approx 85% complete for ERR1442623_1.fastq.gz\n",
      "Approx 15% complete for ERR1442643_2.fastq.gz\n",
      "Approx 65% complete for ERR1442632_1.fastq.gz\n",
      "Approx 25% complete for ERR1442642_1.fastq.gz\n",
      "Approx 40% complete for ERR1442633_1.fastq.gz\n",
      "Approx 20% complete for ERR1442642_2.fastq.gz\n",
      "Approx 70% complete for ERR1442631_1.fastq.gz\n",
      "Approx 80% complete for ERR1442623_2.fastq.gz\n",
      "Approx 25% complete for ERR1442643_1.fastq.gz\n",
      "Approx 60% complete for ERR1442632_2.fastq.gz\n",
      "Approx 30% complete for ERR1442641_2.fastq.gz\n",
      "Approx 85% complete for ERR1442622_2.fastq.gz\n",
      "Approx 35% complete for ERR1442641_1.fastq.gz\n",
      "Analysis complete for ERR1442622_1.fastq.gz\n",
      "Approx 65% complete for ERR1442631_2.fastq.gz\n",
      "Approx 90% complete for ERR1442623_1.fastq.gz\n",
      "Approx 70% complete for ERR1442632_1.fastq.gz\n",
      "Approx 20% complete for ERR1442643_2.fastq.gz\n",
      "Approx 45% complete for ERR1442633_2.fastq.gz\n",
      "Approx 75% complete for ERR1442631_1.fastq.gz\n",
      "Approx 45% complete for ERR1442633_1.fastq.gz\n",
      "Approx 85% complete for ERR1442623_2.fastq.gz\n",
      "Approx 30% complete for ERR1442642_1.fastq.gz\n",
      "Approx 30% complete for ERR1442643_1.fastq.gz\n",
      "Approx 25% complete for ERR1442642_2.fastq.gz\n",
      "Approx 65% complete for ERR1442632_2.fastq.gz\n",
      "Approx 90% complete for ERR1442622_2.fastq.gz\n",
      "Approx 75% complete for ERR1442632_1.fastq.gz\n",
      "Approx 95% complete for ERR1442623_1.fastq.gz\n",
      "Approx 25% complete for ERR1442643_2.fastq.gz\n",
      "Approx 35% complete for ERR1442641_2.fastq.gz\n",
      "Approx 70% complete for ERR1442631_2.fastq.gz\n",
      "Approx 40% complete for ERR1442641_1.fastq.gz\n",
      "Approx 50% complete for ERR1442633_2.fastq.gz\n",
      "Approx 80% complete for ERR1442631_1.fastq.gz\n",
      "Approx 90% complete for ERR1442623_2.fastq.gz\n",
      "Approx 35% complete for ERR1442643_1.fastq.gz\n",
      "Approx 50% complete for ERR1442633_1.fastq.gz\n",
      "Approx 70% complete for ERR1442632_2.fastq.gz\n",
      "Approx 35% complete for ERR1442642_1.fastq.gz\n",
      "Approx 80% complete for ERR1442632_1.fastq.gz\n",
      "Approx 95% complete for ERR1442622_2.fastq.gz\n",
      "Approx 30% complete for ERR1442642_2.fastq.gz\n",
      "Analysis complete for ERR1442623_1.fastq.gz\n",
      "Approx 30% complete for ERR1442643_2.fastq.gz\n",
      "Approx 75% complete for ERR1442631_2.fastq.gz\n",
      "Approx 85% complete for ERR1442631_1.fastq.gz\n",
      "Approx 40% complete for ERR1442641_2.fastq.gz\n",
      "Approx 55% complete for ERR1442633_2.fastq.gz\n",
      "Approx 45% complete for ERR1442641_1.fastq.gz\n",
      "Approx 40% complete for ERR1442643_1.fastq.gz\n",
      "Approx 95% complete for ERR1442623_2.fastq.gz\n",
      "Approx 85% complete for ERR1442632_1.fastq.gz\n",
      "Approx 75% complete for ERR1442632_2.fastq.gz\n",
      "Approx 55% complete for ERR1442633_1.fastq.gz\n",
      "Analysis complete for ERR1442622_2.fastq.gz\n",
      "Approx 35% complete for ERR1442643_2.fastq.gz\n",
      "Approx 40% complete for ERR1442642_1.fastq.gz\n",
      "Approx 90% complete for ERR1442631_1.fastq.gz\n",
      "Approx 80% complete for ERR1442631_2.fastq.gz\n",
      "Approx 35% complete for ERR1442642_2.fastq.gz\n",
      "Approx 45% complete for ERR1442643_1.fastq.gz\n",
      "Analysis complete for ERR1442623_2.fastq.gz\n",
      "Approx 60% complete for ERR1442633_2.fastq.gz\n",
      "Approx 50% complete for ERR1442641_1.fastq.gz\n",
      "Approx 90% complete for ERR1442632_1.fastq.gz\n",
      "Approx 45% complete for ERR1442641_2.fastq.gz\n",
      "Approx 80% complete for ERR1442632_2.fastq.gz\n",
      "Approx 40% complete for ERR1442643_2.fastq.gz\n",
      "Approx 60% complete for ERR1442633_1.fastq.gz\n",
      "Approx 95% complete for ERR1442631_1.fastq.gz\n",
      "Approx 50% complete for ERR1442643_1.fastq.gz\n",
      "Approx 85% complete for ERR1442631_2.fastq.gz\n",
      "Approx 95% complete for ERR1442632_1.fastq.gz\n",
      "Approx 40% complete for ERR1442642_2.fastq.gz\n",
      "Approx 85% complete for ERR1442632_2.fastq.gz\n",
      "Approx 65% complete for ERR1442633_2.fastq.gz\n",
      "Approx 45% complete for ERR1442643_2.fastq.gz\n",
      "Approx 50% complete for ERR1442641_2.fastq.gz\n",
      "Approx 55% complete for ERR1442641_1.fastq.gz\n",
      "Analysis complete for ERR1442631_1.fastq.gz\n",
      "Approx 65% complete for ERR1442633_1.fastq.gz\n",
      "Approx 55% complete for ERR1442643_1.fastq.gz\n",
      "Analysis complete for ERR1442632_1.fastq.gz\n",
      "Approx 90% complete for ERR1442631_2.fastq.gz\n",
      "Approx 50% complete for ERR1442643_2.fastq.gz\n",
      "Approx 90% complete for ERR1442632_2.fastq.gz\n",
      "Approx 70% complete for ERR1442633_2.fastq.gz\n",
      "Approx 45% complete for ERR1442642_2.fastq.gz\n",
      "Approx 55% complete for ERR1442641_2.fastq.gz\n",
      "Approx 45% complete for ERR1442642_1.fastq.gz\n",
      "Approx 60% complete for ERR1442643_1.fastq.gz\n",
      "Approx 60% complete for ERR1442641_1.fastq.gz\n",
      "Approx 70% complete for ERR1442633_1.fastq.gz\n",
      "Approx 95% complete for ERR1442632_2.fastq.gz\n",
      "Approx 55% complete for ERR1442643_2.fastq.gz\n",
      "Approx 95% complete for ERR1442631_2.fastq.gz\n",
      "Approx 75% complete for ERR1442633_2.fastq.gz\n",
      "Approx 50% complete for ERR1442642_2.fastq.gz\n",
      "Approx 65% complete for ERR1442643_1.fastq.gz\n",
      "Analysis complete for ERR1442632_2.fastq.gz\n",
      "Approx 75% complete for ERR1442633_1.fastq.gz\n",
      "Approx 50% complete for ERR1442642_1.fastq.gz\n",
      "Approx 60% complete for ERR1442641_2.fastq.gz\n",
      "Approx 65% complete for ERR1442641_1.fastq.gz\n",
      "Approx 60% complete for ERR1442643_2.fastq.gz\n",
      "Analysis complete for ERR1442631_2.fastq.gz\n",
      "Approx 80% complete for ERR1442633_2.fastq.gz\n",
      "Approx 70% complete for ERR1442643_1.fastq.gz\n",
      "Approx 80% complete for ERR1442633_1.fastq.gz\n",
      "Approx 65% complete for ERR1442643_2.fastq.gz\n",
      "Approx 55% complete for ERR1442642_2.fastq.gz\n",
      "Approx 70% complete for ERR1442641_1.fastq.gz\n",
      "Approx 65% complete for ERR1442641_2.fastq.gz\n",
      "Approx 55% complete for ERR1442642_1.fastq.gz\n",
      "Approx 75% complete for ERR1442643_1.fastq.gz\n",
      "Approx 85% complete for ERR1442633_2.fastq.gz\n",
      "Approx 70% complete for ERR1442643_2.fastq.gz\n",
      "Approx 85% complete for ERR1442633_1.fastq.gz\n",
      "Approx 80% complete for ERR1442643_1.fastq.gz\n",
      "Approx 60% complete for ERR1442642_2.fastq.gz\n",
      "Approx 75% complete for ERR1442641_1.fastq.gz\n",
      "Approx 70% complete for ERR1442641_2.fastq.gz\n",
      "Approx 90% complete for ERR1442633_2.fastq.gz\n",
      "Approx 75% complete for ERR1442643_2.fastq.gz\n",
      "Approx 60% complete for ERR1442642_1.fastq.gz\n",
      "Approx 90% complete for ERR1442633_1.fastq.gz\n",
      "Approx 85% complete for ERR1442643_1.fastq.gz\n",
      "Approx 80% complete for ERR1442641_1.fastq.gz\n",
      "Approx 80% complete for ERR1442643_2.fastq.gz\n",
      "Approx 65% complete for ERR1442642_2.fastq.gz\n",
      "Approx 95% complete for ERR1442633_2.fastq.gz\n",
      "Approx 75% complete for ERR1442641_2.fastq.gz\n",
      "Approx 95% complete for ERR1442633_1.fastq.gz\n",
      "Approx 65% complete for ERR1442642_1.fastq.gz\n",
      "Approx 90% complete for ERR1442643_1.fastq.gz\n",
      "Approx 85% complete for ERR1442643_2.fastq.gz\n",
      "Approx 85% complete for ERR1442641_1.fastq.gz\n",
      "Analysis complete for ERR1442633_2.fastq.gz\n",
      "Approx 95% complete for ERR1442643_1.fastq.gz\n",
      "Approx 80% complete for ERR1442641_2.fastq.gz\n",
      "Analysis complete for ERR1442633_1.fastq.gz\n",
      "Approx 70% complete for ERR1442642_2.fastq.gz\n",
      "Approx 70% complete for ERR1442642_1.fastq.gz\n",
      "Approx 90% complete for ERR1442643_2.fastq.gz\n",
      "Analysis complete for ERR1442643_1.fastq.gz\n",
      "Approx 90% complete for ERR1442641_1.fastq.gz\n",
      "Approx 85% complete for ERR1442641_2.fastq.gz\n",
      "Approx 95% complete for ERR1442643_2.fastq.gz\n",
      "Approx 75% complete for ERR1442642_2.fastq.gz\n",
      "Approx 75% complete for ERR1442642_1.fastq.gz\n",
      "Approx 95% complete for ERR1442641_1.fastq.gz\n",
      "Analysis complete for ERR1442643_2.fastq.gz\n",
      "Approx 90% complete for ERR1442641_2.fastq.gz\n",
      "Approx 80% complete for ERR1442642_2.fastq.gz\n",
      "Approx 80% complete for ERR1442642_1.fastq.gz\n",
      "Analysis complete for ERR1442641_1.fastq.gz\n",
      "Approx 85% complete for ERR1442642_2.fastq.gz\n",
      "Approx 95% complete for ERR1442641_2.fastq.gz\n",
      "Approx 85% complete for ERR1442642_1.fastq.gz\n",
      "Approx 90% complete for ERR1442642_2.fastq.gz\n",
      "Analysis complete for ERR1442641_2.fastq.gz\n",
      "Approx 90% complete for ERR1442642_1.fastq.gz\n",
      "Approx 95% complete for ERR1442642_2.fastq.gz\n",
      "Approx 95% complete for ERR1442642_1.fastq.gz\n",
      "Analysis complete for ERR1442642_2.fastq.gz\n",
      "Analysis complete for ERR1442642_1.fastq.gz\n",
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=347897;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.17\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Prepending directory to sample names\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/heky1803/BIOL343/5_fastq/hw_fastq/qc\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/heky1803/BIOL343/5_fastq/hwtrimmed/qc\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m72/72\u001b[0m  wtrimmed/qc/ERR1442643_1_fastqc.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 36 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc_report.html   (overwritten)\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc_data   (overwritten)\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "!fastqc -t 32 hwtrimmed/*.fastq.gz -o hwtrimmed/qc\n",
    "!multiqc --force -d hw_fastq/qc/ hwtrimmed/qc/"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Report your observations on which QC metrics improved (or got worse): all red fails disappeared from \"Per Base Sequence Content\" and \"Per Sequence GC Content\"\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Snakemake Pipeline"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When running bioinformatics studies, it is wise to use notebooks (such as this one) to tie together code, decisions, and data visualization. The above task is something you might generate and report to a client or supervisor who might be interested in sequencing QC and what kind of pre-processing decisions you made prior to alignment. In addition, the preprocessing should be readily reproducible by all relevant stakeholders, so you should supplement notebooks with pipeline scripts. Throughout the semester, we will be creating snakemake pipelines to reproduce the published analysis that you've selected. Follow the below steps to begin producing this pipeline:\n",
    "\n",
    "1. Make a new `Snakefile` and `config.yaml` in `5_fastq`. \n",
    "2. In the `Snakefile`, produce the following rules:  \n",
    "    a. `fetch_fastq` - use `parallel-fastq-dump` to fetch every FASTQ file listed in `config.yaml`.  \n",
    "    b. `fastq_qc` - use `fastqc` to run QC on the raw FASTQ files.  \n",
    "    c. `trim_filter` - use `cutadapt` to trim and/or filter the FASTQ reads using parameters justified in this notebook. ***Note:*** if you choose to trim/filter, it should be done on *all* of the FASTQ files; do not cherry pick.  \n",
    "    d. `trim_qc` - use `fastqc` to run QC on the trimmed FASTQ files.  \n",
    "    e. `report` - use `multiqc` to aggregate all QC data into a single report.\n",
    "\n",
    "This `Snakefile` should not use any `wrappers` because we already have the relevant software installed in the `biol343` conda environment. The pipeline should run to completion when any instructor or classmate runs `snakemake --use-conda`. You may use the documentation found [here](https://multiqc.info/docs/usage/pipelines/#snakemake) or any other online documentation you may find. You may not use any AI tools to complete this homework."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biol343",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
