{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 5 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 10/18 by 11:59 pm***"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTQ fetch, QC, and trimming/filtering\n",
    "\n",
    "Use the SRA-toolkit to fetch the relevant FASTQ files. You may use a hand-crafted `txt` file to provide the list of SRAs instead of manipulating an SRA Run Table with `cut`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ERR4231022', 'ERR4231010', 'ERR4231012', 'ERR4231013', 'ERR4231014', 'ERR4231015', 'ERR4231016', 'ERR4231017', 'ERR4231018', 'ERR4231019', 'ERR4231020', 'ERR4231021', 'ERR4231023', 'ERR4231011']\n"
     ]
    }
   ],
   "source": [
    "# ERR4231022\n",
    "# ERR4231010\n",
    "# ERR4231012\n",
    "# ERR4231013\n",
    "# ERR4231014\n",
    "# ERR4231015\n",
    "# ERR4231016\n",
    "# ERR4231017\n",
    "# ERR4231018\n",
    "# ERR4231019\n",
    "# ERR4231020\n",
    "# ERR4231021\n",
    "# ERR4231023\n",
    "# ERR4231011\n",
    "# print(\"ERR4231022 ERR4231010 ERR4231012 ERR4231013 ERR4231014 ERR4231015 ERR4231016 ERR4231017 ERR4231018 ERR4231019 ERR4231020 ERR4231021 ERR4231023 ERR4231011\".split(\" \"))\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run QC on the FASTQ files and aggregate the results into a report."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting ERR4231022 from NCBI SRA\n",
      "2024-10-31 15:35:06,751 - SRR ids: ['ERR4231022']\n",
      "2024-10-31 15:35:06,751 - extra args: ['--gzip']\n",
      "2024-10-31 15:35:06,751 - tempdir: /local/scratch/job_102381/pfd_s1fwvzt3\n",
      "2024-10-31 15:35:06,752 - CMD: sra-stat --meta --quick ERR4231022\n",
      "2024-10-31 15:35:09,579 - ERR4231022 spots: 55208970\n",
      "2024-10-31 15:35:09,579 - blocks: [[1, 3450560], [3450561, 6901120], [6901121, 10351680], [10351681, 13802240], [13802241, 17252800], [17252801, 20703360], [20703361, 24153920], [24153921, 27604480], [27604481, 31055040], [31055041, 34505600], [34505601, 37956160], [37956161, 41406720], [41406721, 44857280], [44857281, 48307840], [48307841, 51758400], [51758401, 55208970]]\n",
      "2024-10-31 15:35:09,579 - CMD: fastq-dump -N 1 -X 3450560 -O /local/scratch/job_102381/pfd_s1fwvzt3/0 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,581 - CMD: fastq-dump -N 3450561 -X 6901120 -O /local/scratch/job_102381/pfd_s1fwvzt3/1 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,581 - CMD: fastq-dump -N 6901121 -X 10351680 -O /local/scratch/job_102381/pfd_s1fwvzt3/2 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,582 - CMD: fastq-dump -N 10351681 -X 13802240 -O /local/scratch/job_102381/pfd_s1fwvzt3/3 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,582 - CMD: fastq-dump -N 13802241 -X 17252800 -O /local/scratch/job_102381/pfd_s1fwvzt3/4 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,582 - CMD: fastq-dump -N 17252801 -X 20703360 -O /local/scratch/job_102381/pfd_s1fwvzt3/5 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,583 - CMD: fastq-dump -N 20703361 -X 24153920 -O /local/scratch/job_102381/pfd_s1fwvzt3/6 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,583 - CMD: fastq-dump -N 24153921 -X 27604480 -O /local/scratch/job_102381/pfd_s1fwvzt3/7 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,583 - CMD: fastq-dump -N 27604481 -X 31055040 -O /local/scratch/job_102381/pfd_s1fwvzt3/8 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,584 - CMD: fastq-dump -N 31055041 -X 34505600 -O /local/scratch/job_102381/pfd_s1fwvzt3/9 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,584 - CMD: fastq-dump -N 34505601 -X 37956160 -O /local/scratch/job_102381/pfd_s1fwvzt3/10 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,584 - CMD: fastq-dump -N 37956161 -X 41406720 -O /local/scratch/job_102381/pfd_s1fwvzt3/11 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,584 - CMD: fastq-dump -N 41406721 -X 44857280 -O /local/scratch/job_102381/pfd_s1fwvzt3/12 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,585 - CMD: fastq-dump -N 44857281 -X 48307840 -O /local/scratch/job_102381/pfd_s1fwvzt3/13 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,585 - CMD: fastq-dump -N 48307841 -X 51758400 -O /local/scratch/job_102381/pfd_s1fwvzt3/14 --gzip ERR4231022\n",
      "2024-10-31 15:35:09,585 - CMD: fastq-dump -N 51758401 -X 55208970 -O /local/scratch/job_102381/pfd_s1fwvzt3/15 --gzip ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450570 spots for ERR4231022\n",
      "Written 3450570 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Read 3450560 spots for ERR4231022\n",
      "Written 3450560 spots for ERR4231022\n",
      "Getting ERR4231010 from NCBI SRA\n",
      "2024-10-31 15:40:45,380 - SRR ids: ['ERR4231010']\n",
      "2024-10-31 15:40:45,380 - extra args: ['--gzip']\n",
      "2024-10-31 15:40:45,381 - tempdir: /local/scratch/job_102381/pfd_h50ftnbg\n",
      "2024-10-31 15:40:45,381 - CMD: sra-stat --meta --quick ERR4231010\n",
      "2024-10-31 15:40:47,755 - ERR4231010 spots: 64341577\n",
      "2024-10-31 15:40:47,755 - blocks: [[1, 4021348], [4021349, 8042696], [8042697, 12064044], [12064045, 16085392], [16085393, 20106740], [20106741, 24128088], [24128089, 28149436], [28149437, 32170784], [32170785, 36192132], [36192133, 40213480], [40213481, 44234828], [44234829, 48256176], [48256177, 52277524], [52277525, 56298872], [56298873, 60320220], [60320221, 64341577]]\n",
      "2024-10-31 15:40:47,755 - CMD: fastq-dump -N 1 -X 4021348 -O /local/scratch/job_102381/pfd_h50ftnbg/0 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,756 - CMD: fastq-dump -N 4021349 -X 8042696 -O /local/scratch/job_102381/pfd_h50ftnbg/1 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,756 - CMD: fastq-dump -N 8042697 -X 12064044 -O /local/scratch/job_102381/pfd_h50ftnbg/2 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,756 - CMD: fastq-dump -N 12064045 -X 16085392 -O /local/scratch/job_102381/pfd_h50ftnbg/3 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,757 - CMD: fastq-dump -N 16085393 -X 20106740 -O /local/scratch/job_102381/pfd_h50ftnbg/4 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,757 - CMD: fastq-dump -N 20106741 -X 24128088 -O /local/scratch/job_102381/pfd_h50ftnbg/5 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,757 - CMD: fastq-dump -N 24128089 -X 28149436 -O /local/scratch/job_102381/pfd_h50ftnbg/6 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,757 - CMD: fastq-dump -N 28149437 -X 32170784 -O /local/scratch/job_102381/pfd_h50ftnbg/7 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,758 - CMD: fastq-dump -N 32170785 -X 36192132 -O /local/scratch/job_102381/pfd_h50ftnbg/8 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,758 - CMD: fastq-dump -N 36192133 -X 40213480 -O /local/scratch/job_102381/pfd_h50ftnbg/9 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,758 - CMD: fastq-dump -N 40213481 -X 44234828 -O /local/scratch/job_102381/pfd_h50ftnbg/10 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,759 - CMD: fastq-dump -N 44234829 -X 48256176 -O /local/scratch/job_102381/pfd_h50ftnbg/11 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,759 - CMD: fastq-dump -N 48256177 -X 52277524 -O /local/scratch/job_102381/pfd_h50ftnbg/12 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,759 - CMD: fastq-dump -N 52277525 -X 56298872 -O /local/scratch/job_102381/pfd_h50ftnbg/13 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,760 - CMD: fastq-dump -N 56298873 -X 60320220 -O /local/scratch/job_102381/pfd_h50ftnbg/14 --gzip ERR4231010\n",
      "2024-10-31 15:40:47,773 - CMD: fastq-dump -N 60320221 -X 64341577 -O /local/scratch/job_102381/pfd_h50ftnbg/15 --gzip ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021357 spots for ERR4231010\n",
      "Written 4021357 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Read 4021348 spots for ERR4231010\n",
      "Written 4021348 spots for ERR4231010\n",
      "Getting ERR4231012 from NCBI SRA\n",
      "2024-10-31 15:47:29,767 - SRR ids: ['ERR4231012']\n",
      "2024-10-31 15:47:29,767 - extra args: ['--gzip']\n",
      "2024-10-31 15:47:29,767 - tempdir: /local/scratch/job_102381/pfd_53j8t1pz\n",
      "2024-10-31 15:47:29,767 - CMD: sra-stat --meta --quick ERR4231012\n",
      "2024-10-31 15:47:32,688 - ERR4231012 spots: 73535954\n",
      "2024-10-31 15:47:32,688 - blocks: [[1, 4595997], [4595998, 9191994], [9191995, 13787991], [13787992, 18383988], [18383989, 22979985], [22979986, 27575982], [27575983, 32171979], [32171980, 36767976], [36767977, 41363973], [41363974, 45959970], [45959971, 50555967], [50555968, 55151964], [55151965, 59747961], [59747962, 64343958], [64343959, 68939955], [68939956, 73535954]]\n",
      "2024-10-31 15:47:32,688 - CMD: fastq-dump -N 1 -X 4595997 -O /local/scratch/job_102381/pfd_53j8t1pz/0 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,689 - CMD: fastq-dump -N 4595998 -X 9191994 -O /local/scratch/job_102381/pfd_53j8t1pz/1 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,690 - CMD: fastq-dump -N 9191995 -X 13787991 -O /local/scratch/job_102381/pfd_53j8t1pz/2 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,690 - CMD: fastq-dump -N 13787992 -X 18383988 -O /local/scratch/job_102381/pfd_53j8t1pz/3 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,690 - CMD: fastq-dump -N 18383989 -X 22979985 -O /local/scratch/job_102381/pfd_53j8t1pz/4 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,691 - CMD: fastq-dump -N 22979986 -X 27575982 -O /local/scratch/job_102381/pfd_53j8t1pz/5 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,691 - CMD: fastq-dump -N 27575983 -X 32171979 -O /local/scratch/job_102381/pfd_53j8t1pz/6 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,691 - CMD: fastq-dump -N 32171980 -X 36767976 -O /local/scratch/job_102381/pfd_53j8t1pz/7 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,692 - CMD: fastq-dump -N 36767977 -X 41363973 -O /local/scratch/job_102381/pfd_53j8t1pz/8 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,692 - CMD: fastq-dump -N 41363974 -X 45959970 -O /local/scratch/job_102381/pfd_53j8t1pz/9 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,693 - CMD: fastq-dump -N 45959971 -X 50555967 -O /local/scratch/job_102381/pfd_53j8t1pz/10 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,694 - CMD: fastq-dump -N 50555968 -X 55151964 -O /local/scratch/job_102381/pfd_53j8t1pz/11 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,694 - CMD: fastq-dump -N 55151965 -X 59747961 -O /local/scratch/job_102381/pfd_53j8t1pz/12 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,695 - CMD: fastq-dump -N 59747962 -X 64343958 -O /local/scratch/job_102381/pfd_53j8t1pz/13 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,696 - CMD: fastq-dump -N 64343959 -X 68939955 -O /local/scratch/job_102381/pfd_53j8t1pz/14 --gzip ERR4231012\n",
      "2024-10-31 15:47:32,696 - CMD: fastq-dump -N 68939956 -X 73535954 -O /local/scratch/job_102381/pfd_53j8t1pz/15 --gzip ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595999 spots for ERR4231012\n",
      "Written 4595999 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Read 4595997 spots for ERR4231012\n",
      "Written 4595997 spots for ERR4231012\n",
      "Getting ERR4231013 from NCBI SRA\n",
      "2024-10-31 15:55:02,741 - SRR ids: ['ERR4231013']\n",
      "2024-10-31 15:55:02,741 - extra args: ['--gzip']\n",
      "2024-10-31 15:55:02,742 - tempdir: /local/scratch/job_102381/pfd_ofggmnoq\n",
      "2024-10-31 15:55:02,742 - CMD: sra-stat --meta --quick ERR4231013\n",
      "2024-10-31 15:55:05,130 - ERR4231013 spots: 62889838\n",
      "2024-10-31 15:55:05,130 - blocks: [[1, 3930614], [3930615, 7861228], [7861229, 11791842], [11791843, 15722456], [15722457, 19653070], [19653071, 23583684], [23583685, 27514298], [27514299, 31444912], [31444913, 35375526], [35375527, 39306140], [39306141, 43236754], [43236755, 47167368], [47167369, 51097982], [51097983, 55028596], [55028597, 58959210], [58959211, 62889838]]\n",
      "2024-10-31 15:55:05,130 - CMD: fastq-dump -N 1 -X 3930614 -O /local/scratch/job_102381/pfd_ofggmnoq/0 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,156 - CMD: fastq-dump -N 3930615 -X 7861228 -O /local/scratch/job_102381/pfd_ofggmnoq/1 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,156 - CMD: fastq-dump -N 7861229 -X 11791842 -O /local/scratch/job_102381/pfd_ofggmnoq/2 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,156 - CMD: fastq-dump -N 11791843 -X 15722456 -O /local/scratch/job_102381/pfd_ofggmnoq/3 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,157 - CMD: fastq-dump -N 15722457 -X 19653070 -O /local/scratch/job_102381/pfd_ofggmnoq/4 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,157 - CMD: fastq-dump -N 19653071 -X 23583684 -O /local/scratch/job_102381/pfd_ofggmnoq/5 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,157 - CMD: fastq-dump -N 23583685 -X 27514298 -O /local/scratch/job_102381/pfd_ofggmnoq/6 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,158 - CMD: fastq-dump -N 27514299 -X 31444912 -O /local/scratch/job_102381/pfd_ofggmnoq/7 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,158 - CMD: fastq-dump -N 31444913 -X 35375526 -O /local/scratch/job_102381/pfd_ofggmnoq/8 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,158 - CMD: fastq-dump -N 35375527 -X 39306140 -O /local/scratch/job_102381/pfd_ofggmnoq/9 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,160 - CMD: fastq-dump -N 39306141 -X 43236754 -O /local/scratch/job_102381/pfd_ofggmnoq/10 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,160 - CMD: fastq-dump -N 43236755 -X 47167368 -O /local/scratch/job_102381/pfd_ofggmnoq/11 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,161 - CMD: fastq-dump -N 47167369 -X 51097982 -O /local/scratch/job_102381/pfd_ofggmnoq/12 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,166 - CMD: fastq-dump -N 51097983 -X 55028596 -O /local/scratch/job_102381/pfd_ofggmnoq/13 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,169 - CMD: fastq-dump -N 55028597 -X 58959210 -O /local/scratch/job_102381/pfd_ofggmnoq/14 --gzip ERR4231013\n",
      "2024-10-31 15:55:05,169 - CMD: fastq-dump -N 58959211 -X 62889838 -O /local/scratch/job_102381/pfd_ofggmnoq/15 --gzip ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Read 3930628 spots for ERR4231013\n",
      "Written 3930628 spots for ERR4231013\n",
      "Read 3930614 spots for ERR4231013\n",
      "Written 3930614 spots for ERR4231013\n",
      "Getting ERR4231014 from NCBI SRA\n",
      "2024-10-31 16:01:24,705 - SRR ids: ['ERR4231014']\n",
      "2024-10-31 16:01:24,705 - extra args: ['--gzip']\n",
      "2024-10-31 16:01:24,705 - tempdir: /local/scratch/job_102381/pfd_isdid1wa\n",
      "2024-10-31 16:01:24,705 - CMD: sra-stat --meta --quick ERR4231014\n",
      "2024-10-31 16:01:27,101 - ERR4231014 spots: 36185523\n",
      "2024-10-31 16:01:27,101 - blocks: [[1, 2261595], [2261596, 4523190], [4523191, 6784785], [6784786, 9046380], [9046381, 11307975], [11307976, 13569570], [13569571, 15831165], [15831166, 18092760], [18092761, 20354355], [20354356, 22615950], [22615951, 24877545], [24877546, 27139140], [27139141, 29400735], [29400736, 31662330], [31662331, 33923925], [33923926, 36185523]]\n",
      "2024-10-31 16:01:27,101 - CMD: fastq-dump -N 1 -X 2261595 -O /local/scratch/job_102381/pfd_isdid1wa/0 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,102 - CMD: fastq-dump -N 2261596 -X 4523190 -O /local/scratch/job_102381/pfd_isdid1wa/1 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,102 - CMD: fastq-dump -N 4523191 -X 6784785 -O /local/scratch/job_102381/pfd_isdid1wa/2 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,103 - CMD: fastq-dump -N 6784786 -X 9046380 -O /local/scratch/job_102381/pfd_isdid1wa/3 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,103 - CMD: fastq-dump -N 9046381 -X 11307975 -O /local/scratch/job_102381/pfd_isdid1wa/4 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,103 - CMD: fastq-dump -N 11307976 -X 13569570 -O /local/scratch/job_102381/pfd_isdid1wa/5 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,103 - CMD: fastq-dump -N 13569571 -X 15831165 -O /local/scratch/job_102381/pfd_isdid1wa/6 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,104 - CMD: fastq-dump -N 15831166 -X 18092760 -O /local/scratch/job_102381/pfd_isdid1wa/7 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,105 - CMD: fastq-dump -N 18092761 -X 20354355 -O /local/scratch/job_102381/pfd_isdid1wa/8 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,105 - CMD: fastq-dump -N 20354356 -X 22615950 -O /local/scratch/job_102381/pfd_isdid1wa/9 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,105 - CMD: fastq-dump -N 22615951 -X 24877545 -O /local/scratch/job_102381/pfd_isdid1wa/10 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,106 - CMD: fastq-dump -N 24877546 -X 27139140 -O /local/scratch/job_102381/pfd_isdid1wa/11 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,106 - CMD: fastq-dump -N 27139141 -X 29400735 -O /local/scratch/job_102381/pfd_isdid1wa/12 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,107 - CMD: fastq-dump -N 29400736 -X 31662330 -O /local/scratch/job_102381/pfd_isdid1wa/13 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,109 - CMD: fastq-dump -N 31662331 -X 33923925 -O /local/scratch/job_102381/pfd_isdid1wa/14 --gzip ERR4231014\n",
      "2024-10-31 16:01:27,110 - CMD: fastq-dump -N 33923926 -X 36185523 -O /local/scratch/job_102381/pfd_isdid1wa/15 --gzip ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Read 2261598 spots for ERR4231014\n",
      "Written 2261598 spots for ERR4231014\n",
      "Read 2261595 spots for ERR4231014\n",
      "Written 2261595 spots for ERR4231014\n",
      "Getting ERR4231015 from NCBI SRA\n",
      "2024-10-31 16:05:11,394 - SRR ids: ['ERR4231015']\n",
      "2024-10-31 16:05:11,394 - extra args: ['--gzip']\n",
      "2024-10-31 16:05:11,395 - tempdir: /local/scratch/job_102381/pfd_hnfqes8j\n",
      "2024-10-31 16:05:11,395 - CMD: sra-stat --meta --quick ERR4231015\n",
      "2024-10-31 16:05:13,748 - ERR4231015 spots: 53650335\n",
      "2024-10-31 16:05:13,748 - blocks: [[1, 3353145], [3353146, 6706290], [6706291, 10059435], [10059436, 13412580], [13412581, 16765725], [16765726, 20118870], [20118871, 23472015], [23472016, 26825160], [26825161, 30178305], [30178306, 33531450], [33531451, 36884595], [36884596, 40237740], [40237741, 43590885], [43590886, 46944030], [46944031, 50297175], [50297176, 53650335]]\n",
      "2024-10-31 16:05:13,748 - CMD: fastq-dump -N 1 -X 3353145 -O /local/scratch/job_102381/pfd_hnfqes8j/0 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,749 - CMD: fastq-dump -N 3353146 -X 6706290 -O /local/scratch/job_102381/pfd_hnfqes8j/1 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,750 - CMD: fastq-dump -N 6706291 -X 10059435 -O /local/scratch/job_102381/pfd_hnfqes8j/2 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,750 - CMD: fastq-dump -N 10059436 -X 13412580 -O /local/scratch/job_102381/pfd_hnfqes8j/3 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,750 - CMD: fastq-dump -N 13412581 -X 16765725 -O /local/scratch/job_102381/pfd_hnfqes8j/4 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,751 - CMD: fastq-dump -N 16765726 -X 20118870 -O /local/scratch/job_102381/pfd_hnfqes8j/5 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,751 - CMD: fastq-dump -N 20118871 -X 23472015 -O /local/scratch/job_102381/pfd_hnfqes8j/6 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,751 - CMD: fastq-dump -N 23472016 -X 26825160 -O /local/scratch/job_102381/pfd_hnfqes8j/7 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,752 - CMD: fastq-dump -N 26825161 -X 30178305 -O /local/scratch/job_102381/pfd_hnfqes8j/8 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,752 - CMD: fastq-dump -N 30178306 -X 33531450 -O /local/scratch/job_102381/pfd_hnfqes8j/9 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,753 - CMD: fastq-dump -N 33531451 -X 36884595 -O /local/scratch/job_102381/pfd_hnfqes8j/10 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,753 - CMD: fastq-dump -N 36884596 -X 40237740 -O /local/scratch/job_102381/pfd_hnfqes8j/11 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,754 - CMD: fastq-dump -N 40237741 -X 43590885 -O /local/scratch/job_102381/pfd_hnfqes8j/12 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,755 - CMD: fastq-dump -N 43590886 -X 46944030 -O /local/scratch/job_102381/pfd_hnfqes8j/13 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,755 - CMD: fastq-dump -N 46944031 -X 50297175 -O /local/scratch/job_102381/pfd_hnfqes8j/14 --gzip ERR4231015\n",
      "2024-10-31 16:05:13,756 - CMD: fastq-dump -N 50297176 -X 53650335 -O /local/scratch/job_102381/pfd_hnfqes8j/15 --gzip ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353145 spots for ERR4231015\n",
      "Written 3353145 spots for ERR4231015\n",
      "Read 3353160 spots for ERR4231015\n",
      "Written 3353160 spots for ERR4231015\n",
      "Getting ERR4231016 from NCBI SRA\n",
      "2024-10-31 16:10:01,502 - SRR ids: ['ERR4231016']\n",
      "2024-10-31 16:10:01,503 - extra args: ['--gzip']\n",
      "2024-10-31 16:10:01,503 - tempdir: /local/scratch/job_102381/pfd_ihnuydaw\n",
      "2024-10-31 16:10:01,503 - CMD: sra-stat --meta --quick ERR4231016\n",
      "2024-10-31 16:10:03,945 - ERR4231016 spots: 51945229\n",
      "2024-10-31 16:10:03,946 - blocks: [[1, 3246576], [3246577, 6493152], [6493153, 9739728], [9739729, 12986304], [12986305, 16232880], [16232881, 19479456], [19479457, 22726032], [22726033, 25972608], [25972609, 29219184], [29219185, 32465760], [32465761, 35712336], [35712337, 38958912], [38958913, 42205488], [42205489, 45452064], [45452065, 48698640], [48698641, 51945229]]\n",
      "2024-10-31 16:10:03,946 - CMD: fastq-dump -N 1 -X 3246576 -O /local/scratch/job_102381/pfd_ihnuydaw/0 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,946 - CMD: fastq-dump -N 3246577 -X 6493152 -O /local/scratch/job_102381/pfd_ihnuydaw/1 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,947 - CMD: fastq-dump -N 6493153 -X 9739728 -O /local/scratch/job_102381/pfd_ihnuydaw/2 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,948 - CMD: fastq-dump -N 9739729 -X 12986304 -O /local/scratch/job_102381/pfd_ihnuydaw/3 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,948 - CMD: fastq-dump -N 12986305 -X 16232880 -O /local/scratch/job_102381/pfd_ihnuydaw/4 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,948 - CMD: fastq-dump -N 16232881 -X 19479456 -O /local/scratch/job_102381/pfd_ihnuydaw/5 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,949 - CMD: fastq-dump -N 19479457 -X 22726032 -O /local/scratch/job_102381/pfd_ihnuydaw/6 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,949 - CMD: fastq-dump -N 22726033 -X 25972608 -O /local/scratch/job_102381/pfd_ihnuydaw/7 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,949 - CMD: fastq-dump -N 25972609 -X 29219184 -O /local/scratch/job_102381/pfd_ihnuydaw/8 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,950 - CMD: fastq-dump -N 29219185 -X 32465760 -O /local/scratch/job_102381/pfd_ihnuydaw/9 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,950 - CMD: fastq-dump -N 32465761 -X 35712336 -O /local/scratch/job_102381/pfd_ihnuydaw/10 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,951 - CMD: fastq-dump -N 35712337 -X 38958912 -O /local/scratch/job_102381/pfd_ihnuydaw/11 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,952 - CMD: fastq-dump -N 38958913 -X 42205488 -O /local/scratch/job_102381/pfd_ihnuydaw/12 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,953 - CMD: fastq-dump -N 42205489 -X 45452064 -O /local/scratch/job_102381/pfd_ihnuydaw/13 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,954 - CMD: fastq-dump -N 45452065 -X 48698640 -O /local/scratch/job_102381/pfd_ihnuydaw/14 --gzip ERR4231016\n",
      "2024-10-31 16:10:03,955 - CMD: fastq-dump -N 48698641 -X 51945229 -O /local/scratch/job_102381/pfd_ihnuydaw/15 --gzip ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246589 spots for ERR4231016\n",
      "Written 3246589 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Read 3246576 spots for ERR4231016\n",
      "Written 3246576 spots for ERR4231016\n",
      "Getting ERR4231017 from NCBI SRA\n",
      "2024-10-31 16:15:26,256 - SRR ids: ['ERR4231017']\n",
      "2024-10-31 16:15:26,257 - extra args: ['--gzip']\n",
      "2024-10-31 16:15:26,257 - tempdir: /local/scratch/job_102381/pfd_syth_788\n",
      "2024-10-31 16:15:26,257 - CMD: sra-stat --meta --quick ERR4231017\n",
      "2024-10-31 16:15:28,724 - ERR4231017 spots: 33950855\n",
      "2024-10-31 16:15:28,725 - blocks: [[1, 2121928], [2121929, 4243856], [4243857, 6365784], [6365785, 8487712], [8487713, 10609640], [10609641, 12731568], [12731569, 14853496], [14853497, 16975424], [16975425, 19097352], [19097353, 21219280], [21219281, 23341208], [23341209, 25463136], [25463137, 27585064], [27585065, 29706992], [29706993, 31828920], [31828921, 33950855]]\n",
      "2024-10-31 16:15:28,725 - CMD: fastq-dump -N 1 -X 2121928 -O /local/scratch/job_102381/pfd_syth_788/0 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,726 - CMD: fastq-dump -N 2121929 -X 4243856 -O /local/scratch/job_102381/pfd_syth_788/1 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,727 - CMD: fastq-dump -N 4243857 -X 6365784 -O /local/scratch/job_102381/pfd_syth_788/2 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,727 - CMD: fastq-dump -N 6365785 -X 8487712 -O /local/scratch/job_102381/pfd_syth_788/3 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,727 - CMD: fastq-dump -N 8487713 -X 10609640 -O /local/scratch/job_102381/pfd_syth_788/4 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,727 - CMD: fastq-dump -N 10609641 -X 12731568 -O /local/scratch/job_102381/pfd_syth_788/5 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,728 - CMD: fastq-dump -N 12731569 -X 14853496 -O /local/scratch/job_102381/pfd_syth_788/6 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,728 - CMD: fastq-dump -N 14853497 -X 16975424 -O /local/scratch/job_102381/pfd_syth_788/7 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,728 - CMD: fastq-dump -N 16975425 -X 19097352 -O /local/scratch/job_102381/pfd_syth_788/8 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,729 - CMD: fastq-dump -N 19097353 -X 21219280 -O /local/scratch/job_102381/pfd_syth_788/9 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,729 - CMD: fastq-dump -N 21219281 -X 23341208 -O /local/scratch/job_102381/pfd_syth_788/10 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,729 - CMD: fastq-dump -N 23341209 -X 25463136 -O /local/scratch/job_102381/pfd_syth_788/11 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,730 - CMD: fastq-dump -N 25463137 -X 27585064 -O /local/scratch/job_102381/pfd_syth_788/12 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,730 - CMD: fastq-dump -N 27585065 -X 29706992 -O /local/scratch/job_102381/pfd_syth_788/13 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,730 - CMD: fastq-dump -N 29706993 -X 31828920 -O /local/scratch/job_102381/pfd_syth_788/14 --gzip ERR4231017\n",
      "2024-10-31 16:15:28,731 - CMD: fastq-dump -N 31828921 -X 33950855 -O /local/scratch/job_102381/pfd_syth_788/15 --gzip ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121935 spots for ERR4231017\n",
      "Written 2121935 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Read 2121928 spots for ERR4231017\n",
      "Written 2121928 spots for ERR4231017\n",
      "Getting ERR4231018 from NCBI SRA\n",
      "2024-10-31 16:19:04,275 - SRR ids: ['ERR4231018']\n",
      "2024-10-31 16:19:04,275 - extra args: ['--gzip']\n",
      "2024-10-31 16:19:04,276 - tempdir: /local/scratch/job_102381/pfd_kf3stso3\n",
      "2024-10-31 16:19:04,276 - CMD: sra-stat --meta --quick ERR4231018\n",
      "2024-10-31 16:19:06,693 - ERR4231018 spots: 69462070\n",
      "2024-10-31 16:19:06,693 - blocks: [[1, 4341379], [4341380, 8682758], [8682759, 13024137], [13024138, 17365516], [17365517, 21706895], [21706896, 26048274], [26048275, 30389653], [30389654, 34731032], [34731033, 39072411], [39072412, 43413790], [43413791, 47755169], [47755170, 52096548], [52096549, 56437927], [56437928, 60779306], [60779307, 65120685], [65120686, 69462070]]\n",
      "2024-10-31 16:19:06,693 - CMD: fastq-dump -N 1 -X 4341379 -O /local/scratch/job_102381/pfd_kf3stso3/0 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,694 - CMD: fastq-dump -N 4341380 -X 8682758 -O /local/scratch/job_102381/pfd_kf3stso3/1 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,694 - CMD: fastq-dump -N 8682759 -X 13024137 -O /local/scratch/job_102381/pfd_kf3stso3/2 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,695 - CMD: fastq-dump -N 13024138 -X 17365516 -O /local/scratch/job_102381/pfd_kf3stso3/3 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,695 - CMD: fastq-dump -N 17365517 -X 21706895 -O /local/scratch/job_102381/pfd_kf3stso3/4 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,695 - CMD: fastq-dump -N 21706896 -X 26048274 -O /local/scratch/job_102381/pfd_kf3stso3/5 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,695 - CMD: fastq-dump -N 26048275 -X 30389653 -O /local/scratch/job_102381/pfd_kf3stso3/6 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,696 - CMD: fastq-dump -N 30389654 -X 34731032 -O /local/scratch/job_102381/pfd_kf3stso3/7 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,696 - CMD: fastq-dump -N 34731033 -X 39072411 -O /local/scratch/job_102381/pfd_kf3stso3/8 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,697 - CMD: fastq-dump -N 39072412 -X 43413790 -O /local/scratch/job_102381/pfd_kf3stso3/9 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,697 - CMD: fastq-dump -N 43413791 -X 47755169 -O /local/scratch/job_102381/pfd_kf3stso3/10 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,698 - CMD: fastq-dump -N 47755170 -X 52096548 -O /local/scratch/job_102381/pfd_kf3stso3/11 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,698 - CMD: fastq-dump -N 52096549 -X 56437927 -O /local/scratch/job_102381/pfd_kf3stso3/12 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,699 - CMD: fastq-dump -N 56437928 -X 60779306 -O /local/scratch/job_102381/pfd_kf3stso3/13 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,700 - CMD: fastq-dump -N 60779307 -X 65120685 -O /local/scratch/job_102381/pfd_kf3stso3/14 --gzip ERR4231018\n",
      "2024-10-31 16:19:06,700 - CMD: fastq-dump -N 65120686 -X 69462070 -O /local/scratch/job_102381/pfd_kf3stso3/15 --gzip ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341385 spots for ERR4231018\n",
      "Written 4341385 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Read 4341379 spots for ERR4231018\n",
      "Written 4341379 spots for ERR4231018\n",
      "Getting ERR4231019 from NCBI SRA\n",
      "2024-10-31 16:26:13,409 - SRR ids: ['ERR4231019']\n",
      "2024-10-31 16:26:13,409 - extra args: ['--gzip']\n",
      "2024-10-31 16:26:13,409 - tempdir: /local/scratch/job_102381/pfd_p_rf_lfv\n",
      "2024-10-31 16:26:13,409 - CMD: sra-stat --meta --quick ERR4231019\n",
      "2024-10-31 16:26:15,781 - ERR4231019 spots: 60339006\n",
      "2024-10-31 16:26:15,781 - blocks: [[1, 3771187], [3771188, 7542374], [7542375, 11313561], [11313562, 15084748], [15084749, 18855935], [18855936, 22627122], [22627123, 26398309], [26398310, 30169496], [30169497, 33940683], [33940684, 37711870], [37711871, 41483057], [41483058, 45254244], [45254245, 49025431], [49025432, 52796618], [52796619, 56567805], [56567806, 60339006]]\n",
      "2024-10-31 16:26:15,781 - CMD: fastq-dump -N 1 -X 3771187 -O /local/scratch/job_102381/pfd_p_rf_lfv/0 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,782 - CMD: fastq-dump -N 3771188 -X 7542374 -O /local/scratch/job_102381/pfd_p_rf_lfv/1 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,783 - CMD: fastq-dump -N 7542375 -X 11313561 -O /local/scratch/job_102381/pfd_p_rf_lfv/2 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,783 - CMD: fastq-dump -N 11313562 -X 15084748 -O /local/scratch/job_102381/pfd_p_rf_lfv/3 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,783 - CMD: fastq-dump -N 15084749 -X 18855935 -O /local/scratch/job_102381/pfd_p_rf_lfv/4 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,784 - CMD: fastq-dump -N 18855936 -X 22627122 -O /local/scratch/job_102381/pfd_p_rf_lfv/5 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,784 - CMD: fastq-dump -N 22627123 -X 26398309 -O /local/scratch/job_102381/pfd_p_rf_lfv/6 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,784 - CMD: fastq-dump -N 26398310 -X 30169496 -O /local/scratch/job_102381/pfd_p_rf_lfv/7 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,785 - CMD: fastq-dump -N 30169497 -X 33940683 -O /local/scratch/job_102381/pfd_p_rf_lfv/8 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,785 - CMD: fastq-dump -N 33940684 -X 37711870 -O /local/scratch/job_102381/pfd_p_rf_lfv/9 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,786 - CMD: fastq-dump -N 37711871 -X 41483057 -O /local/scratch/job_102381/pfd_p_rf_lfv/10 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,786 - CMD: fastq-dump -N 41483058 -X 45254244 -O /local/scratch/job_102381/pfd_p_rf_lfv/11 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,787 - CMD: fastq-dump -N 45254245 -X 49025431 -O /local/scratch/job_102381/pfd_p_rf_lfv/12 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,788 - CMD: fastq-dump -N 49025432 -X 52796618 -O /local/scratch/job_102381/pfd_p_rf_lfv/13 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,792 - CMD: fastq-dump -N 52796619 -X 56567805 -O /local/scratch/job_102381/pfd_p_rf_lfv/14 --gzip ERR4231019\n",
      "2024-10-31 16:26:15,793 - CMD: fastq-dump -N 56567806 -X 60339006 -O /local/scratch/job_102381/pfd_p_rf_lfv/15 --gzip ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771201 spots for ERR4231019\n",
      "Written 3771201 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Read 3771187 spots for ERR4231019\n",
      "Written 3771187 spots for ERR4231019\n",
      "Getting ERR4231020 from NCBI SRA\n",
      "2024-10-31 16:32:28,797 - SRR ids: ['ERR4231020']\n",
      "2024-10-31 16:32:28,797 - extra args: ['--gzip']\n",
      "2024-10-31 16:32:28,797 - tempdir: /local/scratch/job_102381/pfd_a3lm6c33\n",
      "2024-10-31 16:32:28,797 - CMD: sra-stat --meta --quick ERR4231020\n",
      "2024-10-31 16:32:31,570 - ERR4231020 spots: 74203167\n",
      "2024-10-31 16:32:31,570 - blocks: [[1, 4637697], [4637698, 9275394], [9275395, 13913091], [13913092, 18550788], [18550789, 23188485], [23188486, 27826182], [27826183, 32463879], [32463880, 37101576], [37101577, 41739273], [41739274, 46376970], [46376971, 51014667], [51014668, 55652364], [55652365, 60290061], [60290062, 64927758], [64927759, 69565455], [69565456, 74203167]]\n",
      "2024-10-31 16:32:31,570 - CMD: fastq-dump -N 1 -X 4637697 -O /local/scratch/job_102381/pfd_a3lm6c33/0 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,571 - CMD: fastq-dump -N 4637698 -X 9275394 -O /local/scratch/job_102381/pfd_a3lm6c33/1 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,571 - CMD: fastq-dump -N 9275395 -X 13913091 -O /local/scratch/job_102381/pfd_a3lm6c33/2 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,572 - CMD: fastq-dump -N 13913092 -X 18550788 -O /local/scratch/job_102381/pfd_a3lm6c33/3 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,572 - CMD: fastq-dump -N 18550789 -X 23188485 -O /local/scratch/job_102381/pfd_a3lm6c33/4 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,572 - CMD: fastq-dump -N 23188486 -X 27826182 -O /local/scratch/job_102381/pfd_a3lm6c33/5 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,573 - CMD: fastq-dump -N 27826183 -X 32463879 -O /local/scratch/job_102381/pfd_a3lm6c33/6 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,573 - CMD: fastq-dump -N 32463880 -X 37101576 -O /local/scratch/job_102381/pfd_a3lm6c33/7 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,573 - CMD: fastq-dump -N 37101577 -X 41739273 -O /local/scratch/job_102381/pfd_a3lm6c33/8 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,574 - CMD: fastq-dump -N 41739274 -X 46376970 -O /local/scratch/job_102381/pfd_a3lm6c33/9 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,574 - CMD: fastq-dump -N 46376971 -X 51014667 -O /local/scratch/job_102381/pfd_a3lm6c33/10 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,574 - CMD: fastq-dump -N 51014668 -X 55652364 -O /local/scratch/job_102381/pfd_a3lm6c33/11 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,575 - CMD: fastq-dump -N 55652365 -X 60290061 -O /local/scratch/job_102381/pfd_a3lm6c33/12 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,576 - CMD: fastq-dump -N 60290062 -X 64927758 -O /local/scratch/job_102381/pfd_a3lm6c33/13 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,577 - CMD: fastq-dump -N 64927759 -X 69565455 -O /local/scratch/job_102381/pfd_a3lm6c33/14 --gzip ERR4231020\n",
      "2024-10-31 16:32:31,577 - CMD: fastq-dump -N 69565456 -X 74203167 -O /local/scratch/job_102381/pfd_a3lm6c33/15 --gzip ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637712 spots for ERR4231020\n",
      "Written 4637712 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Read 4637697 spots for ERR4231020\n",
      "Written 4637697 spots for ERR4231020\n",
      "Getting ERR4231021 from NCBI SRA\n",
      "2024-10-31 16:39:43,128 - SRR ids: ['ERR4231021']\n",
      "2024-10-31 16:39:43,128 - extra args: ['--gzip']\n",
      "2024-10-31 16:39:43,128 - tempdir: /local/scratch/job_102381/pfd_w8v7ehsq\n",
      "2024-10-31 16:39:43,128 - CMD: sra-stat --meta --quick ERR4231021\n",
      "2024-10-31 16:39:45,616 - ERR4231021 spots: 60201088\n",
      "2024-10-31 16:39:45,616 - blocks: [[1, 3762568], [3762569, 7525136], [7525137, 11287704], [11287705, 15050272], [15050273, 18812840], [18812841, 22575408], [22575409, 26337976], [26337977, 30100544], [30100545, 33863112], [33863113, 37625680], [37625681, 41388248], [41388249, 45150816], [45150817, 48913384], [48913385, 52675952], [52675953, 56438520], [56438521, 60201088]]\n",
      "2024-10-31 16:39:45,617 - CMD: fastq-dump -N 1 -X 3762568 -O /local/scratch/job_102381/pfd_w8v7ehsq/0 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,626 - CMD: fastq-dump -N 3762569 -X 7525136 -O /local/scratch/job_102381/pfd_w8v7ehsq/1 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,627 - CMD: fastq-dump -N 7525137 -X 11287704 -O /local/scratch/job_102381/pfd_w8v7ehsq/2 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,627 - CMD: fastq-dump -N 11287705 -X 15050272 -O /local/scratch/job_102381/pfd_w8v7ehsq/3 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,627 - CMD: fastq-dump -N 15050273 -X 18812840 -O /local/scratch/job_102381/pfd_w8v7ehsq/4 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,628 - CMD: fastq-dump -N 18812841 -X 22575408 -O /local/scratch/job_102381/pfd_w8v7ehsq/5 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,628 - CMD: fastq-dump -N 22575409 -X 26337976 -O /local/scratch/job_102381/pfd_w8v7ehsq/6 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,628 - CMD: fastq-dump -N 26337977 -X 30100544 -O /local/scratch/job_102381/pfd_w8v7ehsq/7 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,629 - CMD: fastq-dump -N 30100545 -X 33863112 -O /local/scratch/job_102381/pfd_w8v7ehsq/8 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,629 - CMD: fastq-dump -N 33863113 -X 37625680 -O /local/scratch/job_102381/pfd_w8v7ehsq/9 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,630 - CMD: fastq-dump -N 37625681 -X 41388248 -O /local/scratch/job_102381/pfd_w8v7ehsq/10 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,631 - CMD: fastq-dump -N 41388249 -X 45150816 -O /local/scratch/job_102381/pfd_w8v7ehsq/11 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,632 - CMD: fastq-dump -N 45150817 -X 48913384 -O /local/scratch/job_102381/pfd_w8v7ehsq/12 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,633 - CMD: fastq-dump -N 48913385 -X 52675952 -O /local/scratch/job_102381/pfd_w8v7ehsq/13 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,645 - CMD: fastq-dump -N 52675953 -X 56438520 -O /local/scratch/job_102381/pfd_w8v7ehsq/14 --gzip ERR4231021\n",
      "2024-10-31 16:39:45,651 - CMD: fastq-dump -N 56438521 -X 60201088 -O /local/scratch/job_102381/pfd_w8v7ehsq/15 --gzip ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Read 3762568 spots for ERR4231021\n",
      "Written 3762568 spots for ERR4231021\n",
      "Getting ERR4231023 from NCBI SRA\n",
      "2024-10-31 16:45:42,735 - SRR ids: ['ERR4231023']\n",
      "2024-10-31 16:45:42,736 - extra args: ['--gzip']\n",
      "2024-10-31 16:45:42,736 - tempdir: /local/scratch/job_102381/pfd_ytf6woe6\n",
      "2024-10-31 16:45:42,736 - CMD: sra-stat --meta --quick ERR4231023\n",
      "2024-10-31 16:45:45,333 - ERR4231023 spots: 45049673\n",
      "2024-10-31 16:45:45,333 - blocks: [[1, 2815604], [2815605, 5631208], [5631209, 8446812], [8446813, 11262416], [11262417, 14078020], [14078021, 16893624], [16893625, 19709228], [19709229, 22524832], [22524833, 25340436], [25340437, 28156040], [28156041, 30971644], [30971645, 33787248], [33787249, 36602852], [36602853, 39418456], [39418457, 42234060], [42234061, 45049673]]\n",
      "2024-10-31 16:45:45,333 - CMD: fastq-dump -N 1 -X 2815604 -O /local/scratch/job_102381/pfd_ytf6woe6/0 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,334 - CMD: fastq-dump -N 2815605 -X 5631208 -O /local/scratch/job_102381/pfd_ytf6woe6/1 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,335 - CMD: fastq-dump -N 5631209 -X 8446812 -O /local/scratch/job_102381/pfd_ytf6woe6/2 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,335 - CMD: fastq-dump -N 8446813 -X 11262416 -O /local/scratch/job_102381/pfd_ytf6woe6/3 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,335 - CMD: fastq-dump -N 11262417 -X 14078020 -O /local/scratch/job_102381/pfd_ytf6woe6/4 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,336 - CMD: fastq-dump -N 14078021 -X 16893624 -O /local/scratch/job_102381/pfd_ytf6woe6/5 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,336 - CMD: fastq-dump -N 16893625 -X 19709228 -O /local/scratch/job_102381/pfd_ytf6woe6/6 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,336 - CMD: fastq-dump -N 19709229 -X 22524832 -O /local/scratch/job_102381/pfd_ytf6woe6/7 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,337 - CMD: fastq-dump -N 22524833 -X 25340436 -O /local/scratch/job_102381/pfd_ytf6woe6/8 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,337 - CMD: fastq-dump -N 25340437 -X 28156040 -O /local/scratch/job_102381/pfd_ytf6woe6/9 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,337 - CMD: fastq-dump -N 28156041 -X 30971644 -O /local/scratch/job_102381/pfd_ytf6woe6/10 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,338 - CMD: fastq-dump -N 30971645 -X 33787248 -O /local/scratch/job_102381/pfd_ytf6woe6/11 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,338 - CMD: fastq-dump -N 33787249 -X 36602852 -O /local/scratch/job_102381/pfd_ytf6woe6/12 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,338 - CMD: fastq-dump -N 36602853 -X 39418456 -O /local/scratch/job_102381/pfd_ytf6woe6/13 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,339 - CMD: fastq-dump -N 39418457 -X 42234060 -O /local/scratch/job_102381/pfd_ytf6woe6/14 --gzip ERR4231023\n",
      "2024-10-31 16:45:45,339 - CMD: fastq-dump -N 42234061 -X 45049673 -O /local/scratch/job_102381/pfd_ytf6woe6/15 --gzip ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815613 spots for ERR4231023\n",
      "Written 2815613 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Read 2815604 spots for ERR4231023\n",
      "Written 2815604 spots for ERR4231023\n",
      "Getting ERR4231011 from NCBI SRA\n",
      "2024-10-31 16:50:12,577 - SRR ids: ['ERR4231011']\n",
      "2024-10-31 16:50:12,577 - extra args: ['--gzip']\n",
      "2024-10-31 16:50:12,577 - tempdir: /local/scratch/job_102381/pfd_0axmhtnz\n",
      "2024-10-31 16:50:12,577 - CMD: sra-stat --meta --quick ERR4231011\n",
      "2024-10-31 16:50:14,932 - ERR4231011 spots: 44940444\n",
      "2024-10-31 16:50:14,932 - blocks: [[1, 2808777], [2808778, 5617554], [5617555, 8426331], [8426332, 11235108], [11235109, 14043885], [14043886, 16852662], [16852663, 19661439], [19661440, 22470216], [22470217, 25278993], [25278994, 28087770], [28087771, 30896547], [30896548, 33705324], [33705325, 36514101], [36514102, 39322878], [39322879, 42131655], [42131656, 44940444]]\n",
      "2024-10-31 16:50:14,932 - CMD: fastq-dump -N 1 -X 2808777 -O /local/scratch/job_102381/pfd_0axmhtnz/0 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,933 - CMD: fastq-dump -N 2808778 -X 5617554 -O /local/scratch/job_102381/pfd_0axmhtnz/1 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,934 - CMD: fastq-dump -N 5617555 -X 8426331 -O /local/scratch/job_102381/pfd_0axmhtnz/2 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,934 - CMD: fastq-dump -N 8426332 -X 11235108 -O /local/scratch/job_102381/pfd_0axmhtnz/3 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,934 - CMD: fastq-dump -N 11235109 -X 14043885 -O /local/scratch/job_102381/pfd_0axmhtnz/4 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,935 - CMD: fastq-dump -N 14043886 -X 16852662 -O /local/scratch/job_102381/pfd_0axmhtnz/5 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,935 - CMD: fastq-dump -N 16852663 -X 19661439 -O /local/scratch/job_102381/pfd_0axmhtnz/6 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,935 - CMD: fastq-dump -N 19661440 -X 22470216 -O /local/scratch/job_102381/pfd_0axmhtnz/7 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,936 - CMD: fastq-dump -N 22470217 -X 25278993 -O /local/scratch/job_102381/pfd_0axmhtnz/8 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,936 - CMD: fastq-dump -N 25278994 -X 28087770 -O /local/scratch/job_102381/pfd_0axmhtnz/9 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,936 - CMD: fastq-dump -N 28087771 -X 30896547 -O /local/scratch/job_102381/pfd_0axmhtnz/10 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,936 - CMD: fastq-dump -N 30896548 -X 33705324 -O /local/scratch/job_102381/pfd_0axmhtnz/11 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,937 - CMD: fastq-dump -N 33705325 -X 36514101 -O /local/scratch/job_102381/pfd_0axmhtnz/12 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,937 - CMD: fastq-dump -N 36514102 -X 39322878 -O /local/scratch/job_102381/pfd_0axmhtnz/13 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,937 - CMD: fastq-dump -N 39322879 -X 42131655 -O /local/scratch/job_102381/pfd_0axmhtnz/14 --gzip ERR4231011\n",
      "2024-10-31 16:50:14,938 - CMD: fastq-dump -N 42131656 -X 44940444 -O /local/scratch/job_102381/pfd_0axmhtnz/15 --gzip ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808789 spots for ERR4231011\n",
      "Written 2808789 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n",
      "Read 2808777 spots for ERR4231011\n",
      "Written 2808777 spots for ERR4231011\n"
     ]
    }
   ],
   "source": [
    "# !mkdir fastq_placenta\n",
    "# !cut -d ',' -f 1 SraRunTable_Placenta.txt | tail -n +2 > sra_list_placenta.txt\n",
    "!while IFS= read -r line; do \\\n",
    "    echo \"Getting $line from NCBI SRA\"; \\\n",
    "    parallel-fastq-dump --sra-id $line --threads 16 --outdir fastq_placenta --gzip; \\\n",
    "    done < sra_list_placenta.txt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Edit this block with a short evaluation of each of the report sections. Each evaluation should include your important observations, reasons for potential QC failure, and an opinion on what should be done on the failure (i.e., trimming or filtering). Problem FASTQ files, if any, should be noted.\n",
    "\n",
    "*Per base sequence quality:*\n",
    "\n",
    "    Samples ERR42310[19-21] have a low sequence quality after at base 172 with 177 being where the low points start. This could be due to a poly-A tail.\n",
    "    Samples ERR42310[17&18] drop of in quality at base 187.\n",
    "    The other samples seem to have decent per base sequence qualities.\n",
    "*Per tile sequence quality:*\n",
    "\n",
    "    no report on my multiqc file\n",
    "*Per sequence quality scores:*\n",
    "\n",
    "    The Phred scores are within high quality ranges for every sample. \n",
    "*Per base sequence content:*\n",
    "\n",
    "    Before bp 10 there is a low sequence content score, this is likely due to an adapter or identifier. \n",
    "*Per sequence GC content:*\n",
    "\n",
    "    Samples ERR42310[16-18,20,22-23] have warnings for their GC content with some sequences having 61% gc content.\n",
    "*Per base N content:*\n",
    "\n",
    "    Samples ERR42310[19-21] have a high percentage of per base N content starting at bp 177 which is similar to their quality drop off points, likely due to the same thing that lowered their quality score.   \n",
    "*Sequence length distribution:*\n",
    "\n",
    "    All sequences had a length of 202 bp\n",
    "*Sequence duplication levels:*\n",
    "\n",
    "    Samples ERR42310[11&14] have warnings for their duplication levels while the rest of the samples failed their duplication levels. They all had over 10k sequences that were repeated or duplicated\n",
    "*Overrepresented sequences:*\n",
    "\n",
    "    The top overrepresented sequences were:\n",
    "    CTAAAATCCCTCTTCGGATCCACAGTCAACCGCCCTGAACACATCCTGCA\t4\t246420\t0.0314%\n",
    "    CGAAGCTAAAATCCCTCTTCGGATCCACAGTCAACCGCCCTGAACACATC\t3\t169383\t0.0216%\n",
    "    AGATCGGAAGAGCACACGTCTGAACTCCAGTCACCAGATCATCTCGTATG\t3\t578466\t0.0736%\n",
    "    AGATCGGAAGAGCACACGTCTGAACTCCAGTCACCGATGTATCTCGTATG\t3\t228204\t0.0290%\n",
    "    GATCGGAAGAGCACACGTCTGAACTCCAGTCACCAGATCATCTCGTATGC\t2\t200072\t0.0255%\n",
    "    GGAAAACCCATTCTTCTCCCAGCCGGGTGCCCCAATACTTCAGTGCATGG\t1\t37188\t0.0047%\n",
    "    Most of these are known illumina sequence adapters and the rest are human sequences.\n",
    "\n",
    "*Adapter content:*\n",
    "\n",
    "    Most of the samples have illumina sequence adapters or poly-A tails after bp 92 and some even at bp 67"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If necessary, use `cutadapt` to trim and filter your reads. Provide a justification for the `cutadapt` parameters you used:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.001s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /dev/cpuset.\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of ERR4231010.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of ERR4231011.fastq.gz\n",
      "Started analysis of ERR4231012.fastq.gz\n",
      "Started analysis of ERR4231013.fastq.gz\n",
      "Started analysis of ERR4231014.fastq.gz\n",
      "Started analysis of ERR4231015.fastq.gz\n",
      "Started analysis of ERR4231016.fastq.gz\n",
      "Started analysis of ERR4231017.fastq.gz\n",
      "Started analysis of ERR4231018.fastq.gz\n",
      "Started analysis of ERR4231019.fastq.gz\n",
      "Started analysis of ERR4231020.fastq.gz\n",
      "Started analysis of ERR4231021.fastq.gz\n",
      "Started analysis of ERR4231022.fastq.gz\n",
      "Started analysis of ERR4231023.fastq.gz\n",
      "Approx 5% complete for ERR4231017.fastq.gz\n",
      "Approx 5% complete for ERR4231014.fastq.gz\n",
      "Approx 5% complete for ERR4231011.fastq.gz\n",
      "Approx 5% complete for ERR4231023.fastq.gz\n",
      "Approx 5% complete for ERR4231016.fastq.gz\n",
      "Approx 5% complete for ERR4231015.fastq.gz\n",
      "Approx 5% complete for ERR4231010.fastq.gz\n",
      "Approx 5% complete for ERR4231022.fastq.gz\n",
      "Approx 5% complete for ERR4231013.fastq.gz\n",
      "Approx 5% complete for ERR4231019.fastq.gz\n",
      "Approx 5% complete for ERR4231021.fastq.gz\n",
      "Approx 10% complete for ERR4231017.fastq.gz\n",
      "Approx 5% complete for ERR4231012.fastq.gz\n",
      "Approx 5% complete for ERR4231018.fastq.gz\n",
      "Approx 10% complete for ERR4231014.fastq.gz\n",
      "Approx 5% complete for ERR4231020.fastq.gz\n",
      "Approx 10% complete for ERR4231011.fastq.gz\n",
      "Approx 10% complete for ERR4231023.fastq.gz\n",
      "Approx 15% complete for ERR4231017.fastq.gz\n",
      "Approx 10% complete for ERR4231016.fastq.gz\n",
      "Approx 10% complete for ERR4231015.fastq.gz\n",
      "Approx 15% complete for ERR4231014.fastq.gz\n",
      "Approx 10% complete for ERR4231022.fastq.gz\n",
      "Approx 10% complete for ERR4231010.fastq.gz\n",
      "Approx 10% complete for ERR4231013.fastq.gz\n",
      "Approx 10% complete for ERR4231019.fastq.gz\n",
      "Approx 15% complete for ERR4231011.fastq.gz\n",
      "Approx 10% complete for ERR4231021.fastq.gz\n",
      "Approx 20% complete for ERR4231017.fastq.gz\n",
      "Approx 10% complete for ERR4231018.fastq.gz\n",
      "Approx 15% complete for ERR4231023.fastq.gz\n",
      "Approx 10% complete for ERR4231012.fastq.gz\n",
      "Approx 20% complete for ERR4231014.fastq.gz\n",
      "Approx 10% complete for ERR4231020.fastq.gz\n",
      "Approx 15% complete for ERR4231016.fastq.gz\n",
      "Approx 15% complete for ERR4231015.fastq.gz\n",
      "Approx 15% complete for ERR4231022.fastq.gz\n",
      "Approx 20% complete for ERR4231011.fastq.gz\n",
      "Approx 25% complete for ERR4231017.fastq.gz\n",
      "Approx 15% complete for ERR4231013.fastq.gz\n",
      "Approx 15% complete for ERR4231010.fastq.gz\n",
      "Approx 15% complete for ERR4231019.fastq.gz\n",
      "Approx 25% complete for ERR4231014.fastq.gz\n",
      "Approx 15% complete for ERR4231021.fastq.gz\n",
      "Approx 20% complete for ERR4231023.fastq.gz\n",
      "Approx 30% complete for ERR4231017.fastq.gz\n",
      "Approx 15% complete for ERR4231018.fastq.gz\n",
      "Approx 20% complete for ERR4231016.fastq.gz\n",
      "Approx 25% complete for ERR4231011.fastq.gz\n",
      "Approx 20% complete for ERR4231015.fastq.gz\n",
      "Approx 15% complete for ERR4231012.fastq.gz\n",
      "Approx 30% complete for ERR4231014.fastq.gz\n",
      "Approx 20% complete for ERR4231022.fastq.gz\n",
      "Approx 15% complete for ERR4231020.fastq.gz\n",
      "Approx 25% complete for ERR4231023.fastq.gz\n",
      "Approx 35% complete for ERR4231017.fastq.gz\n",
      "Approx 20% complete for ERR4231019.fastq.gz\n",
      "Approx 20% complete for ERR4231013.fastq.gz\n",
      "Approx 20% complete for ERR4231010.fastq.gz\n",
      "Approx 20% complete for ERR4231021.fastq.gz\n",
      "Approx 35% complete for ERR4231014.fastq.gz\n",
      "Approx 30% complete for ERR4231011.fastq.gz\n",
      "Approx 25% complete for ERR4231016.fastq.gz\n",
      "Approx 25% complete for ERR4231015.fastq.gz\n",
      "Approx 40% complete for ERR4231017.fastq.gz\n",
      "Approx 25% complete for ERR4231022.fastq.gz\n",
      "Approx 30% complete for ERR4231023.fastq.gz\n",
      "Approx 20% complete for ERR4231018.fastq.gz\n",
      "Approx 40% complete for ERR4231014.fastq.gz\n",
      "Approx 20% complete for ERR4231012.fastq.gz\n",
      "Approx 20% complete for ERR4231020.fastq.gz\n",
      "Approx 35% complete for ERR4231011.fastq.gz\n",
      "Approx 25% complete for ERR4231019.fastq.gz\n",
      "Approx 25% complete for ERR4231021.fastq.gz\n",
      "Approx 25% complete for ERR4231013.fastq.gz\n",
      "Approx 45% complete for ERR4231017.fastq.gz\n",
      "Approx 25% complete for ERR4231010.fastq.gz\n",
      "Approx 30% complete for ERR4231016.fastq.gz\n",
      "Approx 35% complete for ERR4231023.fastq.gz\n",
      "Approx 45% complete for ERR4231014.fastq.gz\n",
      "Approx 30% complete for ERR4231015.fastq.gz\n",
      "Approx 30% complete for ERR4231022.fastq.gz\n",
      "Approx 50% complete for ERR4231017.fastq.gz\n",
      "Approx 40% complete for ERR4231011.fastq.gz\n",
      "Approx 25% complete for ERR4231018.fastq.gz\n",
      "Approx 50% complete for ERR4231014.fastq.gz\n",
      "Approx 30% complete for ERR4231019.fastq.gz\n",
      "Approx 25% complete for ERR4231012.fastq.gz\n",
      "Approx 30% complete for ERR4231021.fastq.gz\n",
      "Approx 30% complete for ERR4231013.fastq.gz\n",
      "Approx 35% complete for ERR4231016.fastq.gz\n",
      "Approx 25% complete for ERR4231020.fastq.gz\n",
      "Approx 40% complete for ERR4231023.fastq.gz\n",
      "Approx 30% complete for ERR4231010.fastq.gz\n",
      "Approx 55% complete for ERR4231017.fastq.gz\n",
      "Approx 35% complete for ERR4231015.fastq.gz\n",
      "Approx 35% complete for ERR4231022.fastq.gz\n",
      "Approx 45% complete for ERR4231011.fastq.gz\n",
      "Approx 55% complete for ERR4231014.fastq.gz\n",
      "Approx 60% complete for ERR4231017.fastq.gz\n",
      "Approx 45% complete for ERR4231023.fastq.gz\n",
      "Approx 40% complete for ERR4231016.fastq.gz\n",
      "Approx 30% complete for ERR4231018.fastq.gz\n",
      "Approx 35% complete for ERR4231019.fastq.gz\n",
      "Approx 35% complete for ERR4231021.fastq.gz\n",
      "Approx 35% complete for ERR4231013.fastq.gz\n",
      "Approx 50% complete for ERR4231011.fastq.gz\n",
      "Approx 60% complete for ERR4231014.fastq.gz\n",
      "Approx 35% complete for ERR4231010.fastq.gz\n",
      "Approx 30% complete for ERR4231012.fastq.gz\n",
      "Approx 40% complete for ERR4231015.fastq.gz\n",
      "Approx 40% complete for ERR4231022.fastq.gz\n",
      "Approx 30% complete for ERR4231020.fastq.gz\n",
      "Approx 65% complete for ERR4231017.fastq.gz\n",
      "Approx 50% complete for ERR4231023.fastq.gz\n",
      "Approx 65% complete for ERR4231014.fastq.gz\n",
      "Approx 45% complete for ERR4231016.fastq.gz\n",
      "Approx 55% complete for ERR4231011.fastq.gz\n",
      "Approx 70% complete for ERR4231017.fastq.gz\n",
      "Approx 40% complete for ERR4231019.fastq.gz\n",
      "Approx 40% complete for ERR4231021.fastq.gz\n",
      "Approx 40% complete for ERR4231013.fastq.gz\n",
      "Approx 35% complete for ERR4231018.fastq.gz\n",
      "Approx 45% complete for ERR4231022.fastq.gz\n",
      "Approx 45% complete for ERR4231015.fastq.gz\n",
      "Approx 40% complete for ERR4231010.fastq.gz\n",
      "Approx 70% complete for ERR4231014.fastq.gz\n",
      "Approx 55% complete for ERR4231023.fastq.gz\n",
      "Approx 35% complete for ERR4231012.fastq.gz\n",
      "Approx 35% complete for ERR4231020.fastq.gz\n",
      "Approx 60% complete for ERR4231011.fastq.gz\n",
      "Approx 75% complete for ERR4231017.fastq.gz\n",
      "Approx 50% complete for ERR4231016.fastq.gz\n",
      "Approx 75% complete for ERR4231014.fastq.gz\n",
      "Approx 45% complete for ERR4231019.fastq.gz\n",
      "Approx 45% complete for ERR4231021.fastq.gz\n",
      "Approx 50% complete for ERR4231022.fastq.gz\n",
      "Approx 50% complete for ERR4231015.fastq.gz\n",
      "Approx 60% complete for ERR4231023.fastq.gz\n",
      "Approx 45% complete for ERR4231013.fastq.gz\n",
      "Approx 80% complete for ERR4231017.fastq.gz\n",
      "Approx 45% complete for ERR4231010.fastq.gz\n",
      "Approx 40% complete for ERR4231018.fastq.gz\n",
      "Approx 65% complete for ERR4231011.fastq.gz\n",
      "Approx 80% complete for ERR4231014.fastq.gz\n",
      "Approx 55% complete for ERR4231016.fastq.gz\n",
      "Approx 40% complete for ERR4231020.fastq.gz\n",
      "Approx 40% complete for ERR4231012.fastq.gz\n",
      "Approx 85% complete for ERR4231017.fastq.gz\n",
      "Approx 65% complete for ERR4231023.fastq.gz\n",
      "Approx 50% complete for ERR4231019.fastq.gz\n",
      "Approx 55% complete for ERR4231022.fastq.gz\n",
      "Approx 50% complete for ERR4231021.fastq.gz\n",
      "Approx 55% complete for ERR4231015.fastq.gz\n",
      "Approx 70% complete for ERR4231011.fastq.gz\n",
      "Approx 50% complete for ERR4231013.fastq.gz\n",
      "Approx 85% complete for ERR4231014.fastq.gz\n",
      "Approx 50% complete for ERR4231010.fastq.gz\n",
      "Approx 90% complete for ERR4231017.fastq.gz\n",
      "Approx 45% complete for ERR4231018.fastq.gz\n",
      "Approx 60% complete for ERR4231016.fastq.gz\n",
      "Approx 70% complete for ERR4231023.fastq.gz\n",
      "Approx 90% complete for ERR4231014.fastq.gz\n",
      "Approx 75% complete for ERR4231011.fastq.gz\n",
      "Approx 45% complete for ERR4231020.fastq.gz\n",
      "Approx 45% complete for ERR4231012.fastq.gz\n",
      "Approx 60% complete for ERR4231022.fastq.gz\n",
      "Approx 60% complete for ERR4231015.fastq.gz\n",
      "Approx 95% complete for ERR4231017.fastq.gz\n",
      "Approx 55% complete for ERR4231019.fastq.gz\n",
      "Approx 55% complete for ERR4231021.fastq.gz\n",
      "Approx 55% complete for ERR4231013.fastq.gz\n",
      "Approx 95% complete for ERR4231014.fastq.gz\n",
      "Approx 65% complete for ERR4231016.fastq.gz\n",
      "Approx 55% complete for ERR4231010.fastq.gz\n",
      "Approx 75% complete for ERR4231023.fastq.gz\n",
      "Approx 80% complete for ERR4231011.fastq.gz\n",
      "Analysis complete for ERR4231017.fastq.gz\n",
      "Approx 50% complete for ERR4231018.fastq.gz\n",
      "Approx 65% complete for ERR4231022.fastq.gz\n",
      "Approx 65% complete for ERR4231015.fastq.gz\n",
      "Analysis complete for ERR4231014.fastq.gz\n",
      "Approx 50% complete for ERR4231020.fastq.gz\n",
      "Approx 60% complete for ERR4231019.fastq.gz\n",
      "Approx 50% complete for ERR4231012.fastq.gz\n",
      "Approx 60% complete for ERR4231021.fastq.gz\n",
      "Approx 80% complete for ERR4231023.fastq.gz\n",
      "Approx 85% complete for ERR4231011.fastq.gz\n",
      "Approx 60% complete for ERR4231013.fastq.gz\n",
      "Approx 70% complete for ERR4231016.fastq.gz\n",
      "Approx 60% complete for ERR4231010.fastq.gz\n",
      "Approx 70% complete for ERR4231022.fastq.gz\n",
      "Approx 55% complete for ERR4231018.fastq.gz\n",
      "Approx 70% complete for ERR4231015.fastq.gz\n",
      "Approx 85% complete for ERR4231023.fastq.gz\n",
      "Approx 90% complete for ERR4231011.fastq.gz\n",
      "Approx 65% complete for ERR4231019.fastq.gz\n",
      "Approx 65% complete for ERR4231021.fastq.gz\n",
      "Approx 75% complete for ERR4231016.fastq.gz\n",
      "Approx 65% complete for ERR4231013.fastq.gz\n",
      "Approx 55% complete for ERR4231020.fastq.gz\n",
      "Approx 55% complete for ERR4231012.fastq.gz\n",
      "Approx 65% complete for ERR4231010.fastq.gz\n",
      "Approx 75% complete for ERR4231022.fastq.gz\n",
      "Approx 95% complete for ERR4231011.fastq.gz\n",
      "Approx 75% complete for ERR4231015.fastq.gz\n",
      "Approx 90% complete for ERR4231023.fastq.gz\n",
      "Approx 60% complete for ERR4231018.fastq.gz\n",
      "Approx 80% complete for ERR4231016.fastq.gz\n",
      "Approx 70% complete for ERR4231019.fastq.gz\n",
      "Approx 70% complete for ERR4231021.fastq.gz\n",
      "Approx 70% complete for ERR4231013.fastq.gz\n",
      "Analysis complete for ERR4231011.fastq.gz\n",
      "Approx 70% complete for ERR4231010.fastq.gz\n",
      "Approx 95% complete for ERR4231023.fastq.gz\n",
      "Approx 80% complete for ERR4231022.fastq.gz\n",
      "Approx 60% complete for ERR4231012.fastq.gz\n",
      "Approx 60% complete for ERR4231020.fastq.gz\n",
      "Approx 80% complete for ERR4231015.fastq.gz\n",
      "Approx 85% complete for ERR4231016.fastq.gz\n",
      "Approx 65% complete for ERR4231018.fastq.gz\n",
      "Approx 75% complete for ERR4231019.fastq.gz\n",
      "Approx 75% complete for ERR4231021.fastq.gz\n",
      "Analysis complete for ERR4231023.fastq.gz\n",
      "Approx 75% complete for ERR4231013.fastq.gz\n",
      "Approx 85% complete for ERR4231022.fastq.gz\n",
      "Approx 85% complete for ERR4231015.fastq.gz\n",
      "Approx 75% complete for ERR4231010.fastq.gz\n",
      "Approx 65% complete for ERR4231012.fastq.gz\n",
      "Approx 90% complete for ERR4231016.fastq.gz\n",
      "Approx 65% complete for ERR4231020.fastq.gz\n",
      "Approx 80% complete for ERR4231019.fastq.gz\n",
      "Approx 80% complete for ERR4231021.fastq.gz\n",
      "Approx 70% complete for ERR4231018.fastq.gz\n",
      "Approx 80% complete for ERR4231013.fastq.gz\n",
      "Approx 90% complete for ERR4231022.fastq.gz\n",
      "Approx 90% complete for ERR4231015.fastq.gz\n",
      "Approx 80% complete for ERR4231010.fastq.gz\n",
      "Approx 95% complete for ERR4231016.fastq.gz\n",
      "Approx 70% complete for ERR4231012.fastq.gz\n",
      "Approx 85% complete for ERR4231019.fastq.gz\n",
      "Approx 85% complete for ERR4231021.fastq.gz\n",
      "Approx 70% complete for ERR4231020.fastq.gz\n",
      "Approx 95% complete for ERR4231022.fastq.gz\n",
      "Approx 95% complete for ERR4231015.fastq.gz\n",
      "Approx 85% complete for ERR4231013.fastq.gz\n",
      "Approx 75% complete for ERR4231018.fastq.gz\n",
      "Approx 85% complete for ERR4231010.fastq.gz\n",
      "Analysis complete for ERR4231016.fastq.gz\n",
      "Analysis complete for ERR4231022.fastq.gz\n",
      "Approx 90% complete for ERR4231019.fastq.gz\n",
      "Approx 90% complete for ERR4231021.fastq.gz\n",
      "Analysis complete for ERR4231015.fastq.gz\n",
      "Approx 75% complete for ERR4231012.fastq.gz\n",
      "Approx 90% complete for ERR4231013.fastq.gz\n",
      "Approx 75% complete for ERR4231020.fastq.gz\n",
      "Approx 90% complete for ERR4231010.fastq.gz\n",
      "Approx 80% complete for ERR4231018.fastq.gz\n",
      "Approx 95% complete for ERR4231021.fastq.gz\n",
      "Approx 95% complete for ERR4231019.fastq.gz\n",
      "Approx 80% complete for ERR4231012.fastq.gz\n",
      "Approx 95% complete for ERR4231013.fastq.gz\n",
      "Approx 95% complete for ERR4231010.fastq.gz\n",
      "Approx 80% complete for ERR4231020.fastq.gz\n",
      "Approx 85% complete for ERR4231018.fastq.gz\n",
      "Analysis complete for ERR4231021.fastq.gz\n",
      "Analysis complete for ERR4231019.fastq.gz\n",
      "Analysis complete for ERR4231013.fastq.gz\n",
      "Approx 85% complete for ERR4231012.fastq.gz\n",
      "Analysis complete for ERR4231010.fastq.gz\n",
      "Approx 85% complete for ERR4231020.fastq.gz\n",
      "Approx 90% complete for ERR4231018.fastq.gz\n",
      "Approx 90% complete for ERR4231012.fastq.gz\n",
      "Approx 95% complete for ERR4231018.fastq.gz\n",
      "Approx 90% complete for ERR4231020.fastq.gz\n",
      "Approx 95% complete for ERR4231012.fastq.gz\n",
      "Analysis complete for ERR4231018.fastq.gz\n",
      "Approx 95% complete for ERR4231020.fastq.gz\n",
      "Analysis complete for ERR4231012.fastq.gz\n",
      "Analysis complete for ERR4231020.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "!mkdir fastq_placenta/placenta_qc_tempt_2\n",
    "!fastqc -t 16 fastq_placenta/*.fastq.gz -o fastq_placenta/placenta_qc_tempt_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cutadapt version 4.9\n",
      "\n",
      "Copyright (C) 2010 Marcel Martin <marcel.martin@scilifelab.se> and contributors\n",
      "\n",
      "Cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. All reads from input.fastq will be written to\n",
      "output.fastq with the adapter sequence removed. Adapter matching is\n",
      "error-tolerant. Multiple adapter sequences can be given (use further -a\n",
      "options), but only the best-matching adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Run \"cutadapt --help\" to see all command-line options.\n",
      "See https://cutadapt.readthedocs.io/ for full documentation.\n",
      "\n",
      "Options:\n",
      "  -h, --help            Show this help message and exit\n",
      "  --version             Show version number and exit\n",
      "  --debug               Print debug log. Use twice to also print DP matrices\n",
      "  -j CORES, --cores CORES\n",
      "                        Number of CPU cores to use. Use 0 to auto-detect.\n",
      "                        Default: 1\n",
      "\n",
      "Finding adapters:\n",
      "  Parameters -a, -g, -b specify adapters to be removed from each read (or from\n",
      "  R1 if data is paired-end. If specified multiple times, only the best\n",
      "  matching adapter is trimmed (but see the --times option). Use notation\n",
      "  'file:FILE' to read adapter sequences from a FASTA file.\n",
      "\n",
      "  -a ADAPTER, --adapter ADAPTER\n",
      "                        Sequence of an adapter ligated to the 3' end (paired\n",
      "                        data: of the first read). The adapter and subsequent\n",
      "                        bases are trimmed. If a '$' character is appended\n",
      "                        ('anchoring'), the adapter is only found if it is a\n",
      "                        suffix of the read.\n",
      "  -g ADAPTER, --front ADAPTER\n",
      "                        Sequence of an adapter ligated to the 5' end (paired\n",
      "                        data: of the first read). The adapter and any preceding\n",
      "                        bases are trimmed. Partial matches at the 5' end are\n",
      "                        allowed. If a '^' character is prepended ('anchoring'),\n",
      "                        the adapter is only found if it is a prefix of the read.\n",
      "  -b ADAPTER, --anywhere ADAPTER\n",
      "                        Sequence of an adapter that may be ligated to the 5' or\n",
      "                        3' end (paired data: of the first read). Both types of\n",
      "                        matches as described under -a and -g are allowed. If the\n",
      "                        first base of the read is part of the match, the\n",
      "                        behavior is as with -g, otherwise as with -a. This\n",
      "                        option is mostly for rescuing failed library\n",
      "                        preparations - do not use if you know which end your\n",
      "                        adapter was ligated to!\n",
      "  -e E, --error-rate E, --errors E\n",
      "                        Maximum allowed error rate (if 0 <= E < 1), or absolute\n",
      "                        number of errors for full-length adapter match (if E is\n",
      "                        an integer >= 1). Error rate = no. of errors divided by\n",
      "                        length of matching region. Default: 0.1 (10%)\n",
      "  --no-indels           Allow only mismatches in alignments. Default: allow both\n",
      "                        mismatches and indels\n",
      "  -n COUNT, --times COUNT\n",
      "                        Remove up to COUNT adapters from each read. Default: 1\n",
      "  -O MINLENGTH, --overlap MINLENGTH\n",
      "                        Require MINLENGTH overlap between read and adapter for\n",
      "                        an adapter to be found. Default: 3\n",
      "  --match-read-wildcards\n",
      "                        Interpret IUPAC wildcards in reads. Default: False\n",
      "  -N, --no-match-adapter-wildcards\n",
      "                        Do not interpret IUPAC wildcards in adapters.\n",
      "  --action {trim,retain,mask,lowercase,crop,none}\n",
      "                        What to do if a match was found. trim: trim adapter and\n",
      "                        up- or downstream sequence; retain: trim, but retain\n",
      "                        adapter; mask: replace with 'N' characters; lowercase:\n",
      "                        convert to lowercase; crop: trim up and downstream\n",
      "                        sequence; none: leave unchanged. Default: trim\n",
      "  --rc, --revcomp       Check both the read and its reverse complement for\n",
      "                        adapter matches. If match is on reverse-complemented\n",
      "                        version, output that one. Default: check only read\n",
      "\n",
      "Additional read modifications:\n",
      "  -u LEN, --cut LEN     Remove LEN bases from each read (or R1 if paired; use -U\n",
      "                        option for R2). If LEN is positive, remove bases from\n",
      "                        the beginning. If LEN is negative, remove bases from the\n",
      "                        end. Can be used twice if LENs have different signs.\n",
      "                        Applied *before* adapter trimming.\n",
      "  --nextseq-trim 3'CUTOFF\n",
      "                        NextSeq-specific quality trimming (each read). Trims\n",
      "                        also dark cycles appearing as high-quality G bases.\n",
      "  -q [5'CUTOFF,]3'CUTOFF, --quality-cutoff [5'CUTOFF,]3'CUTOFF\n",
      "                        Trim low-quality bases from 5' and/or 3' ends of each\n",
      "                        read before adapter removal. Applied to both reads if\n",
      "                        data is paired. If one value is given, only the 3' end\n",
      "                        is trimmed. If two comma-separated cutoffs are given,\n",
      "                        the 5' end is trimmed with the first cutoff, the 3' end\n",
      "                        with the second.\n",
      "  --quality-base N      Assume that quality values in FASTQ are encoded as\n",
      "                        ascii(quality + N). This needs to be set to 64 for some\n",
      "                        old Illumina FASTQ files. Default: 33\n",
      "  --poly-a              Trim poly-A tails\n",
      "  --length LENGTH, -l LENGTH\n",
      "                        Shorten reads to LENGTH. Positive values remove bases at\n",
      "                        the end while negative ones remove bases at the\n",
      "                        beginning. This and the following modifications are\n",
      "                        applied after adapter trimming.\n",
      "  --trim-n              Trim N's on ends of reads.\n",
      "  --length-tag TAG      Search for TAG followed by a decimal number in the\n",
      "                        description field of the read. Replace the decimal\n",
      "                        number with the correct length of the trimmed read. For\n",
      "                        example, use --length-tag 'length=' to correct fields\n",
      "                        like 'length=123'.\n",
      "  --strip-suffix STRIP_SUFFIX\n",
      "                        Remove this suffix from read names if present. Can be\n",
      "                        given multiple times.\n",
      "  -x PREFIX, --prefix PREFIX\n",
      "                        Add this prefix to read names. Use {name} to insert the\n",
      "                        name of the matching adapter.\n",
      "  -y SUFFIX, --suffix SUFFIX\n",
      "                        Add this suffix to read names; can also include {name}\n",
      "  --rename TEMPLATE     Rename reads using TEMPLATE containing variables such as\n",
      "                        {id}, {adapter_name} etc. (see documentation)\n",
      "  --zero-cap, -z        Change negative quality values to zero.\n",
      "\n",
      "Filtering of processed reads:\n",
      "  Filters are applied after above read modifications. Paired-end reads are\n",
      "  always discarded pairwise (see also --pair-filter).\n",
      "\n",
      "  -m LEN[:LEN2], --minimum-length LEN[:LEN2]\n",
      "                        Discard reads shorter than LEN. Default: 0\n",
      "  -M LEN[:LEN2], --maximum-length LEN[:LEN2]\n",
      "                        Discard reads longer than LEN. Default: no limit\n",
      "  --max-n COUNT         Discard reads with more than COUNT 'N' bases. If COUNT\n",
      "                        is a number between 0 and 1, it is interpreted as a\n",
      "                        fraction of the read length.\n",
      "  --max-expected-errors ERRORS, --max-ee ERRORS\n",
      "                        Discard reads whose expected number of errors (computed\n",
      "                        from quality values) exceeds ERRORS.\n",
      "  --max-average-error-rate ERROR_RATE, --max-aer ERROR_RATE\n",
      "                        as --max-expected-errors (see above), but divided by\n",
      "                        length to account for reads of varying length.\n",
      "  --discard-trimmed, --discard\n",
      "                        Discard reads that contain an adapter. Use also -O to\n",
      "                        avoid discarding too many randomly matching reads.\n",
      "  --discard-untrimmed, --trimmed-only\n",
      "                        Discard reads that do not contain an adapter.\n",
      "  --discard-casava      Discard reads that did not pass CASAVA filtering (header\n",
      "                        has :Y:).\n",
      "\n",
      "Output:\n",
      "  --quiet               Print only error messages.\n",
      "  --report {full,minimal}\n",
      "                        Which type of report to print: 'full' or 'minimal'.\n",
      "                        Default: full\n",
      "  --json FILE           Dump report in JSON format to FILE\n",
      "  -o FILE, --output FILE\n",
      "                        Write trimmed reads to FILE. FASTQ or FASTA format is\n",
      "                        chosen depending on input. Summary report is sent to\n",
      "                        standard output. Use '{name}' for demultiplexing (see\n",
      "                        docs). Default: write to standard output\n",
      "  --fasta               Output FASTA to standard output even on FASTQ input.\n",
      "  -Z                    Use compression level 1 for gzipped output files\n",
      "                        (faster, but uses more space)\n",
      "  --info-file FILE      Write information about each read and its adapter\n",
      "                        matches into FILE. See the documentation for the file\n",
      "                        format.\n",
      "  -r FILE, --rest-file FILE\n",
      "                        When the adapter matches in the middle of a read, write\n",
      "                        the rest (after the adapter) to FILE.\n",
      "  --wildcard-file FILE  When the adapter has N wildcard bases, write adapter\n",
      "                        bases matching wildcard positions to FILE. (Inaccurate\n",
      "                        with indels.)\n",
      "  --too-short-output FILE\n",
      "                        Write reads that are too short (according to length\n",
      "                        specified by -m) to FILE. Default: discard reads\n",
      "  --too-long-output FILE\n",
      "                        Write reads that are too long (according to length\n",
      "                        specified by -M) to FILE. Default: discard reads\n",
      "  --untrimmed-output FILE\n",
      "                        Write reads that do not contain any adapter to FILE.\n",
      "                        Default: output to same file as trimmed reads\n",
      "\n",
      "Paired-end options:\n",
      "  The -A/-G/-B/-U/-Q options work like their lowercase counterparts, but are\n",
      "  applied to R2 (second read in pair)\n",
      "\n",
      "  -A ADAPTER            3' adapter to be removed from R2\n",
      "  -G ADAPTER            5' adapter to be removed from R2\n",
      "  -B ADAPTER            5'/3 adapter to be removed from R2\n",
      "  -U LENGTH             Remove LENGTH bases from R2\n",
      "  -Q [5'CUTOFF,]3'CUTOFF\n",
      "                        Quality-trimming cutoff for R2. Default: same as for R1\n",
      "  -L LENGTH             Shorten R2 to LENGTH. Default: same as for R1\n",
      "  -p FILE, --paired-output FILE\n",
      "                        Write R2 to FILE.\n",
      "  --pair-adapters       Treat adapters given with -a/-A etc. as pairs. Either\n",
      "                        both or none are removed from each read pair.\n",
      "  --pair-filter {any,both,first}\n",
      "                        Which of the reads in a paired-end read have to match\n",
      "                        the filtering criterion in order for the pair to be\n",
      "                        filtered. Default: any\n",
      "  --interleaved         Read and/or write interleaved paired-end reads.\n",
      "  --untrimmed-paired-output FILE\n",
      "                        Write second read in a pair to this FILE when no adapter\n",
      "                        was found. Use with --untrimmed-output. Default: output\n",
      "                        to same file as trimmed reads\n",
      "  --too-short-paired-output FILE\n",
      "                        Write second read in a pair to this file if pair is too\n",
      "                        short.\n",
      "  --too-long-paired-output FILE\n",
      "                        Write second read in a pair to this file if pair is too\n",
      "                        long.\n"
     ]
    }
   ],
   "source": [
    "# !cutadapt -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231018.fastq.gz fast\n",
    "# f=expand(\"{list_item}\", list_item=config[\"accessions\"])\n",
    "!cutadapt -h\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231010.fastq.gz fastq_placenta/fastq_files/ERR4231010.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:16:24    64,341,577 reads @  15.3 µs/read;   3.92 M reads/minute\n",
      "Finished in 984.252 s (15.297 µs/read; 3.92 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              64,341,577\n",
      "Reads with adapters:                 1,267,627 (2.0%)\n",
      "Reads written (passing filters):    64,341,577 (100.0%)\n",
      "\n",
      "Total basepairs processed: 12,996,998,554 bp\n",
      "Quality-trimmed:             927,005,324 bp (7.1%)\n",
      "Total written (filtered):  11,422,263,145 bp (87.9%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 1267627 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.3%\n",
      "  C: 22.5%\n",
      "  G: 23.7%\n",
      "  T: 18.4%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t888512\t1005337.1\t0\t888512\n",
      "4\t281503\t251334.3\t0\t281503\n",
      "5\t75813\t62833.6\t0\t75813\n",
      "6\t15571\t15708.4\t0\t15571\n",
      "7\t3539\t3927.1\t0\t3539\n",
      "8\t524\t981.8\t0\t524\n",
      "9\t850\t245.4\t0\t59 791\n",
      "10\t860\t61.4\t1\t20 840\n",
      "11\t418\t15.3\t1\t0 418\n",
      "12\t27\t3.8\t1\t0 27\n",
      "13\t8\t1.0\t1\t0 8\n",
      "14\t1\t0.2\t1\t0 1\n",
      "21\t1\t0.0\t2\t1\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231011.fastq.gz fastq_placenta/fastq_files/ERR4231011.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:11:25    44,940,444 reads @  15.3 µs/read;   3.93 M reads/minute\n",
      "Finished in 685.402 s (15.251 µs/read; 3.93 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              44,940,444\n",
      "Reads with adapters:                   888,332 (2.0%)\n",
      "Reads written (passing filters):    44,940,444 (100.0%)\n",
      "\n",
      "Total basepairs processed: 9,077,969,688 bp\n",
      "Quality-trimmed:             690,471,056 bp (7.6%)\n",
      "Total written (filtered):  7,935,067,874 bp (87.4%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 888332 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.6%\n",
      "  C: 22.2%\n",
      "  G: 23.5%\n",
      "  T: 18.6%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t619128\t702194.4\t0\t619128\n",
      "4\t199173\t175548.6\t0\t199173\n",
      "5\t55427\t43887.2\t0\t55427\n",
      "6\t10596\t10971.8\t0\t10596\n",
      "7\t2550\t2742.9\t0\t2550\n",
      "8\t422\t685.7\t0\t422\n",
      "9\t329\t171.4\t0\t42 287\n",
      "10\t478\t42.9\t1\t13 465\n",
      "11\t192\t10.7\t1\t0 192\n",
      "12\t29\t2.7\t1\t0 29\n",
      "13\t8\t0.7\t1\t0 8\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231012.fastq.gz fastq_placenta/fastq_files/ERR4231012.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:18:39    73,535,954 reads @  15.2 µs/read;   3.94 M reads/minute\n",
      "Finished in 1119.827 s (15.228 µs/read; 3.94 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              73,535,954\n",
      "Reads with adapters:                 1,472,162 (2.0%)\n",
      "Reads written (passing filters):    73,535,954 (100.0%)\n",
      "\n",
      "Total basepairs processed: 14,854,262,708 bp\n",
      "Quality-trimmed:           1,073,935,893 bp (7.2%)\n",
      "Total written (filtered):  13,039,950,704 bp (87.8%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 1472162 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.9%\n",
      "  C: 22.2%\n",
      "  G: 23.4%\n",
      "  T: 18.4%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t1025460\t1148999.3\t0\t1025460\n",
      "4\t331197\t287249.8\t0\t331197\n",
      "5\t90383\t71812.5\t0\t90383\n",
      "6\t18481\t17953.1\t0\t18481\n",
      "7\t4144\t4488.3\t0\t4144\n",
      "8\t661\t1122.1\t0\t661\n",
      "9\t543\t280.5\t0\t56 487\n",
      "10\t893\t70.1\t1\t26 867\n",
      "11\t330\t17.5\t1\t0 330\n",
      "12\t57\t4.4\t1\t0 57\n",
      "13\t11\t1.1\t1\t0 11\n",
      "15\t1\t0.1\t1\t0 1\n",
      "17\t1\t0.0\t1\t1\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231013.fastq.gz fastq_placenta/fastq_files/ERR4231013.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:15:54    62,889,838 reads @  15.2 µs/read;   3.95 M reads/minute\n",
      "Finished in 954.643 s (15.180 µs/read; 3.95 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              62,889,838\n",
      "Reads with adapters:                 1,235,190 (2.0%)\n",
      "Reads written (passing filters):    62,889,838 (100.0%)\n",
      "\n",
      "Total basepairs processed: 12,703,747,276 bp\n",
      "Quality-trimmed:             651,303,752 bp (5.1%)\n",
      "Total written (filtered):  11,419,342,313 bp (89.9%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 1235190 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.1%\n",
      "  C: 22.2%\n",
      "  G: 24.2%\n",
      "  T: 18.4%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t867203\t982653.7\t0\t867203\n",
      "4\t271900\t245663.4\t0\t271900\n",
      "5\t75372\t61415.9\t0\t75372\n",
      "6\t14661\t15354.0\t0\t14661\n",
      "7\t3388\t3838.5\t0\t3388\n",
      "8\t507\t959.6\t0\t507\n",
      "9\t1020\t239.9\t0\t54 966\n",
      "10\t744\t60.0\t1\t14 730\n",
      "11\t348\t15.0\t1\t1 347\n",
      "12\t36\t3.7\t1\t0 36\n",
      "13\t10\t0.9\t1\t0 10\n",
      "14\t1\t0.2\t1\t0 1\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231014.fastq.gz fastq_placenta/fastq_files/ERR4231014.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:09:08    36,185,523 reads @  15.1 µs/read;   3.96 M reads/minute\n",
      "Finished in 548.127 s (15.148 µs/read; 3.96 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              36,185,523\n",
      "Reads with adapters:                   674,463 (1.9%)\n",
      "Reads written (passing filters):    36,185,523 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,309,475,646 bp\n",
      "Quality-trimmed:             484,105,735 bp (6.6%)\n",
      "Total written (filtered):  6,461,217,748 bp (88.4%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 674463 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.2%\n",
      "  C: 22.3%\n",
      "  G: 24.1%\n",
      "  T: 18.3%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t470670\t565398.8\t0\t470670\n",
      "4\t151068\t141349.7\t0\t151068\n",
      "5\t41511\t35337.4\t0\t41511\n",
      "6\t8250\t8834.4\t0\t8250\n",
      "7\t1842\t2208.6\t0\t1842\n",
      "8\t276\t552.1\t0\t276\n",
      "9\t283\t138.0\t0\t27 256\n",
      "10\t376\t34.5\t1\t13 363\n",
      "11\t160\t8.6\t1\t1 159\n",
      "12\t20\t2.2\t1\t0 20\n",
      "14\t2\t0.1\t1\t1 1\n",
      "19\t1\t0.0\t1\t0 1\n",
      "21\t1\t0.0\t2\t1\n",
      "24\t1\t0.0\t2\t0 0 1\n",
      "39\t1\t0.0\t3\t0 0 0 1\n",
      "56\t1\t0.0\t5\t0 0 0 1\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231015.fastq.gz fastq_placenta/fastq_files/ERR4231015.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:13:53    53,650,335 reads @  15.5 µs/read;   3.86 M reads/minute\n",
      "Finished in 833.685 s (15.539 µs/read; 3.86 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              53,650,335\n",
      "Reads with adapters:                 1,028,621 (1.9%)\n",
      "Reads written (passing filters):    53,650,335 (100.0%)\n",
      "\n",
      "Total basepairs processed: 10,837,367,670 bp\n",
      "Quality-trimmed:             608,296,775 bp (5.6%)\n",
      "Total written (filtered):  9,689,065,673 bp (89.4%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 1028621 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.4%\n",
      "  C: 22.2%\n",
      "  G: 24.8%\n",
      "  T: 18.5%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t720323\t838286.5\t0\t720323\n",
      "4\t229239\t209571.6\t0\t229239\n",
      "5\t61859\t52392.9\t0\t61859\n",
      "6\t12080\t13098.2\t0\t12080\n",
      "7\t2562\t3274.6\t0\t2562\n",
      "8\t424\t818.6\t0\t424\n",
      "9\t1055\t204.7\t0\t53 1002\n",
      "10\t629\t51.2\t1\t10 619\n",
      "11\t381\t12.8\t1\t7 374\n",
      "12\t47\t3.2\t1\t0 47\n",
      "13\t18\t0.8\t1\t0 18\n",
      "14\t3\t0.2\t1\t1 2\n",
      "30\t1\t0.0\t3\t1\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231016.fastq.gz fastq_placenta/fastq_files/ERR4231016.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:13:15    51,945,229 reads @  15.3 µs/read;   3.92 M reads/minute\n",
      "Finished in 796.070 s (15.325 µs/read; 3.92 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              51,945,229\n",
      "Reads with adapters:                   910,189 (1.8%)\n",
      "Reads written (passing filters):    51,945,229 (100.0%)\n",
      "\n",
      "Total basepairs processed: 10,492,936,258 bp\n",
      "Quality-trimmed:             799,740,675 bp (7.6%)\n",
      "Total written (filtered):  9,170,637,903 bp (87.4%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 910189 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.9%\n",
      "  C: 22.8%\n",
      "  G: 23.7%\n",
      "  T: 17.4%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t618560\t811644.2\t0\t618560\n",
      "4\t229263\t202911.1\t0\t229263\n",
      "5\t49466\t50727.8\t0\t49466\n",
      "6\t9127\t12681.9\t0\t9127\n",
      "7\t2210\t3170.5\t0\t2210\n",
      "8\t304\t792.6\t0\t304\n",
      "9\t363\t198.2\t0\t24 339\n",
      "10\t611\t49.5\t1\t16 595\n",
      "11\t240\t12.4\t1\t1 239\n",
      "12\t32\t3.1\t1\t0 32\n",
      "13\t10\t0.8\t1\t0 10\n",
      "14\t1\t0.2\t1\t0 1\n",
      "19\t1\t0.0\t1\t0 1\n",
      "100\t1\t0.0\t5\t0 0 0 0 1\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231017.fastq.gz fastq_placenta/fastq_files/ERR4231017.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:08:29    33,950,855 reads @  15.0 µs/read;   4.00 M reads/minute\n",
      "Finished in 509.633 s (15.011 µs/read; 4.00 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              33,950,855\n",
      "Reads with adapters:                   594,814 (1.8%)\n",
      "Reads written (passing filters):    33,950,855 (100.0%)\n",
      "\n",
      "Total basepairs processed: 6,858,072,710 bp\n",
      "Quality-trimmed:             576,689,349 bp (8.4%)\n",
      "Total written (filtered):  5,939,842,676 bp (86.6%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 594814 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.7%\n",
      "  C: 23.5%\n",
      "  G: 24.8%\n",
      "  T: 16.9%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t398580\t530482.1\t0\t398580\n",
      "4\t157128\t132620.5\t0\t157128\n",
      "5\t31528\t33155.1\t0\t31528\n",
      "6\t5482\t8288.8\t0\t5482\n",
      "7\t1172\t2072.2\t0\t1172\n",
      "8\t187\t518.0\t0\t187\n",
      "9\t184\t129.5\t0\t8 176\n",
      "10\t341\t32.4\t1\t10 331\n",
      "11\t158\t8.1\t1\t1 157\n",
      "12\t25\t2.0\t1\t0 25\n",
      "13\t28\t0.5\t1\t1 27\n",
      "183\t1\t0.0\t5\t0 0 0 0 1\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231018.fastq.gz fastq_placenta/fastq_files/ERR4231018.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:17:14    69,462,070 reads @  14.9 µs/read;   4.03 M reads/minute\n",
      "Finished in 1034.152 s (14.888 µs/read; 4.03 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              69,462,070\n",
      "Reads with adapters:                 1,235,795 (1.8%)\n",
      "Reads written (passing filters):    69,462,070 (100.0%)\n",
      "\n",
      "Total basepairs processed: 14,031,338,140 bp\n",
      "Quality-trimmed:           1,106,277,931 bp (7.9%)\n",
      "Total written (filtered):  12,226,219,361 bp (87.1%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 1235795 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.7%\n",
      "  C: 23.0%\n",
      "  G: 23.7%\n",
      "  T: 17.5%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t837655\t1085344.8\t0\t837655\n",
      "4\t312298\t271336.2\t0\t312298\n",
      "5\t67925\t67834.1\t0\t67925\n",
      "6\t12717\t16958.5\t0\t12717\n",
      "7\t2992\t4239.6\t0\t2992\n",
      "8\t482\t1059.9\t0\t482\n",
      "9\t530\t265.0\t0\t50 480\n",
      "10\t780\t66.2\t1\t19 761\n",
      "11\t346\t16.6\t1\t0 346\n",
      "12\t43\t4.1\t1\t0 43\n",
      "13\t26\t1.0\t1\t0 26\n",
      "34\t1\t0.0\t3\t0 1\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231019.fastq.gz fastq_placenta/fastq_files/ERR4231019.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:13:51    60,339,006 reads @  13.8 µs/read;   4.36 M reads/minute\n",
      "Finished in 831.273 s (13.777 µs/read; 4.36 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              60,339,006\n",
      "Reads with adapters:                 1,333,000 (2.2%)\n",
      "Reads written (passing filters):    60,339,006 (100.0%)\n",
      "\n",
      "Total basepairs processed: 12,188,479,212 bp\n",
      "Quality-trimmed:           2,718,560,431 bp (22.3%)\n",
      "Total written (filtered):  8,862,037,207 bp (72.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 1333000 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.6%\n",
      "  C: 22.2%\n",
      "  G: 23.2%\n",
      "  T: 18.9%\n",
      "  none/other: 0.2%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t942294\t942797.0\t0\t942294\n",
      "4\t319558\t235699.2\t0\t319558\n",
      "5\t51068\t58924.8\t0\t51068\n",
      "6\t14987\t14731.2\t0\t14987\n",
      "7\t2968\t3682.8\t0\t2968\n",
      "8\t440\t920.7\t0\t440\n",
      "9\t467\t230.2\t0\t40 427\n",
      "10\t840\t57.5\t1\t17 823\n",
      "11\t323\t14.4\t1\t1 322\n",
      "12\t29\t3.6\t1\t0 29\n",
      "13\t26\t0.9\t1\t0 26\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231020.fastq.gz fastq_placenta/fastq_files/ERR4231020.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:16:51    74,203,167 reads @  13.6 µs/read;   4.40 M reads/minute\n",
      "Finished in 1011.635 s (13.633 µs/read; 4.40 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              74,203,167\n",
      "Reads with adapters:                 1,578,504 (2.1%)\n",
      "Reads written (passing filters):    74,203,167 (100.0%)\n",
      "\n",
      "Total basepairs processed: 14,989,039,734 bp\n",
      "Quality-trimmed:           3,659,031,046 bp (24.4%)\n",
      "Total written (filtered):  10,582,660,575 bp (70.6%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 1578504 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.9%\n",
      "  C: 22.6%\n",
      "  G: 22.6%\n",
      "  T: 18.7%\n",
      "  none/other: 0.2%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t1117377\t1159424.5\t0\t1117377\n",
      "4\t376602\t289856.1\t0\t376602\n",
      "5\t60837\t72464.0\t0\t60837\n",
      "6\t18297\t18116.0\t0\t18297\n",
      "7\t2962\t4529.0\t0\t2962\n",
      "8\t567\t1132.3\t0\t567\n",
      "9\t519\t283.1\t0\t57 462\n",
      "10\t925\t70.8\t1\t15 910\n",
      "11\t337\t17.7\t1\t1 336\n",
      "12\t46\t4.4\t1\t0 46\n",
      "13\t32\t1.1\t1\t0 32\n",
      "14\t2\t0.3\t1\t0 2\n",
      "43\t1\t0.0\t4\t0 0 1\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231021.fastq.gz fastq_placenta/fastq_files/ERR4231021.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:13:44    60,201,088 reads @  13.7 µs/read;   4.38 M reads/minute\n",
      "Finished in 824.313 s (13.693 µs/read; 4.38 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              60,201,088\n",
      "Reads with adapters:                 1,288,489 (2.1%)\n",
      "Reads written (passing filters):    60,201,088 (100.0%)\n",
      "\n",
      "Total basepairs processed: 12,160,619,776 bp\n",
      "Quality-trimmed:           2,801,685,149 bp (23.0%)\n",
      "Total written (filtered):  8,752,578,948 bp (72.0%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 1288489 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.4%\n",
      "  C: 22.9%\n",
      "  G: 22.9%\n",
      "  T: 18.6%\n",
      "  none/other: 0.2%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t908338\t940642.0\t0\t908338\n",
      "4\t310910\t235160.5\t0\t310910\n",
      "5\t49331\t58790.1\t0\t49331\n",
      "6\t15323\t14697.5\t0\t15323\n",
      "7\t2444\t3674.4\t0\t2444\n",
      "8\t453\t918.6\t0\t453\n",
      "9\t503\t229.6\t0\t45 458\n",
      "10\t858\t57.4\t1\t22 836\n",
      "11\t267\t14.4\t1\t2 265\n",
      "12\t34\t3.6\t1\t0 34\n",
      "13\t24\t0.9\t1\t0 24\n",
      "14\t4\t0.2\t1\t0 4\n",
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231022.fastq.gz fastq_placenta/fastq_files/ERR4231022.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "[      8<----] 00:00:13       910,000 reads @  14.9 µs/read;   4.03 M reads/minute"
     ]
    }
   ],
   "source": [
    "# %mkdir fastq_placenta/trimmed_placenta\n",
    "!for fastq in fastq_placenta/fastq_files/*.fastq.gz; do \\\n",
    "    base_name=$(basename \"$fastq\" .fastq.gz); \\\n",
    "    cutadapt -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/$base_name.fastq.gz $fastq; \\\n",
    "done\n",
    "#suggestive to remove the empty reads by -m 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231023.fastq.gz fastq_placenta/fastq_files/ERR4231023.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:11:13    45,049,673 reads @  15.0 µs/read;   4.01 M reads/minute\n",
      "Finished in 673.757 s (14.956 µs/read; 4.01 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              45,049,673\n",
      "Reads with adapters:                   781,219 (1.7%)\n",
      "Reads written (passing filters):    45,049,673 (100.0%)\n",
      "\n",
      "Total basepairs processed: 9,100,033,946 bp\n",
      "Quality-trimmed:             603,472,881 bp (6.6%)\n",
      "Total written (filtered):  8,043,403,616 bp (88.4%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 781219 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 33.3%\n",
      "  C: 23.8%\n",
      "  G: 24.9%\n",
      "  T: 17.9%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t535977\t703901.1\t0\t535977\n",
      "4\t190815\t175975.3\t0\t190815\n",
      "5\t43145\t43993.8\t0\t43145\n",
      "6\t8320\t10998.5\t0\t8320\n",
      "7\t1704\t2749.6\t0\t1704\n",
      "8\t307\t687.4\t0\t307\n",
      "9\t298\t171.9\t0\t25 273\n",
      "10\t447\t43.0\t1\t13 434\n",
      "11\t154\t10.7\t1\t0 154\n",
      "12\t39\t2.7\t1\t0 39\n",
      "13\t11\t0.7\t1\t0 11\n",
      "20\t1\t0.0\t2\t0 0 1\n",
      "22\t1\t0.0\t2\t1\n"
     ]
    }
   ],
   "source": [
    "# !cutadapt -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231023.fastq.gz fastq_placenta/fastq_files/ERR4231023.fastq.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 4.9 with Python 3.12.3\n",
      "Command line parameters: --poly-a -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231020_polya.fastq.gz fastq_placenta/fastq_files/ERR4231020.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Done           00:18:21    74,203,167 reads @  14.8 µs/read;   4.04 M reads/minute\n",
      "Finished in 1101.072 s (14.839 µs/read; 4.04 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              74,203,167\n",
      "Reads with adapters:                 1,578,504 (2.1%)\n",
      "Reads written (passing filters):    74,203,167 (100.0%)\n",
      "\n",
      "Total basepairs processed: 14,989,039,734 bp\n",
      "Quality-trimmed:           3,659,031,046 bp (24.4%)\n",
      "Poly-A-trimmed:               15,156,995 bp (0.1%)\n",
      "Total written (filtered):  10,567,503,580 bp (70.5%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT; Type: regular 3'; Length: 58; Trimmed: 1578504 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-58 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.9%\n",
      "  C: 22.6%\n",
      "  G: 22.6%\n",
      "  T: 18.7%\n",
      "  none/other: 0.2%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t1117377\t1159424.5\t0\t1117377\n",
      "4\t376602\t289856.1\t0\t376602\n",
      "5\t60837\t72464.0\t0\t60837\n",
      "6\t18297\t18116.0\t0\t18297\n",
      "7\t2962\t4529.0\t0\t2962\n",
      "8\t567\t1132.3\t0\t567\n",
      "9\t519\t283.1\t0\t57 462\n",
      "10\t925\t70.8\t1\t15 910\n",
      "11\t337\t17.7\t1\t1 336\n",
      "12\t46\t4.4\t1\t0 46\n",
      "13\t32\t1.1\t1\t0 32\n",
      "14\t2\t0.3\t1\t0 2\n",
      "43\t1\t0.0\t4\t0 0 1\n",
      "\n",
      "\n",
      "=== Poly-A trimmed ===\n",
      "\n",
      "length\tcount\n",
      "0\t71843826\n",
      "3\t772037\n",
      "4\t272057\n",
      "5\t466501\n",
      "6\t264880\n",
      "7\t118990\n",
      "8\t56669\n",
      "9\t33107\n",
      "10\t51469\n",
      "11\t37183\n",
      "12\t29218\n",
      "13\t25650\n",
      "14\t24288\n",
      "15\t30579\n",
      "16\t26977\n",
      "17\t22322\n",
      "18\t18300\n",
      "19\t15565\n",
      "20\t16001\n",
      "21\t12710\n",
      "22\t10299\n",
      "23\t8233\n",
      "24\t7050\n",
      "25\t6594\n",
      "26\t5670\n",
      "27\t4472\n",
      "28\t3394\n",
      "29\t2635\n",
      "30\t2403\n",
      "31\t1933\n",
      "32\t1470\n",
      "33\t1329\n",
      "34\t1075\n",
      "35\t1153\n",
      "36\t869\n",
      "37\t752\n",
      "38\t541\n",
      "39\t452\n",
      "40\t466\n",
      "41\t380\n",
      "42\t319\n",
      "43\t275\n",
      "44\t314\n",
      "45\t276\n",
      "46\t241\n",
      "47\t209\n",
      "48\t151\n",
      "49\t101\n",
      "50\t131\n",
      "51\t116\n",
      "52\t92\n",
      "53\t79\n",
      "54\t71\n",
      "55\t59\n",
      "56\t52\n",
      "57\t55\n",
      "58\t50\n",
      "59\t34\n",
      "60\t53\n",
      "61\t57\n",
      "62\t46\n",
      "63\t50\n",
      "64\t64\n",
      "65\t82\n",
      "66\t56\n",
      "67\t42\n",
      "68\t33\n",
      "69\t23\n",
      "70\t64\n",
      "71\t36\n",
      "72\t42\n",
      "73\t285\n",
      "74\t74\n",
      "75\t4\n",
      "76\t3\n",
      "78\t2\n",
      "80\t2\n",
      "81\t2\n",
      "82\t5\n",
      "83\t2\n",
      "84\t1\n",
      "86\t2\n",
      "89\t1\n",
      "91\t2\n",
      "93\t1\n",
      "96\t1\n",
      "100\t1\n",
      "108\t1\n",
      "111\t1\n",
      "113\t1\n",
      "122\t1\n",
      "123\t1\n",
      "132\t1\n",
      "134\t1\n"
     ]
    }
   ],
   "source": [
    "# !cutadapt --poly-a -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT -u 10 -q 28 -o ./fastq_placenta/trimmed_placenta/ERR4231020_polya.fastq.gz fastq_placenta/fastq_files/ERR4231020.fastq.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.001s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /dev/cpuset.\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of ERR4231010.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of ERR4231011.fastq.gz\n",
      "Started analysis of ERR4231012.fastq.gz\n",
      "Started analysis of ERR4231013.fastq.gz\n",
      "Started analysis of ERR4231014.fastq.gz\n",
      "Started analysis of ERR4231015.fastq.gz\n",
      "Started analysis of ERR4231016.fastq.gz\n",
      "Started analysis of ERR4231017.fastq.gz\n",
      "Started analysis of ERR4231018.fastq.gz\n",
      "Started analysis of ERR4231019.fastq.gz\n",
      "Started analysis of ERR4231020.fastq.gz\n",
      "Started analysis of ERR4231021.fastq.gz\n",
      "Started analysis of ERR4231022.fastq.gz\n",
      "Started analysis of ERR4231023.fastq.gz\n",
      "Approx 5% complete for ERR4231014.fastq.gz\n",
      "Approx 5% complete for ERR4231017.fastq.gz\n",
      "Approx 5% complete for ERR4231011.fastq.gz\n",
      "Approx 5% complete for ERR4231016.fastq.gz\n",
      "Approx 5% complete for ERR4231023.fastq.gz\n",
      "Approx 5% complete for ERR4231015.fastq.gz\n",
      "Approx 5% complete for ERR4231021.fastq.gz\n",
      "Approx 5% complete for ERR4231019.fastq.gz\n",
      "Approx 5% complete for ERR4231010.fastq.gz\n",
      "Approx 5% complete for ERR4231022.fastq.gz\n",
      "Approx 5% complete for ERR4231013.fastq.gz\n",
      "Approx 10% complete for ERR4231017.fastq.gz\n",
      "Approx 5% complete for ERR4231020.fastq.gz\n",
      "Approx 10% complete for ERR4231014.fastq.gz\n",
      "Approx 5% complete for ERR4231012.fastq.gz\n",
      "Approx 5% complete for ERR4231018.fastq.gz\n",
      "Approx 10% complete for ERR4231011.fastq.gz\n",
      "Approx 10% complete for ERR4231023.fastq.gz\n",
      "Approx 10% complete for ERR4231016.fastq.gz\n",
      "Approx 15% complete for ERR4231017.fastq.gz\n",
      "Approx 15% complete for ERR4231014.fastq.gz\n",
      "Approx 10% complete for ERR4231015.fastq.gz\n",
      "Approx 10% complete for ERR4231021.fastq.gz\n",
      "Approx 10% complete for ERR4231019.fastq.gz\n",
      "Approx 10% complete for ERR4231022.fastq.gz\n",
      "Approx 10% complete for ERR4231010.fastq.gz\n",
      "Approx 15% complete for ERR4231011.fastq.gz\n",
      "Approx 10% complete for ERR4231013.fastq.gz\n",
      "Approx 10% complete for ERR4231020.fastq.gz\n",
      "Approx 20% complete for ERR4231017.fastq.gz\n",
      "Approx 20% complete for ERR4231014.fastq.gz\n",
      "Approx 15% complete for ERR4231023.fastq.gz\n",
      "Approx 10% complete for ERR4231012.fastq.gz\n",
      "Approx 10% complete for ERR4231018.fastq.gz\n",
      "Approx 15% complete for ERR4231016.fastq.gz\n",
      "Approx 15% complete for ERR4231021.fastq.gz\n",
      "Approx 15% complete for ERR4231015.fastq.gz\n",
      "Approx 15% complete for ERR4231019.fastq.gz\n",
      "Approx 25% complete for ERR4231017.fastq.gz\n",
      "Approx 15% complete for ERR4231022.fastq.gz\n",
      "Approx 25% complete for ERR4231014.fastq.gz\n",
      "Approx 20% complete for ERR4231011.fastq.gz\n",
      "Approx 20% complete for ERR4231023.fastq.gz\n",
      "Approx 15% complete for ERR4231010.fastq.gz\n",
      "Approx 15% complete for ERR4231020.fastq.gz\n",
      "Approx 15% complete for ERR4231013.fastq.gz\n",
      "Approx 20% complete for ERR4231016.fastq.gz\n",
      "Approx 30% complete for ERR4231017.fastq.gz\n",
      "Approx 20% complete for ERR4231021.fastq.gz\n",
      "Approx 20% complete for ERR4231015.fastq.gz\n",
      "Approx 30% complete for ERR4231014.fastq.gz\n",
      "Approx 15% complete for ERR4231012.fastq.gz\n",
      "Approx 15% complete for ERR4231018.fastq.gz\n",
      "Approx 20% complete for ERR4231019.fastq.gz\n",
      "Approx 25% complete for ERR4231011.fastq.gz\n",
      "Approx 20% complete for ERR4231022.fastq.gz\n",
      "Approx 25% complete for ERR4231023.fastq.gz\n",
      "Approx 35% complete for ERR4231017.fastq.gz\n",
      "Approx 35% complete for ERR4231014.fastq.gz\n",
      "Approx 25% complete for ERR4231016.fastq.gz\n",
      "Approx 20% complete for ERR4231010.fastq.gz\n",
      "Approx 25% complete for ERR4231021.fastq.gz\n",
      "Approx 20% complete for ERR4231020.fastq.gz\n",
      "Approx 25% complete for ERR4231015.fastq.gz\n",
      "Approx 20% complete for ERR4231013.fastq.gz\n",
      "Approx 30% complete for ERR4231011.fastq.gz\n",
      "Approx 40% complete for ERR4231017.fastq.gz\n",
      "Approx 25% complete for ERR4231019.fastq.gz\n",
      "Approx 30% complete for ERR4231023.fastq.gz\n",
      "Approx 25% complete for ERR4231022.fastq.gz\n",
      "Approx 40% complete for ERR4231014.fastq.gz\n",
      "Approx 20% complete for ERR4231012.fastq.gz\n",
      "Approx 20% complete for ERR4231018.fastq.gz\n",
      "Approx 30% complete for ERR4231016.fastq.gz\n",
      "Approx 45% complete for ERR4231017.fastq.gz\n",
      "Approx 30% complete for ERR4231021.fastq.gz\n",
      "Approx 35% complete for ERR4231011.fastq.gz\n",
      "Approx 45% complete for ERR4231014.fastq.gz\n",
      "Approx 25% complete for ERR4231010.fastq.gz\n",
      "Approx 30% complete for ERR4231015.fastq.gz\n",
      "Approx 35% complete for ERR4231023.fastq.gz\n",
      "Approx 25% complete for ERR4231020.fastq.gz\n",
      "Approx 30% complete for ERR4231019.fastq.gz\n",
      "Approx 30% complete for ERR4231022.fastq.gz\n",
      "Approx 25% complete for ERR4231013.fastq.gz\n",
      "Approx 50% complete for ERR4231017.fastq.gz\n",
      "Approx 50% complete for ERR4231014.fastq.gz\n",
      "Approx 35% complete for ERR4231016.fastq.gz\n",
      "Approx 40% complete for ERR4231011.fastq.gz\n",
      "Approx 35% complete for ERR4231021.fastq.gz\n",
      "Approx 25% complete for ERR4231018.fastq.gz\n",
      "Approx 25% complete for ERR4231012.fastq.gz\n",
      "Approx 40% complete for ERR4231023.fastq.gz\n",
      "Approx 55% complete for ERR4231017.fastq.gz\n",
      "Approx 35% complete for ERR4231015.fastq.gz\n",
      "Approx 35% complete for ERR4231019.fastq.gz\n",
      "Approx 30% complete for ERR4231010.fastq.gz\n",
      "Approx 35% complete for ERR4231022.fastq.gz\n",
      "Approx 55% complete for ERR4231014.fastq.gz\n",
      "Approx 30% complete for ERR4231020.fastq.gz\n",
      "Approx 45% complete for ERR4231011.fastq.gz\n",
      "Approx 30% complete for ERR4231013.fastq.gz\n",
      "Approx 40% complete for ERR4231016.fastq.gz\n",
      "Approx 40% complete for ERR4231021.fastq.gz\n",
      "Approx 60% complete for ERR4231017.fastq.gz\n",
      "Approx 45% complete for ERR4231023.fastq.gz\n",
      "Approx 60% complete for ERR4231014.fastq.gz\n",
      "Approx 30% complete for ERR4231018.fastq.gz\n",
      "Approx 40% complete for ERR4231019.fastq.gz\n",
      "Approx 30% complete for ERR4231012.fastq.gz\n",
      "Approx 40% complete for ERR4231015.fastq.gz\n",
      "Approx 40% complete for ERR4231022.fastq.gz\n",
      "Approx 50% complete for ERR4231011.fastq.gz\n",
      "Approx 35% complete for ERR4231010.fastq.gz\n",
      "Approx 65% complete for ERR4231017.fastq.gz\n",
      "Approx 35% complete for ERR4231020.fastq.gz\n",
      "Approx 45% complete for ERR4231016.fastq.gz\n",
      "Approx 45% complete for ERR4231021.fastq.gz\n",
      "Approx 50% complete for ERR4231023.fastq.gz\n",
      "Approx 65% complete for ERR4231014.fastq.gz\n",
      "Approx 35% complete for ERR4231013.fastq.gz\n",
      "Approx 70% complete for ERR4231017.fastq.gz\n",
      "Approx 45% complete for ERR4231019.fastq.gz\n",
      "Approx 55% complete for ERR4231011.fastq.gz\n",
      "Approx 45% complete for ERR4231015.fastq.gz\n",
      "Approx 35% complete for ERR4231018.fastq.gz\n",
      "Approx 70% complete for ERR4231014.fastq.gz\n",
      "Approx 45% complete for ERR4231022.fastq.gz\n",
      "Approx 50% complete for ERR4231016.fastq.gz\n",
      "Approx 35% complete for ERR4231012.fastq.gz\n",
      "Approx 55% complete for ERR4231023.fastq.gz\n",
      "Approx 50% complete for ERR4231021.fastq.gz\n",
      "Approx 40% complete for ERR4231010.fastq.gz\n",
      "Approx 40% complete for ERR4231020.fastq.gz\n",
      "Approx 75% complete for ERR4231017.fastq.gz\n",
      "Approx 50% complete for ERR4231019.fastq.gz\n",
      "Approx 75% complete for ERR4231014.fastq.gz\n",
      "Approx 60% complete for ERR4231011.fastq.gz\n",
      "Approx 40% complete for ERR4231013.fastq.gz\n",
      "Approx 50% complete for ERR4231015.fastq.gz\n",
      "Approx 55% complete for ERR4231016.fastq.gz\n",
      "Approx 50% complete for ERR4231022.fastq.gz\n",
      "Approx 60% complete for ERR4231023.fastq.gz\n",
      "Approx 80% complete for ERR4231017.fastq.gz\n",
      "Approx 55% complete for ERR4231021.fastq.gz\n",
      "Approx 40% complete for ERR4231018.fastq.gz\n",
      "Approx 80% complete for ERR4231014.fastq.gz\n",
      "Approx 45% complete for ERR4231020.fastq.gz\n",
      "Approx 45% complete for ERR4231010.fastq.gz\n",
      "Approx 40% complete for ERR4231012.fastq.gz\n",
      "Approx 65% complete for ERR4231011.fastq.gz\n",
      "Approx 55% complete for ERR4231019.fastq.gz\n",
      "Approx 85% complete for ERR4231017.fastq.gz\n",
      "Approx 65% complete for ERR4231023.fastq.gz\n",
      "Approx 60% complete for ERR4231016.fastq.gz\n",
      "Approx 45% complete for ERR4231013.fastq.gz\n",
      "Approx 55% complete for ERR4231015.fastq.gz\n",
      "Approx 85% complete for ERR4231014.fastq.gz\n",
      "Approx 55% complete for ERR4231022.fastq.gz\n",
      "Approx 60% complete for ERR4231021.fastq.gz\n",
      "Approx 90% complete for ERR4231017.fastq.gz\n",
      "Approx 70% complete for ERR4231011.fastq.gz\n",
      "Approx 45% complete for ERR4231018.fastq.gz\n",
      "Approx 50% complete for ERR4231020.fastq.gz\n",
      "Approx 50% complete for ERR4231010.fastq.gz\n",
      "Approx 60% complete for ERR4231019.fastq.gz\n",
      "Approx 70% complete for ERR4231023.fastq.gz\n",
      "Approx 90% complete for ERR4231014.fastq.gz\n",
      "Approx 65% complete for ERR4231016.fastq.gz\n",
      "Approx 45% complete for ERR4231012.fastq.gz\n",
      "Approx 65% complete for ERR4231021.fastq.gz\n",
      "Approx 95% complete for ERR4231017.fastq.gz\n",
      "Approx 60% complete for ERR4231015.fastq.gz\n",
      "Approx 60% complete for ERR4231022.fastq.gz\n",
      "Approx 50% complete for ERR4231013.fastq.gz\n",
      "Approx 75% complete for ERR4231011.fastq.gz\n",
      "Approx 95% complete for ERR4231014.fastq.gz\n",
      "Approx 65% complete for ERR4231019.fastq.gz\n",
      "Approx 75% complete for ERR4231023.fastq.gz\n",
      "Analysis complete for ERR4231017.fastq.gz\n",
      "Approx 70% complete for ERR4231016.fastq.gz\n",
      "Approx 55% complete for ERR4231020.fastq.gz\n",
      "Approx 55% complete for ERR4231010.fastq.gz\n",
      "Approx 50% complete for ERR4231018.fastq.gz\n",
      "Approx 70% complete for ERR4231021.fastq.gz\n",
      "Approx 80% complete for ERR4231011.fastq.gz\n",
      "Analysis complete for ERR4231014.fastq.gz\n",
      "Approx 65% complete for ERR4231015.fastq.gz\n",
      "Approx 65% complete for ERR4231022.fastq.gz\n",
      "Approx 50% complete for ERR4231012.fastq.gz\n",
      "Approx 80% complete for ERR4231023.fastq.gz\n",
      "Approx 55% complete for ERR4231013.fastq.gz\n",
      "Approx 70% complete for ERR4231019.fastq.gz\n",
      "Approx 75% complete for ERR4231016.fastq.gz\n",
      "Approx 85% complete for ERR4231011.fastq.gz\n",
      "Approx 75% complete for ERR4231021.fastq.gz\n",
      "Approx 60% complete for ERR4231020.fastq.gz\n",
      "Approx 60% complete for ERR4231010.fastq.gz\n",
      "Approx 55% complete for ERR4231018.fastq.gz\n",
      "Approx 70% complete for ERR4231015.fastq.gz\n",
      "Approx 70% complete for ERR4231022.fastq.gz\n",
      "Approx 85% complete for ERR4231023.fastq.gz\n",
      "Approx 75% complete for ERR4231019.fastq.gz\n",
      "Approx 55% complete for ERR4231012.fastq.gz\n",
      "Approx 80% complete for ERR4231016.fastq.gz\n",
      "Approx 90% complete for ERR4231011.fastq.gz\n",
      "Approx 60% complete for ERR4231013.fastq.gz\n",
      "Approx 80% complete for ERR4231021.fastq.gz\n",
      "Approx 90% complete for ERR4231023.fastq.gz\n",
      "Approx 65% complete for ERR4231020.fastq.gz\n",
      "Approx 75% complete for ERR4231015.fastq.gz\n",
      "Approx 65% complete for ERR4231010.fastq.gz\n",
      "Approx 75% complete for ERR4231022.fastq.gz\n",
      "Approx 60% complete for ERR4231018.fastq.gz\n",
      "Approx 80% complete for ERR4231019.fastq.gz\n",
      "Approx 95% complete for ERR4231011.fastq.gz\n",
      "Approx 85% complete for ERR4231016.fastq.gz\n",
      "Approx 85% complete for ERR4231021.fastq.gz\n",
      "Approx 65% complete for ERR4231013.fastq.gz\n",
      "Approx 60% complete for ERR4231012.fastq.gz\n",
      "Approx 95% complete for ERR4231023.fastq.gz\n",
      "Approx 80% complete for ERR4231015.fastq.gz\n",
      "Approx 80% complete for ERR4231022.fastq.gz\n",
      "Approx 70% complete for ERR4231020.fastq.gz\n",
      "Analysis complete for ERR4231011.fastq.gz\n",
      "Approx 70% complete for ERR4231010.fastq.gz\n",
      "Approx 85% complete for ERR4231019.fastq.gz\n",
      "Approx 90% complete for ERR4231016.fastq.gz\n",
      "Approx 65% complete for ERR4231018.fastq.gz\n",
      "Analysis complete for ERR4231023.fastq.gz\n",
      "Approx 90% complete for ERR4231021.fastq.gz\n",
      "Approx 70% complete for ERR4231013.fastq.gz\n",
      "Approx 85% complete for ERR4231015.fastq.gz\n",
      "Approx 85% complete for ERR4231022.fastq.gz\n",
      "Approx 65% complete for ERR4231012.fastq.gz\n",
      "Approx 95% complete for ERR4231016.fastq.gz\n",
      "Approx 75% complete for ERR4231020.fastq.gz\n",
      "Approx 90% complete for ERR4231019.fastq.gz\n",
      "Approx 75% complete for ERR4231010.fastq.gz\n",
      "Approx 70% complete for ERR4231018.fastq.gz\n",
      "Approx 95% complete for ERR4231021.fastq.gz\n",
      "Approx 90% complete for ERR4231015.fastq.gz\n",
      "Approx 75% complete for ERR4231013.fastq.gz\n",
      "Approx 90% complete for ERR4231022.fastq.gz\n",
      "Analysis complete for ERR4231016.fastq.gz\n",
      "Approx 95% complete for ERR4231019.fastq.gz\n",
      "Approx 80% complete for ERR4231020.fastq.gz\n",
      "Approx 70% complete for ERR4231012.fastq.gz\n",
      "Approx 80% complete for ERR4231010.fastq.gz\n",
      "Analysis complete for ERR4231021.fastq.gz\n",
      "Approx 75% complete for ERR4231018.fastq.gz\n",
      "Approx 95% complete for ERR4231015.fastq.gz\n",
      "Approx 95% complete for ERR4231022.fastq.gz\n",
      "Approx 100% complete for ERR4231019.fastq.gz\n",
      "Analysis complete for ERR4231019.fastq.gz\n",
      "Approx 80% complete for ERR4231013.fastq.gz\n",
      "Approx 85% complete for ERR4231020.fastq.gz\n",
      "Approx 85% complete for ERR4231010.fastq.gz\n",
      "Approx 75% complete for ERR4231012.fastq.gz\n",
      "Analysis complete for ERR4231015.fastq.gz\n",
      "Analysis complete for ERR4231022.fastq.gz\n",
      "Approx 80% complete for ERR4231018.fastq.gz\n",
      "Approx 85% complete for ERR4231013.fastq.gz\n",
      "Approx 90% complete for ERR4231020.fastq.gz\n",
      "Approx 90% complete for ERR4231010.fastq.gz\n",
      "Approx 80% complete for ERR4231012.fastq.gz\n",
      "Approx 85% complete for ERR4231018.fastq.gz\n",
      "Approx 90% complete for ERR4231013.fastq.gz\n",
      "Approx 95% complete for ERR4231020.fastq.gz\n",
      "Approx 95% complete for ERR4231010.fastq.gz\n",
      "Approx 85% complete for ERR4231012.fastq.gz\n",
      "Approx 90% complete for ERR4231018.fastq.gz\n",
      "Analysis complete for ERR4231020.fastq.gz\n",
      "Approx 95% complete for ERR4231013.fastq.gz\n",
      "Analysis complete for ERR4231010.fastq.gz\n",
      "Approx 90% complete for ERR4231012.fastq.gz\n",
      "Approx 95% complete for ERR4231018.fastq.gz\n",
      "Analysis complete for ERR4231013.fastq.gz\n",
      "Analysis complete for ERR4231018.fastq.gz\n",
      "Approx 95% complete for ERR4231012.fastq.gz\n",
      "Analysis complete for ERR4231012.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "# !mkdir fastq_placenta/placenta_qc_trimmed\n",
    "!fastqc -t 16 fastq_placenta/trimmed_placenta/*.fastq.gz -o fastq_placenta/placenta_qc_trimmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.001s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /dev/cpuset.\n",
      "application/gzip\n",
      "Started analysis of ERR4231020_polya.fastq.gz\n",
      "Approx 5% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 10% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 15% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 20% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 25% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 30% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 35% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 40% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 45% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 50% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 55% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 60% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 65% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 70% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 75% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 80% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 85% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 90% complete for ERR4231020_polya.fastq.gz\n",
      "Approx 95% complete for ERR4231020_polya.fastq.gz\n",
      "Analysis complete for ERR4231020_polya.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "# !fastqc -t 16 fastq_placenta/trimmed_placenta/ERR4231020_polya.fastq.gz -o fastq_placenta/placenta_qc_trimmed"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run QC on the trimmed/filtered reads and generate an aggregated report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=392783;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2m| v1.17\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Prepending directory to sample names\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/mccallke0364/BIOL343/5_fastq/fastq_placenta\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m85/85\u001b[0m  \u001b[0m0m  \n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 28 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc_report_placenta_trim_compare.html   (overwritten)\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc_report_placenta_trim_compare_data   (overwritten)\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "!multiqc --force -d fastq_placenta/ -n multiqc_report_placenta_trim_compare.html"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Report your observations on which QC metrics improved (or got worse):\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Snakemake Pipeline"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When running bioinformatics studies, it is wise to use notebooks (such as this one) to tie together code, decisions, and data visualization. The above task is something you might generate and report to a client or supervisor who might be interested in sequencing QC and what kind of pre-processing decisions you made prior to alignment. In addition, the preprocessing should be readily reproducible by all relevant stakeholders, so you should supplement notebooks with pipeline scripts. Throughout the semester, we will be creating snakemake pipelines to reproduce the published analysis that you've selected. Follow the below steps to begin producing this pipeline:\n",
    "\n",
    "1. Make a new `Snakefile` and `config.yaml` in `5_fastq`. \n",
    "2. In the `Snakefile`, produce the following rules:  \n",
    "    a. `fetch_fastq` - use `parallel-fastq-dump` to fetch every FASTQ file listed in `config.yaml`.  \n",
    "    b. `fastq_qc` - use `fastqc` to run QC on the raw FASTQ files.  \n",
    "    c. `trim_filter` - use `cutadapt` to trim and/or filter the FASTQ reads using parameters justified in this notebook. ***Note:*** if you choose to trim/filter, it should be done on *all* of the FASTQ files; do not cherry pick.  \n",
    "    d. `trim_qc` - use `fastqc` to run QC on the trimmed FASTQ files.  \n",
    "    e. `report` - use `multiqc` to aggregate all QC data into a single report.\n",
    "\n",
    "This `Snakefile` should not use any `wrappers` because we already have the relevant software installed in the `biol343` conda environment. The pipeline should run to completion when any instructor or classmate runs `snakemake --use-conda`. You may use the documentation found [here](https://multiqc.info/docs/usage/pipelines/#snakemake) or any other online documentation you may find. You may not use any AI tools to complete this homework."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "# pd.read_csv(\"SraRunTable_Placenta.csv\", index_col=0).sort_index()[\"Individual\"].to_csv(\"a.csv\", sep=\"\\t\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biol343",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
