{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 5 Homework\n",
    "\n",
    "***Due (pushed to your GitHub branch) on 10/17 by 11:59 pm***"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTQ fetch, QC, and trimming/filtering\n",
    "\n",
    "Use the SRA-toolkit to fetch the relevant FASTQ files. You may use a hand-crafted `txt` file to provide the list of SRAs instead of manipulating an SRA Run Table with `cut`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run QC on the FASTQ files and aggregate the results into a report."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir mouse_hippocampus_fastq\n",
    "!while IFS= read -r line; do \\\n",
    "    echo \"Getting $line from NCBI SRA\"; \\\n",
    "    parallel-fastq-dump --sra-id $line --threads 32 --outdir mouse_hippocampus_fastq --gzip; \\\n",
    "    done < mouse_hippocampus_list.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.141s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /dev/cpuset.\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR2916027.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR2916028.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR2916029.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR2916030.fastq.gz\n",
      "application/gzip\n",
      "Started analysis of SRR2916031.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR2916032.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR2916033.fastq.gz\n",
      "application/gzip\n",
      "Started analysis of SRR2916034.fastq.gz\n",
      "application/gzip\n",
      "Started analysis of SRR2916035.fastq.gz\n",
      "application/gzip\n",
      "Started analysis of SRR2916036.fastq.gz\n",
      "application/gzip\n",
      "Started analysis of SRR2916037.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR2916038.fastq.gz\n",
      "application/gzip\n",
      "Started analysis of SRR2916039.fastq.gz\n",
      "application/gzip\n",
      "Started analysis of SRR2916040.fastq.gz\n",
      "Started analysis of SRR2916041.fastq.gz\n",
      "Started analysis of SRR2916042.fastq.gz\n",
      "Started analysis of SRR2916043.fastq.gz\n",
      "Started analysis of SRR2916044.fastq.gz\n",
      "Started analysis of SRR2916045.fastq.gz\n",
      "Started analysis of SRR2916046.fastq.gz\n",
      "Started analysis of SRR2916047.fastq.gz\n",
      "Started analysis of SRR2916048.fastq.gz\n",
      "Started analysis of SRR2916049.fastq.gz\n",
      "Approx 5% complete for SRR2916049.fastq.gz\n",
      "Approx 5% complete for SRR2916030.fastq.gz\n",
      "Approx 5% complete for SRR2916031.fastq.gz\n",
      "Approx 5% complete for SRR2916032.fastq.gz\n",
      "Approx 5% complete for SRR2916048.fastq.gz\n",
      "Approx 10% complete for SRR2916049.fastq.gz\n",
      "Approx 5% complete for SRR2916027.fastq.gz\n",
      "Approx 5% complete for SRR2916029.fastq.gz\n",
      "Approx 5% complete for SRR2916028.fastq.gz\n",
      "Approx 15% complete for SRR2916049.fastq.gz\n",
      "Approx 10% complete for SRR2916031.fastq.gz\n",
      "Approx 10% complete for SRR2916032.fastq.gz\n",
      "Approx 10% complete for SRR2916030.fastq.gz\n",
      "Approx 10% complete for SRR2916048.fastq.gz\n",
      "Approx 5% complete for SRR2916036.fastq.gz\n",
      "Approx 5% complete for SRR2916035.fastq.gz\n",
      "Approx 20% complete for SRR2916049.fastq.gz\n",
      "Approx 5% complete for SRR2916034.fastq.gz\n",
      "Approx 5% complete for SRR2916038.fastq.gz\n",
      "Approx 5% complete for SRR2916039.fastq.gz\n",
      "Approx 5% complete for SRR2916040.fastq.gz\n",
      "Approx 5% complete for SRR2916037.fastq.gz\n",
      "Approx 5% complete for SRR2916033.fastq.gz\n",
      "Approx 5% complete for SRR2916047.fastq.gz\n",
      "Approx 5% complete for SRR2916043.fastq.gz\n",
      "Approx 5% complete for SRR2916044.fastq.gz\n",
      "Approx 5% complete for SRR2916042.fastq.gz\n",
      "Approx 25% complete for SRR2916049.fastq.gz\n",
      "Approx 15% complete for SRR2916048.fastq.gz\n",
      "Approx 5% complete for SRR2916045.fastq.gz\n",
      "Approx 5% complete for SRR2916041.fastq.gz\n",
      "Approx 15% complete for SRR2916032.fastq.gz\n",
      "Approx 15% complete for SRR2916031.fastq.gz\n",
      "Approx 15% complete for SRR2916030.fastq.gz\n",
      "Approx 5% complete for SRR2916046.fastq.gz\n",
      "Approx 10% complete for SRR2916027.fastq.gz\n",
      "Approx 10% complete for SRR2916029.fastq.gz\n",
      "Approx 30% complete for SRR2916049.fastq.gz\n",
      "Approx 20% complete for SRR2916048.fastq.gz\n",
      "Approx 10% complete for SRR2916028.fastq.gz\n",
      "Approx 20% complete for SRR2916032.fastq.gz\n",
      "Approx 35% complete for SRR2916049.fastq.gz\n",
      "Approx 20% complete for SRR2916031.fastq.gz\n",
      "Approx 20% complete for SRR2916030.fastq.gz\n",
      "Approx 10% complete for SRR2916036.fastq.gz\n",
      "Approx 40% complete for SRR2916049.fastq.gz\n",
      "Approx 10% complete for SRR2916035.fastq.gz\n",
      "Approx 25% complete for SRR2916048.fastq.gz\n",
      "Approx 10% complete for SRR2916038.fastq.gz\n",
      "Approx 10% complete for SRR2916039.fastq.gz\n",
      "Approx 10% complete for SRR2916034.fastq.gz\n",
      "Approx 25% complete for SRR2916032.fastq.gz\n",
      "Approx 10% complete for SRR2916040.fastq.gz\n",
      "Approx 45% complete for SRR2916049.fastq.gz\n",
      "Approx 25% complete for SRR2916031.fastq.gz\n",
      "Approx 10% complete for SRR2916037.fastq.gz\n",
      "Approx 15% complete for SRR2916027.fastq.gz\n",
      "Approx 25% complete for SRR2916030.fastq.gz\n",
      "Approx 15% complete for SRR2916029.fastq.gz\n",
      "Approx 10% complete for SRR2916047.fastq.gz\n",
      "Approx 10% complete for SRR2916033.fastq.gz\n",
      "Approx 10% complete for SRR2916043.fastq.gz\n",
      "Approx 10% complete for SRR2916042.fastq.gz\n",
      "Approx 10% complete for SRR2916044.fastq.gz\n",
      "Approx 50% complete for SRR2916049.fastq.gz\n",
      "Approx 10% complete for SRR2916045.fastq.gz\n",
      "Approx 30% complete for SRR2916048.fastq.gz\n",
      "Approx 10% complete for SRR2916041.fastq.gz\n",
      "Approx 30% complete for SRR2916032.fastq.gz\n",
      "Approx 15% complete for SRR2916028.fastq.gz\n",
      "Approx 55% complete for SRR2916049.fastq.gz\n",
      "Approx 10% complete for SRR2916046.fastq.gz\n",
      "Approx 30% complete for SRR2916031.fastq.gz\n",
      "Approx 30% complete for SRR2916030.fastq.gz\n",
      "Approx 15% complete for SRR2916036.fastq.gz\n",
      "Approx 35% complete for SRR2916048.fastq.gz\n",
      "Approx 60% complete for SRR2916049.fastq.gz\n",
      "Approx 15% complete for SRR2916035.fastq.gz\n",
      "Approx 35% complete for SRR2916032.fastq.gz\n",
      "Approx 20% complete for SRR2916027.fastq.gz\n",
      "Approx 20% complete for SRR2916029.fastq.gz\n",
      "Approx 65% complete for SRR2916049.fastq.gz\n",
      "Approx 15% complete for SRR2916038.fastq.gz\n",
      "Approx 35% complete for SRR2916031.fastq.gz\n",
      "Approx 15% complete for SRR2916039.fastq.gz\n",
      "Approx 15% complete for SRR2916034.fastq.gz\n",
      "Approx 35% complete for SRR2916030.fastq.gz\n",
      "Approx 15% complete for SRR2916040.fastq.gz\n",
      "Approx 40% complete for SRR2916048.fastq.gz\n",
      "Approx 70% complete for SRR2916049.fastq.gz\n",
      "Approx 15% complete for SRR2916037.fastq.gz\n",
      "Approx 15% complete for SRR2916047.fastq.gz\n",
      "Approx 40% complete for SRR2916032.fastq.gz\n",
      "Approx 20% complete for SRR2916028.fastq.gz\n",
      "Approx 15% complete for SRR2916033.fastq.gz\n",
      "Approx 15% complete for SRR2916043.fastq.gz\n",
      "Approx 15% complete for SRR2916042.fastq.gz\n",
      "Approx 75% complete for SRR2916049.fastq.gz\n",
      "Approx 15% complete for SRR2916044.fastq.gz\n",
      "Approx 40% complete for SRR2916031.fastq.gz\n",
      "Approx 15% complete for SRR2916045.fastq.gz\n",
      "Approx 45% complete for SRR2916048.fastq.gz\n",
      "Approx 15% complete for SRR2916041.fastq.gz\n",
      "Approx 40% complete for SRR2916030.fastq.gz\n",
      "Approx 20% complete for SRR2916036.fastq.gz\n",
      "Approx 80% complete for SRR2916049.fastq.gz\n",
      "Approx 25% complete for SRR2916027.fastq.gz\n",
      "Approx 45% complete for SRR2916032.fastq.gz\n",
      "Approx 25% complete for SRR2916029.fastq.gz\n",
      "Approx 15% complete for SRR2916046.fastq.gz\n",
      "Approx 20% complete for SRR2916035.fastq.gz\n",
      "Approx 85% complete for SRR2916049.fastq.gz\n",
      "Approx 50% complete for SRR2916048.fastq.gz\n",
      "Approx 45% complete for SRR2916031.fastq.gz\n",
      "Approx 20% complete for SRR2916038.fastq.gz\n",
      "Approx 45% complete for SRR2916030.fastq.gz\n",
      "Approx 20% complete for SRR2916039.fastq.gz\n",
      "Approx 90% complete for SRR2916049.fastq.gz\n",
      "Approx 20% complete for SRR2916034.fastq.gz\n",
      "Approx 20% complete for SRR2916040.fastq.gz\n",
      "Approx 50% complete for SRR2916032.fastq.gz\n",
      "Approx 25% complete for SRR2916028.fastq.gz\n",
      "Approx 55% complete for SRR2916048.fastq.gz\n",
      "Approx 20% complete for SRR2916037.fastq.gz\n",
      "Approx 95% complete for SRR2916049.fastq.gz\n",
      "Approx 20% complete for SRR2916047.fastq.gz\n",
      "Approx 50% complete for SRR2916031.fastq.gz\n",
      "Approx 30% complete for SRR2916027.fastq.gz\n",
      "Approx 20% complete for SRR2916043.fastq.gz\n",
      "Approx 20% complete for SRR2916044.fastq.gz\n",
      "Approx 20% complete for SRR2916033.fastq.gz\n",
      "Approx 50% complete for SRR2916030.fastq.gz\n",
      "Approx 20% complete for SRR2916042.fastq.gz\n",
      "Approx 30% complete for SRR2916029.fastq.gz\n",
      "Approx 25% complete for SRR2916036.fastq.gz\n",
      "Analysis complete for SRR2916049.fastq.gz\n",
      "Approx 20% complete for SRR2916045.fastq.gz\n",
      "Approx 55% complete for SRR2916032.fastq.gz\n",
      "Approx 60% complete for SRR2916048.fastq.gz\n",
      "Approx 20% complete for SRR2916041.fastq.gz\n",
      "Approx 30% complete for SRR2916028.fastq.gz\n",
      "Approx 25% complete for SRR2916035.fastq.gz\n",
      "Approx 35% complete for SRR2916027.fastq.gz\n",
      "Approx 55% complete for SRR2916031.fastq.gz\n",
      "Approx 25% complete for SRR2916038.fastq.gz\n",
      "Approx 20% complete for SRR2916046.fastq.gz\n",
      "Approx 55% complete for SRR2916030.fastq.gz\n",
      "Approx 65% complete for SRR2916048.fastq.gz\n",
      "Approx 60% complete for SRR2916032.fastq.gz\n",
      "Approx 25% complete for SRR2916039.fastq.gz\n",
      "Approx 25% complete for SRR2916040.fastq.gz\n",
      "Approx 25% complete for SRR2916034.fastq.gz\n",
      "Approx 40% complete for SRR2916027.fastq.gz\n",
      "Approx 35% complete for SRR2916028.fastq.gz\n",
      "Approx 35% complete for SRR2916029.fastq.gz\n",
      "Approx 60% complete for SRR2916031.fastq.gz\n",
      "Approx 25% complete for SRR2916037.fastq.gz\n",
      "Approx 25% complete for SRR2916047.fastq.gz\n",
      "Approx 70% complete for SRR2916048.fastq.gz\n",
      "Approx 30% complete for SRR2916036.fastq.gz\n",
      "Approx 60% complete for SRR2916030.fastq.gz\n",
      "Approx 25% complete for SRR2916043.fastq.gz\n",
      "Approx 65% complete for SRR2916032.fastq.gz\n",
      "Approx 25% complete for SRR2916044.fastq.gz\n",
      "Approx 45% complete for SRR2916027.fastq.gz\n",
      "Approx 25% complete for SRR2916033.fastq.gz\n",
      "Approx 25% complete for SRR2916042.fastq.gz\n",
      "Approx 25% complete for SRR2916045.fastq.gz\n",
      "Approx 40% complete for SRR2916028.fastq.gz\n",
      "Approx 30% complete for SRR2916035.fastq.gz\n",
      "Approx 75% complete for SRR2916048.fastq.gz\n",
      "Approx 65% complete for SRR2916031.fastq.gz\n",
      "Approx 25% complete for SRR2916041.fastq.gz\n",
      "Approx 65% complete for SRR2916030.fastq.gz\n",
      "Approx 70% complete for SRR2916032.fastq.gz\n",
      "Approx 30% complete for SRR2916038.fastq.gz\n",
      "Approx 40% complete for SRR2916029.fastq.gz\n",
      "Approx 80% complete for SRR2916048.fastq.gz\n",
      "Approx 30% complete for SRR2916039.fastq.gz\n",
      "Approx 30% complete for SRR2916040.fastq.gz\n",
      "Approx 50% complete for SRR2916027.fastq.gz\n",
      "Approx 30% complete for SRR2916034.fastq.gz\n",
      "Approx 25% complete for SRR2916046.fastq.gz\n",
      "Approx 70% complete for SRR2916031.fastq.gz\n",
      "Approx 35% complete for SRR2916036.fastq.gz\n",
      "Approx 30% complete for SRR2916037.fastq.gz\n",
      "Approx 75% complete for SRR2916032.fastq.gz\n",
      "Approx 30% complete for SRR2916047.fastq.gz\n",
      "Approx 70% complete for SRR2916030.fastq.gz\n",
      "Approx 85% complete for SRR2916048.fastq.gz\n",
      "Approx 45% complete for SRR2916028.fastq.gz\n",
      "Approx 30% complete for SRR2916043.fastq.gz\n",
      "Approx 30% complete for SRR2916044.fastq.gz\n",
      "Approx 35% complete for SRR2916035.fastq.gz\n",
      "Approx 30% complete for SRR2916033.fastq.gz\n",
      "Approx 75% complete for SRR2916031.fastq.gz\n",
      "Approx 30% complete for SRR2916042.fastq.gz\n",
      "Approx 90% complete for SRR2916048.fastq.gz\n",
      "Approx 80% complete for SRR2916032.fastq.gz\n",
      "Approx 45% complete for SRR2916029.fastq.gz\n",
      "Approx 30% complete for SRR2916045.fastq.gz\n",
      "Approx 75% complete for SRR2916030.fastq.gz\n",
      "Approx 35% complete for SRR2916038.fastq.gz\n",
      "Approx 55% complete for SRR2916027.fastq.gz\n",
      "Approx 30% complete for SRR2916041.fastq.gz\n",
      "Approx 80% complete for SRR2916031.fastq.gz\n",
      "Approx 95% complete for SRR2916048.fastq.gz\n",
      "Approx 35% complete for SRR2916039.fastq.gz\n",
      "Approx 35% complete for SRR2916040.fastq.gz\n",
      "Approx 40% complete for SRR2916036.fastq.gz\n",
      "Approx 85% complete for SRR2916032.fastq.gz\n",
      "Approx 35% complete for SRR2916034.fastq.gz\n",
      "Approx 80% complete for SRR2916030.fastq.gz\n",
      "Approx 35% complete for SRR2916047.fastq.gz\n",
      "Approx 50% complete for SRR2916028.fastq.gz\n",
      "Approx 35% complete for SRR2916037.fastq.gz\n",
      "Approx 30% complete for SRR2916046.fastq.gz\n",
      "Analysis complete for SRR2916048.fastq.gz\n",
      "Approx 85% complete for SRR2916031.fastq.gz\n",
      "Approx 40% complete for SRR2916035.fastq.gz\n",
      "Approx 50% complete for SRR2916029.fastq.gz\n",
      "Approx 35% complete for SRR2916043.fastq.gz\n",
      "Approx 90% complete for SRR2916032.fastq.gz\n",
      "Approx 60% complete for SRR2916027.fastq.gz\n",
      "Approx 85% complete for SRR2916030.fastq.gz\n",
      "Approx 35% complete for SRR2916044.fastq.gz\n",
      "Approx 35% complete for SRR2916033.fastq.gz\n",
      "Approx 35% complete for SRR2916042.fastq.gz\n",
      "Approx 40% complete for SRR2916038.fastq.gz\n",
      "Approx 90% complete for SRR2916031.fastq.gz\n",
      "Approx 35% complete for SRR2916045.fastq.gz\n",
      "Approx 45% complete for SRR2916036.fastq.gz\n",
      "Approx 95% complete for SRR2916032.fastq.gz\n",
      "Approx 40% complete for SRR2916039.fastq.gz\n",
      "Approx 40% complete for SRR2916040.fastq.gz\n",
      "Approx 90% complete for SRR2916030.fastq.gz\n",
      "Approx 55% complete for SRR2916028.fastq.gz\n",
      "Approx 35% complete for SRR2916041.fastq.gz\n",
      "Approx 40% complete for SRR2916034.fastq.gz\n",
      "Approx 55% complete for SRR2916029.fastq.gz\n",
      "Approx 95% complete for SRR2916031.fastq.gz\n",
      "Approx 40% complete for SRR2916047.fastq.gz\n",
      "Approx 40% complete for SRR2916037.fastq.gz\n",
      "Approx 65% complete for SRR2916027.fastq.gz\n",
      "Approx 45% complete for SRR2916035.fastq.gz\n",
      "Analysis complete for SRR2916032.fastq.gz\n",
      "Approx 95% complete for SRR2916030.fastq.gz\n",
      "Approx 35% complete for SRR2916046.fastq.gz\n",
      "Approx 40% complete for SRR2916043.fastq.gz\n",
      "Analysis complete for SRR2916031.fastq.gz\n",
      "Approx 40% complete for SRR2916044.fastq.gz\n",
      "Approx 40% complete for SRR2916033.fastq.gz\n",
      "Approx 50% complete for SRR2916036.fastq.gz\n",
      "Approx 45% complete for SRR2916038.fastq.gz\n",
      "Approx 40% complete for SRR2916042.fastq.gz\n",
      "Analysis complete for SRR2916030.fastq.gz\n",
      "Approx 40% complete for SRR2916045.fastq.gz\n",
      "Approx 60% complete for SRR2916029.fastq.gz\n",
      "Approx 60% complete for SRR2916028.fastq.gz\n",
      "Approx 45% complete for SRR2916039.fastq.gz\n",
      "Approx 45% complete for SRR2916040.fastq.gz\n",
      "Approx 70% complete for SRR2916027.fastq.gz\n",
      "Approx 45% complete for SRR2916034.fastq.gz\n",
      "Approx 50% complete for SRR2916035.fastq.gz\n",
      "Approx 45% complete for SRR2916047.fastq.gz\n",
      "Approx 45% complete for SRR2916037.fastq.gz\n",
      "Approx 40% complete for SRR2916041.fastq.gz\n",
      "Approx 55% complete for SRR2916036.fastq.gz\n",
      "Approx 65% complete for SRR2916029.fastq.gz\n",
      "Approx 45% complete for SRR2916043.fastq.gz\n",
      "Approx 50% complete for SRR2916038.fastq.gz\n",
      "Approx 45% complete for SRR2916033.fastq.gz\n",
      "Approx 45% complete for SRR2916044.fastq.gz\n",
      "Approx 75% complete for SRR2916027.fastq.gz\n",
      "Approx 65% complete for SRR2916028.fastq.gz\n",
      "Approx 40% complete for SRR2916046.fastq.gz\n",
      "Approx 45% complete for SRR2916042.fastq.gz\n",
      "Approx 45% complete for SRR2916045.fastq.gz\n",
      "Approx 50% complete for SRR2916039.fastq.gz\n",
      "Approx 50% complete for SRR2916040.fastq.gz\n",
      "Approx 55% complete for SRR2916035.fastq.gz\n",
      "Approx 50% complete for SRR2916034.fastq.gz\n",
      "Approx 50% complete for SRR2916047.fastq.gz\n",
      "Approx 50% complete for SRR2916037.fastq.gz\n",
      "Approx 70% complete for SRR2916029.fastq.gz\n",
      "Approx 60% complete for SRR2916036.fastq.gz\n",
      "Approx 45% complete for SRR2916041.fastq.gz\n",
      "Approx 80% complete for SRR2916027.fastq.gz\n",
      "Approx 70% complete for SRR2916028.fastq.gz\n",
      "Approx 55% complete for SRR2916038.fastq.gz\n",
      "Approx 50% complete for SRR2916043.fastq.gz\n",
      "Approx 50% complete for SRR2916033.fastq.gz\n",
      "Approx 50% complete for SRR2916044.fastq.gz\n",
      "Approx 55% complete for SRR2916039.fastq.gz\n",
      "Approx 55% complete for SRR2916040.fastq.gz\n",
      "Approx 60% complete for SRR2916035.fastq.gz\n",
      "Approx 50% complete for SRR2916042.fastq.gz\n",
      "Approx 45% complete for SRR2916046.fastq.gz\n",
      "Approx 50% complete for SRR2916045.fastq.gz\n",
      "Approx 75% complete for SRR2916029.fastq.gz\n",
      "Approx 55% complete for SRR2916034.fastq.gz\n",
      "Approx 55% complete for SRR2916037.fastq.gz\n",
      "Approx 85% complete for SRR2916027.fastq.gz\n",
      "Approx 55% complete for SRR2916047.fastq.gz\n",
      "Approx 65% complete for SRR2916036.fastq.gz\n",
      "Approx 75% complete for SRR2916028.fastq.gz\n",
      "Approx 60% complete for SRR2916038.fastq.gz\n",
      "Approx 50% complete for SRR2916041.fastq.gz\n",
      "Approx 55% complete for SRR2916043.fastq.gz\n",
      "Approx 80% complete for SRR2916029.fastq.gz\n",
      "Approx 55% complete for SRR2916033.fastq.gz\n",
      "Approx 55% complete for SRR2916044.fastq.gz\n",
      "Approx 65% complete for SRR2916035.fastq.gz\n",
      "Approx 90% complete for SRR2916027.fastq.gz\n",
      "Approx 60% complete for SRR2916039.fastq.gz\n",
      "Approx 60% complete for SRR2916040.fastq.gz\n",
      "Approx 70% complete for SRR2916036.fastq.gz\n",
      "Approx 55% complete for SRR2916042.fastq.gz\n",
      "Approx 60% complete for SRR2916037.fastq.gz\n",
      "Approx 55% complete for SRR2916045.fastq.gz\n",
      "Approx 60% complete for SRR2916034.fastq.gz\n",
      "Approx 60% complete for SRR2916047.fastq.gz\n",
      "Approx 50% complete for SRR2916046.fastq.gz\n",
      "Approx 80% complete for SRR2916028.fastq.gz\n",
      "Approx 65% complete for SRR2916038.fastq.gz\n",
      "Approx 85% complete for SRR2916029.fastq.gz\n",
      "Approx 95% complete for SRR2916027.fastq.gz\n",
      "Approx 55% complete for SRR2916041.fastq.gz\n",
      "Approx 70% complete for SRR2916035.fastq.gz\n",
      "Approx 60% complete for SRR2916043.fastq.gz\n",
      "Approx 75% complete for SRR2916036.fastq.gz\n",
      "Approx 60% complete for SRR2916033.fastq.gz\n",
      "Approx 65% complete for SRR2916040.fastq.gz\n",
      "Approx 65% complete for SRR2916039.fastq.gz\n",
      "Approx 60% complete for SRR2916044.fastq.gz\n",
      "Approx 65% complete for SRR2916037.fastq.gz\n",
      "Approx 85% complete for SRR2916028.fastq.gz\n",
      "Approx 60% complete for SRR2916042.fastq.gz\n",
      "Approx 65% complete for SRR2916047.fastq.gz\n",
      "Approx 65% complete for SRR2916034.fastq.gz\n",
      "Approx 60% complete for SRR2916045.fastq.gz\n",
      "Approx 90% complete for SRR2916029.fastq.gz\n",
      "Analysis complete for SRR2916027.fastq.gz\n",
      "Approx 55% complete for SRR2916046.fastq.gz\n",
      "Approx 70% complete for SRR2916038.fastq.gz\n",
      "Approx 80% complete for SRR2916036.fastq.gz\n",
      "Approx 75% complete for SRR2916035.fastq.gz\n",
      "Approx 60% complete for SRR2916041.fastq.gz\n",
      "Approx 70% complete for SRR2916039.fastq.gz\n",
      "Approx 70% complete for SRR2916040.fastq.gz\n",
      "Approx 65% complete for SRR2916043.fastq.gz\n",
      "Approx 90% complete for SRR2916028.fastq.gz\n",
      "Approx 65% complete for SRR2916033.fastq.gz\n",
      "Approx 95% complete for SRR2916029.fastq.gz\n",
      "Approx 65% complete for SRR2916044.fastq.gz\n",
      "Approx 70% complete for SRR2916037.fastq.gz\n",
      "Approx 70% complete for SRR2916047.fastq.gz\n",
      "Approx 70% complete for SRR2916034.fastq.gz\n",
      "Approx 65% complete for SRR2916042.fastq.gz\n",
      "Approx 65% complete for SRR2916045.fastq.gz\n",
      "Approx 85% complete for SRR2916036.fastq.gz\n",
      "Approx 75% complete for SRR2916038.fastq.gz\n",
      "Approx 80% complete for SRR2916035.fastq.gz\n",
      "Approx 60% complete for SRR2916046.fastq.gz\n",
      "Analysis complete for SRR2916029.fastq.gz\n",
      "Approx 95% complete for SRR2916028.fastq.gz\n",
      "Approx 75% complete for SRR2916039.fastq.gz\n",
      "Approx 75% complete for SRR2916040.fastq.gz\n",
      "Approx 65% complete for SRR2916041.fastq.gz\n",
      "Approx 70% complete for SRR2916043.fastq.gz\n",
      "Approx 75% complete for SRR2916037.fastq.gz\n",
      "Approx 70% complete for SRR2916033.fastq.gz\n",
      "Approx 70% complete for SRR2916044.fastq.gz\n",
      "Approx 75% complete for SRR2916047.fastq.gz\n",
      "Approx 75% complete for SRR2916034.fastq.gz\n",
      "Approx 90% complete for SRR2916036.fastq.gz\n",
      "Approx 70% complete for SRR2916042.fastq.gz\n",
      "Approx 85% complete for SRR2916035.fastq.gz\n",
      "Approx 80% complete for SRR2916038.fastq.gz\n",
      "Approx 70% complete for SRR2916045.fastq.gz\n",
      "Analysis complete for SRR2916028.fastq.gz\n",
      "Approx 80% complete for SRR2916039.fastq.gz\n",
      "Approx 80% complete for SRR2916040.fastq.gz\n",
      "Approx 80% complete for SRR2916037.fastq.gz\n",
      "Approx 65% complete for SRR2916046.fastq.gz\n",
      "Approx 75% complete for SRR2916043.fastq.gz\n",
      "Approx 75% complete for SRR2916033.fastq.gz\n",
      "Approx 70% complete for SRR2916041.fastq.gz\n",
      "Approx 95% complete for SRR2916036.fastq.gz\n",
      "Approx 75% complete for SRR2916044.fastq.gz\n",
      "Approx 80% complete for SRR2916047.fastq.gz\n",
      "Approx 80% complete for SRR2916034.fastq.gz\n",
      "Approx 90% complete for SRR2916035.fastq.gz\n",
      "Approx 75% complete for SRR2916042.fastq.gz\n",
      "Approx 85% complete for SRR2916038.fastq.gz\n",
      "Approx 75% complete for SRR2916045.fastq.gz\n",
      "Approx 85% complete for SRR2916039.fastq.gz\n",
      "Analysis complete for SRR2916036.fastq.gz\n",
      "Approx 85% complete for SRR2916037.fastq.gz\n",
      "Approx 85% complete for SRR2916040.fastq.gz\n",
      "Approx 80% complete for SRR2916033.fastq.gz\n",
      "Approx 80% complete for SRR2916043.fastq.gz\n",
      "Approx 95% complete for SRR2916035.fastq.gz\n",
      "Approx 70% complete for SRR2916046.fastq.gz\n",
      "Approx 85% complete for SRR2916047.fastq.gz\n",
      "Approx 80% complete for SRR2916044.fastq.gz\n",
      "Approx 75% complete for SRR2916041.fastq.gz\n",
      "Approx 85% complete for SRR2916034.fastq.gz\n",
      "Approx 90% complete for SRR2916038.fastq.gz\n",
      "Approx 80% complete for SRR2916042.fastq.gz\n",
      "Approx 80% complete for SRR2916045.fastq.gz\n",
      "Approx 90% complete for SRR2916039.fastq.gz\n",
      "Approx 90% complete for SRR2916037.fastq.gz\n",
      "Approx 90% complete for SRR2916040.fastq.gz\n",
      "Analysis complete for SRR2916035.fastq.gz\n",
      "Approx 85% complete for SRR2916033.fastq.gz\n",
      "Approx 85% complete for SRR2916043.fastq.gz\n",
      "Approx 90% complete for SRR2916047.fastq.gz\n",
      "Approx 85% complete for SRR2916044.fastq.gz\n",
      "Approx 90% complete for SRR2916034.fastq.gz\n",
      "Approx 80% complete for SRR2916041.fastq.gz\n",
      "Approx 95% complete for SRR2916038.fastq.gz\n",
      "Approx 75% complete for SRR2916046.fastq.gz\n",
      "Approx 85% complete for SRR2916042.fastq.gz\n",
      "Approx 95% complete for SRR2916037.fastq.gz\n",
      "Approx 95% complete for SRR2916039.fastq.gz\n",
      "Approx 85% complete for SRR2916045.fastq.gz\n",
      "Approx 95% complete for SRR2916040.fastq.gz\n",
      "Approx 90% complete for SRR2916033.fastq.gz\n",
      "Approx 90% complete for SRR2916043.fastq.gz\n",
      "Approx 95% complete for SRR2916047.fastq.gz\n",
      "Approx 95% complete for SRR2916034.fastq.gz\n",
      "Approx 90% complete for SRR2916044.fastq.gz\n",
      "Analysis complete for SRR2916038.fastq.gz\n",
      "Approx 85% complete for SRR2916041.fastq.gz\n",
      "Approx 90% complete for SRR2916042.fastq.gz\n",
      "Approx 80% complete for SRR2916046.fastq.gz\n",
      "Analysis complete for SRR2916037.fastq.gz\n",
      "Analysis complete for SRR2916039.fastq.gz\n",
      "Approx 90% complete for SRR2916045.fastq.gz\n",
      "Analysis complete for SRR2916040.fastq.gz\n",
      "Approx 95% complete for SRR2916033.fastq.gz\n",
      "Analysis complete for SRR2916047.fastq.gz\n",
      "Approx 95% complete for SRR2916043.fastq.gz\n",
      "Analysis complete for SRR2916034.fastq.gz\n",
      "Approx 95% complete for SRR2916044.fastq.gz\n",
      "Approx 90% complete for SRR2916041.fastq.gz\n",
      "Approx 95% complete for SRR2916042.fastq.gz\n",
      "Approx 85% complete for SRR2916046.fastq.gz\n",
      "Approx 95% complete for SRR2916045.fastq.gz\n",
      "Analysis complete for SRR2916033.fastq.gz\n",
      "Analysis complete for SRR2916043.fastq.gz\n",
      "Analysis complete for SRR2916044.fastq.gz\n",
      "Approx 95% complete for SRR2916041.fastq.gz\n",
      "Analysis complete for SRR2916042.fastq.gz\n",
      "Analysis complete for SRR2916045.fastq.gz\n",
      "Approx 90% complete for SRR2916046.fastq.gz\n",
      "Analysis complete for SRR2916041.fastq.gz\n",
      "Approx 95% complete for SRR2916046.fastq.gz\n",
      "Analysis complete for SRR2916046.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "!mkdir mouse_hippocampus_fastq/qc\n",
    "!fastqc -t 32 mouse_hippocampus_fastq/*.fastq.gz -o mouse_hippocampus_fastq/qc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                \n",
      " \u001b[91m///\u001b[0m \u001b]8;id=334057;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2m| v1.17\u001b[0m                                                         \n",
      "                                                                                \n",
      " \u001b[33mUsage:\u001b[0m \u001b[1mmultiqc\u001b[0m [\u001b[1;36mOPTIONS\u001b[0m] [\u001b[1;36mANALYSIS\u001b[0m \u001b[1;36mDIRECTORY\u001b[0m]                                  \n",
      "                                                                                \n",
      " MultiQC aggregates results from bioinformatics analyses across many samples    \n",
      " into a single report.                                                          \n",
      " \u001b[2mIt searches a given directory for analysis logs and compiles a HTML report. \u001b[0m   \n",
      " \u001b[2mIt's a general use tool, perfect for summarising the output from numerous \u001b[0m     \n",
      " \u001b[2mbioinformatics tools.\u001b[0m                                                          \n",
      " \u001b[2mTo run, supply with one or more directory to scan for analysis results. For \u001b[0m   \n",
      " \u001b[2mexample, to run in the current working directory, use '\u001b[0m\u001b[1;2;34mmultiqc .\u001b[0m\u001b[2m'\u001b[0m              \n",
      "                                                                                \n",
      "\u001b[2m╭─\u001b[0m\u001b[2m Main options \u001b[0m\u001b[2m──────────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--force\u001b[0m            \u001b[1;32m-f\u001b[0m  Overwrite any existing reports                        \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--config\u001b[0m           \u001b[1;32m-c\u001b[0m  Specific config file to load, after those in MultiQC  \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        dir / home dir / working dir.                         \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--cl-config\u001b[0m            Specify MultiQC config YAML on the command line       \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(TEXT)                                         \u001b[0m       \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--filename\u001b[0m         \u001b[1;32m-n\u001b[0m  Report filename. Use '\u001b[3;33mstdout\u001b[0m' to print to standard    \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        out.                                                  \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--outdir\u001b[0m           \u001b[1;32m-o\u001b[0m  Create report in the specified output directory.      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(TEXT)                                          \u001b[0m      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--ignore\u001b[0m           \u001b[1;32m-x\u001b[0m  Ignore analysis files \u001b[2;33m(GLOB EXPRESSION)\u001b[0m               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--ignore-samples\u001b[0m       Ignore sample names \u001b[2;33m(GLOB EXPRESSION)\u001b[0m                 \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--ignore-symlinks\u001b[0m      Ignore symlinked directories and files                \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--file-list\u001b[0m        \u001b[1;32m-l\u001b[0m  Supply a file containing a list of file paths to be   \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        searched, one per row                                 \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "\u001b[2m╭─\u001b[0m\u001b[2m Choosing modules to run \u001b[0m\u001b[2m───────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--module\u001b[0m     \u001b[1;32m-m\u001b[0m  Use only this module. Can specify multiple times.           \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                  \u001b[2;33m(MODULE NAME)                                    \u001b[0m           \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--exclude\u001b[0m    \u001b[1;32m-e\u001b[0m  Do not use this module. Can specify multiple times.         \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                  \u001b[2;33m(MODULE NAME)                                      \u001b[0m         \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--tag\u001b[0m            Use only modules which tagged with this keyword \u001b[2;33m(TEXT)\u001b[0m      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--view-tags\u001b[0m      View the available tags and which modules they load         \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "\u001b[2m╭─\u001b[0m\u001b[2m Sample handling \u001b[0m\u001b[2m───────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--dirs\u001b[0m           \u001b[1;32m-d\u001b[0m   Prepend directory to sample names                      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--dirs-depth\u001b[0m     \u001b[1;32m-dd\u001b[0m  Prepend \u001b[3;33mn\u001b[0m directories to sample names. Negative number \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                       to take from start of path.                            \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                       \u001b[2;33m(INTEGER)                                             \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--fullnames\u001b[0m      \u001b[1;32m-s\u001b[0m   Do not clean the sample names \u001b[3m(leave as full file \u001b[0m     \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                       \u001b[3mname)\u001b[0m                                                  \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--fn_as_s_name\u001b[0m        Use the log filename as the sample name                \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--replace-names\u001b[0m       TSV file to rename sample names during report          \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                       generation                                             \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                       \u001b[2;33m(PATH)                                                \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "\u001b[2m╭─\u001b[0m\u001b[2m Report customisation \u001b[0m\u001b[2m──────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--title\u001b[0m            \u001b[1;32m-i\u001b[0m  Report title. Printed as page header, used for        \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        filename if not otherwise specified.                  \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--comment\u001b[0m          \u001b[1;32m-b\u001b[0m  Custom comment, will be printed at the top of the     \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        report.                                               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--template\u001b[0m         \u001b[1;32m-t\u001b[0m  Report template to use.                               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(default|default_dev|gathered|geo|sections|simple)\u001b[0m    \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--sample-names\u001b[0m         TSV file containing alternative sample names for      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        renaming buttons in the report                        \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--sample-filters\u001b[0m       TSV file containing show/hide patterns for the report \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--custom-css-file\u001b[0m      Custom CSS file to add to the final report \u001b[2;33m(PATH)\u001b[0m     \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "\u001b[2m╭─\u001b[0m\u001b[2m Output files \u001b[0m\u001b[2m──────────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--flat\u001b[0m          \u001b[1;32m-fp\u001b[0m  Use only flat plots \u001b[3m(static images)\u001b[0m                     \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--interactive\u001b[0m   \u001b[1;32m-ip\u001b[0m  Use only interactive plots \u001b[3m(in-browser Javascript)\u001b[0m      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--export\u001b[0m        \u001b[1;32m-p\u001b[0m   Export plots as static images in addition to the report \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--data-dir\u001b[0m           Force the parsed data directory to be created.          \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--no-data-dir\u001b[0m        Prevent the parsed data directory from being created.   \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--data-format\u001b[0m   \u001b[1;32m-k\u001b[0m   Output parsed data in a different format.               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                      \u001b[2;33m(tsv|json|yaml)                          \u001b[0m               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--zip-data-dir\u001b[0m  \u001b[1;32m-z\u001b[0m   Compress the data directory.                            \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--no-report\u001b[0m          Do not generate a report, only export data and plots    \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--pdf\u001b[0m                Creates PDF report with the \u001b[3m'simple'\u001b[0m template. Requires \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                      \u001b]8;id=520893;https://pandoc.org/\u001b\\Pandoc\u001b]8;;\u001b\\ to be installed.                                 \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "\u001b[2m╭─\u001b[0m\u001b[2m MultiQC behaviour \u001b[0m\u001b[2m─────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--verbose\u001b[0m           \u001b[1;32m-v\u001b[0m  Increase output verbosity. \u001b[2;33m(INTEGER RANGE)\u001b[0m           \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--quiet\u001b[0m             \u001b[1;32m-q\u001b[0m  Only show log warnings                               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--strict\u001b[0m                Don't catch exceptions, run additional code checks   \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                         to help development.                                 \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--require-logs\u001b[0m          Require all explicitly requested modules to have log \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                         files. If not, MultiQC will exit with an error.      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--profile-runtime\u001b[0m       Add analysis of how long MultiQC takes to run to the \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                         report                                               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--no-megaqc-upload\u001b[0m      Don't upload generated report to MegaQC, even if     \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                         MegaQC options are found                             \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--no-ansi\u001b[0m               Disable coloured log output                          \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--version\u001b[0m               Show the version and exit.                           \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--help\u001b[0m              \u001b[1;32m-h\u001b[0m  Show this message and exit.                          \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "                                                                                \n",
      " See \u001b]8;id=214615;http://multiqc.info\u001b\\http://multiqc.info\u001b]8;;\u001b\\ for more details.                                      \n",
      "\n"
     ]
    }
   ],
   "source": [
    "!multiqc -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=633313;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2m| v1.17\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data/users/zhouz6436/BIOL343/5_fastq/mouse_hippocampus_fastq/qc\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m46/46\u001b[0m  \u001b[0m0m  \n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 23 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : mouse_hippocampus_fastq/qc/multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : mouse_hippocampus_fastq/qc/multiqc_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "!multiqc mouse_hippocampus_fastq/qc -o mouse_hippocampus_fastq/qc\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Edit this block with a short evaluation of each of the report sections. Each evaluation should include your important observations, reasons for potential QC failure, and an opinion on what should be done on the failure (i.e., trimming or filtering). Problem FASTQ files, if any, should be noted.\n",
    "\n",
    "*Per base sequence quality:*\n",
    "\n",
    "*Per tile sequence quality:*\n",
    "\n",
    "*Per sequence quality scores:*\n",
    "\n",
    "*Per base sequence content:*\n",
    "\n",
    "*Per sequence GC content:*\n",
    "\n",
    "*Per base N content:*\n",
    "\n",
    "*Sequence length distribution:*\n",
    "\n",
    "*Sequence duplication levels:*\n",
    "\n",
    "*Overrepresented sequences:*\n",
    "\n",
    "*Adapter content:*"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If necessary, use `cutadapt` to trim and filter your reads. Provide a justification for the `cutadapt` parameters you used:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run QC on the trimmed/filtered reads and generate an aggregated report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Report your observations on which QC metrics improved (or got worse):\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Snakemake Pipeline"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When running bioinformatics studies, it is wise to use notebooks (such as this one) to tie together code, decisions, and data visualization. The above task is something you might generate and report to a client or supervisor who might be interested in sequencing QC and what kind of pre-processing decisions you made prior to alignment. In addition, the preprocessing should be readily reproducible by all relevant stakeholders, so you should supplement notebooks with pipeline scripts. Throughout the semester, we will be creating snakemake pipelines to reproduce the published analysis that you've selected. Follow the below steps to begin producing this pipeline:\n",
    "\n",
    "1. Make a new `Snakefile` and `config.yaml` in `5_fastq`. \n",
    "2. In the `Snakefile`, produce the following rules:  \n",
    "    a. `fetch_fastq` - use `parallel-fastq-dump` to fetch every FASTQ file listed in `config.yaml`.  \n",
    "    b. `fastq_qc` - use `fastqc` to run QC on the raw FASTQ files.  \n",
    "    c. `trim_filter` - use `cutadapt` to trim and/or filter the FASTQ reads using parameters justified in this notebook. ***Note:*** if you choose to trim/filter, it should be done on *all* of the FASTQ files; do not cherry pick.  \n",
    "    d. `trim_qc` - use `fastqc` to run QC on the trimmed FASTQ files.  \n",
    "    e. `report` - use `multiqc` to aggregate all QC data into a single report.\n",
    "\n",
    "This `Snakefile` should not use any `wrappers` because we already have the relevant software installed in the `biol343` conda environment. The pipeline should run to completion when any instructor or classmate runs `snakemake --use-conda`. You may use the documentation found [here](https://multiqc.info/docs/usage/pipelines/#snakemake) or any other online documentation you may find. You may not use any AI tools to complete this homework."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biol343",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
